{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7e4387d2-7683-4c42-93b4-5776e7107cac",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "import pandas as pd\n",
    "from cerberus.cerberus import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6ce57d9e-d288-4b76-b4e9-30dd2498734b",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf = 'tests/files/Canx.gtf' \n",
    "mode = 'tss'\n",
    "dist = 50\n",
    "slack = 50"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "366a969d-d90f-4efa-90e0-a553feed0bb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "bed = get_ends_from_gtf(gtf, 'tss', dist, slack)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "013a4a59-3d18-48b5-b70a-5dc3f244f351",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = get_ics_from_gtf(gtf)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9da6ec9c-ad77-4d27-8154-37d00234e428",
   "metadata": {},
   "source": [
    "## New assign triplets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7aaf28cb-826e-44d2-bbff-cab8eb80e0bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# def add_triplets(gtf, ic_file, tss_bed, tes_bed):\n",
    "#     \"\"\"\n",
    "#     Determines which tss, intron chain, and tes are used from a cerberus \n",
    "#     reference are used in a specific gtf.\n",
    "    \n",
    "#     Parameters:\n",
    "#         gtf (str): File path to gtf you want to assign triplets to\n",
    "#         ic_file (str): File path to intron chain tsv (part of cerberus ref.)\n",
    "#         tss_bed (str): File path to tss bed (part of cerberus ref.)\n",
    "#         tes_bed (str): File path to tes bed (part of cerberus ref.)\n",
    "    \n",
    "#     Returns:\n",
    "#         df (pandas DataFrame): File that maps each transcript from gtf to \n",
    "#             a tss, intron chain, and tes given by the cerberus reference.\n",
    "#             Also includes new transcript ids to refer to each transcript\n",
    "#     \"\"\"\n",
    "#     t_df = pr.read_gtf(gtf)\n",
    "#     t_df = t_df.df\n",
    "#     t_df['gene_id'] = t_df.gene_id.str.split('.', n=1, expand=True)[0]\n",
    "#     t_df = pr.PyRanges(t_df)\n",
    "\n",
    "#     ### intron chain annotation ###\n",
    "\n",
    "#     # get unique intron chains from gtf\n",
    "#     df = t_df.copy()\n",
    "#     df = get_ic(df)\n",
    "#     df.rename({'ic': 'Coordinates'},\n",
    "#               axis=1, inplace=True)\n",
    "\n",
    "#     # read in reference set of intron chains and format them as strings \n",
    "#     # to enable merging\n",
    "#     ref = read_ic_ref(ic_file)\n",
    "\n",
    "#     # merge on intron chain, strand, chromosome, and gene id\n",
    "#     df = df.merge(ref, how='left',\n",
    "#                  on=['Chromosome', 'Strand', \n",
    "#                      'Coordinates', 'gene_id'])\n",
    "\n",
    "#     # formatting\n",
    "#     df.rename({'Name': 'ic_id'}, axis=1, inplace=True)\n",
    "#     df = df[['transcript_id', 'ic', 'ic_id']]\n",
    "\n",
    "#     ### end annotation ###\n",
    "\n",
    "#     for mode, bed_file in zip(['tss', 'tes'], [tss_bed, tes_bed]):\n",
    "\n",
    "#         if mode == 'tss':\n",
    "#             ends = t_df.features.tss()\n",
    "#         elif mode == 'tes':\n",
    "#             ends = t_df.features.tes()\n",
    "\n",
    "#         # merge transcriptome ends with reference ends\n",
    "#         ref = read_end_ref(bed_file, mode=mode)\n",
    "#         ref = pr.PyRanges(ref)\n",
    "#         ends = ends.join(ref, \n",
    "#                          strandedness='same', \n",
    "#                          how='left').df\n",
    "#         ends = ends.loc[ends.gene_id == ends.gene_id_b]\n",
    "\n",
    "#         # formatting \n",
    "#         ends.rename({'Name': '{}_id'.format(mode)}, axis=1, inplace=True)\n",
    "#         ends = ends[['transcript_id', mode, '{}_id'.format(mode)]]\n",
    "\n",
    "#         # add ends into map df\n",
    "#         df = df.merge(ends, how='left', on='transcript_id')\n",
    "\n",
    "#     ### creating map file ###\n",
    "\n",
    "#     # get gene id / name and transcript name from original gtf\n",
    "#     t_df = t_df.df\n",
    "#     t_df = t_df.loc[t_df.Feature == 'transcript']\n",
    "#     if 'gene_name' not in t_df.columns:\n",
    "#         t_df['gene_name'] = t_df.transcript_name.str.split('-', n=1, expand=True)[0]\n",
    "#     t_df = t_df[['gene_id', 'gene_name', \n",
    "#                  'transcript_id', 'transcript_name']]\n",
    "#     df = df.merge(t_df, how='left', on='transcript_id')\n",
    "\n",
    "#     # create triplets and rename old ids\n",
    "#     df.rename({'transcript_id': 'original_transcript_id', \n",
    "#                'transcript_name': 'original_transcript_name'},\n",
    "#                axis=1, inplace=True)\n",
    "#     df['transcript_triplet'] = '['+df.tss.astype(str)+', '+\\\n",
    "#                                    df.ic.astype(str)+', '+\\\n",
    "#                                    df.tes.astype(str)+']'\n",
    "#     df['transcript_id'] = df['gene_id']+' '+df.transcript_triplet\n",
    "#     df['transcript_name'] = df['gene_name']+' '+df.transcript_triplet\n",
    "    \n",
    "#     return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "3cbf79a7-5ddf-4b73-8bac-7473b7f2c9aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf = 'tests/files/Canx.gtf'\n",
    "ic_file = 'tests/files/Canx_ics.tsv'\n",
    "tss_bed = 'tests/files/Canx_tss.bed'\n",
    "tes_bed = 'tests/files/Canx_tes.bed'\n",
    "h5 = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "12ba61e2-a9c4-4196-8c25-129656ff84fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = add_triplets(gtf, ic_file, tss_bed, tes_bed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0f2e227a-9017-4c91-acec-b37331cb9f85",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>original_transcript_id</th>\n",
       "      <th>ic</th>\n",
       "      <th>ic_id</th>\n",
       "      <th>tss</th>\n",
       "      <th>tss_id</th>\n",
       "      <th>tes</th>\n",
       "      <th>tes_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>original_transcript_name</th>\n",
       "      <th>transcript_triplet</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUST00000020637.8</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSMUSG00000020368_3</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00000020368_1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000020368_2</td>\n",
       "      <td>ENSMUSG00000020368</td>\n",
       "      <td>Canx</td>\n",
       "      <td>Canx-001</td>\n",
       "      <td>[1, 3, 2]</td>\n",
       "      <td>ENSMUSG00000020368 [1, 3, 2]</td>\n",
       "      <td>Canx [1, 3, 2]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUST00000146979.1</td>\n",
       "      <td>5</td>\n",
       "      <td>ENSMUSG00000020368_5</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSMUSG00000020368_4</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSMUSG00000020368_4</td>\n",
       "      <td>ENSMUSG00000020368</td>\n",
       "      <td>Canx</td>\n",
       "      <td>Canx-003</td>\n",
       "      <td>[4, 5, 4]</td>\n",
       "      <td>ENSMUSG00000020368 [4, 5, 4]</td>\n",
       "      <td>Canx [4, 5, 4]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUST00000153068.1</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSMUSG00000020368_4</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSMUSG00000020368_3</td>\n",
       "      <td>5</td>\n",
       "      <td>ENSMUSG00000020368_5</td>\n",
       "      <td>ENSMUSG00000020368</td>\n",
       "      <td>Canx</td>\n",
       "      <td>Canx-002</td>\n",
       "      <td>[3, 4, 5]</td>\n",
       "      <td>ENSMUSG00000020368 [3, 4, 5]</td>\n",
       "      <td>Canx [3, 4, 5]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUST00000155801.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000020368_2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000020368_2</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSMUSG00000020368_3</td>\n",
       "      <td>ENSMUSG00000020368</td>\n",
       "      <td>Canx</td>\n",
       "      <td>Canx-004</td>\n",
       "      <td>[2, 2, 3]</td>\n",
       "      <td>ENSMUSG00000020368 [2, 2, 3]</td>\n",
       "      <td>Canx [2, 2, 3]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUST00000155801.2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000020368_2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000020368_2</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSMUSG00000020368_3</td>\n",
       "      <td>ENSMUSG00000020368</td>\n",
       "      <td>Canx</td>\n",
       "      <td>Canx-004</td>\n",
       "      <td>[2, 2, 3]</td>\n",
       "      <td>ENSMUSG00000020368 [2, 2, 3]</td>\n",
       "      <td>Canx [2, 2, 3]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSMUST00000179865.6</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00000020368_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00000020368_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00000020368_1</td>\n",
       "      <td>ENSMUSG00000020368</td>\n",
       "      <td>Canx</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>[1, 1, 1]</td>\n",
       "      <td>ENSMUSG00000020368 [1, 1, 1]</td>\n",
       "      <td>Canx [1, 1, 1]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  original_transcript_id ic                 ic_id tss                tss_id  \\\n",
       "0   ENSMUST00000020637.8  3  ENSMUSG00000020368_3   1  ENSMUSG00000020368_1   \n",
       "1   ENSMUST00000146979.1  5  ENSMUSG00000020368_5   4  ENSMUSG00000020368_4   \n",
       "2   ENSMUST00000153068.1  4  ENSMUSG00000020368_4   3  ENSMUSG00000020368_3   \n",
       "3   ENSMUST00000155801.1  2  ENSMUSG00000020368_2   2  ENSMUSG00000020368_2   \n",
       "4   ENSMUST00000155801.2  2  ENSMUSG00000020368_2   2  ENSMUSG00000020368_2   \n",
       "5   ENSMUST00000179865.6  1  ENSMUSG00000020368_1   1  ENSMUSG00000020368_1   \n",
       "\n",
       "  tes                tes_id             gene_id gene_name  \\\n",
       "0   2  ENSMUSG00000020368_2  ENSMUSG00000020368      Canx   \n",
       "1   4  ENSMUSG00000020368_4  ENSMUSG00000020368      Canx   \n",
       "2   5  ENSMUSG00000020368_5  ENSMUSG00000020368      Canx   \n",
       "3   3  ENSMUSG00000020368_3  ENSMUSG00000020368      Canx   \n",
       "4   3  ENSMUSG00000020368_3  ENSMUSG00000020368      Canx   \n",
       "5   1  ENSMUSG00000020368_1  ENSMUSG00000020368      Canx   \n",
       "\n",
       "  original_transcript_name transcript_triplet                 transcript_id  \\\n",
       "0                 Canx-001          [1, 3, 2]  ENSMUSG00000020368 [1, 3, 2]   \n",
       "1                 Canx-003          [4, 5, 4]  ENSMUSG00000020368 [4, 5, 4]   \n",
       "2                 Canx-002          [3, 4, 5]  ENSMUSG00000020368 [3, 4, 5]   \n",
       "3                 Canx-004          [2, 2, 3]  ENSMUSG00000020368 [2, 2, 3]   \n",
       "4                 Canx-004          [2, 2, 3]  ENSMUSG00000020368 [2, 2, 3]   \n",
       "5                 Canx-201          [1, 1, 1]  ENSMUSG00000020368 [1, 1, 1]   \n",
       "\n",
       "  transcript_name  \n",
       "0  Canx [1, 3, 2]  \n",
       "1  Canx [4, 5, 4]  \n",
       "2  Canx [3, 4, 5]  \n",
       "3  Canx [2, 2, 3]  \n",
       "4  Canx [2, 2, 3]  \n",
       "5  Canx [1, 1, 1]  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae8b00cf-6562-4c49-b2c8-919d35f821ca",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
