{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7e4387d2-7683-4c42-93b4-5776e7107cac",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import h5py \n",
    "import os\n",
    "\n",
    "from cerberus.cerberus import *\n",
    "from cerberus.main import *\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "05080a6f-d916-40a1-9a21-4316d12c3169",
   "metadata": {},
   "source": [
    "## convert transcriptome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "816e661c-82f6-4001-a4b3-4cb5b1b88870",
   "metadata": {},
   "outputs": [],
   "source": [
    "# def add_triplets(t_df, ic, tss, tes):\n",
    "#     \"\"\"\n",
    "#     Determines which tss, intron chain, and tes are used from a cerberus\n",
    "#     reference are used in a specific gtf.\n",
    "\n",
    "#     Parameters:\n",
    "#         gtf (str): File path to gtf you want to assign triplets to\n",
    "#         ic_file (str): File path to intron chain tsv (part of cerberus ref.)\n",
    "#         tss_bed (str): File path to tss bed (part of cerberus ref.)\n",
    "#         tes_bed (str): File path to tes bed (part of cerberus ref.)\n",
    "\n",
    "#     Returns:\n",
    "#         df (pandas DataFrame): File that maps each transcript from gtf to\n",
    "#             a tss, intron chain, and tes given by the cerberus reference.\n",
    "#             Also includes new transcript ids to refer to each transcript\n",
    "#     \"\"\"\n",
    "\n",
    "#     ### intron chain annotation ###\n",
    "\n",
    "#     # get unique intron chains from gtf\n",
    "#     df = t_df.copy()\n",
    "#     df = get_ic(df)\n",
    "#     df.rename({'ic': 'Coordinates'},\n",
    "#               axis=1, inplace=True)\n",
    "\n",
    "#     # read in reference set of intron chains\n",
    "#     ref = ic\n",
    "\n",
    "#     # merge on intron chain, strand, chromosome, and gene id\n",
    "#     df = df.merge(ref, how='left',\n",
    "#                  on=['Chromosome', 'Strand',\n",
    "#                      'Coordinates', 'gene_id'])\n",
    "\n",
    "#     # formatting\n",
    "#     df.rename({'Name': 'ic_id'}, axis=1, inplace=True)\n",
    "#     df = df[['transcript_id', 'ic', 'ic_id']]\n",
    "\n",
    "#     ### end annotation ###\n",
    "\n",
    "#     for mode, ref in zip(['tss', 'tes'], [tss, tes]):\n",
    "\n",
    "#         if mode == 'tss':\n",
    "#             ends = t_df.features.tss()\n",
    "#         elif mode == 'tes':\n",
    "#             ends = t_df.features.tes()\n",
    "\n",
    "#         # merge transcriptome ends with reference ends\n",
    "#         ref = read_cerberus_ends(bed_file, mode=mode)\n",
    "#         ref = pr.PyRanges(ref)\n",
    "#         ends = ends.join(ref,\n",
    "#                          strandedness='same',\n",
    "#                          how='left').df\n",
    "#         ends = ends.loc[ends.gene_id == ends.gene_id_b]\n",
    "\n",
    "#         # formatting\n",
    "#         ends.rename({'Name': '{}_id'.format(mode)}, axis=1, inplace=True)\n",
    "#         ends = ends[['transcript_id', mode, '{}_id'.format(mode)]]\n",
    "\n",
    "#         # add ends into map df\n",
    "#         df = df.merge(ends, how='left', on='transcript_id')\n",
    "\n",
    "#     ### creating map file ###\n",
    "\n",
    "#     # get gene id / name and transcript name from original gtf\n",
    "#     t_df = t_df.df\n",
    "#     t_df = t_df.loc[t_df.Feature == 'transcript']\n",
    "#     if 'gene_name' not in t_df.columns:\n",
    "#         t_df['gene_name'] = t_df.transcript_name.str.split('-', n=1, expand=True)[0]\n",
    "#     t_df = t_df[['gene_id', 'gene_name',\n",
    "#                  'transcript_id', 'transcript_name']]\n",
    "#     df = df.merge(t_df, how='left', on='transcript_id')\n",
    "\n",
    "#     # create triplets and rename old ids\n",
    "#     df.rename({'transcript_id': 'original_transcript_id',\n",
    "#                'transcript_name': 'original_transcript_name'},\n",
    "#                axis=1, inplace=True)\n",
    "#     df['transcript_triplet'] = '['+df.tss.astype(str)+', '+\\\n",
    "#                                    df.ic.astype(str)+', '+\\\n",
    "#                                    df.tes.astype(str)+']'\n",
    "#     df['transcript_id'] = df['gene_id']+' '+df.transcript_triplet\n",
    "#     df['transcript_name'] = df['gene_name']+' '+df.transcript_triplet\n",
    "\n",
    "#     return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f989ecbf-a44f-451c-9888-55c9aa49dbc9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def assign_triplets(gtf_df, tss, ic, tes):\n",
    "    \"\"\"\n",
    "    Determines which tss, intron chain, and tes are used from a cerberus\n",
    "    reference are used in a specific gtf.\n",
    "\n",
    "    Parameters:\n",
    "        gtf (pyranges PyRanges): PyRanges GTF object for transcriptome to assign triplets to\n",
    "        ic_file (pandas DataFrame): df of intron chains from cerberus ref.\n",
    "        tss_bed (str): PyRanges obj of tsss from cerberus ref\n",
    "        tes_bed (str): PyRanges obj of tess from cerberus ref\n",
    "\n",
    "    Returns:\n",
    "        df (pandas DataFrame): File that maps each transcript from gtf to\n",
    "            a tss, intron chain, and tes given by the cerberus reference.\n",
    "            Also includes new transcript ids to refer to each transcript\n",
    "    \"\"\"\n",
    "\n",
    "    ### intron chain ###\n",
    "\n",
    "    # get intron chains from input transcriptome\n",
    "    df = gtf_df.copy()\n",
    "    df = get_ic(df)\n",
    "    df.rename({'ic': 'Coordinates'}, axis=1, inplace=True)\n",
    "\n",
    "    # merge on intron chain, strand, chromosome, and gene id\n",
    "    df = df.merge(ic, how='left',\n",
    "                  on=['Chromosome', 'Strand',\n",
    "                      'Coordinates', 'gene_id'])\n",
    "\n",
    "    # formatting\n",
    "    df.rename({'Name': 'ic_id'}, axis=1, inplace=True)\n",
    "    df = df[['transcript_id', 'ic', 'ic_id']]\n",
    "\n",
    "\n",
    "    ### ends ###\n",
    "    for mode, ref in zip(['tss', 'tes'], [tss, tes]):\n",
    "        if mode == 'tss':\n",
    "            ends = gtf_df.features.tss()\n",
    "        elif mode == 'tes':\n",
    "            ends = gtf_df.features.tes()\n",
    "\n",
    "        # limit to relevant columns\n",
    "        ends = ends[['Chromosome', 'Start', 'End', 'Strand',\n",
    "                     'gene_id', 'transcript_id']]\n",
    "\n",
    "        # get only the relevant columns and deduplicate\n",
    "        ends = ends.df\n",
    "        t_ends = ends.copy(deep=True)\n",
    "        ends.drop('transcript_id', axis=1, inplace=True)\n",
    "        ends.drop_duplicates(inplace=True)\n",
    "        ends = pr.PyRanges(ends)\n",
    "\n",
    "        # find closest interval in ref\n",
    "        ends = ends.nearest(ref, \n",
    "                            strandedness=None)\n",
    "\n",
    "        # fix the ends with mismatching gene ids - this part can be slow :(\n",
    "        ends = ends.df\n",
    "        fix_ends = ends.loc[ends.gene_id != ends.gene_id_b]\n",
    "        fix_ends = fix_ends[['Chromosome', 'Start', 'End', 'Strand',\n",
    "                             'gene_id']]\n",
    "        ends = ends.loc[ends.gene_id == ends.gene_id_b]\n",
    "\n",
    "        for i, gid in enumerate(fix_ends.gene_id.unique().tolist()):\n",
    "            gene_ends = fix_ends.loc[fix_ends.gene_id == gid].copy(deep=True)\n",
    "            gene_ends = pr.PyRanges(gene_ends)\n",
    "            gene_refs = ref.df.loc[ref.df.gene_id == gid].copy(deep=True)\n",
    "            gene_refs = pr.PyRanges(gene_refs)\n",
    "            gene_ends = gene_ends.nearest(ref, \n",
    "                                          strandedness=None)\n",
    "            gene_ends = gene_ends.df\n",
    "            ends = pd.concat([ends, gene_ends])\n",
    "\n",
    "            if i % 100 == 0:\n",
    "                print('Processed {} / {} genes'.format(i, len(fix_ends.gene_id.unique().tolist())))\n",
    "\n",
    "        # merge back in to get transcript ids\n",
    "        t_ends = t_ends.merge(ends, how='left',\n",
    "                          on=['Chromosome', 'Start', 'End', 'Strand', 'gene_id'])\n",
    "\n",
    "        # formatting\n",
    "        t_ends.rename({'Name': '{}_id'.format(mode)}, axis=1, inplace=True)\n",
    "        t_ends = t_ends[['transcript_id', '{}_id'.format(mode), mode]]\n",
    "\n",
    "        # merge with ic ids\n",
    "        df = df.merge(t_ends, how='left', on='transcript_id')\n",
    "        \n",
    "    ### creating map file ###\n",
    "\n",
    "    # get gene id / name and transcript name from original gtf\n",
    "    gtf_df = gtf_df.df\n",
    "    gtf_df = gtf_df.loc[gtf_df.Feature == 'transcript']\n",
    "    if 'gene_name' not in gtf_df.columns:\n",
    "        gtf_df['gene_name'] = gtf_df.transcript_name.str.split('-', n=1, expand=True)[0]\n",
    "    gtf_df = gtf_df[['gene_id', 'gene_name',\n",
    "                 'transcript_id', 'transcript_name']]\n",
    "    df = df.merge(gtf_df, how='left', on='transcript_id')\n",
    "\n",
    "    # create triplets and rename old ids\n",
    "    df.rename({'transcript_id': 'original_transcript_id',\n",
    "               'transcript_name': 'original_transcript_name'},\n",
    "               axis=1, inplace=True)\n",
    "    df['transcript_triplet'] = '['+df.tss.astype(str)+','+\\\n",
    "                                   df.ic.astype(str)+','+\\\n",
    "                                   df.tes.astype(str)+']'\n",
    "    df['transcript_id'] = df['gene_id']+' '+df.transcript_triplet\n",
    "    df['transcript_name'] = df['gene_name']+' '+df.transcript_triplet\n",
    "\n",
    "    return df\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a1cc99e2-f14f-4fe3-aeba-c2d88eae0d4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_transcriptome_command(gtf, h5, o):\n",
    "    convert_transcriptome(gtf, h5, o)\n",
    "    \n",
    "def convert_transcriptome(gtf, h5, o):\n",
    "    \n",
    "    # read in / format existing reference\n",
    "    ic, tss, tes, _ = read_h5(h5, as_pyranges=False)\n",
    "    ic = split_cerberus_id(ic, 'ic')\n",
    "    tss = split_cerberus_id(tss, 'tss')\n",
    "    tes = split_cerberus_id(tes, 'tes')\n",
    "    tss = pr.PyRanges(tss)\n",
    "    tes = pr.PyRanges(tes)\n",
    "    \n",
    "    # read in transcriptome to convert to cerberus\n",
    "    gtf_df = pr.read_gtf(gtf).df\n",
    "    gtf_df['gene_id'] = gtf_df.gene_id.str.split('.', expand=True)[0]\n",
    "    gtf_df = pr.PyRanges(gtf_df)\n",
    "    \n",
    "    df = assign_triplets(gtf_df, tss, ic, tes)\n",
    "    \n",
    "    # write h5 file\n",
    "    tss = tss.df\n",
    "    tes = tes.df\n",
    "    write_h5(ic, tss, tes, h5, m=df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "aaf8c257-6dce-471a-9233-b50f8624c0f9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processed 0 / 2025 genes\n",
      "Processed 100 / 2025 genes\n",
      "Processed 200 / 2025 genes\n",
      "Processed 300 / 2025 genes\n",
      "Processed 400 / 2025 genes\n",
      "Processed 500 / 2025 genes\n",
      "Processed 600 / 2025 genes\n",
      "Processed 700 / 2025 genes\n",
      "Processed 800 / 2025 genes\n",
      "Processed 900 / 2025 genes\n",
      "Processed 1000 / 2025 genes\n",
      "Processed 1100 / 2025 genes\n",
      "Processed 1200 / 2025 genes\n",
      "Processed 1300 / 2025 genes\n",
      "Processed 1400 / 2025 genes\n",
      "Processed 1500 / 2025 genes\n",
      "Processed 1600 / 2025 genes\n",
      "Processed 1700 / 2025 genes\n",
      "Processed 1800 / 2025 genes\n",
      "Processed 1900 / 2025 genes\n",
      "Processed 2000 / 2025 genes\n",
      "Processed 0 / 1423 genes\n",
      "Processed 100 / 1423 genes\n",
      "Processed 200 / 1423 genes\n",
      "Processed 300 / 1423 genes\n",
      "Processed 400 / 1423 genes\n",
      "Processed 500 / 1423 genes\n",
      "Processed 600 / 1423 genes\n",
      "Processed 700 / 1423 genes\n",
      "Processed 800 / 1423 genes\n",
      "Processed 900 / 1423 genes\n",
      "Processed 1000 / 1423 genes\n",
      "Processed 1100 / 1423 genes\n",
      "Processed 1200 / 1423 genes\n",
      "Processed 1300 / 1423 genes\n",
      "Processed 1400 / 1423 genes\n"
     ]
    }
   ],
   "source": [
    "h5 = '/Users/fairliereese/mortazavi_lab/data/rnawg/lr_bulk/cerberus/cerberus_ref.h5'\n",
    "gtf = '/Users/fairliereese/mortazavi_lab/data/rnawg/lr_bulk/talon/human_known_nic_nnc_talon.gtf'\n",
    "# gtf_df, tss, ic, tes = convert_transcriptome_command(gtf, h5)\n",
    "df, ic, tss, tes = convert_transcriptome(gtf, h5)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "a5d4dcfa-5e44-4a30-8ef4-f98190d43072",
   "metadata": {},
   "outputs": [],
   "source": [
    "fname = 'talon.h5'\n",
    "# tss = tss.df\n",
    "# tes = tes.df\n",
    "\n",
    "write_h5(ic, tss, tes, fname, m=df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6916dc67-dc98-4a1d-8ad1-dbb90fda2e34",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>original_transcript_id</th>\n",
       "      <th>ic</th>\n",
       "      <th>ic_id</th>\n",
       "      <th>tss_id</th>\n",
       "      <th>tss</th>\n",
       "      <th>tes_id</th>\n",
       "      <th>tes</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>original_transcript_name</th>\n",
       "      <th>transcript_triplet</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENCODEHT000206942</td>\n",
       "      <td>1</td>\n",
       "      <td>ENCODEHG000058846_1</td>\n",
       "      <td>ENCODEHG000058846_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENCODEHG000058846_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>ENCODEHT000206942</td>\n",
       "      <td>[1,1,1]</td>\n",
       "      <td>ENCODEHG000058846 [1,1,1]</td>\n",
       "      <td>ENCODEHG000058846 [1,1,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENCODEHT000206867</td>\n",
       "      <td>4</td>\n",
       "      <td>ENCODEHG000058837_4</td>\n",
       "      <td>ENCODEHG000058837_2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENCODEHG000058837_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>ENCODEHT000206867</td>\n",
       "      <td>[2,4,1]</td>\n",
       "      <td>ENCODEHG000058837 [2,4,1]</td>\n",
       "      <td>ENCODEHG000058837 [2,4,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENCODEHT000206868</td>\n",
       "      <td>2</td>\n",
       "      <td>ENCODEHG000058837_2</td>\n",
       "      <td>ENCODEHG000058837_2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENCODEHG000058837_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>ENCODEHT000206868</td>\n",
       "      <td>[2,2,1]</td>\n",
       "      <td>ENCODEHG000058837 [2,2,1]</td>\n",
       "      <td>ENCODEHG000058837 [2,2,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENCODEHT000206870</td>\n",
       "      <td>3</td>\n",
       "      <td>ENCODEHG000058837_3</td>\n",
       "      <td>ENCODEHG000058837_2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENCODEHG000058837_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>ENCODEHT000206870</td>\n",
       "      <td>[2,3,1]</td>\n",
       "      <td>ENCODEHG000058837 [2,3,1]</td>\n",
       "      <td>ENCODEHG000058837 [2,3,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENCODEHT000206886</td>\n",
       "      <td>1</td>\n",
       "      <td>ENCODEHG000058837_1</td>\n",
       "      <td>ENCODEHG000058837_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENCODEHG000058837_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>ENCODEHT000206886</td>\n",
       "      <td>[1,1,1]</td>\n",
       "      <td>ENCODEHG000058837 [1,1,1]</td>\n",
       "      <td>ENCODEHG000058837 [1,1,1]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  original_transcript_id  ic                ic_id               tss_id  tss  \\\n",
       "0      ENCODEHT000206942   1  ENCODEHG000058846_1  ENCODEHG000058846_1    1   \n",
       "1      ENCODEHT000206867   4  ENCODEHG000058837_4  ENCODEHG000058837_2    2   \n",
       "2      ENCODEHT000206868   2  ENCODEHG000058837_2  ENCODEHG000058837_2    2   \n",
       "3      ENCODEHT000206870   3  ENCODEHG000058837_3  ENCODEHG000058837_2    2   \n",
       "4      ENCODEHT000206886   1  ENCODEHG000058837_1  ENCODEHG000058837_1    1   \n",
       "\n",
       "                tes_id  tes            gene_id          gene_name  \\\n",
       "0  ENCODEHG000058846_1    1  ENCODEHG000058846  ENCODEHG000058846   \n",
       "1  ENCODEHG000058837_1    1  ENCODEHG000058837  ENCODEHG000058837   \n",
       "2  ENCODEHG000058837_1    1  ENCODEHG000058837  ENCODEHG000058837   \n",
       "3  ENCODEHG000058837_1    1  ENCODEHG000058837  ENCODEHG000058837   \n",
       "4  ENCODEHG000058837_1    1  ENCODEHG000058837  ENCODEHG000058837   \n",
       "\n",
       "  original_transcript_name transcript_triplet              transcript_id  \\\n",
       "0        ENCODEHT000206942            [1,1,1]  ENCODEHG000058846 [1,1,1]   \n",
       "1        ENCODEHT000206867            [2,4,1]  ENCODEHG000058837 [2,4,1]   \n",
       "2        ENCODEHT000206868            [2,2,1]  ENCODEHG000058837 [2,2,1]   \n",
       "3        ENCODEHT000206870            [2,3,1]  ENCODEHG000058837 [2,3,1]   \n",
       "4        ENCODEHT000206886            [1,1,1]  ENCODEHG000058837 [1,1,1]   \n",
       "\n",
       "             transcript_name  \n",
       "0  ENCODEHG000058846 [1,1,1]  \n",
       "1  ENCODEHG000058837 [2,4,1]  \n",
       "2  ENCODEHG000058837 [2,2,1]  \n",
       "3  ENCODEHG000058837 [2,3,1]  \n",
       "4  ENCODEHG000058837 [1,1,1]  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "0baceb12-8dc8-49f5-aa37-ac6912d776a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "### creating map file ###\n",
    "\n",
    "# get gene id / name and transcript name from original gtf\n",
    "gtf_df = gtf_df.df\n",
    "gtf_df = gtf_df.loc[gtf_df.Feature == 'transcript']\n",
    "if 'gene_name' not in gtf_df.columns:\n",
    "    gtf_df['gene_name'] = gtf_df.transcript_name.str.split('-', n=1, expand=True)[0]\n",
    "gtf_df = gtf_df[['gene_id', 'gene_name',\n",
    "             'transcript_id', 'transcript_name']]\n",
    "df = df.merge(gtf_df, how='left', on='transcript_id')\n",
    "\n",
    "# create triplets and rename old ids\n",
    "df.rename({'transcript_id': 'original_transcript_id',\n",
    "           'transcript_name': 'original_transcript_name'},\n",
    "           axis=1, inplace=True)\n",
    "df['transcript_triplet'] = '['+df.tss.astype(str)+','+\\\n",
    "                               df.ic.astype(str)+','+\\\n",
    "                               df.tes.astype(str)+']'\n",
    "df['transcript_id'] = df['gene_id']+' '+df.transcript_triplet\n",
    "df['transcript_name'] = df['gene_name']+' '+df.transcript_triplet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "583767ca-6db2-4d8f-a240-23b2ed002546",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>original_transcript_id</th>\n",
       "      <th>ic</th>\n",
       "      <th>ic_id</th>\n",
       "      <th>tss_id</th>\n",
       "      <th>tss</th>\n",
       "      <th>tes_id</th>\n",
       "      <th>tes</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>original_transcript_name</th>\n",
       "      <th>transcript_triplet</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>132936</th>\n",
       "      <td>ENST00000612565.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000278384_1</td>\n",
       "      <td>ENSG00000278384_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000278384_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000278384</td>\n",
       "      <td>AL354822.1</td>\n",
       "      <td>AL354822.1-201</td>\n",
       "      <td>[1, 1, 1]</td>\n",
       "      <td>ENSG00000278384 [1, 1, 1]</td>\n",
       "      <td>AL354822.1 [1, 1, 1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132937</th>\n",
       "      <td>ENCODEHT000218666</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000273748_2</td>\n",
       "      <td>ENSG00000273748_2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000273748_2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000273748</td>\n",
       "      <td>AL592183.1</td>\n",
       "      <td>ENCODEHT000218666</td>\n",
       "      <td>[2, 2, 2]</td>\n",
       "      <td>ENSG00000273748 [2, 2, 2]</td>\n",
       "      <td>AL592183.1 [2, 2, 2]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132938</th>\n",
       "      <td>ENCODEHT000218669</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSG00000273748_3</td>\n",
       "      <td>ENSG00000273748_2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000273748_2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000273748</td>\n",
       "      <td>AL592183.1</td>\n",
       "      <td>ENCODEHT000218669</td>\n",
       "      <td>[2, 3, 2]</td>\n",
       "      <td>ENSG00000273748 [2, 3, 2]</td>\n",
       "      <td>AL592183.1 [2, 3, 2]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132939</th>\n",
       "      <td>ENCODEHT001163615</td>\n",
       "      <td>1</td>\n",
       "      <td>ENCODEHG000060713_1</td>\n",
       "      <td>ENCODEHG000060713_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENCODEHG000060713_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENCODEHG000060713</td>\n",
       "      <td>ENCODEHG000060713</td>\n",
       "      <td>ENCODEHT001163615</td>\n",
       "      <td>[1, 1, 1]</td>\n",
       "      <td>ENCODEHG000060713 [1, 1, 1]</td>\n",
       "      <td>ENCODEHG000060713 [1, 1, 1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132940</th>\n",
       "      <td>ENST00000616830.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000278625_1</td>\n",
       "      <td>ENSG00000278625_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000278625_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000278625</td>\n",
       "      <td>RF00026</td>\n",
       "      <td>RF00026.88-201</td>\n",
       "      <td>[1, 1, 1]</td>\n",
       "      <td>ENSG00000278625 [1, 1, 1]</td>\n",
       "      <td>RF00026 [1, 1, 1]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       original_transcript_id  ic                ic_id               tss_id  \\\n",
       "132936      ENST00000612565.1   1    ENSG00000278384_1    ENSG00000278384_1   \n",
       "132937      ENCODEHT000218666   2    ENSG00000273748_2    ENSG00000273748_2   \n",
       "132938      ENCODEHT000218669   3    ENSG00000273748_3    ENSG00000273748_2   \n",
       "132939      ENCODEHT001163615   1  ENCODEHG000060713_1  ENCODEHG000060713_1   \n",
       "132940      ENST00000616830.1   1    ENSG00000278625_1    ENSG00000278625_1   \n",
       "\n",
       "        tss               tes_id  tes            gene_id          gene_name  \\\n",
       "132936    1    ENSG00000278384_1    1    ENSG00000278384         AL354822.1   \n",
       "132937    2    ENSG00000273748_2    2    ENSG00000273748         AL592183.1   \n",
       "132938    2    ENSG00000273748_2    2    ENSG00000273748         AL592183.1   \n",
       "132939    1  ENCODEHG000060713_1    1  ENCODEHG000060713  ENCODEHG000060713   \n",
       "132940    1    ENSG00000278625_1    1    ENSG00000278625            RF00026   \n",
       "\n",
       "       original_transcript_name transcript_triplet  \\\n",
       "132936           AL354822.1-201          [1, 1, 1]   \n",
       "132937        ENCODEHT000218666          [2, 2, 2]   \n",
       "132938        ENCODEHT000218669          [2, 3, 2]   \n",
       "132939        ENCODEHT001163615          [1, 1, 1]   \n",
       "132940           RF00026.88-201          [1, 1, 1]   \n",
       "\n",
       "                      transcript_id              transcript_name  \n",
       "132936    ENSG00000278384 [1, 1, 1]         AL354822.1 [1, 1, 1]  \n",
       "132937    ENSG00000273748 [2, 2, 2]         AL592183.1 [2, 2, 2]  \n",
       "132938    ENSG00000273748 [2, 3, 2]         AL592183.1 [2, 3, 2]  \n",
       "132939  ENCODEHG000060713 [1, 1, 1]  ENCODEHG000060713 [1, 1, 1]  \n",
       "132940    ENSG00000278625 [1, 1, 1]            RF00026 [1, 1, 1]  "
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "68d6acb2-bccc-4959-8db8-76983469df7f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "93818\n",
      "1871\n"
     ]
    }
   ],
   "source": [
    "print(len(ends.index))\n",
    "print(len(ends.loc[ends.gene_id != ends.gene_id_b].index))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "4c8bb193-97da-4388-9e5c-0c84a5b4e4e6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1423\n"
     ]
    }
   ],
   "source": [
    "fix_ends = ends.loc[ends.gene_id != ends.gene_id_b]\n",
    "fix_ends = fix_ends[['Chromosome', 'Start', 'End', 'Strand',\n",
    "                     'gene_id']]\n",
    "print(len(fix_ends.gene_id.unique().tolist()))\n",
    "ends = ends.loc[ends.gene_id == ends.gene_id_b]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "1077a2e7-4661-4561-a7da-40bb260ad64b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processed 0 / 1423 genes\n",
      "Processed 100 / 1423 genes\n",
      "Processed 200 / 1423 genes\n",
      "Processed 300 / 1423 genes\n",
      "Processed 400 / 1423 genes\n",
      "Processed 500 / 1423 genes\n",
      "Processed 600 / 1423 genes\n",
      "Processed 700 / 1423 genes\n",
      "Processed 800 / 1423 genes\n",
      "Processed 900 / 1423 genes\n",
      "Processed 1000 / 1423 genes\n",
      "Processed 1100 / 1423 genes\n",
      "Processed 1200 / 1423 genes\n",
      "Processed 1300 / 1423 genes\n",
      "Processed 1400 / 1423 genes\n"
     ]
    }
   ],
   "source": [
    "# fix the ends with mismatching gene ids - this part can be slow :(\n",
    "for i, gid in enumerate(fix_ends.gene_id.unique().tolist()):\n",
    "    gene_ends = fix_ends.loc[fix_ends.gene_id == gid].copy(deep=True)\n",
    "    gene_ends = pr.PyRanges(gene_ends)\n",
    "    gene_refs = ref.df.loc[ref.df.gene_id == gid].copy(deep=True)\n",
    "    gene_refs = pr.PyRanges(gene_refs)\n",
    "    gene_ends = gene_ends.nearest(ref, \n",
    "                                  strandedness=None)\n",
    "    gene_ends = gene_ends.df\n",
    "    ends = pd.concat([ends, gene_ends])\n",
    "    \n",
    "    if i % 100 == 0:\n",
    "        print('Processed {} / {} genes'.format(i, len(fix_ends.gene_id.unique().tolist())))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "9459ca18-8f92-4d2f-a707-5672799d4b37",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "93818\n"
     ]
    }
   ],
   "source": [
    "print(len(ends.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "66b18e31-83a8-49d4-b9fe-6dbdcb2b2a2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "t_ends = t_ends.merge(ends, how='left',\n",
    "                      on=['Chromosome', 'Start', 'End', 'Strand', 'gene_id'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "b4dc4282-9a55-497d-b152-6073886096cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# formatting\n",
    "t_ends.rename({'Name': '{}_id'.format(mode)}, axis=1, inplace=True)\n",
    "t_ends = t_ends[['transcript_id', '{}_id'.format(mode), mode]]\n",
    "\n",
    "# merge with ic ids\n",
    "df = df.merge(t_ends, how='left', on='transcript_id')\n",
    "\n",
    "# t_ends.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "3ba8af5d-6e78-41e3-8d26-8f2178965d62",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>ic</th>\n",
       "      <th>ic_id</th>\n",
       "      <th>tes_id</th>\n",
       "      <th>tes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENCODEHT000206942</td>\n",
       "      <td>1</td>\n",
       "      <td>ENCODEHG000058846_1</td>\n",
       "      <td>ENCODEHG000058846_1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENCODEHT000206867</td>\n",
       "      <td>4</td>\n",
       "      <td>ENCODEHG000058837_4</td>\n",
       "      <td>ENCODEHG000058837_1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENCODEHT000206868</td>\n",
       "      <td>2</td>\n",
       "      <td>ENCODEHG000058837_2</td>\n",
       "      <td>ENCODEHG000058837_1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENCODEHT000206870</td>\n",
       "      <td>3</td>\n",
       "      <td>ENCODEHG000058837_3</td>\n",
       "      <td>ENCODEHG000058837_1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENCODEHT000206886</td>\n",
       "      <td>1</td>\n",
       "      <td>ENCODEHG000058837_1</td>\n",
       "      <td>ENCODEHG000058837_1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132936</th>\n",
       "      <td>ENST00000612565.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000278384_1</td>\n",
       "      <td>ENSG00000278384_1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132937</th>\n",
       "      <td>ENCODEHT000218666</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000273748_2</td>\n",
       "      <td>ENSG00000273748_2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132938</th>\n",
       "      <td>ENCODEHT000218669</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSG00000273748_3</td>\n",
       "      <td>ENSG00000273748_2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132939</th>\n",
       "      <td>ENCODEHT001163615</td>\n",
       "      <td>1</td>\n",
       "      <td>ENCODEHG000060713_1</td>\n",
       "      <td>ENCODEHG000060713_1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132940</th>\n",
       "      <td>ENST00000616830.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000278625_1</td>\n",
       "      <td>ENSG00000278625_1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>132941 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            transcript_id  ic                ic_id               tes_id  tes\n",
       "0       ENCODEHT000206942   1  ENCODEHG000058846_1  ENCODEHG000058846_1    1\n",
       "1       ENCODEHT000206867   4  ENCODEHG000058837_4  ENCODEHG000058837_1    1\n",
       "2       ENCODEHT000206868   2  ENCODEHG000058837_2  ENCODEHG000058837_1    1\n",
       "3       ENCODEHT000206870   3  ENCODEHG000058837_3  ENCODEHG000058837_1    1\n",
       "4       ENCODEHT000206886   1  ENCODEHG000058837_1  ENCODEHG000058837_1    1\n",
       "...                   ...  ..                  ...                  ...  ...\n",
       "132936  ENST00000612565.1   1    ENSG00000278384_1    ENSG00000278384_1    1\n",
       "132937  ENCODEHT000218666   2    ENSG00000273748_2    ENSG00000273748_2    2\n",
       "132938  ENCODEHT000218669   3    ENSG00000273748_3    ENSG00000273748_2    2\n",
       "132939  ENCODEHT001163615   1  ENCODEHG000060713_1  ENCODEHG000060713_1    1\n",
       "132940  ENST00000616830.1   1    ENSG00000278625_1    ENSG00000278625_1    1\n",
       "\n",
       "[132941 rows x 5 columns]"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "8695ef34-cb93-4bf9-b734-9954e93c336d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "132941"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t_ends.head()\n",
    "len(t_ends.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "133b9ec1-97e7-4851-b05e-ce6adaa3b8b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_df = gtf_df.df\n",
    "gtf_df['gene_id'] = gtf_df.gene_id.str.split('.', n=1, expand=True)[0]\n",
    "gtf_df = pr.PyRanges(gtf_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "5ed8c2bf-849b-4984-974c-f5a1c29f2f2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "m_df = get_ic(gtf_df)\n",
    "# print(m_df.head())\n",
    "m_df.rename({'ic': 'Coordinates'}, axis=1, inplace=True)\n",
    "m_df = m_df.merge(ic, how='left',\n",
    "                  on=['Chromosome', 'Strand', \n",
    "                      'Coordinates', 'gene_id'])\n",
    "\n",
    "# # formatting\n",
    "# m_df.rename({'Name': 'ic_id'}, axis=1, inplace=True)\n",
    "# m_df = m_df[['transcript_id', 'ic_id']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "6d19f84a-6443-4508-bfb4-3a472870c3f1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>Coordinates</th>\n",
       "      <th>Name</th>\n",
       "      <th>source</th>\n",
       "      <th>ic</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENCODEHT000206942</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>10791-10882-11057-11434</td>\n",
       "      <td>ENCODEHG000058846_1</td>\n",
       "      <td>talon</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENCODEHT000206867</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>10647-10366-10282-7814-7552-6813-6560-6473-633...</td>\n",
       "      <td>ENCODEHG000058837_4</td>\n",
       "      <td>talon</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENCODEHT000206868</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>10444-10366-10282-7814-7552-6813-6560-6473-633...</td>\n",
       "      <td>ENCODEHG000058837_2</td>\n",
       "      <td>talon</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENCODEHT000206870</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>10553-7808-7552-1484</td>\n",
       "      <td>ENCODEHG000058837_3</td>\n",
       "      <td>talon</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENCODEHT000206886</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>10282-7814-7552-6813-6337-1484</td>\n",
       "      <td>ENCODEHG000058837_1</td>\n",
       "      <td>talon</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome Strand      transcript_id            gene_id  \\\n",
       "0      SIRV1      +  ENCODEHT000206942  ENCODEHG000058846   \n",
       "1      SIRV1      -  ENCODEHT000206867  ENCODEHG000058837   \n",
       "2      SIRV1      -  ENCODEHT000206868  ENCODEHG000058837   \n",
       "3      SIRV1      -  ENCODEHT000206870  ENCODEHG000058837   \n",
       "4      SIRV1      -  ENCODEHT000206886  ENCODEHG000058837   \n",
       "\n",
       "                                         Coordinates                 Name  \\\n",
       "0                            10791-10882-11057-11434  ENCODEHG000058846_1   \n",
       "1  10647-10366-10282-7814-7552-6813-6560-6473-633...  ENCODEHG000058837_4   \n",
       "2  10444-10366-10282-7814-7552-6813-6560-6473-633...  ENCODEHG000058837_2   \n",
       "3                               10553-7808-7552-1484  ENCODEHG000058837_3   \n",
       "4                     10282-7814-7552-6813-6337-1484  ENCODEHG000058837_1   \n",
       "\n",
       "  source  ic  \n",
       "0  talon   1  \n",
       "1  talon   4  \n",
       "2  talon   2  \n",
       "3  talon   3  \n",
       "4  talon   1  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "m_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9bacb2b0-3b88-4c18-948d-9b1f1a33b687",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  Chromosome Strand              Coordinates            gene_id  \\\n",
      "0      SIRV1      +  10791-10882-11057-11434  ENCODEHG000058846   \n",
      "\n",
      "       transcript_id  MANE_Select  basic_set  appris_principal  ic_num  \\\n",
      "0  ENCODEHT000206942        False      False               NaN       1   \n",
      "\n",
      "                  Name  \n",
      "0  ENCODEHG000058846_1  \n"
     ]
    }
   ],
   "source": [
    "gtf_to_ics('/Users/fairliereese/mortazavi_lab/data/rnawg/lr_bulk/talon/human_known_nic_nnc_talon.gtf', o='test.tsv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7cab1fa3-5072-43fb-9482-2947d096635d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "81a7ee9c-bb06-4dc0-877f-cf43ea348efb",
   "metadata": {},
   "outputs": [],
   "source": [
    "h5 = '/Users/fairliereese/mortazavi_lab/data/rnawg/lr_bulk/cerberus/cerberus_ref.h5'\n",
    "ic, tss, tes, _ = read_h5(h5)\n",
    "write_h5(ic, tss, tes, h5)\n",
    "ic, tss, tes, _ = read_h5(h5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "8d947c10-a738-4621-a19a-e55fcea1797b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       Chromosome Strand              Coordinates                 Name source\n",
      "241882      SIRV1      +  10791-10882-11057-11434             SIRV1B_1    v29\n",
      "0           SIRV1      +  10791-10882-11057-11434  ENCODEHG000058846_1  talon\n",
      "283060\n",
      "441\n",
      "175\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Coordinates</th>\n",
       "      <th>Name</th>\n",
       "      <th>source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>241790</th>\n",
       "      <td>ERCC-00002</td>\n",
       "      <td>+</td>\n",
       "      <td>1045</td>\n",
       "      <td>ERCC-00002A_1</td>\n",
       "      <td>v29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>241791</th>\n",
       "      <td>ERCC-00003</td>\n",
       "      <td>+</td>\n",
       "      <td>1007</td>\n",
       "      <td>ERCC-00003A_1</td>\n",
       "      <td>v29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>241792</th>\n",
       "      <td>ERCC-00004</td>\n",
       "      <td>+</td>\n",
       "      <td>507</td>\n",
       "      <td>ERCC-00004A_1</td>\n",
       "      <td>v29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>241793</th>\n",
       "      <td>ERCC-00009</td>\n",
       "      <td>+</td>\n",
       "      <td>968</td>\n",
       "      <td>ERCC-00009A_1</td>\n",
       "      <td>v29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>241794</th>\n",
       "      <td>ERCC-00012</td>\n",
       "      <td>+</td>\n",
       "      <td>978</td>\n",
       "      <td>ERCC-00012A_1</td>\n",
       "      <td>v29</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chromosome Strand Coordinates           Name source\n",
       "241790  ERCC-00002      +        1045  ERCC-00002A_1    v29\n",
       "241791  ERCC-00003      +        1007  ERCC-00003A_1    v29\n",
       "241792  ERCC-00004      +         507  ERCC-00004A_1    v29\n",
       "241793  ERCC-00009      +         968  ERCC-00009A_1    v29\n",
       "241794  ERCC-00012      +         978  ERCC-00012A_1    v29"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(ic.loc[ic.Coordinates=='10791-10882-11057-11434'])\n",
    "print(len(ic.loc[ic.Name.str.contains('ENSG')]))\n",
    "print(len(ic.loc[ic.Name.str.contains('ENCODE')]))\n",
    "print(len(ic.loc[~(ic.Name.str.contains('ENSG'))&~(ic.Name.str.contains('ENCODE'))]))\n",
    "ic.loc[~(ic.Name.str.contains('ENSG'))&~(ic.Name.str.contains('ENCODE'))].head()\n",
    "# uhhhh I don't think that there are any sirv /ercc / long sirv genes in our talon stuff oops\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "326cc1dd-867f-40da-bbe9-53d90b7213f1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\u001b[0;31mSignature:\u001b[0m \u001b[0mwrite_h5\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mic\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtss\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtes\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0moname\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mm\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
       "\u001b[0;31mDocstring:\u001b[0m\n",
       "Write a cerberus transcriptome as an h5df file\n",
       "\n",
       "Parameters:\n",
       "    ic (pandas DataFrame): DataFrame of intron chains\n",
       "    tss (pandas DataFrame): DataFrame in bed format of tsss\n",
       "    tes (pandas DataFrame): DataFrame in bed format of tess\n",
       "    oname (str): Output file name ending in '.h5'\n",
       "    m (pandas DataFrame): DataFrame of map file\n",
       "\u001b[0;31mFile:\u001b[0m      ~/Documents/programming/mortazavi_lab/bin/cerberus/cerberus/cerberus.py\n",
       "\u001b[0;31mType:\u001b[0m      function\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# ?write_h5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a01f460-4322-4fc7-9427-b62869a67e4e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
