{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7e4387d2-7683-4c42-93b4-5776e7107cac",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import h5py \n",
    "import os\n",
    "\n",
    "from cerberus.cerberus import *\n",
    "from cerberus.main import *\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e3c24a57-cb3a-4fe0-84ca-59b2926c949a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def replace_gtf_ids(gtf, h5, agg):\n",
    "    \"\"\"\n",
    "    Replace transcript ids and names in a gtf with the triplets\n",
    "    calculated from assign_triplets\n",
    "\n",
    "    Parameters:\n",
    "        gtf (str): Path to gtf file\n",
    "        h5 (str): Path to h5 annotation (output from assign)\n",
    "        agg (bool): Whether or not to collapse transcripts with\n",
    "            duplicate triplets\n",
    "\n",
    "    Returns:\n",
    "        df (pyranges PyRanges): PyRanges gtf table with updated ids\n",
    "    \"\"\"\n",
    "    \n",
    "    # keep only the gene, exon, and transcript entries\n",
    "    df = pr.read_gtf(gtf).df\n",
    "    entry_types = ['gene', 'transcript', 'exon']\n",
    "    df = df.loc[df.Feature.isin(entry_types)]\n",
    "   \n",
    "    if not update_ends:\n",
    "        _, _, _, m_df = read_h5(h5)\n",
    "    else:\n",
    "        _, tss, tes, m_df = read_h5(h5)\n",
    "\n",
    "#     # groupby transcripts that are the same\n",
    "#     gb_cols = ['gene_name', 'gene_id', 'transcript_triplet',\n",
    "#                'transcript_id', 'transcript_name']\n",
    "#     temp = m_df[['transcript_id',\n",
    "#                  'original_transcript_id',\n",
    "#                  'original_transcript_name']].copy(deep=True)\n",
    "#     m_df = m_df.groupby(gb_cols).agg({'original_transcript_id': ','.join,\n",
    "#                                       'original_transcript_name': ','.join}).reset_index()\n",
    "#     m_df = m_df.merge(temp, on='transcript_id', suffixes=('','_merge'))\n",
    "#     m_df.drop(['gene_name', 'gene_id', 'transcript_triplet'],\n",
    "#               axis=1, inplace=True)\n",
    "\n",
    "#     # add new transcript ids\n",
    "#     df = df.merge(m_df, left_on=['transcript_id', 'transcript_name'],\n",
    "#                   right_on=['original_transcript_id_merge',\n",
    "#                             'original_transcript_name_merge'],\n",
    "#                  suffixes=('_x', ''))\n",
    "\n",
    "#     # drop old tids\n",
    "#     df.drop(['transcript_id_x', 'transcript_name_x',\n",
    "#              'original_transcript_name_merge'],\n",
    "#             axis=1, inplace=True)\n",
    "\n",
    "#     # remove duplicated transcripts; just keeping the first one\n",
    "#     if agg:\n",
    "#         temp = df[['transcript_id', 'original_transcript_id_merge']].drop_duplicates()\n",
    "#         dupe_old_tids = temp.loc[temp.transcript_id.duplicated(keep='first'), 'original_transcript_id_merge']\n",
    "#         df = df.loc[~df.original_transcript_id_merge.isin(dupe_old_tids)]\n",
    "\n",
    "#     # drop last column\n",
    "#     df.drop('original_transcript_id_merge', axis=1, inplace=True)\n",
    "\n",
    "#     df = pr.PyRanges(df)\n",
    "\n",
    "#     return df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0faa232-0c52-4e30-9591-96406c346324",
   "metadata": {},
   "source": [
    "## modify / collapse transcripts in TALON GTF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 499,
   "id": "4c6b0b01-d85e-4d2c-af62-08d3122061c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "h5 = '/Users/fairliereese/mortazavi_lab/data/rnawg/lr_bulk/cerberus/human_cerberus.h5'\n",
    "gtf = '/Users/fairliereese/mortazavi_lab/data/rnawg/lr_bulk/talon/human_known_nic_nnc_talon.gtf'\n",
    "agg = True\n",
    "update_ends = True\n",
    "\n",
    "# replace_gtf_ids(gtf, h5, agg, update_ends)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 534,
   "id": "7db16831-2564-4356-ba8d-799ef02cde42",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1216320\n"
     ]
    }
   ],
   "source": [
    "h5 = '/Users/fairliereese/mortazavi_lab/data/rnawg/lr_bulk/cerberus/human_cerberus.h5'\n",
    "gtf = '/Users/fairliereese/mortazavi_lab/data/rnawg/lr_bulk/talon/human_known_nic_nnc_talon.gtf'\n",
    "agg = True\n",
    "update_ends = True\n",
    "\n",
    "df = pr.read_gtf(gtf).df\n",
    "print(len(df.index))\n",
    "entry_types = ['gene', 'transcript', 'exon']\n",
    "df = df.loc[df.Feature.isin(entry_types)]\n",
    "\n",
    "if not update_ends:\n",
    "    _, _, _, m_df = read_h5(h5)\n",
    "else:\n",
    "    _, tss, tes, m_df = read_h5(h5)\n",
    "    # hack to remove duplicates for now\n",
    "    tss = pr.PyRanges(tss.df[['Chromosome', 'Start', 'End', 'Strand', 'gene_id', 'tss']].drop_duplicates())\n",
    "    tes = pr.PyRanges(tes.df[['Chromosome', 'Start', 'End', 'Strand', 'gene_id', 'tes']].drop_duplicates())\n",
    "    tss = tss.df\n",
    "    tes = tes.df\n",
    "    tss['tss_id'] = tss.gene_id+'_'+tss.tss.astype(str)\n",
    "    tes['tes_id'] = tes.gene_id+'_'+tes.tes.astype(str)\n",
    "    tss = pr.PyRanges(tss)\n",
    "    tes = pr.PyRanges(tes)\n",
    "\n",
    "\n",
    "m_df.drop(['transcript_triplet', \n",
    "           'gene_name', 'gene_id'], axis=1, inplace=True)\n",
    "\n",
    "df = map_gtf_transcripts(df, m_df)\n",
    "\n",
    "if update_ends:\n",
    "    gtf = update_gtf_ends(df, tss, tes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 536,
   "id": "05531444-a639-4e09-9ae9-3d728bea855b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # after just tss on fwd strand\n",
    "# gtf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 537,
   "id": "6f1d37d1-0e4d-48be-ab3d-e7c0503d367f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# spot check - fwd strand"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 538,
   "id": "e29c4838-41a1-47ae-8bdb-5bd0ac193e0a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>tss_id</th>\n",
       "      <th>tes_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>gene</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>10597.0</td>\n",
       "      <td>11656.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>transcript</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>ENCODEHG000058846[1,1,1]</td>\n",
       "      <td>+</td>\n",
       "      <td>10597.0</td>\n",
       "      <td>11656.0</td>\n",
       "      <td>ENCODEHG000058846_1</td>\n",
       "      <td>ENCODEHG000058846_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>ENCODEHG000058846[1,1,1]</td>\n",
       "      <td>+</td>\n",
       "      <td>10597.0</td>\n",
       "      <td>10791.0</td>\n",
       "      <td>ENCODEHG000058846_1</td>\n",
       "      <td>ENCODEHG000058846_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>ENCODEHG000058846[1,1,1]</td>\n",
       "      <td>+</td>\n",
       "      <td>10882.0</td>\n",
       "      <td>11057.0</td>\n",
       "      <td>ENCODEHG000058846_1</td>\n",
       "      <td>ENCODEHG000058846_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>ENCODEHG000058846[1,1,1]</td>\n",
       "      <td>+</td>\n",
       "      <td>11434.0</td>\n",
       "      <td>11656.0</td>\n",
       "      <td>ENCODEHG000058846_1</td>\n",
       "      <td>ENCODEHG000058846_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Feature            gene_id             transcript_id Strand    Start  \\\n",
       "29        gene  ENCODEHG000058846                       NaN      +  10597.0   \n",
       "30  transcript  ENCODEHG000058846  ENCODEHG000058846[1,1,1]      +  10597.0   \n",
       "31        exon  ENCODEHG000058846  ENCODEHG000058846[1,1,1]      +  10597.0   \n",
       "32        exon  ENCODEHG000058846  ENCODEHG000058846[1,1,1]      +  10882.0   \n",
       "33        exon  ENCODEHG000058846  ENCODEHG000058846[1,1,1]      +  11434.0   \n",
       "\n",
       "        End               tss_id               tes_id  \n",
       "29  11656.0                  NaN                  NaN  \n",
       "30  11656.0  ENCODEHG000058846_1  ENCODEHG000058846_1  \n",
       "31  10791.0  ENCODEHG000058846_1  ENCODEHG000058846_1  \n",
       "32  11057.0  ENCODEHG000058846_1  ENCODEHG000058846_1  \n",
       "33  11656.0  ENCODEHG000058846_1  ENCODEHG000058846_1  "
      ]
     },
     "execution_count": 538,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf.loc[gtf.gene_id == 'ENCODEHG000058846', ['Feature', 'gene_id', 'transcript_id', 'Strand', 'Start', 'End', 'tss_id', 'tes_id']] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 539,
   "id": "b172e3c6-fa43-4f21-9151-3e61a3d5af96",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>tss</th>\n",
       "      <th>tss_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>10597</td>\n",
       "      <td>10698</td>\n",
       "      <td>+</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>1</td>\n",
       "      <td>ENCODEHG000058846_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome  Start    End Strand            gene_id  tss               tss_id\n",
       "0      SIRV1  10597  10698      +  ENCODEHG000058846    1  ENCODEHG000058846_1"
      ]
     },
     "execution_count": 539,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tss_df = tss.df\n",
    "tss_df.loc[tss_df.tss_id == 'ENCODEHG000058846_1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 540,
   "id": "36fcf784-9f3e-4303-bb39-b8af5be43880",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>tes</th>\n",
       "      <th>tes_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>11555</td>\n",
       "      <td>11656</td>\n",
       "      <td>+</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>1</td>\n",
       "      <td>ENCODEHG000058846_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome  Start    End Strand            gene_id  tes               tes_id\n",
       "0      SIRV1  11555  11656      +  ENCODEHG000058846    1  ENCODEHG000058846_1"
      ]
     },
     "execution_count": 540,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tes_df = tes.df\n",
    "tes_df.loc[tes_df.tes_id == 'ENCODEHG000058846_1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 541,
   "id": "0f2be942-040b-439a-85af-16b33f712e80",
   "metadata": {},
   "outputs": [],
   "source": [
    "# spot check - rev strand"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 547,
   "id": "63ad4aa8-300d-472f-9d1b-2e6c92152a5f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>tss_id</th>\n",
       "      <th>tes_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gene</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>945.0</td>\n",
       "      <td>5956.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>transcript</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784[1,1,2]</td>\n",
       "      <td>-</td>\n",
       "      <td>3615.0</td>\n",
       "      <td>4814.0</td>\n",
       "      <td>ENCODEHG000058784_1</td>\n",
       "      <td>ENCODEHG000058784_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784[1,1,2]</td>\n",
       "      <td>-</td>\n",
       "      <td>4687.0</td>\n",
       "      <td>4814.0</td>\n",
       "      <td>ENCODEHG000058784_1</td>\n",
       "      <td>ENCODEHG000058784_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784[1,1,2]</td>\n",
       "      <td>-</td>\n",
       "      <td>3966.0</td>\n",
       "      <td>4479.0</td>\n",
       "      <td>ENCODEHG000058784_1</td>\n",
       "      <td>ENCODEHG000058784_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784[1,1,2]</td>\n",
       "      <td>-</td>\n",
       "      <td>3615.0</td>\n",
       "      <td>3825.0</td>\n",
       "      <td>ENCODEHG000058784_1</td>\n",
       "      <td>ENCODEHG000058784_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>transcript</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784[2,2,1]</td>\n",
       "      <td>-</td>\n",
       "      <td>945.0</td>\n",
       "      <td>5956.0</td>\n",
       "      <td>ENCODEHG000058784_2</td>\n",
       "      <td>ENCODEHG000058784_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784[2,2,1]</td>\n",
       "      <td>-</td>\n",
       "      <td>5788.0</td>\n",
       "      <td>5956.0</td>\n",
       "      <td>ENCODEHG000058784_2</td>\n",
       "      <td>ENCODEHG000058784_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784[2,2,1]</td>\n",
       "      <td>-</td>\n",
       "      <td>4687.0</td>\n",
       "      <td>4800.0</td>\n",
       "      <td>ENCODEHG000058784_2</td>\n",
       "      <td>ENCODEHG000058784_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784[2,2,1]</td>\n",
       "      <td>-</td>\n",
       "      <td>4338.0</td>\n",
       "      <td>4479.0</td>\n",
       "      <td>ENCODEHG000058784_2</td>\n",
       "      <td>ENCODEHG000058784_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784[2,2,1]</td>\n",
       "      <td>-</td>\n",
       "      <td>3966.0</td>\n",
       "      <td>4094.0</td>\n",
       "      <td>ENCODEHG000058784_2</td>\n",
       "      <td>ENCODEHG000058784_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784[2,2,1]</td>\n",
       "      <td>-</td>\n",
       "      <td>3665.0</td>\n",
       "      <td>3825.0</td>\n",
       "      <td>ENCODEHG000058784_2</td>\n",
       "      <td>ENCODEHG000058784_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784[2,2,1]</td>\n",
       "      <td>-</td>\n",
       "      <td>3105.0</td>\n",
       "      <td>3325.0</td>\n",
       "      <td>ENCODEHG000058784_2</td>\n",
       "      <td>ENCODEHG000058784_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784[2,2,1]</td>\n",
       "      <td>-</td>\n",
       "      <td>2881.0</td>\n",
       "      <td>3010.0</td>\n",
       "      <td>ENCODEHG000058784_2</td>\n",
       "      <td>ENCODEHG000058784_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784[2,2,1]</td>\n",
       "      <td>-</td>\n",
       "      <td>2674.0</td>\n",
       "      <td>2802.0</td>\n",
       "      <td>ENCODEHG000058784_2</td>\n",
       "      <td>ENCODEHG000058784_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784[2,2,1]</td>\n",
       "      <td>-</td>\n",
       "      <td>1973.0</td>\n",
       "      <td>2064.0</td>\n",
       "      <td>ENCODEHG000058784_2</td>\n",
       "      <td>ENCODEHG000058784_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784[2,2,1]</td>\n",
       "      <td>-</td>\n",
       "      <td>1741.0</td>\n",
       "      <td>1853.0</td>\n",
       "      <td>ENCODEHG000058784_2</td>\n",
       "      <td>ENCODEHG000058784_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784[2,2,1]</td>\n",
       "      <td>-</td>\n",
       "      <td>945.0</td>\n",
       "      <td>1661.0</td>\n",
       "      <td>ENCODEHG000058784_2</td>\n",
       "      <td>ENCODEHG000058784_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Feature            gene_id             transcript_id Strand   Start  \\\n",
       "0         gene  ENCODEHG000058784                       NaN      -   945.0   \n",
       "1   transcript  ENCODEHG000058784  ENCODEHG000058784[1,1,2]      -  3615.0   \n",
       "2         exon  ENCODEHG000058784  ENCODEHG000058784[1,1,2]      -  4687.0   \n",
       "3         exon  ENCODEHG000058784  ENCODEHG000058784[1,1,2]      -  3966.0   \n",
       "4         exon  ENCODEHG000058784  ENCODEHG000058784[1,1,2]      -  3615.0   \n",
       "5   transcript  ENCODEHG000058784  ENCODEHG000058784[2,2,1]      -   945.0   \n",
       "6         exon  ENCODEHG000058784  ENCODEHG000058784[2,2,1]      -  5788.0   \n",
       "7         exon  ENCODEHG000058784  ENCODEHG000058784[2,2,1]      -  4687.0   \n",
       "8         exon  ENCODEHG000058784  ENCODEHG000058784[2,2,1]      -  4338.0   \n",
       "9         exon  ENCODEHG000058784  ENCODEHG000058784[2,2,1]      -  3966.0   \n",
       "10        exon  ENCODEHG000058784  ENCODEHG000058784[2,2,1]      -  3665.0   \n",
       "11        exon  ENCODEHG000058784  ENCODEHG000058784[2,2,1]      -  3105.0   \n",
       "12        exon  ENCODEHG000058784  ENCODEHG000058784[2,2,1]      -  2881.0   \n",
       "13        exon  ENCODEHG000058784  ENCODEHG000058784[2,2,1]      -  2674.0   \n",
       "14        exon  ENCODEHG000058784  ENCODEHG000058784[2,2,1]      -  1973.0   \n",
       "15        exon  ENCODEHG000058784  ENCODEHG000058784[2,2,1]      -  1741.0   \n",
       "16        exon  ENCODEHG000058784  ENCODEHG000058784[2,2,1]      -   945.0   \n",
       "\n",
       "       End               tss_id               tes_id  \n",
       "0   5956.0                  NaN                  NaN  \n",
       "1   4814.0  ENCODEHG000058784_1  ENCODEHG000058784_2  \n",
       "2   4814.0  ENCODEHG000058784_1  ENCODEHG000058784_2  \n",
       "3   4479.0  ENCODEHG000058784_1  ENCODEHG000058784_2  \n",
       "4   3825.0  ENCODEHG000058784_1  ENCODEHG000058784_2  \n",
       "5   5956.0  ENCODEHG000058784_2  ENCODEHG000058784_1  \n",
       "6   5956.0  ENCODEHG000058784_2  ENCODEHG000058784_1  \n",
       "7   4800.0  ENCODEHG000058784_2  ENCODEHG000058784_1  \n",
       "8   4479.0  ENCODEHG000058784_2  ENCODEHG000058784_1  \n",
       "9   4094.0  ENCODEHG000058784_2  ENCODEHG000058784_1  \n",
       "10  3825.0  ENCODEHG000058784_2  ENCODEHG000058784_1  \n",
       "11  3325.0  ENCODEHG000058784_2  ENCODEHG000058784_1  \n",
       "12  3010.0  ENCODEHG000058784_2  ENCODEHG000058784_1  \n",
       "13  2802.0  ENCODEHG000058784_2  ENCODEHG000058784_1  \n",
       "14  2064.0  ENCODEHG000058784_2  ENCODEHG000058784_1  \n",
       "15  1853.0  ENCODEHG000058784_2  ENCODEHG000058784_1  \n",
       "16  1661.0  ENCODEHG000058784_2  ENCODEHG000058784_1  "
      ]
     },
     "execution_count": 547,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf.loc[gtf.gene_id == 'ENCODEHG000058784', ['Feature', 'gene_id', 'transcript_id', 'Strand', 'Start', 'End', 'tss_id', 'tes_id']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 545,
   "id": "486bdae8-a015-46a1-b976-aaa5b5a4c75c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>tss</th>\n",
       "      <th>tss_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SIRV2</td>\n",
       "      <td>4713</td>\n",
       "      <td>4814</td>\n",
       "      <td>-</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>1</td>\n",
       "      <td>ENCODEHG000058784_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome  Start   End Strand            gene_id  tss               tss_id\n",
       "3      SIRV2   4713  4814      -  ENCODEHG000058784    1  ENCODEHG000058784_1"
      ]
     },
     "execution_count": 545,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tss_df = tss.df\n",
    "tss_df.loc[tss_df.tss_id == 'ENCODEHG000058784_1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 546,
   "id": "9dc81d67-dfd6-49cf-bcd9-96a489b44a0e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>tes</th>\n",
       "      <th>tes_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SIRV2</td>\n",
       "      <td>3615</td>\n",
       "      <td>3716</td>\n",
       "      <td>-</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>2</td>\n",
       "      <td>ENCODEHG000058784_2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome  Start   End Strand            gene_id  tes               tes_id\n",
       "3      SIRV2   3615  3716      -  ENCODEHG000058784    2  ENCODEHG000058784_2"
      ]
     },
     "execution_count": 546,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tes_df = tes.df\n",
    "tes_df.loc[tes_df.tes_id == 'ENCODEHG000058784_2']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 533,
   "id": "4aeab32f-3e12-42d1-9a78-8df666cea71c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def update_gtf_ends(gtf, tss, tes):\n",
    "    \"\"\"\n",
    "    Update gene, transcript, and exon boundaries to be \n",
    "    furthest upstream or downstream entry for end used\n",
    "    \"\"\"\n",
    "    gtf = gtf.copy(deep=True)\n",
    "    gtf = sort_gtf(gtf)\n",
    "    # print(gtf.loc[gtf.gene_id == 'ENCODEHG000058844', ['Feature', 'gene_id', 'transcript_id', 'Start', 'End']])\n",
    "\n",
    "    for mode, ends in zip(['tss', 'tes'], [tss, tes]):\n",
    "        ends = ends.df\n",
    "        ends = ends[['Start', 'End', '{}_id'.format(mode)]]\n",
    "        gtf = gtf.merge(ends, how='left',\n",
    "                        on='{}_id'.format(mode),\n",
    "                        suffixes=('', '_end'))\n",
    "\n",
    "        fwd, rev = get_stranded_gtf_dfs(gtf)\n",
    "        df = pd.DataFrame()\n",
    "        for strand, temp in zip(['+', '-'], [fwd, rev]):\n",
    "    \n",
    "            # fix exon, transcript, and gene boundaries\n",
    "            temp = update_transcript_ends(temp, mode, strand)\n",
    "            temp = update_gene_ends(temp, mode, strand)\n",
    "            df = pd.concat([df, temp])\n",
    "        \n",
    "        gtf = df.copy(deep=True)\n",
    "        gtf.drop(['Start_end', 'End_end'], axis=1, inplace=True)\n",
    "        \n",
    "    return gtf\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 530,
   "id": "c0b46624-a1a1-43ca-b1dd-0262ef3c1180",
   "metadata": {},
   "outputs": [],
   "source": [
    "# print(rev[['Feature', 'gene_id', 'transcript_id', 'Strand', 'Start', 'End', 'Start_end', 'End_end']].head(20))\n",
    "\n",
    "def get_update_ends_settings(strand, mode):\n",
    "    if mode == 'tss': \n",
    "        if strand == '+':\n",
    "            old_end = 'Start'\n",
    "            new_end = 'Start_end'\n",
    "            gene_func = 'min'\n",
    "        elif strand == '-':\n",
    "            old_end = 'End'\n",
    "            new_end = 'End_end'\n",
    "            gene_func = 'max'\n",
    "    elif mode == 'tes':\n",
    "        if strand == '+':\n",
    "            old_end = 'End'\n",
    "            new_end = 'End_end'\n",
    "            gene_func = 'max'\n",
    "        elif strand == '-':\n",
    "            old_end = 'Start'\n",
    "            new_end  = 'Start_end'\n",
    "            gene_func = 'min'\n",
    "            \n",
    "    return old_end, new_end, gene_func\n",
    "\n",
    "def update_transcript_ends(df, mode, strand):\n",
    "    \"\"\"\n",
    "    Update the ends of transcripts and the first / last exon\n",
    "    in a GTF. GTF must be sorted!\n",
    "    \"\"\"\n",
    "    old_col, new_col, gene_func = get_update_ends_settings(strand, mode)\n",
    "    \n",
    "    temp = df[['Feature', 'gene_id', 'transcript_id', 'Strand', 'Start', 'End', 'Start_end', 'End_end']].copy(deep=True)\n",
    "    temp = temp.loc[temp.Feature != 'gene']\n",
    "    if mode == 'tss':\n",
    "        inds = temp.groupby('transcript_id').head(2).index.tolist()\n",
    "    elif mode == 'tes':\n",
    "        inds = temp.groupby('transcript_id').head(1).index.tolist()\n",
    "        inds += temp.groupby('transcript_id').tail(1).index.tolist()\n",
    "    \n",
    "    # print('transcript')\n",
    "    # print(df.columns)\n",
    "    # print(old_col)\n",
    "    # print(new_col)\n",
    "    # print()\n",
    "    df.loc[inds, old_col] = df.loc[inds, new_col]\n",
    "    \n",
    "    return df\n",
    "\n",
    "def update_gene_ends(df, mode, strand):\n",
    "    \"\"\"\n",
    "    Update the ends of genes in a GTF. \n",
    "    \"\"\"\n",
    "    old_col, new_col, gene_func = get_update_ends_settings(strand, mode)\n",
    "    temp = df[['Feature', 'gene_id', old_col]].copy(deep=True)\n",
    "    temp = temp.loc[temp.Feature == 'transcript']\n",
    "    # print(temp.loc[temp.gene_id == 'ENCODEHG000058784'])\n",
    "    temp = temp.groupby(['gene_id', 'Feature'], observed=True).agg(gene_func).reset_index()\n",
    "    temp.drop('Feature', axis=1, inplace=True)\n",
    "    # print(temp.loc[temp.gene_id == 'ENCODEHG000058784'])\n",
    "    df = df.merge(temp, on='gene_id', suffixes=('', '_gene'))\n",
    "    inds = df.loc[df.Feature == 'gene'].index.tolist()\n",
    "    # print('gene')\n",
    "    # print(df.columns)\n",
    "    # print(old_col)\n",
    "    # print('{}_gene'.format(old_col))\n",
    "    # print()\n",
    "    \n",
    "    df.loc[inds, old_col] = df.loc[inds, '{}_gene'.format(old_col)]\n",
    "    df.drop('{}_gene'.format(old_col), axis=1, inplace=True)\n",
    "    \n",
    "    return df\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 490,
   "id": "758adc58-36fc-4101-8a9c-b9d4de6bf028",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_stranded_gtf_dfs(df):\n",
    "    \"\"\"\n",
    "    Split a GTF df into fwd and rev strands\n",
    "    \"\"\"\n",
    "    rev = df.loc[df.Strand == '-'].copy(deep=True)\n",
    "    fwd = df.loc[df.Strand == '+'].copy(deep=True)\n",
    "    \n",
    "    return fwd, rev\n",
    "\n",
    "def sort_gtf(df):\n",
    "    \"\"\"\n",
    "    Sort a GTF \n",
    "    \"\"\"\n",
    "    df['feature_rank'] = df.Feature.map({'gene':0, 'transcript':1, 'exon':2})\n",
    "    df.feature_rank = df.feature_rank.astype(int)    \n",
    "     \n",
    "    fwd, rev = get_stranded_gtf_dfs(df)\n",
    "\n",
    "    df = pd.DataFrame()\n",
    "    for temp in [fwd, rev]:\n",
    "        if len(temp.index) > 0:\n",
    "            strand = temp.Strand.values.tolist()[0]\n",
    "            if strand == '+':\n",
    "                ascending = True\n",
    "            elif strand == '-':\n",
    "                ascending = False\n",
    "            temp.sort_values(by=['gene_id', 'transcript_id', 'feature_rank', 'Start'], \n",
    "                             ascending=[True, True, True, ascending],\n",
    "                             na_position='first', inplace=True)\n",
    "            \n",
    "            df = pd.concat([df, temp])\n",
    "    df.drop('feature_rank', axis=1, inplace=True)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "084cc90d-6c95-4733-bdf7-cfac1d96c06c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a0814ee-7254-43fd-aa29-bb7203423cd3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 480,
   "id": "0b998d29-dc5a-4b3c-8013-e7447e9843ca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'tes'"
      ]
     },
     "execution_count": 480,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mode"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 471,
   "id": "1e7a3ba7-08e5-43cc-9adb-980b6a0a089f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Chromosome', 'Source', 'Feature', 'Start', 'End', 'Score', 'Strand',\n",
       "       'Frame', 'gene_id', 'gene_name', 'gene_status', 'gene_type',\n",
       "       'talon_gene', 'havana_gene', 'level', 'transcript_status',\n",
       "       'talon_transcript', 'NIC_transcript', 'exon_number', 'exon_id',\n",
       "       'talon_exon', 'exon_status', 'ont', 'source', 'tag', 'transcript_type',\n",
       "       'transcript_support_level', 'havana_transcript', 'NNC_transcript',\n",
       "       'protein_id', 'ccdsid', 'intergenic_novel', 'antisense_gene',\n",
       "       'gene_antisense_to_IDs', 'original_transcript_id', 'ic', 'ic_id',\n",
       "       'tss_id', 'tss', 'tes_id', 'tes', 'original_transcript_name',\n",
       "       'transcript_id', 'transcript_name', 'Start_end', 'End_end'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 471,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 474,
   "id": "60db94ab-f2f6-4f24-9840-7901aece7dd5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Start_end</th>\n",
       "      <th>End_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gene</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>945.0</td>\n",
       "      <td>4764</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>transcript</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784[1,1,2]</td>\n",
       "      <td>-</td>\n",
       "      <td>3615.0</td>\n",
       "      <td>4764</td>\n",
       "      <td>3615.0</td>\n",
       "      <td>3716.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784[1,1,2]</td>\n",
       "      <td>-</td>\n",
       "      <td>4687.0</td>\n",
       "      <td>4764</td>\n",
       "      <td>3615.0</td>\n",
       "      <td>3716.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784[1,1,2]</td>\n",
       "      <td>-</td>\n",
       "      <td>3966.0</td>\n",
       "      <td>4479</td>\n",
       "      <td>3615.0</td>\n",
       "      <td>3716.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784[1,1,2]</td>\n",
       "      <td>-</td>\n",
       "      <td>3615.0</td>\n",
       "      <td>3825</td>\n",
       "      <td>3615.0</td>\n",
       "      <td>3716.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>transcript</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784[2,2,1]</td>\n",
       "      <td>-</td>\n",
       "      <td>945.0</td>\n",
       "      <td>5906</td>\n",
       "      <td>945.0</td>\n",
       "      <td>1046.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784[2,2,1]</td>\n",
       "      <td>-</td>\n",
       "      <td>5788.0</td>\n",
       "      <td>5906</td>\n",
       "      <td>945.0</td>\n",
       "      <td>1046.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784[2,2,1]</td>\n",
       "      <td>-</td>\n",
       "      <td>4687.0</td>\n",
       "      <td>4800</td>\n",
       "      <td>945.0</td>\n",
       "      <td>1046.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784[2,2,1]</td>\n",
       "      <td>-</td>\n",
       "      <td>4338.0</td>\n",
       "      <td>4479</td>\n",
       "      <td>945.0</td>\n",
       "      <td>1046.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784[2,2,1]</td>\n",
       "      <td>-</td>\n",
       "      <td>3966.0</td>\n",
       "      <td>4094</td>\n",
       "      <td>945.0</td>\n",
       "      <td>1046.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784[2,2,1]</td>\n",
       "      <td>-</td>\n",
       "      <td>3665.0</td>\n",
       "      <td>3825</td>\n",
       "      <td>945.0</td>\n",
       "      <td>1046.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784[2,2,1]</td>\n",
       "      <td>-</td>\n",
       "      <td>3105.0</td>\n",
       "      <td>3325</td>\n",
       "      <td>945.0</td>\n",
       "      <td>1046.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784[2,2,1]</td>\n",
       "      <td>-</td>\n",
       "      <td>2881.0</td>\n",
       "      <td>3010</td>\n",
       "      <td>945.0</td>\n",
       "      <td>1046.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784[2,2,1]</td>\n",
       "      <td>-</td>\n",
       "      <td>2674.0</td>\n",
       "      <td>2802</td>\n",
       "      <td>945.0</td>\n",
       "      <td>1046.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784[2,2,1]</td>\n",
       "      <td>-</td>\n",
       "      <td>1973.0</td>\n",
       "      <td>2064</td>\n",
       "      <td>945.0</td>\n",
       "      <td>1046.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784[2,2,1]</td>\n",
       "      <td>-</td>\n",
       "      <td>1741.0</td>\n",
       "      <td>1853</td>\n",
       "      <td>945.0</td>\n",
       "      <td>1046.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784[2,2,1]</td>\n",
       "      <td>-</td>\n",
       "      <td>945.0</td>\n",
       "      <td>1661</td>\n",
       "      <td>945.0</td>\n",
       "      <td>1046.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Feature            gene_id             transcript_id Strand   Start  \\\n",
       "0         gene  ENCODEHG000058784                       NaN      -   945.0   \n",
       "1   transcript  ENCODEHG000058784  ENCODEHG000058784[1,1,2]      -  3615.0   \n",
       "2         exon  ENCODEHG000058784  ENCODEHG000058784[1,1,2]      -  4687.0   \n",
       "3         exon  ENCODEHG000058784  ENCODEHG000058784[1,1,2]      -  3966.0   \n",
       "4         exon  ENCODEHG000058784  ENCODEHG000058784[1,1,2]      -  3615.0   \n",
       "5   transcript  ENCODEHG000058784  ENCODEHG000058784[2,2,1]      -   945.0   \n",
       "6         exon  ENCODEHG000058784  ENCODEHG000058784[2,2,1]      -  5788.0   \n",
       "7         exon  ENCODEHG000058784  ENCODEHG000058784[2,2,1]      -  4687.0   \n",
       "8         exon  ENCODEHG000058784  ENCODEHG000058784[2,2,1]      -  4338.0   \n",
       "9         exon  ENCODEHG000058784  ENCODEHG000058784[2,2,1]      -  3966.0   \n",
       "10        exon  ENCODEHG000058784  ENCODEHG000058784[2,2,1]      -  3665.0   \n",
       "11        exon  ENCODEHG000058784  ENCODEHG000058784[2,2,1]      -  3105.0   \n",
       "12        exon  ENCODEHG000058784  ENCODEHG000058784[2,2,1]      -  2881.0   \n",
       "13        exon  ENCODEHG000058784  ENCODEHG000058784[2,2,1]      -  2674.0   \n",
       "14        exon  ENCODEHG000058784  ENCODEHG000058784[2,2,1]      -  1973.0   \n",
       "15        exon  ENCODEHG000058784  ENCODEHG000058784[2,2,1]      -  1741.0   \n",
       "16        exon  ENCODEHG000058784  ENCODEHG000058784[2,2,1]      -   945.0   \n",
       "\n",
       "     End  Start_end  End_end  \n",
       "0   4764        NaN      NaN  \n",
       "1   4764     3615.0   3716.0  \n",
       "2   4764     3615.0   3716.0  \n",
       "3   4479     3615.0   3716.0  \n",
       "4   3825     3615.0   3716.0  \n",
       "5   5906      945.0   1046.0  \n",
       "6   5906      945.0   1046.0  \n",
       "7   4800      945.0   1046.0  \n",
       "8   4479      945.0   1046.0  \n",
       "9   4094      945.0   1046.0  \n",
       "10  3825      945.0   1046.0  \n",
       "11  3325      945.0   1046.0  \n",
       "12  3010      945.0   1046.0  \n",
       "13  2802      945.0   1046.0  \n",
       "14  2064      945.0   1046.0  \n",
       "15  1853      945.0   1046.0  \n",
       "16  1661      945.0   1046.0  "
      ]
     },
     "execution_count": 474,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df.gene_id == 'ENCODEHG000058784', ['Feature', 'gene_id', 'transcript_id', 'Strand', 'Start', 'End', 'Start_end', 'End_end']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 429,
   "id": "a8098d32-1d03-498e-88cf-6f023bd78363",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.loc[df.gene_id == 'ENCODEHG000058844', ['Feature', 'gene_id', 'transcript_id', 'Strand', 'Start', 'End', 'Start_end', 'End_end']]\n",
    "# df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 420,
   "id": "2195dbcb-04e3-44e1-b434-5383b1370264",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENCODEHG000058844</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1930.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             gene_id     Feature   Start\n",
       "0  ENCODEHG000058844  transcript  1930.0"
      ]
     },
     "execution_count": 420,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp.loc[temp.gene_id == 'ENCODEHG000058844']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 395,
   "id": "773ec76b-5afd-426c-8dd6-3331a81dc4cc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Start_end</th>\n",
       "      <th>End_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>gene</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>995</td>\n",
       "      <td>4764.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>transcript</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784[1,1,2]</td>\n",
       "      <td>-</td>\n",
       "      <td>3665</td>\n",
       "      <td>4814.0</td>\n",
       "      <td>4713.0</td>\n",
       "      <td>4814.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784[1,1,2]</td>\n",
       "      <td>-</td>\n",
       "      <td>4687</td>\n",
       "      <td>4814.0</td>\n",
       "      <td>4713.0</td>\n",
       "      <td>4814.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784[1,1,2]</td>\n",
       "      <td>-</td>\n",
       "      <td>3966</td>\n",
       "      <td>4479.0</td>\n",
       "      <td>4713.0</td>\n",
       "      <td>4814.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784[1,1,2]</td>\n",
       "      <td>-</td>\n",
       "      <td>3665</td>\n",
       "      <td>3825.0</td>\n",
       "      <td>4713.0</td>\n",
       "      <td>4814.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>transcript</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784[2,2,1]</td>\n",
       "      <td>-</td>\n",
       "      <td>995</td>\n",
       "      <td>5956.0</td>\n",
       "      <td>5855.0</td>\n",
       "      <td>5956.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784[2,2,1]</td>\n",
       "      <td>-</td>\n",
       "      <td>5788</td>\n",
       "      <td>5956.0</td>\n",
       "      <td>5855.0</td>\n",
       "      <td>5956.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784[2,2,1]</td>\n",
       "      <td>-</td>\n",
       "      <td>4687</td>\n",
       "      <td>4800.0</td>\n",
       "      <td>5855.0</td>\n",
       "      <td>5956.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784[2,2,1]</td>\n",
       "      <td>-</td>\n",
       "      <td>4338</td>\n",
       "      <td>4479.0</td>\n",
       "      <td>5855.0</td>\n",
       "      <td>5956.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784[2,2,1]</td>\n",
       "      <td>-</td>\n",
       "      <td>3966</td>\n",
       "      <td>4094.0</td>\n",
       "      <td>5855.0</td>\n",
       "      <td>5956.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784[2,2,1]</td>\n",
       "      <td>-</td>\n",
       "      <td>3665</td>\n",
       "      <td>3825.0</td>\n",
       "      <td>5855.0</td>\n",
       "      <td>5956.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784[2,2,1]</td>\n",
       "      <td>-</td>\n",
       "      <td>3105</td>\n",
       "      <td>3325.0</td>\n",
       "      <td>5855.0</td>\n",
       "      <td>5956.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784[2,2,1]</td>\n",
       "      <td>-</td>\n",
       "      <td>2881</td>\n",
       "      <td>3010.0</td>\n",
       "      <td>5855.0</td>\n",
       "      <td>5956.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784[2,2,1]</td>\n",
       "      <td>-</td>\n",
       "      <td>2674</td>\n",
       "      <td>2802.0</td>\n",
       "      <td>5855.0</td>\n",
       "      <td>5956.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784[2,2,1]</td>\n",
       "      <td>-</td>\n",
       "      <td>1973</td>\n",
       "      <td>2064.0</td>\n",
       "      <td>5855.0</td>\n",
       "      <td>5956.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784[2,2,1]</td>\n",
       "      <td>-</td>\n",
       "      <td>1741</td>\n",
       "      <td>1853.0</td>\n",
       "      <td>5855.0</td>\n",
       "      <td>5956.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784[2,2,1]</td>\n",
       "      <td>-</td>\n",
       "      <td>995</td>\n",
       "      <td>1661.0</td>\n",
       "      <td>5855.0</td>\n",
       "      <td>5956.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>gene</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>999</td>\n",
       "      <td>10366.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>transcript</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>ENCODEHG000058837[1,1,1]</td>\n",
       "      <td>-</td>\n",
       "      <td>999</td>\n",
       "      <td>10416.0</td>\n",
       "      <td>10315.0</td>\n",
       "      <td>10416.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>ENCODEHG000058837[1,1,1]</td>\n",
       "      <td>-</td>\n",
       "      <td>10282</td>\n",
       "      <td>10416.0</td>\n",
       "      <td>10315.0</td>\n",
       "      <td>10416.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Feature            gene_id             transcript_id Strand  Start  \\\n",
       "29        gene  ENCODEHG000058784                       NaN      -    995   \n",
       "42  transcript  ENCODEHG000058784  ENCODEHG000058784[1,1,2]      -   3665   \n",
       "43        exon  ENCODEHG000058784  ENCODEHG000058784[1,1,2]      -   4687   \n",
       "44        exon  ENCODEHG000058784  ENCODEHG000058784[1,1,2]      -   3966   \n",
       "45        exon  ENCODEHG000058784  ENCODEHG000058784[1,1,2]      -   3665   \n",
       "30  transcript  ENCODEHG000058784  ENCODEHG000058784[2,2,1]      -    995   \n",
       "31        exon  ENCODEHG000058784  ENCODEHG000058784[2,2,1]      -   5788   \n",
       "32        exon  ENCODEHG000058784  ENCODEHG000058784[2,2,1]      -   4687   \n",
       "33        exon  ENCODEHG000058784  ENCODEHG000058784[2,2,1]      -   4338   \n",
       "34        exon  ENCODEHG000058784  ENCODEHG000058784[2,2,1]      -   3966   \n",
       "35        exon  ENCODEHG000058784  ENCODEHG000058784[2,2,1]      -   3665   \n",
       "36        exon  ENCODEHG000058784  ENCODEHG000058784[2,2,1]      -   3105   \n",
       "37        exon  ENCODEHG000058784  ENCODEHG000058784[2,2,1]      -   2881   \n",
       "38        exon  ENCODEHG000058784  ENCODEHG000058784[2,2,1]      -   2674   \n",
       "39        exon  ENCODEHG000058784  ENCODEHG000058784[2,2,1]      -   1973   \n",
       "40        exon  ENCODEHG000058784  ENCODEHG000058784[2,2,1]      -   1741   \n",
       "41        exon  ENCODEHG000058784  ENCODEHG000058784[2,2,1]      -    995   \n",
       "5         gene  ENCODEHG000058837                       NaN      -    999   \n",
       "24  transcript  ENCODEHG000058837  ENCODEHG000058837[1,1,1]      -    999   \n",
       "25        exon  ENCODEHG000058837  ENCODEHG000058837[1,1,1]      -  10282   \n",
       "\n",
       "        End  Start_end  End_end  \n",
       "29   4764.0        NaN      NaN  \n",
       "42   4814.0     4713.0   4814.0  \n",
       "43   4814.0     4713.0   4814.0  \n",
       "44   4479.0     4713.0   4814.0  \n",
       "45   3825.0     4713.0   4814.0  \n",
       "30   5956.0     5855.0   5956.0  \n",
       "31   5956.0     5855.0   5956.0  \n",
       "32   4800.0     5855.0   5956.0  \n",
       "33   4479.0     5855.0   5956.0  \n",
       "34   4094.0     5855.0   5956.0  \n",
       "35   3825.0     5855.0   5956.0  \n",
       "36   3325.0     5855.0   5956.0  \n",
       "37   3010.0     5855.0   5956.0  \n",
       "38   2802.0     5855.0   5956.0  \n",
       "39   2064.0     5855.0   5956.0  \n",
       "40   1853.0     5855.0   5956.0  \n",
       "41   1661.0     5855.0   5956.0  \n",
       "5   10366.0        NaN      NaN  \n",
       "24  10416.0    10315.0  10416.0  \n",
       "25  10416.0    10315.0  10416.0  "
      ]
     },
     "execution_count": 395,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rev[['Feature', 'gene_id', 'transcript_id', 'Strand', 'Start', 'End', 'Start_end', 'End_end']].head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 344,
   "id": "ac858343-0c67-43bd-9a04-f670899e8971",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Start_end</th>\n",
       "      <th>End_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>gene</td>\n",
       "      <td>ENCODEHG000058844</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>1980.0</td>\n",
       "      <td>6780</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>transcript</td>\n",
       "      <td>ENCODEHG000058844</td>\n",
       "      <td>ENCODEHG000058844[1,1,3]</td>\n",
       "      <td>+</td>\n",
       "      <td>1930.0</td>\n",
       "      <td>8837</td>\n",
       "      <td>1930.0</td>\n",
       "      <td>2031.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058844</td>\n",
       "      <td>ENCODEHG000058844[1,1,3]</td>\n",
       "      <td>+</td>\n",
       "      <td>1930.0</td>\n",
       "      <td>2005</td>\n",
       "      <td>1930.0</td>\n",
       "      <td>2031.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058844</td>\n",
       "      <td>ENCODEHG000058844[1,1,3]</td>\n",
       "      <td>+</td>\n",
       "      <td>4003.0</td>\n",
       "      <td>4080</td>\n",
       "      <td>1930.0</td>\n",
       "      <td>2031.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058844</td>\n",
       "      <td>ENCODEHG000058844[1,1,3]</td>\n",
       "      <td>+</td>\n",
       "      <td>4568.0</td>\n",
       "      <td>4779</td>\n",
       "      <td>1930.0</td>\n",
       "      <td>2031.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058844</td>\n",
       "      <td>ENCODEHG000058844[1,1,3]</td>\n",
       "      <td>+</td>\n",
       "      <td>6057.0</td>\n",
       "      <td>6333</td>\n",
       "      <td>1930.0</td>\n",
       "      <td>2031.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058844</td>\n",
       "      <td>ENCODEHG000058844[1,1,3]</td>\n",
       "      <td>+</td>\n",
       "      <td>7270.0</td>\n",
       "      <td>7366</td>\n",
       "      <td>1930.0</td>\n",
       "      <td>2031.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058844</td>\n",
       "      <td>ENCODEHG000058844[1,1,3]</td>\n",
       "      <td>+</td>\n",
       "      <td>7872.0</td>\n",
       "      <td>7988</td>\n",
       "      <td>1930.0</td>\n",
       "      <td>2031.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058844</td>\n",
       "      <td>ENCODEHG000058844[1,1,3]</td>\n",
       "      <td>+</td>\n",
       "      <td>8124.0</td>\n",
       "      <td>8207</td>\n",
       "      <td>1930.0</td>\n",
       "      <td>2031.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058844</td>\n",
       "      <td>ENCODEHG000058844[1,1,3]</td>\n",
       "      <td>+</td>\n",
       "      <td>8755.0</td>\n",
       "      <td>8837</td>\n",
       "      <td>1930.0</td>\n",
       "      <td>2031.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>transcript</td>\n",
       "      <td>ENCODEHG000058844</td>\n",
       "      <td>ENCODEHG000058844[1,2,3]</td>\n",
       "      <td>+</td>\n",
       "      <td>1930.0</td>\n",
       "      <td>8837</td>\n",
       "      <td>1930.0</td>\n",
       "      <td>2031.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058844</td>\n",
       "      <td>ENCODEHG000058844[1,2,3]</td>\n",
       "      <td>+</td>\n",
       "      <td>1930.0</td>\n",
       "      <td>2005</td>\n",
       "      <td>1930.0</td>\n",
       "      <td>2031.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058844</td>\n",
       "      <td>ENCODEHG000058844[1,2,3]</td>\n",
       "      <td>+</td>\n",
       "      <td>4003.0</td>\n",
       "      <td>4080</td>\n",
       "      <td>1930.0</td>\n",
       "      <td>2031.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058844</td>\n",
       "      <td>ENCODEHG000058844[1,2,3]</td>\n",
       "      <td>+</td>\n",
       "      <td>4574.0</td>\n",
       "      <td>4774</td>\n",
       "      <td>1930.0</td>\n",
       "      <td>2031.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058844</td>\n",
       "      <td>ENCODEHG000058844[1,2,3]</td>\n",
       "      <td>+</td>\n",
       "      <td>6057.0</td>\n",
       "      <td>6333</td>\n",
       "      <td>1930.0</td>\n",
       "      <td>2031.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058844</td>\n",
       "      <td>ENCODEHG000058844[1,2,3]</td>\n",
       "      <td>+</td>\n",
       "      <td>8755.0</td>\n",
       "      <td>8837</td>\n",
       "      <td>1930.0</td>\n",
       "      <td>2031.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>transcript</td>\n",
       "      <td>ENCODEHG000058844</td>\n",
       "      <td>ENCODEHG000058844[1,3,3]</td>\n",
       "      <td>+</td>\n",
       "      <td>1930.0</td>\n",
       "      <td>8937</td>\n",
       "      <td>1930.0</td>\n",
       "      <td>2031.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058844</td>\n",
       "      <td>ENCODEHG000058844[1,3,3]</td>\n",
       "      <td>+</td>\n",
       "      <td>1930.0</td>\n",
       "      <td>2005</td>\n",
       "      <td>1930.0</td>\n",
       "      <td>2031.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058844</td>\n",
       "      <td>ENCODEHG000058844[1,3,3]</td>\n",
       "      <td>+</td>\n",
       "      <td>4568.0</td>\n",
       "      <td>4779</td>\n",
       "      <td>1930.0</td>\n",
       "      <td>2031.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058844</td>\n",
       "      <td>ENCODEHG000058844[1,3,3]</td>\n",
       "      <td>+</td>\n",
       "      <td>6057.0</td>\n",
       "      <td>7988</td>\n",
       "      <td>1930.0</td>\n",
       "      <td>2031.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Feature            gene_id             transcript_id Strand   Start  \\\n",
       "46        gene  ENCODEHG000058844                       NaN      +  1980.0   \n",
       "56  transcript  ENCODEHG000058844  ENCODEHG000058844[1,1,3]      +  1930.0   \n",
       "57        exon  ENCODEHG000058844  ENCODEHG000058844[1,1,3]      +  1930.0   \n",
       "58        exon  ENCODEHG000058844  ENCODEHG000058844[1,1,3]      +  4003.0   \n",
       "59        exon  ENCODEHG000058844  ENCODEHG000058844[1,1,3]      +  4568.0   \n",
       "60        exon  ENCODEHG000058844  ENCODEHG000058844[1,1,3]      +  6057.0   \n",
       "61        exon  ENCODEHG000058844  ENCODEHG000058844[1,1,3]      +  7270.0   \n",
       "62        exon  ENCODEHG000058844  ENCODEHG000058844[1,1,3]      +  7872.0   \n",
       "63        exon  ENCODEHG000058844  ENCODEHG000058844[1,1,3]      +  8124.0   \n",
       "64        exon  ENCODEHG000058844  ENCODEHG000058844[1,1,3]      +  8755.0   \n",
       "65  transcript  ENCODEHG000058844  ENCODEHG000058844[1,2,3]      +  1930.0   \n",
       "66        exon  ENCODEHG000058844  ENCODEHG000058844[1,2,3]      +  1930.0   \n",
       "67        exon  ENCODEHG000058844  ENCODEHG000058844[1,2,3]      +  4003.0   \n",
       "68        exon  ENCODEHG000058844  ENCODEHG000058844[1,2,3]      +  4574.0   \n",
       "69        exon  ENCODEHG000058844  ENCODEHG000058844[1,2,3]      +  6057.0   \n",
       "70        exon  ENCODEHG000058844  ENCODEHG000058844[1,2,3]      +  8755.0   \n",
       "47  transcript  ENCODEHG000058844  ENCODEHG000058844[1,3,3]      +  1930.0   \n",
       "48        exon  ENCODEHG000058844  ENCODEHG000058844[1,3,3]      +  1930.0   \n",
       "49        exon  ENCODEHG000058844  ENCODEHG000058844[1,3,3]      +  4568.0   \n",
       "50        exon  ENCODEHG000058844  ENCODEHG000058844[1,3,3]      +  6057.0   \n",
       "\n",
       "     End  Start_end  End_end  \n",
       "46  6780        NaN      NaN  \n",
       "56  8837     1930.0   2031.0  \n",
       "57  2005     1930.0   2031.0  \n",
       "58  4080     1930.0   2031.0  \n",
       "59  4779     1930.0   2031.0  \n",
       "60  6333     1930.0   2031.0  \n",
       "61  7366     1930.0   2031.0  \n",
       "62  7988     1930.0   2031.0  \n",
       "63  8207     1930.0   2031.0  \n",
       "64  8837     1930.0   2031.0  \n",
       "65  8837     1930.0   2031.0  \n",
       "66  2005     1930.0   2031.0  \n",
       "67  4080     1930.0   2031.0  \n",
       "68  4774     1930.0   2031.0  \n",
       "69  6333     1930.0   2031.0  \n",
       "70  8837     1930.0   2031.0  \n",
       "47  8937     1930.0   2031.0  \n",
       "48  2005     1930.0   2031.0  \n",
       "49  4779     1930.0   2031.0  \n",
       "50  7988     1930.0   2031.0  "
      ]
     },
     "execution_count": 344,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fwd[['Feature', 'gene_id', 'transcript_id', 'Strand', 'Start', 'End', 'Start_end', 'End_end']].head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 311,
   "id": "f88799bb-4bb0-4f6e-802b-09d6dbbf20b7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Start_end</th>\n",
       "      <th>End_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>transcript</td>\n",
       "      <td>ENCODEHG000058844</td>\n",
       "      <td>ENCODEHG000058844[1,1,3]</td>\n",
       "      <td>+</td>\n",
       "      <td>1980</td>\n",
       "      <td>8837</td>\n",
       "      <td>1930.0</td>\n",
       "      <td>2031.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058844</td>\n",
       "      <td>ENCODEHG000058844[1,1,3]</td>\n",
       "      <td>+</td>\n",
       "      <td>1980</td>\n",
       "      <td>2005</td>\n",
       "      <td>1930.0</td>\n",
       "      <td>2031.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058844</td>\n",
       "      <td>ENCODEHG000058844[1,1,3]</td>\n",
       "      <td>+</td>\n",
       "      <td>4003</td>\n",
       "      <td>4080</td>\n",
       "      <td>1930.0</td>\n",
       "      <td>2031.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058844</td>\n",
       "      <td>ENCODEHG000058844[1,1,3]</td>\n",
       "      <td>+</td>\n",
       "      <td>4568</td>\n",
       "      <td>4779</td>\n",
       "      <td>1930.0</td>\n",
       "      <td>2031.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058844</td>\n",
       "      <td>ENCODEHG000058844[1,1,3]</td>\n",
       "      <td>+</td>\n",
       "      <td>6057</td>\n",
       "      <td>6333</td>\n",
       "      <td>1930.0</td>\n",
       "      <td>2031.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Feature            gene_id             transcript_id Strand  Start  \\\n",
       "56  transcript  ENCODEHG000058844  ENCODEHG000058844[1,1,3]      +   1980   \n",
       "57        exon  ENCODEHG000058844  ENCODEHG000058844[1,1,3]      +   1980   \n",
       "58        exon  ENCODEHG000058844  ENCODEHG000058844[1,1,3]      +   4003   \n",
       "59        exon  ENCODEHG000058844  ENCODEHG000058844[1,1,3]      +   4568   \n",
       "60        exon  ENCODEHG000058844  ENCODEHG000058844[1,1,3]      +   6057   \n",
       "\n",
       "     End  Start_end  End_end  \n",
       "56  8837     1930.0   2031.0  \n",
       "57  2005     1930.0   2031.0  \n",
       "58  4080     1930.0   2031.0  \n",
       "59  4779     1930.0   2031.0  \n",
       "60  6333     1930.0   2031.0  "
      ]
     },
     "execution_count": 311,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 293,
   "id": "cc780b78-2466-4449-939a-c53b735213de",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>tss</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Chromosome, Start, End, Strand, gene_id, tss]\n",
       "Index: []"
      ]
     },
     "execution_count": 293,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp = tss.df\n",
    "# temp.loc[(temp.gene_id == 'ENSG00000285976')&(beep.tss == 2)]\n",
    "temp.loc[(temp.tss == 'ENSG00000285976')&(temp.tss==2)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 294,
   "id": "b0bc242b-1836-438b-bf21-cb6eec5e003b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "tss    int64\n",
       "dtype: object"
      ]
     },
     "execution_count": 294,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp.tss.to_frame().dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 287,
   "id": "37de5929-7770-4e36-a15e-eaf80e1a0b9d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>original_transcript_id</th>\n",
       "      <th>ic</th>\n",
       "      <th>ic_id</th>\n",
       "      <th>tss_id</th>\n",
       "      <th>tss</th>\n",
       "      <th>tes_id</th>\n",
       "      <th>tes</th>\n",
       "      <th>original_transcript_name</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>106791</th>\n",
       "      <td>ENCODEHT000789200</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSG00000285976_3</td>\n",
       "      <td>ENSG00000285976_2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000285976_3</td>\n",
       "      <td>3</td>\n",
       "      <td>ENCODEHT000789200</td>\n",
       "      <td>ENSG00000285976[2,3,3]</td>\n",
       "      <td>AL135905.2[2,3,3]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       original_transcript_id  ic              ic_id             tss_id  tss  \\\n",
       "106791      ENCODEHT000789200   3  ENSG00000285976_3  ENSG00000285976_2    2   \n",
       "\n",
       "                   tes_id  tes original_transcript_name  \\\n",
       "106791  ENSG00000285976_3    3        ENCODEHT000789200   \n",
       "\n",
       "                 transcript_id    transcript_name  \n",
       "106791  ENSG00000285976[2,3,3]  AL135905.2[2,3,3]  "
      ]
     },
     "execution_count": 287,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "m_df.loc[m_df.transcript_id == 'ENSG00000285976[2,3,3]']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 276,
   "id": "7fb3a9b2-64f1-4f27-921b-b0d6813b7b52",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Name</th>\n",
       "      <th>source</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>tss</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chr1</td>\n",
       "      <td>30216</td>\n",
       "      <td>30317</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000243485_2</td>\n",
       "      <td>v39,v29,talon</td>\n",
       "      <td>ENSG00000243485</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>chr1</td>\n",
       "      <td>30216</td>\n",
       "      <td>30317</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000243485_2</td>\n",
       "      <td>v39,talon</td>\n",
       "      <td>ENSG00000243485</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>chr1</td>\n",
       "      <td>30315</td>\n",
       "      <td>30416</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000284332_1</td>\n",
       "      <td>v39,v29</td>\n",
       "      <td>ENSG00000284332</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr1</td>\n",
       "      <td>30315</td>\n",
       "      <td>30416</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000284332_1</td>\n",
       "      <td>v39</td>\n",
       "      <td>ENSG00000284332</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>492</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10430333</td>\n",
       "      <td>10430485</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000251503_2</td>\n",
       "      <td>v39,v29,talon</td>\n",
       "      <td>ENSG00000251503</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156310</th>\n",
       "      <td>chrY</td>\n",
       "      <td>2615296</td>\n",
       "      <td>2615397</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000230542_1</td>\n",
       "      <td>v39,v29</td>\n",
       "      <td>ENSG00000230542</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156487</th>\n",
       "      <td>chrY</td>\n",
       "      <td>18404995</td>\n",
       "      <td>18405096</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000226362_1</td>\n",
       "      <td>v39,v29</td>\n",
       "      <td>ENSG00000226362</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156488</th>\n",
       "      <td>chrY</td>\n",
       "      <td>18404995</td>\n",
       "      <td>18405096</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000282909_2</td>\n",
       "      <td>v39,v29</td>\n",
       "      <td>ENSG00000282909</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156489</th>\n",
       "      <td>chrY</td>\n",
       "      <td>18404995</td>\n",
       "      <td>18405096</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000226362_1</td>\n",
       "      <td>v39</td>\n",
       "      <td>ENSG00000226362</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156490</th>\n",
       "      <td>chrY</td>\n",
       "      <td>18404995</td>\n",
       "      <td>18405096</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000282909_2</td>\n",
       "      <td>v39</td>\n",
       "      <td>ENSG00000282909</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9310 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome     Start       End Strand               Name  \\\n",
       "18           chr1     30216     30317      +  ENSG00000243485_2   \n",
       "19           chr1     30216     30317      +  ENSG00000243485_2   \n",
       "24           chr1     30315     30416      +  ENSG00000284332_1   \n",
       "25           chr1     30315     30416      +  ENSG00000284332_1   \n",
       "492          chr1  10430333  10430485      +  ENSG00000251503_2   \n",
       "...           ...       ...       ...    ...                ...   \n",
       "156310       chrY   2615296   2615397      -  ENSG00000230542_1   \n",
       "156487       chrY  18404995  18405096      -  ENSG00000226362_1   \n",
       "156488       chrY  18404995  18405096      -  ENSG00000282909_2   \n",
       "156489       chrY  18404995  18405096      -  ENSG00000226362_1   \n",
       "156490       chrY  18404995  18405096      -  ENSG00000282909_2   \n",
       "\n",
       "               source          gene_id  tss  \n",
       "18      v39,v29,talon  ENSG00000243485    2  \n",
       "19          v39,talon  ENSG00000243485    2  \n",
       "24            v39,v29  ENSG00000284332    1  \n",
       "25                v39  ENSG00000284332    1  \n",
       "492     v39,v29,talon  ENSG00000251503    2  \n",
       "...               ...              ...  ...  \n",
       "156310        v39,v29  ENSG00000230542    1  \n",
       "156487        v39,v29  ENSG00000226362    1  \n",
       "156488        v39,v29  ENSG00000282909    2  \n",
       "156489            v39  ENSG00000226362    1  \n",
       "156490            v39  ENSG00000282909    2  \n",
       "\n",
       "[9310 rows x 8 columns]"
      ]
     },
     "execution_count": 276,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp.loc[temp[['Start', 'End', 'Strand', 'Name', 'gene_id', 'tss']].duplicated(keep=False)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 268,
   "id": "57ee5bcd-e66d-4dfb-a45d-b347cd604c02",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Start_end</th>\n",
       "      <th>End_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>transcript</td>\n",
       "      <td>ENCODEHG000058844</td>\n",
       "      <td>ENCODEHG000058844[1,1,3]</td>\n",
       "      <td>+</td>\n",
       "      <td>1980</td>\n",
       "      <td>8837</td>\n",
       "      <td>1930.0</td>\n",
       "      <td>2031.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058844</td>\n",
       "      <td>ENCODEHG000058844[1,1,3]</td>\n",
       "      <td>+</td>\n",
       "      <td>1980</td>\n",
       "      <td>2005</td>\n",
       "      <td>1930.0</td>\n",
       "      <td>2031.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058844</td>\n",
       "      <td>ENCODEHG000058844[1,1,3]</td>\n",
       "      <td>+</td>\n",
       "      <td>4003</td>\n",
       "      <td>4080</td>\n",
       "      <td>1930.0</td>\n",
       "      <td>2031.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058844</td>\n",
       "      <td>ENCODEHG000058844[1,1,3]</td>\n",
       "      <td>+</td>\n",
       "      <td>4568</td>\n",
       "      <td>4779</td>\n",
       "      <td>1930.0</td>\n",
       "      <td>2031.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058844</td>\n",
       "      <td>ENCODEHG000058844[1,1,3]</td>\n",
       "      <td>+</td>\n",
       "      <td>6057</td>\n",
       "      <td>6333</td>\n",
       "      <td>1930.0</td>\n",
       "      <td>2031.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058844</td>\n",
       "      <td>ENCODEHG000058844[1,1,3]</td>\n",
       "      <td>+</td>\n",
       "      <td>7270</td>\n",
       "      <td>7366</td>\n",
       "      <td>1930.0</td>\n",
       "      <td>2031.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058844</td>\n",
       "      <td>ENCODEHG000058844[1,1,3]</td>\n",
       "      <td>+</td>\n",
       "      <td>7872</td>\n",
       "      <td>7988</td>\n",
       "      <td>1930.0</td>\n",
       "      <td>2031.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058844</td>\n",
       "      <td>ENCODEHG000058844[1,1,3]</td>\n",
       "      <td>+</td>\n",
       "      <td>8124</td>\n",
       "      <td>8207</td>\n",
       "      <td>1930.0</td>\n",
       "      <td>2031.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058844</td>\n",
       "      <td>ENCODEHG000058844[1,1,3]</td>\n",
       "      <td>+</td>\n",
       "      <td>8755</td>\n",
       "      <td>8837</td>\n",
       "      <td>1930.0</td>\n",
       "      <td>2031.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>transcript</td>\n",
       "      <td>ENCODEHG000058844</td>\n",
       "      <td>ENCODEHG000058844[1,2,3]</td>\n",
       "      <td>+</td>\n",
       "      <td>1980</td>\n",
       "      <td>8837</td>\n",
       "      <td>1930.0</td>\n",
       "      <td>2031.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Feature            gene_id             transcript_id Strand  Start  \\\n",
       "56  transcript  ENCODEHG000058844  ENCODEHG000058844[1,1,3]      +   1980   \n",
       "57        exon  ENCODEHG000058844  ENCODEHG000058844[1,1,3]      +   1980   \n",
       "58        exon  ENCODEHG000058844  ENCODEHG000058844[1,1,3]      +   4003   \n",
       "59        exon  ENCODEHG000058844  ENCODEHG000058844[1,1,3]      +   4568   \n",
       "60        exon  ENCODEHG000058844  ENCODEHG000058844[1,1,3]      +   6057   \n",
       "61        exon  ENCODEHG000058844  ENCODEHG000058844[1,1,3]      +   7270   \n",
       "62        exon  ENCODEHG000058844  ENCODEHG000058844[1,1,3]      +   7872   \n",
       "63        exon  ENCODEHG000058844  ENCODEHG000058844[1,1,3]      +   8124   \n",
       "64        exon  ENCODEHG000058844  ENCODEHG000058844[1,1,3]      +   8755   \n",
       "65  transcript  ENCODEHG000058844  ENCODEHG000058844[1,2,3]      +   1980   \n",
       "\n",
       "     End  Start_end  End_end  \n",
       "56  8837     1930.0   2031.0  \n",
       "57  2005     1930.0   2031.0  \n",
       "58  4080     1930.0   2031.0  \n",
       "59  4779     1930.0   2031.0  \n",
       "60  6333     1930.0   2031.0  \n",
       "61  7366     1930.0   2031.0  \n",
       "62  7988     1930.0   2031.0  \n",
       "63  8207     1930.0   2031.0  \n",
       "64  8837     1930.0   2031.0  \n",
       "65  8837     1930.0   2031.0  "
      ]
     },
     "execution_count": 268,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 249,
   "id": "785d5d1a-2198-4565-9412-5da57b085672",
   "metadata": {},
   "outputs": [],
   "source": [
    "# test = df.loc[df.gene_id == 'ENCODEHG000058846'].copy(deep=True)\n",
    "# test = sort_gtf(test)\n",
    "# # test.sort_values(by=['transcript_id', 'gene_id', 'feature_rank', 'Start'],\n",
    "# #                 ascending=[True, True, True, True],\n",
    "# #                 na_position='first')[['Feature', 'transcript_id', 'gene_id', 'feature_rank', 'Start']]\n",
    "\n",
    "# # ascending = True\n",
    "# # test.sort_values(by=['transcript_id', 'gene_id', 'feature_rank', 'Start'],\n",
    "# #                 ascending=[True, True, True, ascending],\n",
    "# #                 na_position='first')[['Feature', 'transcript_id', 'gene_id', 'feature_rank', 'Start', 'End']]\n",
    "# test[['Feature', 'transcript_id', 'gene_id', 'feature_rank', 'Start', 'End']]\n",
    "# # test.loc[test[['transcript_id', 'gene_id']].duplicated(keep=False)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 250,
   "id": "be7c7e64-6ac1-4456-ae88-88a78b9b359d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# test = df.loc[df.gene_id.isin(['ENCODEHG000058844', 'ENCODEHG000058784', 'ENSG00000285631.1'])].copy(deep=True)\n",
    "# test = sort_gtf(test)\n",
    "# # test.sort_values(by=['transcript_id', 'gene_id', 'feature_rank', 'Start'],\n",
    "# #                 ascending=[True, True, True, True],\n",
    "# #                 na_position='first')[['Feature', 'transcript_id', 'gene_id', 'feature_rank', 'Start']]\n",
    "\n",
    "# ascending = False\n",
    "# # test.sort_values(by=['gene_id', 'transcript_id', 'feature_rank', 'Start'],\n",
    "# #                 ascending=[True, True, True, ascending],\n",
    "# #                 na_position='first')[['Feature', 'gene_id', 'transcript_id', 'feature_rank', 'Start', 'End']]\n",
    "# test[['Feature', 'transcript_id', 'gene_id', 'feature_rank', 'Start', 'End']]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 248,
   "id": "7b250c15-1c90-443e-8091-08b819e9f61d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       Feature            gene_id             transcript_id  feature_rank  \\\n",
      "0         gene  ENCODEHG000058844                       NaN             0   \n",
      "1   transcript  ENCODEHG000058844  ENCODEHG000058844[1,1,3]             1   \n",
      "2         exon  ENCODEHG000058844  ENCODEHG000058844[1,1,3]             2   \n",
      "3         exon  ENCODEHG000058844  ENCODEHG000058844[1,1,3]             2   \n",
      "4         exon  ENCODEHG000058844  ENCODEHG000058844[1,1,3]             2   \n",
      "5         exon  ENCODEHG000058844  ENCODEHG000058844[1,1,3]             2   \n",
      "6         exon  ENCODEHG000058844  ENCODEHG000058844[1,1,3]             2   \n",
      "7         exon  ENCODEHG000058844  ENCODEHG000058844[1,1,3]             2   \n",
      "8         exon  ENCODEHG000058844  ENCODEHG000058844[1,1,3]             2   \n",
      "9         exon  ENCODEHG000058844  ENCODEHG000058844[1,1,3]             2   \n",
      "10  transcript  ENCODEHG000058844  ENCODEHG000058844[1,2,3]             1   \n",
      "11        exon  ENCODEHG000058844  ENCODEHG000058844[1,2,3]             2   \n",
      "12        exon  ENCODEHG000058844  ENCODEHG000058844[1,2,3]             2   \n",
      "13        exon  ENCODEHG000058844  ENCODEHG000058844[1,2,3]             2   \n",
      "14        exon  ENCODEHG000058844  ENCODEHG000058844[1,2,3]             2   \n",
      "15        exon  ENCODEHG000058844  ENCODEHG000058844[1,2,3]             2   \n",
      "16  transcript  ENCODEHG000058844  ENCODEHG000058844[1,3,3]             1   \n",
      "17        exon  ENCODEHG000058844  ENCODEHG000058844[1,3,3]             2   \n",
      "18        exon  ENCODEHG000058844  ENCODEHG000058844[1,3,3]             2   \n",
      "19        exon  ENCODEHG000058844  ENCODEHG000058844[1,3,3]             2   \n",
      "20        exon  ENCODEHG000058844  ENCODEHG000058844[1,3,3]             2   \n",
      "21        exon  ENCODEHG000058844  ENCODEHG000058844[1,3,3]             2   \n",
      "22  transcript  ENCODEHG000058844  ENCODEHG000058844[1,4,2]             1   \n",
      "23        exon  ENCODEHG000058844  ENCODEHG000058844[1,4,2]             2   \n",
      "24        exon  ENCODEHG000058844  ENCODEHG000058844[1,4,2]             2   \n",
      "25  transcript  ENCODEHG000058844  ENCODEHG000058844[2,5,1]             1   \n",
      "26        exon  ENCODEHG000058844  ENCODEHG000058844[2,5,1]             2   \n",
      "27        exon  ENCODEHG000058844  ENCODEHG000058844[2,5,1]             2   \n",
      "28        exon  ENCODEHG000058844  ENCODEHG000058844[2,5,1]             2   \n",
      "\n",
      "    Start   End  \n",
      "0    1980  6780  \n",
      "1    1980  8837  \n",
      "2    1980  2005  \n",
      "3    4003  4080  \n",
      "4    4568  4779  \n",
      "5    6057  6333  \n",
      "6    7270  7366  \n",
      "7    7872  7988  \n",
      "8    8124  8207  \n",
      "9    8755  8837  \n",
      "10   1980  8837  \n",
      "11   1980  2005  \n",
      "12   4003  4080  \n",
      "13   4574  4774  \n",
      "14   6057  6333  \n",
      "15   8755  8837  \n",
      "16   1980  8937  \n",
      "17   1980  2005  \n",
      "18   4568  4779  \n",
      "19   6057  7988  \n",
      "20   8127  8207  \n",
      "21   8755  8937  \n",
      "22   1980  7674  \n",
      "23   1980  2005  \n",
      "24   6057  7674  \n",
      "25   4003  6780  \n",
      "26   4003  4080  \n",
      "27   4568  4779  \n",
      "28   6570  6780  \n"
     ]
    }
   ],
   "source": [
    "gtf.reset_index(inplace=True, drop=True)\n",
    "print(gtf.loc[gtf.gene_id == 'ENCODEHG000058844', ['Feature', 'gene_id', 'transcript_id', 'feature_rank', 'Start', 'End']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 232,
   "id": "e3097f19-132c-4dd4-b31e-b2ea5ba58213",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>feature_rank</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gene</td>\n",
       "      <td>ENCODEHG000058844</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1980</td>\n",
       "      <td>6780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>transcript</td>\n",
       "      <td>ENCODEHG000058844</td>\n",
       "      <td>ENCODEHG000058844[1,1,3]</td>\n",
       "      <td>1</td>\n",
       "      <td>1980</td>\n",
       "      <td>8837</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058844</td>\n",
       "      <td>ENCODEHG000058844[1,1,3]</td>\n",
       "      <td>2</td>\n",
       "      <td>1980</td>\n",
       "      <td>2005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058844</td>\n",
       "      <td>ENCODEHG000058844[1,1,3]</td>\n",
       "      <td>2</td>\n",
       "      <td>4003</td>\n",
       "      <td>4080</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058844</td>\n",
       "      <td>ENCODEHG000058844[1,1,3]</td>\n",
       "      <td>2</td>\n",
       "      <td>4568</td>\n",
       "      <td>4779</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1216315</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENSG00000285991.1</td>\n",
       "      <td>ENSG00000285991[1,2,2]</td>\n",
       "      <td>2</td>\n",
       "      <td>149890816</td>\n",
       "      <td>149891034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1216316</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENSG00000285991.1</td>\n",
       "      <td>ENSG00000285991[1,2,2]</td>\n",
       "      <td>2</td>\n",
       "      <td>149889884</td>\n",
       "      <td>149890145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1216317</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENSG00000285991.1</td>\n",
       "      <td>ENSG00000285991[1,2,2]</td>\n",
       "      <td>2</td>\n",
       "      <td>149889347</td>\n",
       "      <td>149889623</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1216318</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENSG00000285991.1</td>\n",
       "      <td>ENSG00000285991[1,2,2]</td>\n",
       "      <td>2</td>\n",
       "      <td>149888538</td>\n",
       "      <td>149888667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1216319</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENSG00000285991.1</td>\n",
       "      <td>ENSG00000285991[1,2,2]</td>\n",
       "      <td>2</td>\n",
       "      <td>149882886</td>\n",
       "      <td>149884504</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1216320 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Feature            gene_id             transcript_id  \\\n",
       "0              gene  ENCODEHG000058844                       NaN   \n",
       "1        transcript  ENCODEHG000058844  ENCODEHG000058844[1,1,3]   \n",
       "2              exon  ENCODEHG000058844  ENCODEHG000058844[1,1,3]   \n",
       "3              exon  ENCODEHG000058844  ENCODEHG000058844[1,1,3]   \n",
       "4              exon  ENCODEHG000058844  ENCODEHG000058844[1,1,3]   \n",
       "...             ...                ...                       ...   \n",
       "1216315        exon  ENSG00000285991.1    ENSG00000285991[1,2,2]   \n",
       "1216316        exon  ENSG00000285991.1    ENSG00000285991[1,2,2]   \n",
       "1216317        exon  ENSG00000285991.1    ENSG00000285991[1,2,2]   \n",
       "1216318        exon  ENSG00000285991.1    ENSG00000285991[1,2,2]   \n",
       "1216319        exon  ENSG00000285991.1    ENSG00000285991[1,2,2]   \n",
       "\n",
       "         feature_rank      Start        End  \n",
       "0                   0       1980       6780  \n",
       "1                   1       1980       8837  \n",
       "2                   2       1980       2005  \n",
       "3                   2       4003       4080  \n",
       "4                   2       4568       4779  \n",
       "...               ...        ...        ...  \n",
       "1216315             2  149890816  149891034  \n",
       "1216316             2  149889884  149890145  \n",
       "1216317             2  149889347  149889623  \n",
       "1216318             2  149888538  149888667  \n",
       "1216319             2  149882886  149884504  \n",
       "\n",
       "[1216320 rows x 6 columns]"
      ]
     },
     "execution_count": 232,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# # gtf.loc[gtf.transcript_id.isnull(), ['Feature', 'transcript_id', 'original_transcript_id']]\n",
    "gtf.reset_index(inplace=True, drop=True)\n",
    "gtf.head()\n",
    "gtf[['Feature', 'gene_id', 'transcript_id', 'feature_rank', 'Start', 'End']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "id": "1028b043-acfb-4592-b549-ad430484f1e3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Feature</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>feature_rank</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>transcript</td>\n",
       "      <td>ENCODEHG000058844[1,1,3]</td>\n",
       "      <td>ENCODEHG000058844</td>\n",
       "      <td>1</td>\n",
       "      <td>1980</td>\n",
       "      <td>8837</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058844[1,1,3]</td>\n",
       "      <td>ENCODEHG000058844</td>\n",
       "      <td>2</td>\n",
       "      <td>1980</td>\n",
       "      <td>2005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058844[1,1,3]</td>\n",
       "      <td>ENCODEHG000058844</td>\n",
       "      <td>2</td>\n",
       "      <td>4003</td>\n",
       "      <td>4080</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058844[1,1,3]</td>\n",
       "      <td>ENCODEHG000058844</td>\n",
       "      <td>2</td>\n",
       "      <td>4568</td>\n",
       "      <td>4779</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>exon</td>\n",
       "      <td>ENCODEHG000058844[1,1,3]</td>\n",
       "      <td>ENCODEHG000058844</td>\n",
       "      <td>2</td>\n",
       "      <td>6057</td>\n",
       "      <td>6333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>358883</th>\n",
       "      <td>gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000285967.1</td>\n",
       "      <td>0</td>\n",
       "      <td>36864424</td>\n",
       "      <td>36876207</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>616385</th>\n",
       "      <td>gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000285972.1</td>\n",
       "      <td>0</td>\n",
       "      <td>84167227</td>\n",
       "      <td>84172093</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>378346</th>\n",
       "      <td>gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000285978.1</td>\n",
       "      <td>0</td>\n",
       "      <td>178694604</td>\n",
       "      <td>178729335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>699465</th>\n",
       "      <td>gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000285980.1</td>\n",
       "      <td>0</td>\n",
       "      <td>131564543</td>\n",
       "      <td>131581145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>429315</th>\n",
       "      <td>gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000285991.1</td>\n",
       "      <td>0</td>\n",
       "      <td>149882886</td>\n",
       "      <td>149896011</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1216320 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Feature             transcript_id            gene_id  feature_rank  \\\n",
       "56      transcript  ENCODEHG000058844[1,1,3]  ENCODEHG000058844             1   \n",
       "57            exon  ENCODEHG000058844[1,1,3]  ENCODEHG000058844             2   \n",
       "58            exon  ENCODEHG000058844[1,1,3]  ENCODEHG000058844             2   \n",
       "59            exon  ENCODEHG000058844[1,1,3]  ENCODEHG000058844             2   \n",
       "60            exon  ENCODEHG000058844[1,1,3]  ENCODEHG000058844             2   \n",
       "...            ...                       ...                ...           ...   \n",
       "358883        gene                       NaN  ENSG00000285967.1             0   \n",
       "616385        gene                       NaN  ENSG00000285972.1             0   \n",
       "378346        gene                       NaN  ENSG00000285978.1             0   \n",
       "699465        gene                       NaN  ENSG00000285980.1             0   \n",
       "429315        gene                       NaN  ENSG00000285991.1             0   \n",
       "\n",
       "            Start        End  \n",
       "56           1980       8837  \n",
       "57           1980       2005  \n",
       "58           4003       4080  \n",
       "59           4568       4779  \n",
       "60           6057       6333  \n",
       "...           ...        ...  \n",
       "358883   36864424   36876207  \n",
       "616385   84167227   84172093  \n",
       "378346  178694604  178729335  \n",
       "699465  131564543  131581145  \n",
       "429315  149882886  149896011  \n",
       "\n",
       "[1216320 rows x 6 columns]"
      ]
     },
     "execution_count": 216,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf[['Feature', 'transcript_id', 'gene_id', 'feature_rank', 'Start', 'End']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "id": "38a6486f-0163-4f25-a60e-0affab7502bb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1216320"
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "599084+617236"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "id": "1b43d825-9634-46a0-975b-a7780166dac0",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf.feature_rank = gtf.feature_rank.astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "68836ad1-fe62-4e1f-9915-7a25a26df523",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>...</th>\n",
       "      <th>tss</th>\n",
       "      <th>tes_id</th>\n",
       "      <th>tes</th>\n",
       "      <th>original_transcript_name</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>Start_end</th>\n",
       "      <th>End_end</th>\n",
       "      <th>Start_end</th>\n",
       "      <th>End_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>10647</td>\n",
       "      <td>10791</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENCODEHG000058846_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENCODEHT000206942</td>\n",
       "      <td>ENCODEHG000058846[1,1,1]</td>\n",
       "      <td>ENCODEHG000058846[1,1,1]</td>\n",
       "      <td>10597</td>\n",
       "      <td>10698</td>\n",
       "      <td>11555</td>\n",
       "      <td>11656</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>10882</td>\n",
       "      <td>11057</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENCODEHG000058846_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENCODEHT000206942</td>\n",
       "      <td>ENCODEHG000058846[1,1,1]</td>\n",
       "      <td>ENCODEHG000058846[1,1,1]</td>\n",
       "      <td>10597</td>\n",
       "      <td>10698</td>\n",
       "      <td>11555</td>\n",
       "      <td>11656</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>11434</td>\n",
       "      <td>11606</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENCODEHG000058846_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENCODEHT000206942</td>\n",
       "      <td>ENCODEHG000058846[1,1,1]</td>\n",
       "      <td>ENCODEHG000058846[1,1,1]</td>\n",
       "      <td>10597</td>\n",
       "      <td>10698</td>\n",
       "      <td>11555</td>\n",
       "      <td>11656</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>TALON</td>\n",
       "      <td>transcript</td>\n",
       "      <td>10647</td>\n",
       "      <td>11606</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENCODEHG000058846_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENCODEHT000206942</td>\n",
       "      <td>ENCODEHG000058846[1,1,1]</td>\n",
       "      <td>ENCODEHG000058846[1,1,1]</td>\n",
       "      <td>10597</td>\n",
       "      <td>10698</td>\n",
       "      <td>11555</td>\n",
       "      <td>11656</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chromosome Source     Feature  Start    End Score Strand Frame  \\\n",
       "28      SIRV1  TALON        exon  10647  10791     .      +     .   \n",
       "29      SIRV1  TALON        exon  10882  11057     .      +     .   \n",
       "30      SIRV1  TALON        exon  11434  11606     .      +     .   \n",
       "31      SIRV1  TALON  transcript  10647  11606     .      +     .   \n",
       "\n",
       "              gene_id          gene_name  ...  tss               tes_id  tes  \\\n",
       "28  ENCODEHG000058846  ENCODEHG000058846  ...  1.0  ENCODEHG000058846_1  1.0   \n",
       "29  ENCODEHG000058846  ENCODEHG000058846  ...  1.0  ENCODEHG000058846_1  1.0   \n",
       "30  ENCODEHG000058846  ENCODEHG000058846  ...  1.0  ENCODEHG000058846_1  1.0   \n",
       "31  ENCODEHG000058846  ENCODEHG000058846  ...  1.0  ENCODEHG000058846_1  1.0   \n",
       "\n",
       "   original_transcript_name             transcript_id  \\\n",
       "28        ENCODEHT000206942  ENCODEHG000058846[1,1,1]   \n",
       "29        ENCODEHT000206942  ENCODEHG000058846[1,1,1]   \n",
       "30        ENCODEHT000206942  ENCODEHG000058846[1,1,1]   \n",
       "31        ENCODEHT000206942  ENCODEHG000058846[1,1,1]   \n",
       "\n",
       "             transcript_name Start_end End_end Start_end End_end  \n",
       "28  ENCODEHG000058846[1,1,1]     10597   10698     11555   11656  \n",
       "29  ENCODEHG000058846[1,1,1]     10597   10698     11555   11656  \n",
       "30  ENCODEHG000058846[1,1,1]     10597   10698     11555   11656  \n",
       "31  ENCODEHG000058846[1,1,1]     10597   10698     11555   11656  \n",
       "\n",
       "[4 rows x 48 columns]"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf.loc[gtf.gene_id == 'ENCODEHG000058846']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7a7d3ce-e8f0-4000-9430-4d00799901eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "if update_ends:\n",
    "    df = update_ends(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "48722731-36a9-47e1-a082-745cfbff846e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>...</th>\n",
       "      <th>intergenic_novel</th>\n",
       "      <th>antisense_gene</th>\n",
       "      <th>gene_antisense_to_IDs</th>\n",
       "      <th>original_transcript_id</th>\n",
       "      <th>ic_id</th>\n",
       "      <th>tss_id</th>\n",
       "      <th>tes_id</th>\n",
       "      <th>original_transcript_name</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>698</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1232225</td>\n",
       "      <td>1235041</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000176022.5</td>\n",
       "      <td>B3GALT6</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000647651.1</td>\n",
       "      <td>ENSG00000176022_1</td>\n",
       "      <td>ENSG00000176022_1</td>\n",
       "      <td>ENSG00000176022_1</td>\n",
       "      <td>B3GALT6-202</td>\n",
       "      <td>ENSG00000176022[1,1,1]</td>\n",
       "      <td>B3GALT6[1,1,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>700</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1232258</td>\n",
       "      <td>1235041</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000176022.5</td>\n",
       "      <td>B3GALT6</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000379198.4</td>\n",
       "      <td>ENSG00000176022_1</td>\n",
       "      <td>ENSG00000176022_1</td>\n",
       "      <td>ENSG00000176022_1</td>\n",
       "      <td>B3GALT6-201</td>\n",
       "      <td>ENSG00000176022[1,1,1]</td>\n",
       "      <td>B3GALT6[1,1,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>380352</th>\n",
       "      <td>chr6</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1609971</td>\n",
       "      <td>1613897</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000054598.7</td>\n",
       "      <td>FOXC1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000380874.3</td>\n",
       "      <td>ENSG00000054598_1</td>\n",
       "      <td>ENSG00000054598_1</td>\n",
       "      <td>ENSG00000054598_1</td>\n",
       "      <td>FOXC1-201</td>\n",
       "      <td>ENSG00000054598[1,1,1]</td>\n",
       "      <td>FOXC1[1,1,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>380354</th>\n",
       "      <td>chr6</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1610066</td>\n",
       "      <td>1613897</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000054598.7</td>\n",
       "      <td>FOXC1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000645831.1</td>\n",
       "      <td>ENSG00000054598_1</td>\n",
       "      <td>ENSG00000054598_1</td>\n",
       "      <td>ENSG00000054598_1</td>\n",
       "      <td>FOXC1-202</td>\n",
       "      <td>ENSG00000054598[1,1,1]</td>\n",
       "      <td>FOXC1[1,1,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1189029</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>83508260</td>\n",
       "      <td>83512127</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000196767.7</td>\n",
       "      <td>POU3F4</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000373200.5</td>\n",
       "      <td>ENSG00000196767_1</td>\n",
       "      <td>ENSG00000196767_1</td>\n",
       "      <td>ENSG00000196767_1</td>\n",
       "      <td>POU3F4-201</td>\n",
       "      <td>ENSG00000196767[1,1,1]</td>\n",
       "      <td>POU3F4[1,1,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1189031</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>83508292</td>\n",
       "      <td>83512127</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000196767.7</td>\n",
       "      <td>POU3F4</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000644024.1</td>\n",
       "      <td>ENSG00000196767_1</td>\n",
       "      <td>ENSG00000196767_1</td>\n",
       "      <td>ENSG00000196767_1</td>\n",
       "      <td>POU3F4-202</td>\n",
       "      <td>ENSG00000196767[1,1,1]</td>\n",
       "      <td>POU3F4[1,1,1]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6 rows × 41 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chromosome  Source     Feature     Start       End Score Strand Frame  \\\n",
       "698           chr1  HAVANA  transcript   1232225   1235041     .      +     .   \n",
       "700           chr1  HAVANA  transcript   1232258   1235041     .      +     .   \n",
       "380352        chr6  HAVANA  transcript   1609971   1613897     .      +     .   \n",
       "380354        chr6  HAVANA  transcript   1610066   1613897     .      +     .   \n",
       "1189029       chrX  HAVANA  transcript  83508260  83512127     .      +     .   \n",
       "1189031       chrX  HAVANA  transcript  83508292  83512127     .      +     .   \n",
       "\n",
       "                   gene_id gene_name  ... intergenic_novel antisense_gene  \\\n",
       "698      ENSG00000176022.5   B3GALT6  ...              NaN            NaN   \n",
       "700      ENSG00000176022.5   B3GALT6  ...              NaN            NaN   \n",
       "380352   ENSG00000054598.7     FOXC1  ...              NaN            NaN   \n",
       "380354   ENSG00000054598.7     FOXC1  ...              NaN            NaN   \n",
       "1189029  ENSG00000196767.7    POU3F4  ...              NaN            NaN   \n",
       "1189031  ENSG00000196767.7    POU3F4  ...              NaN            NaN   \n",
       "\n",
       "        gene_antisense_to_IDs original_transcript_id              ic_id  \\\n",
       "698                       NaN      ENST00000647651.1  ENSG00000176022_1   \n",
       "700                       NaN      ENST00000379198.4  ENSG00000176022_1   \n",
       "380352                    NaN      ENST00000380874.3  ENSG00000054598_1   \n",
       "380354                    NaN      ENST00000645831.1  ENSG00000054598_1   \n",
       "1189029                   NaN      ENST00000373200.5  ENSG00000196767_1   \n",
       "1189031                   NaN      ENST00000644024.1  ENSG00000196767_1   \n",
       "\n",
       "                    tss_id             tes_id original_transcript_name  \\\n",
       "698      ENSG00000176022_1  ENSG00000176022_1              B3GALT6-202   \n",
       "700      ENSG00000176022_1  ENSG00000176022_1              B3GALT6-201   \n",
       "380352   ENSG00000054598_1  ENSG00000054598_1                FOXC1-201   \n",
       "380354   ENSG00000054598_1  ENSG00000054598_1                FOXC1-202   \n",
       "1189029  ENSG00000196767_1  ENSG00000196767_1               POU3F4-201   \n",
       "1189031  ENSG00000196767_1  ENSG00000196767_1               POU3F4-202   \n",
       "\n",
       "                  transcript_id transcript_name  \n",
       "698      ENSG00000176022[1,1,1]  B3GALT6[1,1,1]  \n",
       "700      ENSG00000176022[1,1,1]  B3GALT6[1,1,1]  \n",
       "380352   ENSG00000054598[1,1,1]    FOXC1[1,1,1]  \n",
       "380354   ENSG00000054598[1,1,1]    FOXC1[1,1,1]  \n",
       "1189029  ENSG00000196767[1,1,1]   POU3F4[1,1,1]  \n",
       "1189031  ENSG00000196767[1,1,1]   POU3F4[1,1,1]  \n",
       "\n",
       "[6 rows x 41 columns]"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp = df.loc[df.Feature =='transcript']\n",
    "b = temp.loc[temp.transcript_id.duplicated(keep=False)]\n",
    "b"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "6e220b14-0f93-4fa3-acc0-e318897656c2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1216320"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "9dc6dedd-c42c-4a76-985d-d8eab21cc46a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def map_gtf_transcripts(gtf, m_df):\n",
    "    \"\"\"\n",
    "    Add cerberus transcript id to each entry in gtf\n",
    "    \"\"\"\n",
    "    gtf = df.copy(deep=True)\n",
    "    gtf = gtf.merge(m_df, how='left', \n",
    "                    left_on=['transcript_name', 'transcript_id'],\n",
    "                    right_on=['original_transcript_name', 'original_transcript_id'],\n",
    "                    suffixes=('', '_cerberus'))\n",
    "\n",
    "    gtf.drop(['transcript_id', 'transcript_name'], axis=1, inplace=True)\n",
    "    gtf.rename({'transcript_id_cerberus': 'transcript_id',\n",
    "                'transcript_name_cerberus': 'transcript_name'},\n",
    "               axis=1, inplace=True)\n",
    "    \n",
    "    return gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "007b0c3f-2769-4fe9-b086-d8e0d0259d50",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df.loc[df.transcript_id == 'ENST00000469289.1', ['gene_id', 'gene_name', 'transcript_name', 'transcript_id']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "2b9ddcd6-533f-4332-b161-067434c53f0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# m_df.loc[m_df.original_transcript_id == 'ENST00000469289.1', ['gene_id', 'gene_name', 'original_transcript_name', 'original_transcript_id']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "fe688570-f210-4620-8004-345570c1d2d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df.loc[df.transcript_id == 'ENST00000469289.1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "730f2e7f-2da6-4d40-943e-290d0abc0169",
   "metadata": {},
   "outputs": [],
   "source": [
    "# m_df.loc[m_df.original_transcript_id == 'ENST00000469289.1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "a66b833f-f242-4e74-92fc-08342684e26e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>...</th>\n",
       "      <th>intergenic_novel</th>\n",
       "      <th>antisense_gene</th>\n",
       "      <th>gene_antisense_to_IDs</th>\n",
       "      <th>original_transcript_id</th>\n",
       "      <th>ic_id</th>\n",
       "      <th>tss_id</th>\n",
       "      <th>tes_id</th>\n",
       "      <th>original_transcript_name</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>698</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1232225</td>\n",
       "      <td>1235041</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000176022.5</td>\n",
       "      <td>B3GALT6</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000647651.1</td>\n",
       "      <td>ENSG00000176022_1</td>\n",
       "      <td>ENSG00000176022_1</td>\n",
       "      <td>ENSG00000176022_1</td>\n",
       "      <td>B3GALT6-202</td>\n",
       "      <td>ENSG00000176022[1,1,1]</td>\n",
       "      <td>B3GALT6[1,1,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>700</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1232258</td>\n",
       "      <td>1235041</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000176022.5</td>\n",
       "      <td>B3GALT6</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000379198.4</td>\n",
       "      <td>ENSG00000176022_1</td>\n",
       "      <td>ENSG00000176022_1</td>\n",
       "      <td>ENSG00000176022_1</td>\n",
       "      <td>B3GALT6-201</td>\n",
       "      <td>ENSG00000176022[1,1,1]</td>\n",
       "      <td>B3GALT6[1,1,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>380352</th>\n",
       "      <td>chr6</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1609971</td>\n",
       "      <td>1613897</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000054598.7</td>\n",
       "      <td>FOXC1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000380874.3</td>\n",
       "      <td>ENSG00000054598_1</td>\n",
       "      <td>ENSG00000054598_1</td>\n",
       "      <td>ENSG00000054598_1</td>\n",
       "      <td>FOXC1-201</td>\n",
       "      <td>ENSG00000054598[1,1,1]</td>\n",
       "      <td>FOXC1[1,1,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>380354</th>\n",
       "      <td>chr6</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1610066</td>\n",
       "      <td>1613897</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000054598.7</td>\n",
       "      <td>FOXC1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000645831.1</td>\n",
       "      <td>ENSG00000054598_1</td>\n",
       "      <td>ENSG00000054598_1</td>\n",
       "      <td>ENSG00000054598_1</td>\n",
       "      <td>FOXC1-202</td>\n",
       "      <td>ENSG00000054598[1,1,1]</td>\n",
       "      <td>FOXC1[1,1,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1189029</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>83508260</td>\n",
       "      <td>83512127</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000196767.7</td>\n",
       "      <td>POU3F4</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000373200.5</td>\n",
       "      <td>ENSG00000196767_1</td>\n",
       "      <td>ENSG00000196767_1</td>\n",
       "      <td>ENSG00000196767_1</td>\n",
       "      <td>POU3F4-201</td>\n",
       "      <td>ENSG00000196767[1,1,1]</td>\n",
       "      <td>POU3F4[1,1,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1189031</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>83508292</td>\n",
       "      <td>83512127</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000196767.7</td>\n",
       "      <td>POU3F4</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000644024.1</td>\n",
       "      <td>ENSG00000196767_1</td>\n",
       "      <td>ENSG00000196767_1</td>\n",
       "      <td>ENSG00000196767_1</td>\n",
       "      <td>POU3F4-202</td>\n",
       "      <td>ENSG00000196767[1,1,1]</td>\n",
       "      <td>POU3F4[1,1,1]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6 rows × 41 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chromosome  Source     Feature     Start       End Score Strand Frame  \\\n",
       "698           chr1  HAVANA  transcript   1232225   1235041     .      +     .   \n",
       "700           chr1  HAVANA  transcript   1232258   1235041     .      +     .   \n",
       "380352        chr6  HAVANA  transcript   1609971   1613897     .      +     .   \n",
       "380354        chr6  HAVANA  transcript   1610066   1613897     .      +     .   \n",
       "1189029       chrX  HAVANA  transcript  83508260  83512127     .      +     .   \n",
       "1189031       chrX  HAVANA  transcript  83508292  83512127     .      +     .   \n",
       "\n",
       "                   gene_id gene_name  ... intergenic_novel antisense_gene  \\\n",
       "698      ENSG00000176022.5   B3GALT6  ...              NaN            NaN   \n",
       "700      ENSG00000176022.5   B3GALT6  ...              NaN            NaN   \n",
       "380352   ENSG00000054598.7     FOXC1  ...              NaN            NaN   \n",
       "380354   ENSG00000054598.7     FOXC1  ...              NaN            NaN   \n",
       "1189029  ENSG00000196767.7    POU3F4  ...              NaN            NaN   \n",
       "1189031  ENSG00000196767.7    POU3F4  ...              NaN            NaN   \n",
       "\n",
       "        gene_antisense_to_IDs original_transcript_id              ic_id  \\\n",
       "698                       NaN      ENST00000647651.1  ENSG00000176022_1   \n",
       "700                       NaN      ENST00000379198.4  ENSG00000176022_1   \n",
       "380352                    NaN      ENST00000380874.3  ENSG00000054598_1   \n",
       "380354                    NaN      ENST00000645831.1  ENSG00000054598_1   \n",
       "1189029                   NaN      ENST00000373200.5  ENSG00000196767_1   \n",
       "1189031                   NaN      ENST00000644024.1  ENSG00000196767_1   \n",
       "\n",
       "                    tss_id             tes_id original_transcript_name  \\\n",
       "698      ENSG00000176022_1  ENSG00000176022_1              B3GALT6-202   \n",
       "700      ENSG00000176022_1  ENSG00000176022_1              B3GALT6-201   \n",
       "380352   ENSG00000054598_1  ENSG00000054598_1                FOXC1-201   \n",
       "380354   ENSG00000054598_1  ENSG00000054598_1                FOXC1-202   \n",
       "1189029  ENSG00000196767_1  ENSG00000196767_1               POU3F4-201   \n",
       "1189031  ENSG00000196767_1  ENSG00000196767_1               POU3F4-202   \n",
       "\n",
       "                  transcript_id transcript_name  \n",
       "698      ENSG00000176022[1,1,1]  B3GALT6[1,1,1]  \n",
       "700      ENSG00000176022[1,1,1]  B3GALT6[1,1,1]  \n",
       "380352   ENSG00000054598[1,1,1]    FOXC1[1,1,1]  \n",
       "380354   ENSG00000054598[1,1,1]    FOXC1[1,1,1]  \n",
       "1189029  ENSG00000196767[1,1,1]   POU3F4[1,1,1]  \n",
       "1189031  ENSG00000196767[1,1,1]   POU3F4[1,1,1]  \n",
       "\n",
       "[6 rows x 41 columns]"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf.head()\n",
    "temp = gtf.loc[gtf.Feature =='transcript']\n",
    "b = temp.loc[temp.transcript_id.duplicated(keep=False)]\n",
    "# print(len(b.index))\n",
    "# b.transcript_id.head()\n",
    "b"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "2a3abbe7-9088-422a-9a35-480ab76d1647",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>...</th>\n",
       "      <th>intergenic_novel</th>\n",
       "      <th>antisense_gene</th>\n",
       "      <th>gene_antisense_to_IDs</th>\n",
       "      <th>original_transcript_id</th>\n",
       "      <th>ic_id</th>\n",
       "      <th>tss_id</th>\n",
       "      <th>tes_id</th>\n",
       "      <th>original_transcript_name</th>\n",
       "      <th>transcript_id_cerberus</th>\n",
       "      <th>transcript_name_cerberus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows × 43 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Chromosome, Source, Feature, Start, End, Score, Strand, Frame, gene_id, gene_name, gene_status, gene_type, talon_gene, havana_gene, level, transcript_id, transcript_status, transcript_name, talon_transcript, NIC_transcript, exon_number, exon_id, talon_exon, exon_status, ont, source, tag, transcript_type, transcript_support_level, havana_transcript, NNC_transcript, protein_id, ccdsid, intergenic_novel, antisense_gene, gene_antisense_to_IDs, original_transcript_id, ic_id, tss_id, tes_id, original_transcript_name, transcript_id_cerberus, transcript_name_cerberus]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 43 columns]"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "d = gtf.loc[~gtf.transcript_id_cerberus.isnull()]\n",
    "d.loc[d.gene_id.str.contains('ENSG')]\n",
    "# gtf.loc[gtf.transcript_id_cerberus == 'ENSG00000243485[3,2,1]']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "847ceff7-09b1-4d89-89a4-be24a275d024",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>original_transcript_id</th>\n",
       "      <th>ic_id</th>\n",
       "      <th>tss_id</th>\n",
       "      <th>tes_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>original_transcript_name</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4588</th>\n",
       "      <td>ENST00000469289.1</td>\n",
       "      <td>ENSG00000243485_2</td>\n",
       "      <td>ENSG00000243485_3</td>\n",
       "      <td>ENSG00000243485_1</td>\n",
       "      <td>ENSG00000243485</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>MIR1302-2HG-201</td>\n",
       "      <td>ENSG00000243485[3,2,1]</td>\n",
       "      <td>MIR1302-2HG[3,2,1]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     original_transcript_id              ic_id             tss_id  \\\n",
       "4588      ENST00000469289.1  ENSG00000243485_2  ENSG00000243485_3   \n",
       "\n",
       "                 tes_id          gene_id    gene_name  \\\n",
       "4588  ENSG00000243485_1  ENSG00000243485  MIR1302-2HG   \n",
       "\n",
       "     original_transcript_name           transcript_id     transcript_name  \n",
       "4588          MIR1302-2HG-201  ENSG00000243485[3,2,1]  MIR1302-2HG[3,2,1]  "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "m_df.loc[m_df.original_transcript_id == 'ENST00000469289.1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "1e4df3fb-bae0-4a23-8112-ec6a3d7bccca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>original_transcript_id</th>\n",
       "      <th>ic</th>\n",
       "      <th>ic_id</th>\n",
       "      <th>tss_id</th>\n",
       "      <th>tss</th>\n",
       "      <th>tes_id</th>\n",
       "      <th>tes</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>original_transcript_name</th>\n",
       "      <th>transcript_triplet</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENCODEHT000206942</td>\n",
       "      <td>1</td>\n",
       "      <td>ENCODEHG000058846_1</td>\n",
       "      <td>ENCODEHG000058846_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENCODEHG000058846_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>ENCODEHT000206942</td>\n",
       "      <td>[1,1,1]</td>\n",
       "      <td>ENCODEHG000058846[1,1,1]</td>\n",
       "      <td>ENCODEHG000058846[1,1,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENCODEHT000206867</td>\n",
       "      <td>4</td>\n",
       "      <td>ENCODEHG000058837_4</td>\n",
       "      <td>ENCODEHG000058837_2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENCODEHG000058837_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>ENCODEHT000206867</td>\n",
       "      <td>[2,4,1]</td>\n",
       "      <td>ENCODEHG000058837[2,4,1]</td>\n",
       "      <td>ENCODEHG000058837[2,4,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENCODEHT000206868</td>\n",
       "      <td>2</td>\n",
       "      <td>ENCODEHG000058837_2</td>\n",
       "      <td>ENCODEHG000058837_2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENCODEHG000058837_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>ENCODEHT000206868</td>\n",
       "      <td>[2,2,1]</td>\n",
       "      <td>ENCODEHG000058837[2,2,1]</td>\n",
       "      <td>ENCODEHG000058837[2,2,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENCODEHT000206870</td>\n",
       "      <td>3</td>\n",
       "      <td>ENCODEHG000058837_3</td>\n",
       "      <td>ENCODEHG000058837_2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENCODEHG000058837_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>ENCODEHT000206870</td>\n",
       "      <td>[2,3,1]</td>\n",
       "      <td>ENCODEHG000058837[2,3,1]</td>\n",
       "      <td>ENCODEHG000058837[2,3,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENCODEHT000206886</td>\n",
       "      <td>1</td>\n",
       "      <td>ENCODEHG000058837_1</td>\n",
       "      <td>ENCODEHG000058837_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENCODEHG000058837_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>ENCODEHT000206886</td>\n",
       "      <td>[1,1,1]</td>\n",
       "      <td>ENCODEHG000058837[1,1,1]</td>\n",
       "      <td>ENCODEHG000058837[1,1,1]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  original_transcript_id  ic                ic_id               tss_id  tss  \\\n",
       "0      ENCODEHT000206942   1  ENCODEHG000058846_1  ENCODEHG000058846_1    1   \n",
       "1      ENCODEHT000206867   4  ENCODEHG000058837_4  ENCODEHG000058837_2    2   \n",
       "2      ENCODEHT000206868   2  ENCODEHG000058837_2  ENCODEHG000058837_2    2   \n",
       "3      ENCODEHT000206870   3  ENCODEHG000058837_3  ENCODEHG000058837_2    2   \n",
       "4      ENCODEHT000206886   1  ENCODEHG000058837_1  ENCODEHG000058837_1    1   \n",
       "\n",
       "                tes_id  tes            gene_id          gene_name  \\\n",
       "0  ENCODEHG000058846_1    1  ENCODEHG000058846  ENCODEHG000058846   \n",
       "1  ENCODEHG000058837_1    1  ENCODEHG000058837  ENCODEHG000058837   \n",
       "2  ENCODEHG000058837_1    1  ENCODEHG000058837  ENCODEHG000058837   \n",
       "3  ENCODEHG000058837_1    1  ENCODEHG000058837  ENCODEHG000058837   \n",
       "4  ENCODEHG000058837_1    1  ENCODEHG000058837  ENCODEHG000058837   \n",
       "\n",
       "  original_transcript_name transcript_triplet             transcript_id  \\\n",
       "0        ENCODEHT000206942            [1,1,1]  ENCODEHG000058846[1,1,1]   \n",
       "1        ENCODEHT000206867            [2,4,1]  ENCODEHG000058837[2,4,1]   \n",
       "2        ENCODEHT000206868            [2,2,1]  ENCODEHG000058837[2,2,1]   \n",
       "3        ENCODEHT000206870            [2,3,1]  ENCODEHG000058837[2,3,1]   \n",
       "4        ENCODEHT000206886            [1,1,1]  ENCODEHG000058837[1,1,1]   \n",
       "\n",
       "            transcript_name  \n",
       "0  ENCODEHG000058846[1,1,1]  \n",
       "1  ENCODEHG000058837[2,4,1]  \n",
       "2  ENCODEHG000058837[2,2,1]  \n",
       "3  ENCODEHG000058837[2,3,1]  \n",
       "4  ENCODEHG000058837[1,1,1]  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "m_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "b0d6c29f-5bf5-4992-8f58-8504568b6b98",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>...</th>\n",
       "      <th>tag</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>NNC_transcript</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>intergenic_novel</th>\n",
       "      <th>antisense_gene</th>\n",
       "      <th>gene_antisense_to_IDs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>TALON</td>\n",
       "      <td>gene</td>\n",
       "      <td>10647</td>\n",
       "      <td>11606</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>TALON</td>\n",
       "      <td>transcript</td>\n",
       "      <td>10647</td>\n",
       "      <td>11606</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>10647</td>\n",
       "      <td>10791</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>10882</td>\n",
       "      <td>11057</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>11434</td>\n",
       "      <td>11606</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 36 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome Source     Feature  Start    End Score Strand Frame  \\\n",
       "0      SIRV1  TALON        gene  10647  11606     .      +     .   \n",
       "1      SIRV1  TALON  transcript  10647  11606     .      +     .   \n",
       "2      SIRV1  TALON        exon  10647  10791     .      +     .   \n",
       "3      SIRV1  TALON        exon  10882  11057     .      +     .   \n",
       "4      SIRV1  TALON        exon  11434  11606     .      +     .   \n",
       "\n",
       "             gene_id          gene_name  ...  tag transcript_type  \\\n",
       "0  ENCODEHG000058846  ENCODEHG000058846  ...  NaN             NaN   \n",
       "1  ENCODEHG000058846  ENCODEHG000058846  ...  NaN             NaN   \n",
       "2  ENCODEHG000058846  ENCODEHG000058846  ...  NaN             NaN   \n",
       "3  ENCODEHG000058846  ENCODEHG000058846  ...  NaN             NaN   \n",
       "4  ENCODEHG000058846  ENCODEHG000058846  ...  NaN             NaN   \n",
       "\n",
       "  transcript_support_level havana_transcript NNC_transcript protein_id ccdsid  \\\n",
       "0                      NaN               NaN            NaN        NaN    NaN   \n",
       "1                      NaN               NaN           TRUE        NaN    NaN   \n",
       "2                      NaN               NaN            NaN        NaN    NaN   \n",
       "3                      NaN               NaN            NaN        NaN    NaN   \n",
       "4                      NaN               NaN            NaN        NaN    NaN   \n",
       "\n",
       "  intergenic_novel antisense_gene gene_antisense_to_IDs  \n",
       "0             TRUE            NaN                   NaN  \n",
       "1              NaN            NaN                   NaN  \n",
       "2              NaN            NaN                   NaN  \n",
       "3              NaN            NaN                   NaN  \n",
       "4              NaN            NaN                   NaN  \n",
       "\n",
       "[5 rows x 36 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c0b15b0b-029d-477b-8a8c-d52d9cedb6dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "t_df = df.loc[df.Feature == 'transcript']\n",
    "keep_cols = ['Chromosome', 'Source', 'Feature',\n",
    "             'Start', 'End', 'Score', 'Strand',\n",
    "             'Frame', 'gene_id', 'gene_name', "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "d924101b-eb3c-493e-957d-bfa6e57ba130",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[nan, '2', '1', '3']"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t_df.level.unique().tolist()[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "9a4894f1-6e89-4bc2-8085-0e6bc788d1dc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[nan, '5', 'NA', '1', '4']"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t_df.transcript_support_level.unique().tolist()[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "78d48003-07ee-4177-b9e3-d971c642ce46",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Chromosome', 'Source', 'Feature', 'Start', 'End', 'Score', 'Strand',\n",
       "       'Frame', 'gene_id', 'gene_name', 'gene_status', 'gene_type',\n",
       "       'talon_gene', 'havana_gene', 'level', 'transcript_id',\n",
       "       'transcript_status', 'transcript_name', 'talon_transcript',\n",
       "       'NIC_transcript', 'exon_number', 'exon_id', 'talon_exon', 'exon_status',\n",
       "       'ont', 'source', 'tag', 'transcript_type', 'transcript_support_level',\n",
       "       'havana_transcript', 'NNC_transcript', 'protein_id', 'ccdsid',\n",
       "       'intergenic_novel', 'antisense_gene', 'gene_antisense_to_IDs'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# everything related to gene groupby\n",
    "# drop everything that we don't care about\n",
    "# for other we take set of unique values\n",
    "t_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "b1d67a98-5c14-4bc9-9c85-db274e54eda6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'b;a'"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "';'.join(set(['a','a', 'b']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "f2a38eb1-73ad-451a-a7fd-e21f2a271760",
   "metadata": {},
   "outputs": [],
   "source": [
    "test = map_gtf_transcripts(df, m_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "688c8dfb-31cd-45ec-b696-2722134639aa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Chromosome', 'Source', 'Feature', 'Start', 'End', 'Score', 'Strand',\n",
       "       'Frame', 'gene_id', 'gene_name', 'gene_status', 'gene_type',\n",
       "       'talon_gene', 'havana_gene', 'level', 'transcript_status',\n",
       "       'talon_transcript', 'NIC_transcript', 'exon_number', 'exon_id',\n",
       "       'talon_exon', 'exon_status', 'ont', 'source', 'tag', 'transcript_type',\n",
       "       'transcript_support_level', 'havana_transcript', 'NNC_transcript',\n",
       "       'protein_id', 'ccdsid', 'intergenic_novel', 'antisense_gene',\n",
       "       'gene_antisense_to_IDs', 'transcript_id', 'transcript_name',\n",
       "       'original_transcript_id', 'original_transcript_name',\n",
       "       'original_transcript_id_merge'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "86c6060b-0669-4df1-b5ff-a5fa64f4d7ff",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>original_transcript_id</th>\n",
       "      <th>original_transcript_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1181901</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>25393153</td>\n",
       "      <td>25394056</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000233944.1</td>\n",
       "      <td>LINC00265-3P</td>\n",
       "      <td>ENSG00000233944[1,1,1]</td>\n",
       "      <td>LINC00265-3P[1,1,1]</td>\n",
       "      <td>ENST00000447588.1</td>\n",
       "      <td>LINC00265-3P-201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1181902</th>\n",
       "      <td>chrY</td>\n",
       "      <td>transcript</td>\n",
       "      <td>16379178</td>\n",
       "      <td>16387459</td>\n",
       "      <td>-</td>\n",
       "      <td>ENCODEHG000377496</td>\n",
       "      <td>ENCODEHG000377496</td>\n",
       "      <td>ENCODEHG000377496[1,1,1]</td>\n",
       "      <td>ENCODEHG000377496[1,1,1]</td>\n",
       "      <td>ENCODEHT005023774</td>\n",
       "      <td>ENCODEHT005023774</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1181903</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>16387141</td>\n",
       "      <td>16387459</td>\n",
       "      <td>-</td>\n",
       "      <td>ENCODEHG000377496</td>\n",
       "      <td>ENCODEHG000377496</td>\n",
       "      <td>ENCODEHG000377496[1,1,1]</td>\n",
       "      <td>ENCODEHG000377496[1,1,1]</td>\n",
       "      <td>ENCODEHT005023774</td>\n",
       "      <td>ENCODEHT005023774</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1181904</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>16383092</td>\n",
       "      <td>16383318</td>\n",
       "      <td>-</td>\n",
       "      <td>ENCODEHG000377496</td>\n",
       "      <td>ENCODEHG000377496</td>\n",
       "      <td>ENCODEHG000377496[1,1,1]</td>\n",
       "      <td>ENCODEHG000377496[1,1,1]</td>\n",
       "      <td>ENCODEHT005023774</td>\n",
       "      <td>ENCODEHT005023774</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1181905</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>16379178</td>\n",
       "      <td>16381894</td>\n",
       "      <td>-</td>\n",
       "      <td>ENCODEHG000377496</td>\n",
       "      <td>ENCODEHG000377496</td>\n",
       "      <td>ENCODEHG000377496[1,1,1]</td>\n",
       "      <td>ENCODEHG000377496[1,1,1]</td>\n",
       "      <td>ENCODEHT005023774</td>\n",
       "      <td>ENCODEHT005023774</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chromosome     Feature     Start       End Strand            gene_id  \\\n",
       "1181901       chrY        exon  25393153  25394056      -  ENSG00000233944.1   \n",
       "1181902       chrY  transcript  16379178  16387459      -  ENCODEHG000377496   \n",
       "1181903       chrY        exon  16387141  16387459      -  ENCODEHG000377496   \n",
       "1181904       chrY        exon  16383092  16383318      -  ENCODEHG000377496   \n",
       "1181905       chrY        exon  16379178  16381894      -  ENCODEHG000377496   \n",
       "\n",
       "                 gene_name             transcript_id  \\\n",
       "1181901       LINC00265-3P    ENSG00000233944[1,1,1]   \n",
       "1181902  ENCODEHG000377496  ENCODEHG000377496[1,1,1]   \n",
       "1181903  ENCODEHG000377496  ENCODEHG000377496[1,1,1]   \n",
       "1181904  ENCODEHG000377496  ENCODEHG000377496[1,1,1]   \n",
       "1181905  ENCODEHG000377496  ENCODEHG000377496[1,1,1]   \n",
       "\n",
       "                  transcript_name original_transcript_id  \\\n",
       "1181901       LINC00265-3P[1,1,1]      ENST00000447588.1   \n",
       "1181902  ENCODEHG000377496[1,1,1]      ENCODEHT005023774   \n",
       "1181903  ENCODEHG000377496[1,1,1]      ENCODEHT005023774   \n",
       "1181904  ENCODEHG000377496[1,1,1]      ENCODEHT005023774   \n",
       "1181905  ENCODEHG000377496[1,1,1]      ENCODEHT005023774   \n",
       "\n",
       "        original_transcript_name  \n",
       "1181901         LINC00265-3P-201  \n",
       "1181902        ENCODEHT005023774  \n",
       "1181903        ENCODEHT005023774  \n",
       "1181904        ENCODEHT005023774  \n",
       "1181905        ENCODEHT005023774  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test[['Chromosome', 'Feature', 'Start', 'End', 'Strand', \n",
    "      'gene_id', 'gene_name', 'transcript_id', 'transcript_name', \n",
    "      'original_transcript_id', 'original_transcript_name']].tail()\n",
    "\n",
    "# how to handle transcript novelty????"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "id": "699db26d-e7cf-4fba-8705-1e25e60b3f9e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>original_transcript_id</th>\n",
       "      <th>original_transcript_name</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENCODEHT000206942</td>\n",
       "      <td>ENCODEHT000206942</td>\n",
       "      <td>ENCODEHG000058846[1,1,1]</td>\n",
       "      <td>ENCODEHG000058846[1,1,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENCODEHT000206942</td>\n",
       "      <td>ENCODEHT000206942</td>\n",
       "      <td>ENCODEHG000058846[1,1,1]</td>\n",
       "      <td>ENCODEHG000058846[1,1,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENCODEHT000206942</td>\n",
       "      <td>ENCODEHT000206942</td>\n",
       "      <td>ENCODEHG000058846[1,1,1]</td>\n",
       "      <td>ENCODEHG000058846[1,1,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENCODEHT000206942</td>\n",
       "      <td>ENCODEHT000206942</td>\n",
       "      <td>ENCODEHG000058846[1,1,1]</td>\n",
       "      <td>ENCODEHG000058846[1,1,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENCODEHT000206867</td>\n",
       "      <td>ENCODEHT000206867</td>\n",
       "      <td>ENCODEHG000058837[2,4,1]</td>\n",
       "      <td>ENCODEHG000058837[2,4,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1181901</th>\n",
       "      <td>ENST00000447588.1</td>\n",
       "      <td>LINC00265-3P-201</td>\n",
       "      <td>ENSG00000233944[1,1,1]</td>\n",
       "      <td>LINC00265-3P[1,1,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1181902</th>\n",
       "      <td>ENCODEHT005023774</td>\n",
       "      <td>ENCODEHT005023774</td>\n",
       "      <td>ENCODEHG000377496[1,1,1]</td>\n",
       "      <td>ENCODEHG000377496[1,1,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1181903</th>\n",
       "      <td>ENCODEHT005023774</td>\n",
       "      <td>ENCODEHT005023774</td>\n",
       "      <td>ENCODEHG000377496[1,1,1]</td>\n",
       "      <td>ENCODEHG000377496[1,1,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1181904</th>\n",
       "      <td>ENCODEHT005023774</td>\n",
       "      <td>ENCODEHT005023774</td>\n",
       "      <td>ENCODEHG000377496[1,1,1]</td>\n",
       "      <td>ENCODEHG000377496[1,1,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1181905</th>\n",
       "      <td>ENCODEHT005023774</td>\n",
       "      <td>ENCODEHT005023774</td>\n",
       "      <td>ENCODEHG000377496[1,1,1]</td>\n",
       "      <td>ENCODEHG000377496[1,1,1]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1181906 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        original_transcript_id original_transcript_name  \\\n",
       "0            ENCODEHT000206942        ENCODEHT000206942   \n",
       "1            ENCODEHT000206942        ENCODEHT000206942   \n",
       "2            ENCODEHT000206942        ENCODEHT000206942   \n",
       "3            ENCODEHT000206942        ENCODEHT000206942   \n",
       "4            ENCODEHT000206867        ENCODEHT000206867   \n",
       "...                        ...                      ...   \n",
       "1181901      ENST00000447588.1         LINC00265-3P-201   \n",
       "1181902      ENCODEHT005023774        ENCODEHT005023774   \n",
       "1181903      ENCODEHT005023774        ENCODEHT005023774   \n",
       "1181904      ENCODEHT005023774        ENCODEHT005023774   \n",
       "1181905      ENCODEHT005023774        ENCODEHT005023774   \n",
       "\n",
       "                    transcript_id           transcript_name  \n",
       "0        ENCODEHG000058846[1,1,1]  ENCODEHG000058846[1,1,1]  \n",
       "1        ENCODEHG000058846[1,1,1]  ENCODEHG000058846[1,1,1]  \n",
       "2        ENCODEHG000058846[1,1,1]  ENCODEHG000058846[1,1,1]  \n",
       "3        ENCODEHG000058846[1,1,1]  ENCODEHG000058846[1,1,1]  \n",
       "4        ENCODEHG000058837[2,4,1]  ENCODEHG000058837[2,4,1]  \n",
       "...                           ...                       ...  \n",
       "1181901    ENSG00000233944[1,1,1]       LINC00265-3P[1,1,1]  \n",
       "1181902  ENCODEHG000377496[1,1,1]  ENCODEHG000377496[1,1,1]  \n",
       "1181903  ENCODEHG000377496[1,1,1]  ENCODEHG000377496[1,1,1]  \n",
       "1181904  ENCODEHG000377496[1,1,1]  ENCODEHG000377496[1,1,1]  \n",
       "1181905  ENCODEHG000377496[1,1,1]  ENCODEHG000377496[1,1,1]  \n",
       "\n",
       "[1181906 rows x 4 columns]"
      ]
     },
     "execution_count": 182,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test[['original_transcript_id', 'original_transcript_name', 'transcript_id', 'transcript_name']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "d748e55d-2caf-4853-8f9b-9686889f2bb3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def map_gtf_transcripts(df, m_df):\n",
    "    \"\"\"\n",
    "    Add cerberus transcript id to each entry in gtf\n",
    "    \"\"\"\n",
    "    # groupby transcripts that are the same\n",
    "    gb_cols = ['gene_name', 'gene_id', 'transcript_triplet',\n",
    "               'transcript_id', 'transcript_name']\n",
    "    temp = m_df[['transcript_id',\n",
    "                 'original_transcript_id',\n",
    "                 'original_transcript_name']].copy(deep=True)\n",
    "    m_df = m_df.groupby(gb_cols).agg({'original_transcript_id': ','.join,\n",
    "                                      'original_transcript_name': ','.join}).reset_index()\n",
    "    m_df = m_df.merge(temp, on='transcript_id', suffixes=('','_merge'))\n",
    "    m_df.drop(['gene_name', 'gene_id', 'transcript_triplet'],\n",
    "              axis=1, inplace=True)\n",
    "\n",
    "    # add new transcript ids\n",
    "    df = df.merge(m_df, left_on=['transcript_id', 'transcript_name'],\n",
    "                  right_on=['original_transcript_id_merge',\n",
    "                            'original_transcript_name_merge'],\n",
    "                 suffixes=('_x', ''))\n",
    "\n",
    "    # drop old tids\n",
    "    df.drop(['transcript_id_x', 'transcript_name_x',\n",
    "             'original_transcript_name_merge'],\n",
    "            axis=1, inplace=True)\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "id": "8e0ab4fa-925a-46c3-8e1f-13f7dc1d1ac1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# tests for map_gtf_ids\n",
    "# - transcript that does not have duplicate\n",
    "# - transcript that does have a duplicate\n",
    "\n",
    "def make_exon_df(n,c,e,s,g,t):\n",
    "    df = pd.DataFrame()\n",
    "    df['Chromosome'] = c\n",
    "    df['Start'] = [i[0] for i in e]\n",
    "    df['End'] = [i[1] for i in e]\n",
    "    df['Strand'] = s\n",
    "    df['Feature'] = 'exon'\n",
    "    cols = ['gene_name', 'gene_id']\n",
    "    for c in cols:\n",
    "        df[c] = g\n",
    "    cols = ['transcript_id', 'transcript_name']\n",
    "    for c in cols:\n",
    "        df[c] = t\n",
    "\n",
    "    # reorder exons and starts/ stops if needed\n",
    "    df['new_Start'] = df[['Start', 'End']].min(axis=1)\n",
    "    df['new_End'] = df[['Start', 'End']].max(axis=1)\n",
    "    df.drop(['Start', 'End'], axis=1, inplace=True)\n",
    "    df.rename({'new_Start':'Start',\n",
    "               'new_End':'End'}, axis=1, inplace=True)\n",
    "    s = s[0]\n",
    "    if s == '+':\n",
    "        ascending = True\n",
    "    elif s == '-':\n",
    "        ascending = False\n",
    "    df.sort_values(by='Start', ascending=ascending, inplace=True)\n",
    "    return df\n",
    "\n",
    "def make_hier_entry(df, how='t'):\n",
    "    \"\"\"\n",
    "    kind {'g','t'}\n",
    "    \"\"\"\n",
    "    agg_dict = {'min_coord': 'min', 'max_coord': 'max'}\n",
    "    t_df = df.copy(deep=True)\n",
    "    t_df['min_coord'] = t_df[['Start', 'End']].min(axis=1)\n",
    "    t_df['max_coord'] = t_df[['Start', 'End']].max(axis=1)\n",
    "    if how == 't':\n",
    "        gb_cols = ['Chromosome', 'Strand', 'gene_name',\n",
    "                   'gene_id', 'transcript_id', 'transcript_name']\n",
    "    elif how == 'g':\n",
    "        gb_cols = ['Chromosome', 'Strand', 'gene_name',\n",
    "                   'gene_id']\n",
    "        \n",
    "    cols = gb_cols + ['min_coord', 'max_coord']\n",
    "    t_df = t_df[cols]\n",
    "    t_df = t_df.groupby(gb_cols).agg(agg_dict).reset_index()\n",
    "    t_df.rename({'min_coord': 'Start', 'max_coord': 'End'}, axis=1, inplace=True)\n",
    "    if how == 't':\n",
    "        t_df['Feature'] = 'transcript'\n",
    "    elif how == 'g':\n",
    "        t_df['Feature'] = 'gene'\n",
    "        \n",
    "    return t_df\n",
    "\n",
    "def make_test_gtf(ts):\n",
    "    df = pd.concat(ts)\n",
    "        # make transcript entries\n",
    "    t_df = make_hier_entry(df, how='t')\n",
    "    # make gene entries\n",
    "    g_df = make_hier_entry(df, how='g')\n",
    "\n",
    "    # concat everything and sort by gene id, transcript id, feature rank (gene =0, t =1, exon=2), then start coords\n",
    "    df = pd.concat([df, t_df, g_df])\n",
    "    df['feature_rank'] = df.Feature.map({'gene':0, 'transcript':1, 'exon':2})\n",
    "\n",
    "    rev = df.loc[df.Strand == '-'].copy(deep=True)\n",
    "    fwd = df.loc[df.Strand == '+'].copy(deep=True)\n",
    "\n",
    "    df = pd.DataFrame()\n",
    "    for temp in [fwd, rev]:\n",
    "        if len(temp.index) > 0:\n",
    "            strand = temp.Strand.values.tolist()[0]\n",
    "            if strand == '+':\n",
    "                ascending = True\n",
    "            elif strand == '-':\n",
    "                ascending = False\n",
    "            temp.sort_values(by=['gene_id', 'transcript_id', 'feature_rank', 'Start'], \n",
    "                             ascending=[True, True, True, ascending],\n",
    "                             na_position='first', inplace=True)\n",
    "            \n",
    "            df = pd.concat([df, temp])\n",
    "    # df.drop('feature_rank', axis=1, inplace=True)\n",
    "    return df\n",
    "\n",
    "ts = []\n",
    "# t1 - transcript that doesn't need to be merged\n",
    "n = 3\n",
    "c = ['1' for i in range(n)]\n",
    "e = [[1,10], [14,20], [25,30]]\n",
    "s = ['+' for i in range(n)]\n",
    "g = 'g1'\n",
    "t = 'g1_t1'\n",
    "ts.append(make_exon_df(n,c,e,s,g,t))\n",
    "\n",
    "# t2 - rev. strand transcript that needs to be merged\n",
    "n = 3\n",
    "c = ['1' for i in range(n)]\n",
    "e = [[90,60], [45,30], [10,8]]\n",
    "s = ['-' for i in range(n)]\n",
    "g = 'g2'\n",
    "t = 'g2_t1'\n",
    "ts.append(make_exon_df(n,c,e,s,g,t))\n",
    "\n",
    "# t3 - rev. strand transcript that needs to be merged\n",
    "n = 3\n",
    "c = ['1' for i in range(n)]\n",
    "e = [[95,60], [45,30], [10,6]]\n",
    "s = ['-' for i in range(n)]\n",
    "g = 'g2'\n",
    "t = 'g2_t2'\n",
    "ts.append(make_exon_df(n,c,e,s,g,t))\n",
    "\n",
    "test_df = make_test_gtf(ts)\n",
    "\n",
    "# map file\n",
    "otid = ['g1_t1', 'g2_t1', 'g2_t2',]\n",
    "otname = [o+'n' for o in otid]\n",
    "tid = ['g1[1,1,1]', 'g2[1,1,1]', 'g2[1,1,1]']\n",
    "tname = [o.split('[')[0]+'n['+o.split('[')[1] for o in tid]\n",
    "gid = ['g1', 'g2', 'g2']\n",
    "gname = ['g1n', 'g2n', 'g2n']\n",
    "m_df = pd.DataFrame()\n",
    "m_df['original_transcript_id'] = otid\n",
    "m_df['original_transcript_name'] = otname\n",
    "m_df['transcript_id'] = tid\n",
    "m_df['transcript_name'] = tname\n",
    "m_df['gene_name'] = gname\n",
    "m_df['transcript_triplet'] = m_df.transcript_id.str.slice(2)\n",
    "m_df['gene_id'] = gid\n",
    "\n",
    "# control\n",
    "otid = ['g1_t1', 'g2_t2']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "id": "d0044c70-1a2a-41bd-ad01-21da1137b860",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>original_transcript_id</th>\n",
       "      <th>original_transcript_name</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_triplet</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>g1_t1</td>\n",
       "      <td>g1_t1n</td>\n",
       "      <td>g1[1,1,1]</td>\n",
       "      <td>g1n[1,1,1]</td>\n",
       "      <td>g1n</td>\n",
       "      <td>[1,1,1]</td>\n",
       "      <td>g1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>g2_t1</td>\n",
       "      <td>g2_t1n</td>\n",
       "      <td>g2[1,1,1]</td>\n",
       "      <td>g2n[1,1,1]</td>\n",
       "      <td>g2n</td>\n",
       "      <td>[1,1,1]</td>\n",
       "      <td>g2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>g2_t2</td>\n",
       "      <td>g2_t2n</td>\n",
       "      <td>g2[1,1,1]</td>\n",
       "      <td>g2n[1,1,1]</td>\n",
       "      <td>g2n</td>\n",
       "      <td>[1,1,1]</td>\n",
       "      <td>g2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  original_transcript_id original_transcript_name transcript_id  \\\n",
       "0                  g1_t1                   g1_t1n     g1[1,1,1]   \n",
       "1                  g2_t1                   g2_t1n     g2[1,1,1]   \n",
       "2                  g2_t2                   g2_t2n     g2[1,1,1]   \n",
       "\n",
       "  transcript_name gene_name transcript_triplet gene_id  \n",
       "0      g1n[1,1,1]       g1n            [1,1,1]      g1  \n",
       "1      g2n[1,1,1]       g2n            [1,1,1]      g2  \n",
       "2      g2n[1,1,1]       g2n            [1,1,1]      g2  "
      ]
     },
     "execution_count": 209,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "m_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "id": "6d54891e-6260-4a1b-9782-3836b53b1005",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>feature_rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>gene</td>\n",
       "      <td>g1</td>\n",
       "      <td>g1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>transcript</td>\n",
       "      <td>g1</td>\n",
       "      <td>g1</td>\n",
       "      <td>g1_t1</td>\n",
       "      <td>g1_t1</td>\n",
       "      <td>1</td>\n",
       "      <td>30</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>g1</td>\n",
       "      <td>g1</td>\n",
       "      <td>g1_t1</td>\n",
       "      <td>g1_t1</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>g1</td>\n",
       "      <td>g1</td>\n",
       "      <td>g1_t1</td>\n",
       "      <td>g1_t1</td>\n",
       "      <td>14</td>\n",
       "      <td>20</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>g1</td>\n",
       "      <td>g1</td>\n",
       "      <td>g1_t1</td>\n",
       "      <td>g1_t1</td>\n",
       "      <td>25</td>\n",
       "      <td>30</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>gene</td>\n",
       "      <td>g2</td>\n",
       "      <td>g2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6</td>\n",
       "      <td>95</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>transcript</td>\n",
       "      <td>g2</td>\n",
       "      <td>g2</td>\n",
       "      <td>g2_t1</td>\n",
       "      <td>g2_t1</td>\n",
       "      <td>8</td>\n",
       "      <td>90</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>g2</td>\n",
       "      <td>g2</td>\n",
       "      <td>g2_t1</td>\n",
       "      <td>g2_t1</td>\n",
       "      <td>60</td>\n",
       "      <td>90</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>g2</td>\n",
       "      <td>g2</td>\n",
       "      <td>g2_t1</td>\n",
       "      <td>g2_t1</td>\n",
       "      <td>30</td>\n",
       "      <td>45</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>g2</td>\n",
       "      <td>g2</td>\n",
       "      <td>g2_t1</td>\n",
       "      <td>g2_t1</td>\n",
       "      <td>8</td>\n",
       "      <td>10</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>transcript</td>\n",
       "      <td>g2</td>\n",
       "      <td>g2</td>\n",
       "      <td>g2_t2</td>\n",
       "      <td>g2_t2</td>\n",
       "      <td>6</td>\n",
       "      <td>95</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>g2</td>\n",
       "      <td>g2</td>\n",
       "      <td>g2_t2</td>\n",
       "      <td>g2_t2</td>\n",
       "      <td>60</td>\n",
       "      <td>95</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>g2</td>\n",
       "      <td>g2</td>\n",
       "      <td>g2_t2</td>\n",
       "      <td>g2_t2</td>\n",
       "      <td>30</td>\n",
       "      <td>45</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>g2</td>\n",
       "      <td>g2</td>\n",
       "      <td>g2_t2</td>\n",
       "      <td>g2_t2</td>\n",
       "      <td>6</td>\n",
       "      <td>10</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome Strand     Feature gene_name gene_id transcript_id  \\\n",
       "0          1      +        gene        g1      g1           NaN   \n",
       "0          1      +  transcript        g1      g1         g1_t1   \n",
       "0          1      +        exon        g1      g1         g1_t1   \n",
       "1          1      +        exon        g1      g1         g1_t1   \n",
       "2          1      +        exon        g1      g1         g1_t1   \n",
       "1          1      -        gene        g2      g2           NaN   \n",
       "1          1      -  transcript        g2      g2         g2_t1   \n",
       "0          1      -        exon        g2      g2         g2_t1   \n",
       "1          1      -        exon        g2      g2         g2_t1   \n",
       "2          1      -        exon        g2      g2         g2_t1   \n",
       "2          1      -  transcript        g2      g2         g2_t2   \n",
       "0          1      -        exon        g2      g2         g2_t2   \n",
       "1          1      -        exon        g2      g2         g2_t2   \n",
       "2          1      -        exon        g2      g2         g2_t2   \n",
       "\n",
       "  transcript_name  Start  End  feature_rank  \n",
       "0             NaN      1   30             0  \n",
       "0           g1_t1      1   30             1  \n",
       "0           g1_t1      1   10             2  \n",
       "1           g1_t1     14   20             2  \n",
       "2           g1_t1     25   30             2  \n",
       "1             NaN      6   95             0  \n",
       "1           g2_t1      8   90             1  \n",
       "0           g2_t1     60   90             2  \n",
       "1           g2_t1     30   45             2  \n",
       "2           g2_t1      8   10             2  \n",
       "2           g2_t2      6   95             1  \n",
       "0           g2_t2     60   95             2  \n",
       "1           g2_t2     30   45             2  \n",
       "2           g2_t2      6   10             2  "
      ]
     },
     "execution_count": 210,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "id": "61e4c6a4-7e68-429c-9813-862c9e5eafb8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Chromosome         object\n",
       "Strand             object\n",
       "Feature            object\n",
       "gene_name          object\n",
       "gene_id            object\n",
       "transcript_id      object\n",
       "transcript_name    object\n",
       "Start               int64\n",
       "End                 int64\n",
       "feature_rank        int64\n",
       "dtype: object"
      ]
     },
     "execution_count": 163,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_df.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61f503ed-de47-4d30-94b6-9cfb9b500f14",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # tests for update_gtf_ends()\n",
    "# def make_gtf(c,e,s,g,t):\n",
    "\n",
    "\n",
    "# # t1 - transcript that doesn't need to be merged\n",
    "# n = 3\n",
    "# c = ['1' for i in range(n)]\n",
    "# e [[1,10], [14,20], [25,30]]\n",
    "# s = ['+' for i in range(n)]\n",
    "# g = 'g1'\n",
    "# t = 'g1_t1'\n",
    "\n",
    "# # t2 - rev. strand transcript that needs ends to be updated\n",
    "# n = 3\n",
    "# c = ['1' for i in range(n)]\n",
    "# e = [[90,60], [45,30], [10,8]]\n",
    "# s = ['-' for i in range(n)]\n",
    "# g = 'g2'\n",
    "# t = 'g1_t1'\n",
    "\n",
    "# df = pd.DataFrame()\n",
    "# df['Chromosome'] = c\n",
    "# df['Start'] = [i[0] for i in e]\n",
    "# df['End'] = [i[1] for i in e]\n",
    "# df['Strand'] = s\n",
    "# cols = ['gene_name', 'gene_id']\n",
    "# for c in cols:\n",
    "#     df[c] = g\n",
    "# cols = []\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
