{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7e4387d2-7683-4c42-93b4-5776e7107cac",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import h5py \n",
    "import os\n",
    "\n",
    "from cerberus.cerberus import *\n",
    "from cerberus.main import *\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "05080a6f-d916-40a1-9a21-4316d12c3169",
   "metadata": {},
   "source": [
    "## annotate transcriptome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1cc99e2-f14f-4fe3-aeba-c2d88eae0d4f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "8ebf6196-fe8a-48b9-9844-93450c4d35a8",
   "metadata": {},
   "source": [
    "## wrapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a9e7dfe2-47c2-4f50-b8cb-baa159d89a45",
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_files(files, sources):\n",
    "    \"\"\"\n",
    "    Check if a series of files exists and whether their \n",
    "    associated sources are all unique. Raise an exception\n",
    "    if not.\n",
    "    \n",
    "    Parameters:\n",
    "        files (list of str): List of file paths \n",
    "        sources (list of str): List of sources\n",
    "    \"\"\"\n",
    "    \n",
    "    for f in files:\n",
    "        if not os.path.exists(f):\n",
    "            raise Exception('File {} does not exist'.format(bed))\n",
    "\n",
    "    if len(set(sources)) < len(sources):\n",
    "        raise Exception('Sources must be unique')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "a5908498-9bf6-4402-972a-e5402eae6d46",
   "metadata": {},
   "outputs": [],
   "source": [
    "# def gen_reference(ref_gtf, o, ref_tss, ref_tes,\n",
    "#                   gtf_tss_dist, gtf_tss_slack,\n",
    "#                   gtf_tes_dist, gtf_tes_slack,\n",
    "#                   tss_slack, tes_slack,\n",
    "#                   verbosity, tmp_dir,\n",
    "#                   keep_tmp):\n",
    "    \n",
    "#     # parse config files and check if the input files exist\n",
    "#     gtfs, gtf_add_ends, gtf_sources = parse_gtf_config(ref_gtf)\n",
    "#     tss_beds, tss_add_ends, tss_sources = parse_agg_ends_config(ref_tss)\n",
    "#     tes_beds, tes_add_ends, tes_sources = parse_agg_ends_config(ref_tes)\n",
    "    \n",
    "#     check_files(gtfs+tss_beds, gtf_sources+tss_sources)\n",
    "#     check_files(gtfs+tes_beds, gtf_sources+tss_sources)\n",
    "    \n",
    "#     # get ends / ics from gtfs\n",
    "#     gtf_tss_beds = []\n",
    "#     gtf_tes_beds = []\n",
    "#     ics = []\n",
    "#     for gtf, source in zip(gtfs, gtf_sources):\n",
    "        \n",
    "#         if verbosity > 0:\n",
    "#             print('Finding TSSs from {}'.format(source))\n",
    "            \n",
    "#         tss_fname = '{}/{}_tss.bed'.format(tmp_dir, source)\n",
    "#         gtf_to_bed(gtf, 'tss', tss_fname, gtf_tss_dist, gtf_tss_slack)\n",
    "\n",
    "#         if verbosity > 0:\n",
    "#             print('Finding TSSs from {}'.format(source))\n",
    "            \n",
    "#         tes_fname = '{}/{}_tes.bed'.format(tmp_dir, source)\n",
    "#         gtf_to_bed(gtf, 'tes', tes_fname, gtf_tes_dist, gtf_tes_slack) \n",
    "\n",
    "#         if verbosity > 0:\n",
    "#             print('Finding intron chains from {}'.format(source))\n",
    "        \n",
    "#         ic_fname = '{}/{}_ic.tsv'.format(tmp_dir, source)\n",
    "#         ic = gtf_to_ics(gtf, ic_fname)\n",
    "        \n",
    "#         gtf_tss_beds.append(tss_fname)\n",
    "#         gtf_tes_beds.append(tes_fname)\n",
    "#         ics.append(ic_fname)\n",
    "    \n",
    "#     # get list of tmp files to (maybe) remove later\n",
    "#     tmp_files = gtf_tss_beds + gtf_tes_beds + ics\n",
    "    \n",
    "#     # aggregate ends\n",
    "#     tss_beds = gtf_tss_beds + tss_beds\n",
    "#     tss_add_ends = gtf_add_ends + tss_add_ends\n",
    "#     tss_sources = gtf_sources + tss_sources\n",
    "    \n",
    "#     if verbosity > 0:\n",
    "#         print('Aggregating TSSs')\n",
    "    \n",
    "#     tss = aggregate_ends(tss_beds, \n",
    "#                          tss_sources,\n",
    "#                          tss_add_ends, \n",
    "#                          tss_slack,\n",
    "#                          'tss')\n",
    "    \n",
    "#     tes_beds = gtf_tes_beds + tes_beds\n",
    "#     tes_add_ends = gtf_add_ends + tes_add_ends\n",
    "#     tes_sources = gtf_sources + tes_sources\n",
    "    \n",
    "#     if verbosity > 0:\n",
    "#         print('Aggregating TESs')\n",
    "    \n",
    "#     tes = aggregate_ends(tes_beds,\n",
    "#                          tes_sources, \n",
    "#                          tes_add_ends,\n",
    "#                          tes_slack,\n",
    "#                          'tes')\n",
    "    \n",
    "#     # aggregate ics\n",
    "#     if verbosity > 0:\n",
    "#         print('Aggregating intron chains')\n",
    "        \n",
    "#     ics = aggregate_ics(ics, gtf_sources)\n",
    "    \n",
    "#     # write h5df\n",
    "#     write_h5(ics, tss, tes, o)\n",
    "    \n",
    "#     # delete tmp files\n",
    "#     if not keep_tmp:\n",
    "#         for f in tmp_files:\n",
    "#             os.remove(f)\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "fb3968cf-c98d-441c-bd62-9c0423dff108",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a5f9e8b0-c2d2-4526-8346-1e67c95017e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_gtf = '/Users/fairliereese/mortazavi_lab/data/rnawg/lr_bulk/cerberus/gtf_config.csv'\n",
    "ref_tss = '/Users/fairliereese/mortazavi_lab/data/rnawg/lr_bulk/cerberus/tss_no_gtf.csv'\n",
    "ref_tes = None\n",
    "gtf_tss_slack = 50\n",
    "gtf_tss_dist = 50\n",
    "gtf_tes_dist = 50\n",
    "gtf_tes_slack = 50\n",
    "tss_slack = 20\n",
    "tes_slack = 20\n",
    "tmp_dir = '/Users/fairliereese/mortazavi_lab/data/rnawg/lr_bulk/cerberus/'\n",
    "o = '/Users/fairliereese/mortazavi_lab/data/rnawg/lr_bulk/cerberus/bepis.h5'\n",
    "\n",
    "gen_reference(ref_gtf, ref_tss, ref_tes,\n",
    "                  gtf_tss_dist, gtf_tss_slack,\n",
    "                  gtf_tes_dist, gtf_tes_slack,\n",
    "                  tss_slack, tes_slack,\n",
    "                  tmp_dir,\n",
    "                  o)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "940142ee-0378-4e4b-a474-7325db2f4df8",
   "metadata": {},
   "outputs": [],
   "source": [
    "ic, tss, tes, _ = read_h5(o)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "71b4453a-812c-4a48-b8ac-e96df729ff48",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+--------------+-----------+-----------+--------------+---------------------+------------+\n",
      "| Chromosome   |     Start |       End | Strand       | Name                | source     |\n",
      "| (category)   |   (int32) |   (int32) | (category)   | (object)            | (object)   |\n",
      "|--------------+-----------+-----------+--------------+---------------------+------------|\n",
      "| SIRV1        |     10597 |     10698 | +            | ENCODEHG000058846_1 | talon      |\n",
      "| SIRV1        |     10315 |     10416 | -            | ENCODEHG000058837_1 | talon      |\n",
      "| SIRV1        |     10739 |     10840 | -            | ENCODEHG000058837_2 | talon      |\n",
      "| SIRV2        |      4713 |      4814 | -            | ENCODEHG000058784_1 | talon      |\n",
      "| SIRV2        |      5855 |      5956 | -            | ENCODEHG000058784_2 | talon      |\n",
      "| SIRV3        |      1930 |      2031 | +            | ENCODEHG000058844_1 | talon      |\n",
      "| SIRV3        |      3953 |      4054 | +            | ENCODEHG000058844_2 | talon      |\n",
      "| SIRV3        |      9862 |      9963 | -            | ENCODEHG000058850_1 | talon      |\n",
      "+--------------+-----------+-----------+--------------+---------------------+------------+\n",
      "Stranded PyRanges object has 8 rows and 6 columns from 3 chromosomes.\n",
      "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
      "['talon' 'v39,v29,ccre' 'v39,v29,talon,ccre' 'v39,talon,ccre'\n",
      " 'v29,talon,ccre' 'talon,ccre' 'v39,ccre' 'v29,ccre' 'v29' 'v29,talon']\n",
      "+--------------+-----------+-----------+--------------+---------------------+------------+\n",
      "| Chromosome   |     Start |       End | Strand       | Name                | source     |\n",
      "| (category)   |   (int32) |   (int32) | (category)   | (object)            | (object)   |\n",
      "|--------------+-----------+-----------+--------------+---------------------+------------|\n",
      "| SIRV1        |     11555 |     11656 | +            | ENCODEHG000058846_1 | talon      |\n",
      "| SIRV1        |       949 |      1050 | -            | ENCODEHG000058837_1 | talon      |\n",
      "| SIRV2        |       945 |      1046 | -            | ENCODEHG000058784_1 | talon      |\n",
      "| SIRV2        |      3615 |      3716 | -            | ENCODEHG000058784_2 | talon      |\n",
      "| SIRV3        |      6729 |      6830 | +            | ENCODEHG000058844_1 | talon      |\n",
      "| SIRV3        |      7623 |      7724 | +            | ENCODEHG000058844_2 | talon      |\n",
      "| SIRV3        |      8786 |      8987 | +            | ENCODEHG000058844_3 | talon      |\n",
      "| SIRV3        |      8786 |      8887 | -            | ENCODEHG000058850_1 | talon      |\n",
      "+--------------+-----------+-----------+--------------+---------------------+------------+\n",
      "Stranded PyRanges object has 8 rows and 6 columns from 3 chromosomes.\n",
      "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
      "['talon' 'v39,v29' 'v39,v29,talon' 'v29' 'v39' 'v29,talon' 'v39,talon']\n",
      "  Chromosome Strand                                        Coordinates  \\\n",
      "0       chr1      +  100038316-100049908-100050004-100058665-100058...   \n",
      "1       chr1      +            100058728-100059877-100060005-100061834   \n",
      "2       chr1      +                                          100099490   \n",
      "3       chr1      +                    1001263-1008193-1008279-1013983   \n",
      "4       chr1      +                    1001281-1008193-1008279-1013983   \n",
      "\n",
      "                Name         source  \n",
      "0  ENSG00000156875_1  v39,v29,talon  \n",
      "1  ENSG00000156875_2        v39,v29  \n",
      "2  ENSG00000226952_1  v39,v29,talon  \n",
      "3  ENSG00000187608_3  v39,v29,talon  \n",
      "4  ENSG00000187608_2  v39,v29,talon  \n",
      "['v39,v29,talon' 'v39,v29' 'v39' 'v39,talon' 'v29' 'v29,talon' 'talon']\n"
     ]
    }
   ],
   "source": [
    "print(tss.head())\n",
    "print(tss.source.unique())\n",
    "print(tes.head())\n",
    "print(tes.source.unique())\n",
    "print(ic.head())\n",
    "print(ic.source.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "b31774c5-3e66-40dc-a03d-e9823d0f8fce",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "c8f1fba8-ef66-4670-a54e-9154cd682ccd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  Chromosome Strand                                        Coordinates  \\\n",
      "0       chr1      +  100038316-100049908-100050004-100058665-100058...   \n",
      "1       chr1      +            100058728-100059877-100060005-100061834   \n",
      "2       chr1      +                                          100099490   \n",
      "3       chr1      +                    1001263-1008193-1008279-1013983   \n",
      "4       chr1      +                    1001281-1008193-1008279-1013983   \n",
      "\n",
      "                Name         source  \n",
      "0  ENSG00000156875_1  v39,v29,talon  \n",
      "1  ENSG00000156875_2        v39,v29  \n",
      "2  ENSG00000226952_1  v39,v29,talon  \n",
      "3  ENSG00000187608_3  v39,v29,talon  \n",
      "4  ENSG00000187608_2  v39,v29,talon  \n",
      "  Chromosome  Start    End Strand                 Name source\n",
      "1      SIRV1  10315  10416      -  ENCODEHG000058837_1  talon\n",
      "0      SIRV1  10597  10698      +  ENCODEHG000058846_1  talon\n",
      "2      SIRV1  10739  10840      -  ENCODEHG000058837_2  talon\n",
      "3      SIRV2   4713   4814      -  ENCODEHG000058784_1  talon\n",
      "4      SIRV2   5855   5956      -  ENCODEHG000058784_2  talon\n",
      "      Chromosome  Start    End Strand                 Name source\n",
      "17350      SIRV1    949   1050      -  ENCODEHG000058837_1  talon\n",
      "17349      SIRV1  11555  11656      +  ENCODEHG000058846_1  talon\n",
      "17351      SIRV2    945   1046      -  ENCODEHG000058784_1  talon\n",
      "17352      SIRV2   3615   3716      -  ENCODEHG000058784_2  talon\n",
      "17353      SIRV3   6729   6830      +  ENCODEHG000058844_1  talon\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "fcd9a9ea-d345-41e5-8dc5-042b18423116",
   "metadata": {},
   "outputs": [],
   "source": [
    "# gtfs\n",
    "v39_gtf = '/Users/fairliereese/mortazavi_lab/ref/gencode.v39/gencode.v39.annotation.gtf'\n",
    "v29_gtf = '/Users/fairliereese/mortazavi_lab/ref/gencode.v29/gencode.v29.annotation.gtf'\n",
    "talon_gtf = '/Users/fairliereese/mortazavi_lab/data/rnawg/lr_bulk/talon/human_known_nic_nnc_talon.gtf'\n",
    "\n",
    "gtfs = [v39_gtf, v29_gtf, talon_gtf]\n",
    "add_ends = [True, True, True]\n",
    "sources = ['v39', 'v29', 'talon']\n",
    "\n",
    "fname = '/Users/fairliereese/mortazavi_lab/data/rnawg/lr_bulk/cerberus/gtf_config.csv'\n",
    "df = pd.DataFrame()\n",
    "\n",
    "df['fname'] = gtfs\n",
    "df['add_ends'] = add_ends\n",
    "df['source'] = sources\n",
    "\n",
    "df.to_csv(fname, sep=',', header=None, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "974fbadb-23a7-4558-a568-acc81ab917ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "# tss \n",
    "ccre_bed = '/Users/fairliereese/mortazavi_lab/data/rnawg/lr_bulk/cerberus/test_no_gid_no_strand.bed'\n",
    "sources = ['ccre']\n",
    "add_ends = [True]\n",
    "\n",
    "fname = '/Users/fairliereese/mortazavi_lab/data/rnawg/lr_bulk/cerberus/tss_no_gtf.csv'\n",
    "df = pd.DataFrame()\n",
    "\n",
    "df['fname'] = [ccre_bed]\n",
    "df['add_ends'] = add_ends\n",
    "df['source'] = sources\n",
    "\n",
    "df.to_csv(fname, sep=',', header=None, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee01f286-e361-476a-9f16-24e907edb3f9",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
