{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7e4387d2-7683-4c42-93b4-5776e7107cac",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "ecc814db-392c-4c13-aed7-bde1fcaa1cf1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_transcript_ref(fname):\n",
    "    \"\"\"\n",
    "    Get transcripts with necessary tags to order based on basic set etc.\n",
    "\n",
    "    Parameters:\n",
    "        fname (str): Path to GTF file\n",
    "\n",
    "    Returns:\n",
    "        df (pandas DataFrame): DataFrame of only transcript entries\n",
    "            with tags parsed out\n",
    "    \"\"\"\n",
    "    df = pr.read_gtf(fname, duplicate_attr=True).df\n",
    "    df = df.loc[df.Feature == 'transcript']\n",
    "    df['MANE_Select'] = df.tag.str.contains('MANE_Select')\n",
    "    df['basic_set'] = df.tag.str.contains('basic')\n",
    "    df['temp'] = df.tag.str.split('appris_principal_', n=1, expand=True)[1]\n",
    "    df['appris_principal'] = df.temp.str.split(',', n=1, expand=True)[0]\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "6ce57d9e-d288-4b76-b4e9-30dd2498734b",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf = 'tests/files/Canx.gtf' \n",
    "mode = 'tss'\n",
    "dist = 50\n",
    "slack = 50"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "id": "3ac7567e-c276-4b96-b4d4-ca62b599bcee",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_ends_from_gtf(gtf, mode, dist, slack):\n",
    "    \"\"\"\n",
    "    Create bed regions for each end in a gtf and number them \n",
    "    based on tags indicating priority for each gene w/i the gtf\n",
    "\n",
    "    Parameters:\n",
    "        gtf (str): Path to gtf\n",
    "        dist (int): Distance by which to extend regions on either side\n",
    "        slack (int): Distance allowable for merging nearby regions\n",
    "\n",
    "    Returns:\n",
    "        bed (pyranges PyRanges): PyRanges object containing extended regions\n",
    "    \"\"\"\n",
    "    gr_gtf = pr.read_gtf(gtf)\n",
    "    if 'gene_id' not in gr_gtf.columns:\n",
    "        raise ValueError('No gene_id field found in {}'.format(gtf))\n",
    "        \n",
    "    # get and extend ends\n",
    "    if mode == 'tss':\n",
    "        bed = gr_gtf.features.tss().extend(dist).df[[\n",
    "            'Chromosome', 'Start', 'End', 'Strand', 'gene_id',\n",
    "        ]].drop_duplicates()\n",
    "    elif mode == 'tes':\n",
    "        bed = gr_gtf.features.tes().extend(dist).df[[\n",
    "            'Chromosome', 'Start', 'End', 'Strand', 'gene_id',\n",
    "        ]].drop_duplicates()\n",
    "\n",
    "    bed = pr.PyRanges(bed.rename({'ThickStart': 'gene_id'}, axis=1))\n",
    "    bed = bed.merge(strand=True,\n",
    "                by='gene_id',\n",
    "                slack=slack)\n",
    "    \n",
    "    bed = number_gtf_ends(bed, gtf, mode)\n",
    "\n",
    "    return bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "id": "6a099ec5-8948-467c-ab58-7c248e28b275",
   "metadata": {},
   "outputs": [],
   "source": [
    "def number_gtf_ends(bed, gtf, mode):\n",
    "    \"\"\"\n",
    "    As a part of calling ends from a gtf, use the tags and gene id\n",
    "    in the gtf to assign each region a name / number\n",
    "    \n",
    "    Parameters:\n",
    "        bed (pyranges PyRanges): Output of get_ends_from_gtf\n",
    "        gtf (str): Path to gtf file\n",
    "        mode (str): {'tss', 'tes'}\n",
    "    \n",
    "    Returns: \n",
    "        bed (pyranges PyRanges): Bed file with Name field \n",
    "            consisting of stable gene id_end region number\n",
    "    \"\"\"\n",
    "    \n",
    "    # read gtf with extra tag info\n",
    "    gr_gtf = get_transcript_ref(gtf)\n",
    "    gr_gtf = pr.PyRanges(gr_gtf)\n",
    "    \n",
    "    # add a temporary unique identifier for each end\n",
    "    bed = bed.df\n",
    "    bed[mode] = [i for i in range(len(bed.index))]\n",
    "    bed = pr.PyRanges(bed)\n",
    "\n",
    "    # get ends from reference gtf and join with called ends\n",
    "    cols = ['Chromosome', 'Start', 'End', 'Strand',\n",
    "             'gene_id', 'transcript_id', 'MANE_Select',\n",
    "             'basic_set', 'appris_principal']\n",
    "    if mode == 'tss':\n",
    "        temp = pr.PyRanges(gr_gtf.features.tss().df[cols])\n",
    "    elif mode == 'tes':\n",
    "        temp = pr.PyRanges(gr_gtf.features.tes().df[cols])\n",
    "    bed = bed.join(temp,\n",
    "               strandedness='same',\n",
    "               how='left',\n",
    "               slack=0)\n",
    "    bed = bed.df\n",
    "    bed = bed.loc[bed.gene_id == bed.gene_id_b]\n",
    "    cols = ['Start_b', 'End_b', 'Strand_b', 'gene_id_b']\n",
    "    bed.drop(cols, axis=1, inplace=True)\n",
    "    \n",
    "    # use the tags from the gtf to assign an actual end id\n",
    "    bed = number_tss_ic_tes(bed, mode=mode)\n",
    "    bed['gene_id'] = bed.gene_id.str.split(pat='.', n=1, expand=True)[0]\n",
    "    c = '{}_num'.format(mode)\n",
    "    bed['Name'] = bed.gene_id+'_'+bed[c].astype(str)\n",
    "    cols = [mode, 'transcript_id', 'MANE_Select',\n",
    "            'basic_set', 'appris_principal', c, 'gene_id']\n",
    "    bed.drop(cols, axis=1, inplace=True)\n",
    "    bed = pr.PyRanges(bed)\n",
    "    \n",
    "    return bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "id": "e1a20e8a-748b-465d-bf27-497549c4ad3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def number_tss_ic_tes(df, mode):\n",
    "    \"\"\"\n",
    "    Assign a number to each tss, intron chain, or tes per gene based on the\n",
    "    status of the transcript\n",
    "\n",
    "    Parameters:\n",
    "        df (pandas DataFrame): DataFrame derived from PyRanges\n",
    "            with gene id, tss/ic/tes id, and tags to order transcripts\n",
    "            with\n",
    "        mode (str): {'tss', 'ic', 'tes'}\n",
    "\n",
    "    Returns:\n",
    "        df (pandas DataFrame): DataFrame with an identifier for each unique\n",
    "            tss/ic/tes per gene\n",
    "    \"\"\"\n",
    "    # groupby feature but record which feature\n",
    "    # each transcript id uses\n",
    "    cols = ['transcript_id', 'Chromosome', 'Strand',\n",
    "            'Start', 'End',\n",
    "            mode, 'basic_set', 'MANE_Select',\n",
    "            'appris_principal', 'gene_id']\n",
    "    df = df[cols].groupby(['Chromosome', 'Strand',\n",
    "                           'Start', 'End',\n",
    "                           mode, 'gene_id'],\n",
    "                           observed=True).agg({'transcript_id': ','.join,\n",
    "                                     'MANE_Select': 'max',\n",
    "                                     'basic_set': 'max',\n",
    "                                     'appris_principal': 'min'}).reset_index()\n",
    "\n",
    "    # compute feature number based on tags \n",
    "    df['{}_num'.format(mode)] = df.sort_values(by=['gene_id', 'MANE_Select',\n",
    "                                                   'appris_principal', 'basic_set'],\n",
    "                                 ascending=[True, False, True, False],\n",
    "                                 na_position='last')\\\n",
    "                                 .groupby(['gene_id'])\\\n",
    "                                 .cumcount() + 1\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "id": "366a969d-d90f-4efa-90e0-a553feed0bb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "bed = get_ends_from_gtf(gtf, 'tss', dist, slack)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "053530cb-a377-4db4-9936-c561dd1ef617",
   "metadata": {},
   "source": [
    "## Getting ICs from GTF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "id": "caf17756-b855-4f77-a71a-87493efdc9f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_ic(gtf_pr):\n",
    "    \"\"\"\n",
    "    Get a hyphen-separated representation of each transcript's intron chain\n",
    "    from a PyRanges GTF\n",
    "\n",
    "    Parameters:\n",
    "        gtf_pr (pyranges PyRanges): GTF PyRanges object\n",
    "\n",
    "    Returns:\n",
    "        df (pandas DataFrame): DataFrame detailing intron chain, gene, strand,\n",
    "            chromosome, and transcript that intron chain was seen in\n",
    "    \"\"\"\n",
    "    df = gtf_pr.df.copy(deep=True)\n",
    "\n",
    "    # restrict to exon entries\n",
    "    df = df.loc[df.Feature == 'exon']\n",
    "    cols = ['Chromosome', 'Strand', 'Start', 'End', 'transcript_id', 'gene_id']\n",
    "    df = df[cols]\n",
    "\n",
    "    # melt to isolate individual coordinates\n",
    "    df = pd.melt(df, id_vars=['Chromosome', 'Strand', 'transcript_id', 'gene_id'],\n",
    "                value_vars=['Start', 'End'],\n",
    "                value_name='Coord')\n",
    "    df.drop('variable', axis=1, inplace=True)\n",
    "\n",
    "    # sort to order coordinates correctly\n",
    "    df.Coord = df.Coord.astype(int)\n",
    "    fwd = df.loc[df.Strand == '+'].copy(deep=True)\n",
    "    rev = df.loc[df.Strand == '-'].copy(deep=True)\n",
    "\n",
    "    fwd.sort_values(by=['Chromosome', 'transcript_id', 'Coord'],\n",
    "                    ascending=[True, True, True], inplace=True)\n",
    "    rev.sort_values(by=['Chromosome', 'transcript_id', 'Coord'],\n",
    "                    ascending=[True, True, False], inplace=True)\n",
    "    df = pd.concat([fwd, rev])\n",
    "\n",
    "    # create intron chain strings\n",
    "    df.Coord = df.Coord.astype(str)\n",
    "    df = df.groupby(['Chromosome', 'Strand',\n",
    "                     'transcript_id', 'gene_id'], observed=True)['Coord'].apply('-'.join).reset_index()\n",
    "\n",
    "    # remove tss and tes from intron chain\n",
    "    df['temp'] = df.Coord.str.split('-', n=1, expand=True)[1]\n",
    "    df['ic'] = df.temp.str.rsplit('-', n=1, expand=True)[0]\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "id": "2424d5b4-5c1f-4e88-84b1-b0552683cec5",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "\"['Start', 'End'] not in index\"",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-171-154b2f4c6367>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     20\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     21\u001b[0m \u001b[0;31m# add number for each unique intron chain\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 22\u001b[0;31m \u001b[0mdf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnumber_tss_ic_tes\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmode\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'ic'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     23\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     24\u001b[0m \u001b[0;31m# # make coords into tuple and perform additional\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-156-7d77ffd2a2af>\u001b[0m in \u001b[0;36mnumber_tss_ic_tes\u001b[0;34m(df, mode)\u001b[0m\n\u001b[1;32m     20\u001b[0m             \u001b[0mmode\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'basic_set'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'MANE_Select'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     21\u001b[0m             'appris_principal', 'gene_id']\n\u001b[0;32m---> 22\u001b[0;31m     df = df[cols].groupby(['Chromosome', 'Strand',\n\u001b[0m\u001b[1;32m     23\u001b[0m                            \u001b[0;34m'Start'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'End'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     24\u001b[0m                            mode, 'gene_id'],\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   2906\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mis_iterator\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2907\u001b[0m                 \u001b[0mkey\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2908\u001b[0;31m             \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_listlike_indexer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mraise_missing\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2909\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2910\u001b[0m         \u001b[0;31m# take() does not accept boolean indexers\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m_get_listlike_indexer\u001b[0;34m(self, key, axis, raise_missing)\u001b[0m\n\u001b[1;32m   1252\u001b[0m             \u001b[0mkeyarr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindexer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnew_indexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0max\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_reindex_non_unique\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkeyarr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1253\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1254\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_validate_read_indexer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkeyarr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindexer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mraise_missing\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mraise_missing\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1255\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mkeyarr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindexer\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1256\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m_validate_read_indexer\u001b[0;34m(self, key, indexer, axis, raise_missing)\u001b[0m\n\u001b[1;32m   1302\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mraise_missing\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1303\u001b[0m                 \u001b[0mnot_found\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m-\u001b[0m \u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0max\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1304\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"{not_found} not in index\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1305\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1306\u001b[0m             \u001b[0;31m# we skip the warning on Categorical\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: \"['Start', 'End'] not in index\""
     ]
    }
   ],
   "source": [
    "# def get_ics_from_gtf(gtf):\n",
    "#     \"\"\"\n",
    "#     Get a file for each intron chain in a gtf and number them \n",
    "#     based on tags indicating priority for each gene w/i the gtf \n",
    "    \n",
    "#     Parameters:\n",
    "#         gtf (str): Filename for input gtf\n",
    "#     \"\"\"\n",
    "    \n",
    "# get basic status and appris_principal tag for each transcript\n",
    "t_df = get_transcript_ref(gtf)\n",
    "t_df = t_df[['transcript_id', 'MANE_Select', 'basic_set', 'appris_principal']]\n",
    "\n",
    "# get unique intron chains from gtf\n",
    "df = pr.read_gtf(gtf)\n",
    "df = get_ic(df)\n",
    "\n",
    "# add basic annotation, appris principal number, and gene id\n",
    "df = df.merge(t_df, on='transcript_id', how='left')\n",
    "\n",
    "# add number for each unique intron chain\n",
    "df = number_tss_ic_tes(df, mode='ic')\n",
    "\n",
    "# # make coords into tuple and perform additional\n",
    "# # formatting for this table\n",
    "# df['ic'] = df.ic.str.split('-')\n",
    "# df['ic'] = [tuple(c) for c in df.ic.tolist()]\n",
    "# ic = df.copy(deep=True)\n",
    "# ic.rename({'ic': 'coordinates',\n",
    "#            'Chromosome': 'chrom',\n",
    "#            'Strand': 'strand'},\n",
    "#            axis=1, inplace=True)\n",
    "# ic['ic_id'] = ic['gene_id']+'_'+ic.ic_num.astype(str)\n",
    "# ic.drop(['MANE_Select', 'basic_set', 'appris_principal',\n",
    "#          'gene_id', 'ic_num'],\n",
    "#         axis=1, inplace=True)\n",
    "?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "id": "013a4a59-3d18-48b5-b70a-5dc3f244f351",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>Coord</th>\n",
       "      <th>temp</th>\n",
       "      <th>ic</th>\n",
       "      <th>MANE_Select</th>\n",
       "      <th>basic_set</th>\n",
       "      <th>appris_principal</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr11</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000020637.8</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>50325673-50325492-50311784-50311603-50310867-5...</td>\n",
       "      <td>50325492-50311784-50311603-50310867-50310793-5...</td>\n",
       "      <td>50325492-50311784-50311603-50310867-50310793-5...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr11</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000146979.1</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>50299378-50299128-50298236-50298116-50297394-5...</td>\n",
       "      <td>50299128-50298236-50298116-50297394-50297273-5...</td>\n",
       "      <td>50299128-50298236-50298116-50297394-50297273-5...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr11</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000153068.1</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>50298319-50298116-50297394-50297363</td>\n",
       "      <td>50298116-50297394-50297363</td>\n",
       "      <td>50298116-50297394</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr11</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000155801.1</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>50300990-50300897-50299344-50298837</td>\n",
       "      <td>50300897-50299344-50298837</td>\n",
       "      <td>50300897-50299344</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr11</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000155801.2</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>50300990-50300897-50299344-50298837</td>\n",
       "      <td>50300897-50299344-50298837</td>\n",
       "      <td>50300897-50299344</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome Strand         transcript_id                gene_id  \\\n",
       "0      chr11      -  ENSMUST00000020637.8  ENSMUSG00000020368.14   \n",
       "1      chr11      -  ENSMUST00000146979.1  ENSMUSG00000020368.14   \n",
       "2      chr11      -  ENSMUST00000153068.1  ENSMUSG00000020368.14   \n",
       "3      chr11      -  ENSMUST00000155801.1  ENSMUSG00000020368.14   \n",
       "4      chr11      -  ENSMUST00000155801.2  ENSMUSG00000020368.14   \n",
       "\n",
       "                                               Coord  \\\n",
       "0  50325673-50325492-50311784-50311603-50310867-5...   \n",
       "1  50299378-50299128-50298236-50298116-50297394-5...   \n",
       "2                50298319-50298116-50297394-50297363   \n",
       "3                50300990-50300897-50299344-50298837   \n",
       "4                50300990-50300897-50299344-50298837   \n",
       "\n",
       "                                                temp  \\\n",
       "0  50325492-50311784-50311603-50310867-50310793-5...   \n",
       "1  50299128-50298236-50298116-50297394-50297273-5...   \n",
       "2                         50298116-50297394-50297363   \n",
       "3                         50300897-50299344-50298837   \n",
       "4                         50300897-50299344-50298837   \n",
       "\n",
       "                                                  ic MANE_Select basic_set  \\\n",
       "0  50325492-50311784-50311603-50310867-50310793-5...       False      True   \n",
       "1  50299128-50298236-50298116-50297394-50297273-5...         NaN       NaN   \n",
       "2                                  50298116-50297394         NaN       NaN   \n",
       "3                                  50300897-50299344       False     False   \n",
       "4                                  50300897-50299344       False      True   \n",
       "\n",
       "  appris_principal  \n",
       "0                1  \n",
       "1              NaN  \n",
       "2              NaN  \n",
       "3                2  \n",
       "4                1  "
      ]
     },
     "execution_count": 170,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64a3afed-7ef1-4b28-887c-be4a787400cd",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
