{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "7e4387d2-7683-4c42-93b4-5776e7107cac",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import h5py \n",
    "import os\n",
    "\n",
    "from cerberus.cerberus import *\n",
    "from cerberus.main import *\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "bf6bd877-d9b4-40a8-80ab-7cbb190450d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def merge_ics(df, ic):\n",
    "    \"\"\"\n",
    "    Assign each transcript in df to an intron chain in ic\n",
    "    \n",
    "    Parameters:\n",
    "        df (pandas DataFrame): DataFrame w/ intron chain for each transcript\n",
    "        ic (pandas DataFrame): DataFrame from cerberus reference with intron chains\n",
    "        \n",
    "    Returns:\n",
    "        df (pandas DataFrame): \n",
    "    \"\"\"\n",
    "    # merge on intron chain, strand, chromosome, and gene id\n",
    "    df = df.merge(ic, how='left',\n",
    "                  on=['Chromosome', 'Strand',\n",
    "                      'Coordinates', 'gene_id'])\n",
    "\n",
    "    # formatting\n",
    "    df.rename({'Name': 'ic_id'}, axis=1, inplace=True)\n",
    "    df = df[['transcript_id', 'ic', 'ic_id']]\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "b2ee2704-0314-4cfd-8dc3-9e4c8ef43817",
   "metadata": {},
   "outputs": [],
   "source": [
    "def merge_ends(ends, ref, mode):\n",
    "    \"\"\"\n",
    "    Merge ends from a GTF with those already annotated in a cerberus reference\n",
    "    \n",
    "    Parameters:\n",
    "        ends (pyranges PyRanges): PyRanges object of ends from GTF annotation\n",
    "        ref (pyranges PyRanges): PyRanges object of reference ends from cerberus\n",
    "            annotation\n",
    "        mode (str): {'tss', 'tes'}\n",
    "        \n",
    "    Returns:\n",
    "        df (pandas DataFrame): DataFrame detailing which transcript uses which \n",
    "            end from the cerberus reference\n",
    "    \"\"\"\n",
    "    \n",
    "    \n",
    "    # limit to relevant columns\n",
    "    ends = ends[['Chromosome', 'Start', 'End', 'Strand',\n",
    "                 'gene_id', 'transcript_id']]\n",
    "\n",
    "    # get only the relevant columns and deduplicate\n",
    "    ends = ends.df\n",
    "    t_ends = ends.copy(deep=True)\n",
    "    ends.drop('transcript_id', axis=1, inplace=True)\n",
    "    ends.drop_duplicates(inplace=True)\n",
    "    ends = pr.PyRanges(ends)\n",
    "\n",
    "    # find closest interval in ref\n",
    "    ends = ends.nearest(ref,\n",
    "                        strandedness=None)\n",
    "\n",
    "    # fix the ends with mismatching gene ids - this part can be slow :(\n",
    "    ends = ends.df\n",
    "    fix_ends = ends.loc[ends.gene_id != ends.gene_id_b]\n",
    "    fix_ends = fix_ends[['Chromosome', 'Start', 'End', 'Strand',\n",
    "                         'gene_id']]\n",
    "    # print('ends that need to be fixed')\n",
    "    # print(fix_ends)\n",
    "    ends = ends.loc[ends.gene_id == ends.gene_id_b]\n",
    "    # print('okie dokie ends')\n",
    "    # print(ends)\n",
    "\n",
    "    for i, gid in enumerate(fix_ends.gene_id.unique().tolist()):\n",
    "        gene_ends = fix_ends.loc[fix_ends.gene_id == gid].copy(deep=True)\n",
    "        gene_ends = pr.PyRanges(gene_ends)\n",
    "        gene_refs = ref.df.loc[ref.df.gene_id == gid].copy(deep=True)\n",
    "        gene_refs = pr.PyRanges(gene_refs)\n",
    "        gene_ends = gene_ends.nearest(gene_refs,\n",
    "                                      strandedness=None)\n",
    "        gene_ends = gene_ends.df\n",
    "        ends = pd.concat([ends, gene_ends])\n",
    "\n",
    "        if i % 100 == 0:\n",
    "            print('Processed {} / {} genes'.format(i, len(fix_ends.gene_id.unique().tolist())))\n",
    "\n",
    "    # merge back in to get transcript ids\n",
    "    t_ends = t_ends.merge(ends, how='left',\n",
    "                      on=['Chromosome', 'Start', 'End', 'Strand', 'gene_id'])\n",
    "\n",
    "    # formatting\n",
    "    t_ends.rename({'Name': '{}_id'.format(mode)}, axis=1, inplace=True)`\n",
    "    t_ends = t_ends[['transcript_id', '{}_id'.format(mode), mode]]\n",
    "\n",
    "    return t_ends"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "05080a6f-d916-40a1-9a21-4316d12c3169",
   "metadata": {},
   "source": [
    "## convert transcriptome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "0fdbe97b-ad51-4bf2-85b0-d0446ed6ddd1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processed 0 / 2 genes\n",
      "  Chromosome  Start  End Strand gene_id transcript_id  Start_b  End_b  \\\n",
      "0          1      1    2      +   gene1            t1        1     40   \n",
      "1          1    100  101      +   gene2            t3       90     98   \n",
      "2          1    200  201      +   gene1            t2      150    160   \n",
      "3          1    300  301      +   gene3            t4      200    250   \n",
      "\n",
      "  Strand_b   tss_id source gene_id_b tss  Distance  \n",
      "0        +  gene1_1   test     gene1   1         0  \n",
      "1        +  gene2_1   test     gene2   1         3  \n",
      "2        +  gene1_2   test     gene1   2        41  \n",
      "3        +  gene3_1   test     gene3   1        51  \n",
      "\n",
      "test\n",
      "  transcript_id   tss_id tss\n",
      "0            t1  gene1_1   1\n",
      "1            t2  gene1_2   2\n",
      "2            t3  gene2_1   1\n",
      "3            t4  gene3_1   1\n",
      "RangeIndex(start=0, stop=4, step=1)\n",
      "transcript_id    object\n",
      "tss_id           object\n",
      "tss              object\n",
      "dtype: object\n",
      "ctrl\n",
      "  transcript_id   tss_id tss\n",
      "0            t1  gene1_1   1\n",
      "1            t2  gene1_2   2\n",
      "2            t3  gene2_1   1\n",
      "3            t4  gene3_1   1\n",
      "RangeIndex(start=0, stop=4, step=1)\n",
      "transcript_id    object\n",
      "tss_id           object\n",
      "tss              object\n",
      "dtype: object\n",
      "Processed 0 / 2 genes\n",
      "  Chromosome  Start  End Strand gene_id transcript_id  Start_b  End_b  \\\n",
      "0          1      1    2      +   gene1            t1        1     40   \n",
      "1          1    100  101      +   gene2            t3       90     98   \n",
      "2          1    200  201      +   gene1            t2      150    160   \n",
      "3          1    300  301      +   gene3            t4      200    250   \n",
      "\n",
      "  Strand_b   tes_id source gene_id_b tes  Distance  \n",
      "0        +  gene1_1   test     gene1   1         0  \n",
      "1        +  gene2_1   test     gene2   1         3  \n",
      "2        +  gene1_2   test     gene1   2        41  \n",
      "3        +  gene3_1   test     gene3   1        51  \n",
      "\n",
      "test\n",
      "  transcript_id   tes_id tes\n",
      "0            t1  gene1_1   1\n",
      "1            t2  gene1_2   2\n",
      "2            t3  gene2_1   1\n",
      "3            t4  gene3_1   1\n",
      "RangeIndex(start=0, stop=4, step=1)\n",
      "transcript_id    object\n",
      "tes_id           object\n",
      "tes              object\n",
      "dtype: object\n",
      "ctrl\n",
      "  transcript_id   tes_id tes\n",
      "0            t1  gene1_1   1\n",
      "1            t2  gene1_2   2\n",
      "2            t3  gene2_1   1\n",
      "3            t4  gene3_1   1\n",
      "RangeIndex(start=0, stop=4, step=1)\n",
      "transcript_id    object\n",
      "tes_id           object\n",
      "tes              object\n",
      "dtype: object\n"
     ]
    }
   ],
   "source": [
    "def make_end_gtf_df(c,s,st,e,g,t):\n",
    "    df = pd.DataFrame()\n",
    "    cols = ['Chromosome', 'Strand', 'Start', 'End', 'gene_id', 'transcript_id']\n",
    "    var = [c,s,st,e,g,t]\n",
    "    for col, var in zip(cols, var):\n",
    "        if type(var) == list:\n",
    "            df[col] = var\n",
    "\n",
    "    df = format_end_df(df)\n",
    "\n",
    "    return df\n",
    "\n",
    "def make_end_df(c,s,st,e,n, source,mode, add_id=True):\n",
    "    df = pd.DataFrame()\n",
    "    cols = ['Chromosome', 'Strand', 'Start', 'End', 'Name']\n",
    "    var = [c,s,st,e,n]\n",
    "    for col, var in zip(cols, var):\n",
    "        if type(var) == list:\n",
    "            df[col] = var\n",
    "\n",
    "    # add source\n",
    "    df['source'] = source\n",
    "\n",
    "    df = format_end_df(df)\n",
    "\n",
    "    # get end # and gene id\n",
    "    if any(df.Name.isnull()):\n",
    "        df['gene_id'] = np.nan\n",
    "        df[mode] = np.nan\n",
    "    else:\n",
    "        df['gene_id'] = df.Name.str.split('_', expand=True)[0]\n",
    "        df[mode] = df.Name.str.split('_', expand=True)[1]\n",
    "\n",
    "    # get arbitrary unique ids\n",
    "    if add_id:\n",
    "        df['id'] = [i for i in range(len(df.index))]\n",
    "\n",
    "    return df\n",
    "\n",
    "def format_end_df(df):\n",
    "    sort_cols = ['Chromosome', 'Start', 'End', 'Strand']\n",
    "    df = df.sort_values(by=sort_cols)\n",
    "    order = ['Chromosome', 'Start', 'End', 'Strand', 'Name', 'gene_id', 'transcript_id', 'start', 'source']\n",
    "    order = [o for o in order if o in df.columns]\n",
    "    df = df[order]\n",
    "    df.reset_index(drop=True, inplace=True)\n",
    "    return df\n",
    "\n",
    "def test_merge_ends(print_dfs=True):\n",
    "    # cases where\n",
    "    # - ends overlap existing ends as well as gene id\n",
    "    # - ends do not overlap existing ends and we need to choose closest\n",
    "    # - ends overlap existing ends but gene ids don't match \n",
    "    # - ends overlap but are not on same strand\n",
    "    # - equidistant matches?\n",
    "\n",
    "    for mode in ['tss', 'tes']:\n",
    "\n",
    "        # annot\n",
    "        n = 4\n",
    "        c = ['1' for i in range(n)]\n",
    "        s = ['+' for i in range(n)]\n",
    "        st = [1, 200, 100, 300]\n",
    "        e = [2, 201, 101, 301]\n",
    "        g = ['gene1', 'gene1', 'gene2', 'gene3']\n",
    "        t = ['t1', 't2', 't3', 't4']\n",
    "        annot_df = make_end_gtf_df(c,s,st,e,g,t)\n",
    "        annot_df = pr.PyRanges(annot_df)\n",
    "\n",
    "        # ref \n",
    "        n = 5\n",
    "        c = ['1' for i in range(n)]\n",
    "        s = ['+' for i in range(n)]\n",
    "        s[-1] = '-'\n",
    "        st = [1, 150, 90, 200, 290]\n",
    "        e = [40, 160, 98, 250, 340]\n",
    "        n = ['gene1_1', # entry in annot intersects\n",
    "             'gene1_2', # entry in annot needs to choose closest\n",
    "             'gene2_1', # entry in annot matches but wrong gene\n",
    "             'gene3_1', #\n",
    "             'gene4_1'] # entry overlaps on the wrong strand\n",
    "        source = 'test'\n",
    "        ref_df = make_end_df(c,s,st,e,n, source, mode, add_id=False)\n",
    "        ref_df = pr.PyRanges(ref_df)\n",
    "\n",
    "        # ctrl\n",
    "        t = ['t1', 't2', 't3', 't4']\n",
    "        i = ['gene1_1', 'gene1_2', 'gene2_1', 'gene3_1']\n",
    "        ends = ['1', '2', '1', '1']\n",
    "        ctrl = pd.DataFrame()\n",
    "        ctrl['transcript_id'] = t\n",
    "        ctrl['{}_id'.format(mode)] = i\n",
    "        ctrl[mode] = ends\n",
    "\n",
    "        test = merge_ends(annot_df, ref_df, mode)\n",
    "\n",
    "        def order_df(df):\n",
    "            df.sort_values(by='transcript_id', inplace=True)\n",
    "            df.reset_index(drop=True, inplace=True)\n",
    "            return df\n",
    "\n",
    "        ctrl = order_df(ctrl)\n",
    "        test = order_df(test)\n",
    "\n",
    "        if print_dfs:\n",
    "            print('test')\n",
    "            print(test)\n",
    "            print(test.index)\n",
    "            print(test.dtypes)\n",
    "            print('ctrl')\n",
    "            print(ctrl)\n",
    "            print(ctrl.index)\n",
    "            print(ctrl.dtypes)\n",
    "\n",
    "        pd.testing.assert_frame_equal(ctrl, test, check_like=True)\n",
    "\n",
    "        assert len(ctrl.index) == len(test.index)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "e1ea4c37-b798-4626-b2f6-1c07f1ca31db",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>+</td>\n",
       "      <td>gene1</td>\n",
       "      <td>t1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>101</td>\n",
       "      <td>+</td>\n",
       "      <td>gene2</td>\n",
       "      <td>t3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>201</td>\n",
       "      <td>+</td>\n",
       "      <td>gene1</td>\n",
       "      <td>t2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>300</td>\n",
       "      <td>301</td>\n",
       "      <td>+</td>\n",
       "      <td>gene3</td>\n",
       "      <td>t4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+-----------------+\n",
       "|   Chromosome |     Start |       End | Strand       | gene_id    | transcript_id   |\n",
       "|   (category) |   (int32) |   (int32) | (category)   | (object)   | (object)        |\n",
       "|--------------+-----------+-----------+--------------+------------+-----------------|\n",
       "|            1 |         1 |         2 | +            | gene1      | t1              |\n",
       "|            1 |       100 |       101 | +            | gene2      | t3              |\n",
       "|            1 |       200 |       201 | +            | gene1      | t2              |\n",
       "|            1 |       300 |       301 | +            | gene3      | t4              |\n",
       "+--------------+-----------+-----------+--------------+------------+-----------------+\n",
       "Stranded PyRanges object has 4 rows and 6 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annot_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "c2b80e7d-0b14-4149-a5f6-e19b805b51d5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Name</th>\n",
       "      <th>source</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>tss</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>40</td>\n",
       "      <td>+</td>\n",
       "      <td>gene1_1</td>\n",
       "      <td>test</td>\n",
       "      <td>gene1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>90</td>\n",
       "      <td>98</td>\n",
       "      <td>+</td>\n",
       "      <td>gene2_1</td>\n",
       "      <td>test</td>\n",
       "      <td>gene2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>150</td>\n",
       "      <td>160</td>\n",
       "      <td>+</td>\n",
       "      <td>gene1_2</td>\n",
       "      <td>test</td>\n",
       "      <td>gene1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>250</td>\n",
       "      <td>+</td>\n",
       "      <td>gene3_1</td>\n",
       "      <td>test</td>\n",
       "      <td>gene3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>290</td>\n",
       "      <td>340</td>\n",
       "      <td>-</td>\n",
       "      <td>gene4_1</td>\n",
       "      <td>test</td>\n",
       "      <td>gene4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+------------+------------+------------+\n",
       "|   Chromosome |     Start |       End | Strand       | Name       | source     | gene_id    |        tss |\n",
       "|   (category) |   (int32) |   (int32) | (category)   | (object)   | (object)   | (object)   |   (object) |\n",
       "|--------------+-----------+-----------+--------------+------------+------------+------------+------------|\n",
       "|            1 |         1 |        40 | +            | gene1_1    | test       | gene1      |          1 |\n",
       "|            1 |        90 |        98 | +            | gene2_1    | test       | gene2      |          1 |\n",
       "|            1 |       150 |       160 | +            | gene1_2    | test       | gene1      |          2 |\n",
       "|            1 |       200 |       250 | +            | gene3_1    | test       | gene3      |          1 |\n",
       "|            1 |       290 |       340 | -            | gene4_1    | test       | gene4      |          1 |\n",
       "+--------------+-----------+-----------+--------------+------------+------------+------------+------------+\n",
       "Stranded PyRanges object has 5 rows and 8 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "70e01c4c-5aed-4597-8d14-993a6af1f5b5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>tss_id</th>\n",
       "      <th>tss</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>t1</td>\n",
       "      <td>gene1_1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>t3</td>\n",
       "      <td>gene2_1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>t2</td>\n",
       "      <td>gene1_2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>t4</td>\n",
       "      <td>gene3_1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  transcript_id   tss_id tss\n",
       "0            t1  gene1_1   1\n",
       "1            t3  gene2_1   1\n",
       "2            t2  gene1_2   2\n",
       "3            t4  gene3_1   1"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "2950502c-7779-48dc-a929-fe244e82bb4f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>tss_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>t1</td>\n",
       "      <td>gene1_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>t2</td>\n",
       "      <td>gene1_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>t3</td>\n",
       "      <td>gene2_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>t4</td>\n",
       "      <td>gene3_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  transcript_id   tss_id\n",
       "0            t1  gene1_1\n",
       "1            t2  gene1_2\n",
       "2            t3  gene2_1\n",
       "3            t4  gene3_1"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ctrl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "6a01f460-4322-4fc7-9427-b62869a67e4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_ics_gtf_df(c,s,co,g,t):\n",
    "    df = pd.DataFrame()\n",
    "    cols = ['Chromosome', 'Strand', 'Coordinates', 'gene_id', 'transcript_id']\n",
    "    var = [c,s,co,g,t]\n",
    "    for col, var in zip(cols, var):\n",
    "        if type(var) == list:\n",
    "            df[col] = var\n",
    "\n",
    "    df = format_ics_df(df)\n",
    "    \n",
    "    return df   \n",
    "\n",
    "def make_ics_map_df(t, ic, i):\n",
    "    df = pd.DataFrame()\n",
    "    cols = ['transcript_id', 'ic', 'ic_id'] \n",
    "    var = [t, ic, i]\n",
    "    for col, var in zip(cols, var):\n",
    "        df[col] = var\n",
    "    return df\n",
    "\n",
    "def make_ics_df(c,s,co,n, source):\n",
    "    df = pd.DataFrame()\n",
    "    cols = ['Chromosome', 'Strand', 'Coordinates', 'Name']\n",
    "    var = [c,s,co,n]\n",
    "    for col, var in zip(cols, var):\n",
    "        if type(var) == list:\n",
    "            df[col] = var\n",
    "\n",
    "    df = format_ics_df(df)\n",
    "\n",
    "    df['gene_id'] = df.Name.str.split('_', expand=True)[0]\n",
    "    df['ic'] = df.Name.str.split('_', expand=True)[1]\n",
    "    df['source'] = source\n",
    "    df.ic = df.ic.astype(int)\n",
    "\n",
    "    return df\n",
    "\n",
    "def format_ics_df(df):\n",
    "    sort_cols = ['Chromosome', 'Strand', 'Coordinates']\n",
    "    df = df.sort_values(by=sort_cols)\n",
    "    order = ['Chromosome', 'Strand', 'Coordinates', 'Name', 'gene_id', 'transcript_id', 'ic', 'source']\n",
    "    order = [o for o in order if o in df.columns]\n",
    "    df = df[order]\n",
    "    df.reset_index(drop=True, inplace=True)\n",
    "    return df\n",
    "\n",
    "# merge ics\n",
    "def test_merge_ics(print_dfs=True):\n",
    "\n",
    "    # ics to annotate\n",
    "    n = 3\n",
    "    c = ['1' for i in range(n)]\n",
    "    s = ['+' for i in range(n)]\n",
    "    co = ['1-2-3', '1-2-3', '4-5-6']\n",
    "    g = ['gene1', 'gene1', 'gene2']\n",
    "    t = ['t1', 't2', 't3']\n",
    "    annot_df = make_ics_gtf_df(c,s,co,g,t)\n",
    "\n",
    "\n",
    "    # ref ics\n",
    "    n = 3\n",
    "    c = ['1' for i in range(n)]\n",
    "    s = ['+' for i in range(n)]\n",
    "    co = ['1-2-3', '4-5-6', '4-5-6']\n",
    "    n = ['gene1_1', 'gene1_2', 'gene2_1']\n",
    "    source = 'test'\n",
    "    ref_df = make_ics_df(c,s,co,n, source)\n",
    "\n",
    "    # ctrl \n",
    "    t = ['t1', 't2', 't3']\n",
    "    ic = [1, 1, 1]\n",
    "    i = ['gene1_1', 'gene1_1', 'gene2_1']\n",
    "    ctrl = make_ics_map_df(t, ic, i)\n",
    "\n",
    "    test = merge_ics(annot_df, ref_df)\n",
    "\n",
    "    if print_dfs:\n",
    "        print('test')\n",
    "        print(test)\n",
    "        print(test.index)\n",
    "        print(test.dtypes)\n",
    "        print('ctrl')\n",
    "        print(ctrl)\n",
    "        print(ctrl.index)\n",
    "        print(ctrl.dtypes)\n",
    "\n",
    "    pd.testing.assert_frame_equal(ctrl, test, check_like=True)\n",
    "    assert len(ctrl.index) == len(test.index)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "ea0af334-761b-47c0-9862-3f97e0b1ca81",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "test\n",
      "  transcript_id  ic    ic_id\n",
      "0            t1   1  gene1_1\n",
      "1            t2   1  gene1_1\n",
      "2            t3   1  gene2_1\n",
      "Int64Index([0, 1, 2], dtype='int64')\n",
      "transcript_id    object\n",
      "ic                int64\n",
      "ic_id            object\n",
      "dtype: object\n",
      "ctrl\n",
      "  transcript_id  ic    ic_id\n",
      "0            t1   1  gene1_1\n",
      "1            t2   1  gene1_1\n",
      "2            t3   1  gene2_1\n",
      "RangeIndex(start=0, stop=3, step=1)\n",
      "transcript_id    object\n",
      "ic                int64\n",
      "ic_id            object\n",
      "dtype: object\n"
     ]
    }
   ],
   "source": [
    "test_merge_ics()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "678f6b28-289e-4a43-88e4-71ea23b6643c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
