{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7e4387d2-7683-4c42-93b4-5776e7107cac",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "ecc814db-392c-4c13-aed7-bde1fcaa1cf1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_transcript_ref(fname):\n",
    "    \"\"\"\n",
    "    Get transcripts with necessary tags to order based on basic set etc.\n",
    "\n",
    "    Parameters:\n",
    "        fname (str): Path to GTF file\n",
    "\n",
    "    Returns:\n",
    "        df (pandas DataFrame): DataFrame of only transcript entries\n",
    "            with tags parsed out\n",
    "    \"\"\"\n",
    "    df = pr.read_gtf(fname, duplicate_attr=True).df\n",
    "    df = df.loc[df.Feature == 'transcript']\n",
    "    df['MANE_Select'] = df.tag.str.contains('MANE_Select')\n",
    "    df['basic_set'] = df.tag.str.contains('basic')\n",
    "    df['temp'] = df.tag.str.split('appris_principal_', n=1, expand=True)[1]\n",
    "    df['appris_principal'] = df.temp.str.split(',', n=1, expand=True)[0]\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "6ce57d9e-d288-4b76-b4e9-30dd2498734b",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf = 'tests/files/Canx.gtf' \n",
    "mode = 'tss'\n",
    "dist = 50\n",
    "slack = 50"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "id": "3ac7567e-c276-4b96-b4d4-ca62b599bcee",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_ends_from_gtf(gtf, mode, dist, slack):\n",
    "    \"\"\"\n",
    "    Create bed regions for each tes in a gtf and number them \n",
    "    based on tags indicating priority for each gene w/i the gtf\n",
    "\n",
    "    Parameters:\n",
    "        gtf (str): Path to gtf\n",
    "        dist (int): Distance by which to extend regions on either side\n",
    "        slack (int): Distance allowable for merging nearby regions\n",
    "\n",
    "    Returns:\n",
    "        bed (pyranges PyRanges): PyRanges object containing extended regions\n",
    "    \"\"\"\n",
    "    gr_gtf = pr.read_gtf(gtf)\n",
    "    if 'gene_id' not in gr_gtf.columns:\n",
    "        raise ValueError('No gene_id field found in {}'.format(gtf))\n",
    "        \n",
    "    # get and extend ends\n",
    "    if mode == 'tss':\n",
    "        bed = gr_gtf.features.tss().extend(dist).df[[\n",
    "            'Chromosome', 'Start', 'End', 'Strand', 'gene_id',\n",
    "        ]].drop_duplicates()\n",
    "    elif mode == 'tes':\n",
    "        bed = gr_gtf.features.tes().extend(dist).df[[\n",
    "            'Chromosome', 'Start', 'End', 'Strand', 'gene_id',\n",
    "        ]].drop_duplicates()\n",
    "\n",
    "    bed = pr.PyRanges(bed.rename({'ThickStart': 'gene_id'}, axis=1))\n",
    "    bed = bed.merge(strand=True,\n",
    "                by='gene_id',\n",
    "                slack=slack)\n",
    "    \n",
    "    bed = number_gtf_ends(bed, gtf, mode)\n",
    "\n",
    "    return bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "id": "6a099ec5-8948-467c-ab58-7c248e28b275",
   "metadata": {},
   "outputs": [],
   "source": [
    "def number_gtf_ends(bed, gtf, mode):\n",
    "    \"\"\"\n",
    "    As a part of calling ends from a gtf, use the tags and gene id\n",
    "    in the gtf to assign each region a name / number\n",
    "    \n",
    "    Parameters:\n",
    "        bed (pyranges PyRanges): Output of get_ends_from_gtf\n",
    "        gtf (str): Path to gtf file\n",
    "        mode (str): {'tss', 'tes'}\n",
    "    \n",
    "    Returns: \n",
    "        bed (pyranges PyRanges): Bed file with Name field \n",
    "            consisting of stable gene id_end region number\n",
    "    \"\"\"\n",
    "    \n",
    "    # read gtf with extra tag info\n",
    "    gr_gtf = get_transcript_ref(gtf)\n",
    "    gr_gtf = pr.PyRanges(gr_gtf)\n",
    "    \n",
    "    # add a temporary unique identifier for each end\n",
    "    bed = bed.df\n",
    "    bed[mode] = [i for i in range(len(bed.index))]\n",
    "    bed = pr.PyRanges(bed)\n",
    "\n",
    "    # get ends from reference gtf and join with called ends\n",
    "    cols = ['Chromosome', 'Start', 'End', 'Strand',\n",
    "             'gene_id', 'transcript_id', 'MANE_Select',\n",
    "             'basic_set', 'appris_principal']\n",
    "    if mode == 'tss':\n",
    "        temp = pr.PyRanges(gr_gtf.features.tss().df[cols])\n",
    "    elif mode == 'tes':\n",
    "        temp = pr.PyRanges(gr_gtf.features.tes().df[cols])\n",
    "    bed = bed.join(temp,\n",
    "               strandedness='same',\n",
    "               how='left',\n",
    "               slack=0)\n",
    "    bed = bed.df\n",
    "    bed = bed.loc[bed.gene_id == bed.gene_id_b]\n",
    "    cols = ['Start_b', 'End_b', 'Strand_b', 'gene_id_b']\n",
    "    bed.drop(cols, axis=1, inplace=True)\n",
    "    \n",
    "    # use the tags from the gtf to assign an actual end id\n",
    "    bed = number_tss_ic_tes(bed, mode=mode)\n",
    "    bed['gene_id'] = bed.gene_id.str.split(pat='.', n=1, expand=True)[0]\n",
    "    c = '{}_num'.format(mode)\n",
    "    bed['Name'] = bed.gene_id+'_'+bed[c].astype(str)\n",
    "    cols = [mode, 'transcript_id', 'MANE_Select',\n",
    "            'basic_set', 'appris_principal', c, 'gene_id']\n",
    "    bed.drop(cols, axis=1, inplace=True)\n",
    "    bed = pr.PyRanges(bed)\n",
    "    \n",
    "    return bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "id": "e1a20e8a-748b-465d-bf27-497549c4ad3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def number_tss_ic_tes(df, mode):\n",
    "    \"\"\"\n",
    "    Assign a number to each tss, intron chain, or tes per gene based on the\n",
    "    status of the transcript\n",
    "\n",
    "    Parameters:\n",
    "        df (pandas DataFrame): DataFrame derived from PyRanges\n",
    "            with gene id, tss/ic/tes id, and tags to order transcripts\n",
    "            with\n",
    "        mode (str): {'tss', 'ic', 'tes'}\n",
    "\n",
    "    Returns:\n",
    "        df (pandas DataFrame): DataFrame with an identifier for each unique\n",
    "            tss/ic/tes per gene\n",
    "    \"\"\"\n",
    "    # groupby feature but record which feature\n",
    "    # each transcript id uses\n",
    "    cols = ['transcript_id', 'Chromosome', 'Strand',\n",
    "            'Start', 'End',\n",
    "            mode, 'basic_set', 'MANE_Select',\n",
    "            'appris_principal', 'gene_id']\n",
    "    df = df[cols].groupby(['Chromosome', 'Strand',\n",
    "                           'Start', 'End',\n",
    "                           mode, 'gene_id'],\n",
    "                           observed=True).agg({'transcript_id': ','.join,\n",
    "                                     'MANE_Select': 'max',\n",
    "                                     'basic_set': 'max',\n",
    "                                     'appris_principal': 'min'}).reset_index()\n",
    "\n",
    "    # compute feature number based on tags \n",
    "    df['{}_num'.format(mode)] = df.sort_values(by=['gene_id', 'MANE_Select',\n",
    "                                                   'appris_principal', 'basic_set'],\n",
    "                                 ascending=[True, False, True, False],\n",
    "                                 na_position='last')\\\n",
    "                                 .groupby(['gene_id'])\\\n",
    "                                 .cumcount() + 1\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "id": "366a969d-d90f-4efa-90e0-a553feed0bb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "bed = get_ends_from_gtf(gtf, 'tss', dist, slack)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "id": "92ce7f00-8464-4bed-98de-cb76cd99dde1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr11</td>\n",
       "      <td>-</td>\n",
       "      <td>50298268</td>\n",
       "      <td>50298369</td>\n",
       "      <td>ENSMUSG00000020368_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr11</td>\n",
       "      <td>-</td>\n",
       "      <td>50299327</td>\n",
       "      <td>50299428</td>\n",
       "      <td>ENSMUSG00000020368_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr11</td>\n",
       "      <td>-</td>\n",
       "      <td>50300939</td>\n",
       "      <td>50301040</td>\n",
       "      <td>ENSMUSG00000020368_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr11</td>\n",
       "      <td>-</td>\n",
       "      <td>50325622</td>\n",
       "      <td>50325723</td>\n",
       "      <td>ENSMUSG00000020368_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+--------------+-----------+-----------+----------------------+\n",
       "| Chromosome   | Strand       |     Start |       End | Name                 |\n",
       "| (category)   | (category)   |   (int32) |   (int32) | (object)             |\n",
       "|--------------+--------------+-----------+-----------+----------------------|\n",
       "| chr11        | -            |  50298268 |  50298369 | ENSMUSG00000020368_3 |\n",
       "| chr11        | -            |  50299327 |  50299428 | ENSMUSG00000020368_4 |\n",
       "| chr11        | -            |  50300939 |  50301040 | ENSMUSG00000020368_2 |\n",
       "| chr11        | -            |  50325622 |  50325723 | ENSMUSG00000020368_1 |\n",
       "+--------------+--------------+-----------+-----------+----------------------+\n",
       "Stranded PyRanges object has 4 rows and 5 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 164,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bed.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "id": "ecaba561-0edc-4fb3-9200-7a7c381ada07",
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "('PyRanges object has no attribute', 'gene_id')",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-158-0e083dee15ec>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'gene_id'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgene_id\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpat\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'.'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mn\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mexpand\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'Name'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgene_id\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;34m'_'\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtss_num\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mastype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m cols = [mode, 'transcript_id', 'MANE_Select',\n\u001b[1;32m      4\u001b[0m         'basic_set', 'appris_principal', 'tss_num', 'gene_id']\n\u001b[1;32m      5\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdrop\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcols\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0minplace\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pyranges/pyranges.py\u001b[0m in \u001b[0;36m__getattr__\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m    267\u001b[0m         \u001b[0;32mfrom\u001b[0m \u001b[0mpyranges\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmethods\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mattr\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0m_getattr\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    268\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 269\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0m_getattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    270\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    271\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__setattr__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcolumn_name\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcolumn\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pyranges/methods/attr.py\u001b[0m in \u001b[0;36m_getattr\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m     65\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mconcat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mdf\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     66\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 67\u001b[0;31m         \u001b[0;32mraise\u001b[0m \u001b[0mAttributeError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"PyRanges object has no attribute\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m: ('PyRanges object has no attribute', 'gene_id')"
     ]
    }
   ],
   "source": [
    "df['gene_id'] = df.gene_id.str.split(pat='.', n=1, expand=True)[0]\n",
    "df['Name'] = df.gene_id+'_'+df.tss_num.astype(str)\n",
    "cols = [mode, 'transcript_id', 'MANE_Select',\n",
    "        'basic_set', 'appris_principal', 'tss_num', 'gene_id']\n",
    "df.drop(cols, axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "id": "ce873060-b83f-4e8f-9be4-b73a65fe7ebd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr11</td>\n",
       "      <td>-</td>\n",
       "      <td>50298268</td>\n",
       "      <td>50298369</td>\n",
       "      <td>ENSMUSG00000020368_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr11</td>\n",
       "      <td>-</td>\n",
       "      <td>50299327</td>\n",
       "      <td>50299428</td>\n",
       "      <td>ENSMUSG00000020368_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr11</td>\n",
       "      <td>-</td>\n",
       "      <td>50300939</td>\n",
       "      <td>50301040</td>\n",
       "      <td>ENSMUSG00000020368_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr11</td>\n",
       "      <td>-</td>\n",
       "      <td>50325622</td>\n",
       "      <td>50325723</td>\n",
       "      <td>ENSMUSG00000020368_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome Strand     Start       End                  Name\n",
       "0      chr11      -  50298268  50298369  ENSMUSG00000020368_3\n",
       "1      chr11      -  50299327  50299428  ENSMUSG00000020368_4\n",
       "2      chr11      -  50300939  50301040  ENSMUSG00000020368_2\n",
       "3      chr11      -  50325622  50325723  ENSMUSG00000020368_1"
      ]
     },
     "execution_count": 130,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "04f5ebf0-dc39-4e6b-b17a-898f28e07e0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pr.PyRanges(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "08485b25-583c-463a-9cd0-27dce9c5685f",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_bed('test_tss_id.bed')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2424d5b4-5c1f-4e88-84b1-b0552683cec5",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
