{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7e4387d2-7683-4c42-93b4-5776e7107cac",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import h5py \n",
    "import os\n",
    "\n",
    "from cerberus.cerberus import *\n",
    "from cerberus.main import *"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fda3e212-2a3a-45ea-9eb9-877d2a5bee4d",
   "metadata": {},
   "source": [
    "## write h5 ref from input beds and ics tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c97c6fff-29c7-4e14-8d92-37ff668f732e",
   "metadata": {},
   "outputs": [],
   "source": [
    "ic = '/Users/fairliereese/Documents/programming/mortazavi_lab/data/rnawg/lr_bulk/cerberus/temp/talon_ic.tsv'\n",
    "tes = '/Users/fairliereese/Documents/programming/mortazavi_lab/data/rnawg/lr_bulk/cerberus/test_tes.bed'\n",
    "tss = '/Users/fairliereese/Documents/programming/mortazavi_lab/data/rnawg/lr_bulk/cerberus/test_tss.bed'\n",
    "tss_map = '/Users/fairliereese/Documents/programming/mortazavi_lab/data/rnawg/lr_bulk/cerberus/test_tes_source_map.bed'\n",
    "tes_map = '/Users/fairliereese/Documents/programming/mortazavi_lab/data/rnawg/lr_bulk/cerberus/test_tss_source_map.bed'\n",
    "\n",
    "# df = read_ic_ref(ic)\n",
    "# df.head()\n",
    "\n",
    "write_reference(tss, tes, ic, 'test.h5')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "841494bc-6ccb-4391-879a-ec9bc582b4ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df = read_cerberus_ends(tss, mode='tss')\n",
    "# df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1d04fb33-d613-4688-b5ec-10fcf1efdd26",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>source</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>169804296</td>\n",
       "      <td>169804436</td>\n",
       "      <td>+</td>\n",
       "      <td>v40</td>\n",
       "      <td>ENSG00000000460_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>169807786</td>\n",
       "      <td>169807887</td>\n",
       "      <td>+</td>\n",
       "      <td>v40</td>\n",
       "      <td>ENSG00000000460_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>169821668</td>\n",
       "      <td>169821769</td>\n",
       "      <td>+</td>\n",
       "      <td>v40</td>\n",
       "      <td>ENSG00000000460_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>169852986</td>\n",
       "      <td>169853135</td>\n",
       "      <td>+</td>\n",
       "      <td>v40</td>\n",
       "      <td>ENSG00000000460_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>169854029</td>\n",
       "      <td>169854130</td>\n",
       "      <td>+</td>\n",
       "      <td>v40</td>\n",
       "      <td>ENSG00000000460_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome      Start        End Strand source               Name\n",
       "0       chr1  169804296  169804436      +    v40  ENSG00000000460_2\n",
       "1       chr1  169807786  169807887      +    v40  ENSG00000000460_4\n",
       "2       chr1  169821668  169821769      +    v40  ENSG00000000460_5\n",
       "3       chr1  169852986  169853135      +    v40  ENSG00000000460_3\n",
       "4       chr1  169854029  169854130      +    v40  ENSG00000000460_1"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = read_cerberus_source_map(tss_map)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0faa232-0c52-4e30-9591-96406c346324",
   "metadata": {},
   "source": [
    "## modify / collapse transcripts in TALON GTF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "id": "c0b46624-a1a1-43ca-b1dd-0262ef3c1180",
   "metadata": {},
   "outputs": [],
   "source": [
    "# def map_gtf_transcripts(gtf, m_df):\n",
    "#     \"\"\"\n",
    "#     Add cerberus transcript id and ids for the ends used to each entry in gtf\n",
    "    \n",
    "#     Parameters:\n",
    "#         gtf (pandas DataFrame): DF of GTF\n",
    "#         m_df (pandas DataFrame): Map file from cerberus reference matching\n",
    "#             each transcript id in `gtf` to a cerberus transcript id\n",
    "    \n",
    "#     Returns:\n",
    "#         gtf (pandas DataFrame): DF of GTF with cerberus ids added\n",
    "#     \"\"\"\n",
    "#     gtf = df.copy(deep=True)\n",
    "#     gtf = gtf.merge(m_df, how='left', \n",
    "#                     left_on=['transcript_name', 'transcript_id'],\n",
    "#                     right_on=['original_transcript_name', 'original_transcript_id'],\n",
    "#                     suffixes=('', '_cerberus'))\n",
    "\n",
    "#     # gtf.drop(['transcript_id', 'transcript_name'], axis=1, inplace=True)\n",
    "#     # gtf.rename({'transcript_id_cerberus': 'transcript_id',\n",
    "#     #             'transcript_name_cerberus': 'transcript_name'},\n",
    "#     #            axis=1, inplace=True)\n",
    "    \n",
    "#     return gtf\n",
    "\n",
    "def get_stranded_gtf_dfs(df):\n",
    "    \"\"\"\n",
    "    Split a GTF df into fwd and rev strands\n",
    "    \n",
    "    Parameters:\n",
    "        df (pandas DataFrame): DF of gtf\n",
    "    \n",
    "    Returns:\n",
    "        fwd (pandas DataFrame): DF of all forward-stranded entries from GTF\n",
    "        rev (pandas DataFrame): DF of all reverse-stranded entries from GTF\n",
    "    \"\"\"\n",
    "    rev = df.loc[df.Strand == '-'].copy(deep=True)\n",
    "    fwd = df.loc[df.Strand == '+'].copy(deep=True)\n",
    "    \n",
    "    return fwd, rev\n",
    "\n",
    "def sort_gtf(df):\n",
    "    \"\"\"\n",
    "    Sort a GTF into its proper ordering\n",
    "    \n",
    "    Parameters:\n",
    "        df (pandas DataFrame): DF of GTF\n",
    "    \n",
    "    Returns:\n",
    "        df (pandas DataFrame): DF of GTF, sorted\n",
    "    \"\"\"\n",
    "    df['feature_rank'] = df.Feature.map({'gene':0, 'transcript':1, 'exon':2})\n",
    "    df.feature_rank = df.feature_rank.astype(int)    \n",
    "     \n",
    "    fwd, rev = get_stranded_gtf_dfs(df)\n",
    "\n",
    "    df = pd.DataFrame()\n",
    "    for temp in [fwd, rev]:\n",
    "        if len(temp.index) > 0:\n",
    "            strand = temp.Strand.values.tolist()[0]\n",
    "            if strand == '+':\n",
    "                ascending = True\n",
    "            elif strand == '-':\n",
    "                ascending = False\n",
    "            temp.sort_values(by=['gene_id', 'transcript_id', 'feature_rank', 'Start'], \n",
    "                             ascending=[True, True, True, ascending],\n",
    "                             na_position='first', inplace=True)\n",
    "            \n",
    "            df = pd.concat([df, temp])\n",
    "    # print(temp[['gene_id', 'transcript_id', 'feature_rank', 'Start']].dtypes)\n",
    "    # tids = ['ENST00000380874.3', 'ENST00000645831.1', 'ENST00000647651.1', 'ENST00000379198.4', 'ENST00000373200.5', 'ENST00000644024.1']\n",
    "    # if 'original_transcript_id' in df.columns:\n",
    "    #     test = df.loc[df.original_transcript_id.isin(tids)].copy(deep=True)\n",
    "    # else:\n",
    "    #     test = df.loc[df.transcript_id.isin(tids)].copy(deep=True)\n",
    "    # print(test[['Feature', 'gene_id', 'transcript_id', 'feature_rank', 'Start']])\n",
    "    df.drop('feature_rank', axis=1, inplace=True)\n",
    "    return df\n",
    "\n",
    "def get_update_ends_settings(strand, mode):\n",
    "    \"\"\"\n",
    "    Returns which columns to refer to and which min/max function\n",
    "    to use depending on looking at forward / rev strand or \n",
    "    tss / tes\n",
    "    \n",
    "    Parameters:\n",
    "        strand (str): {'+', '-'}\n",
    "        mode (str): {'tss', 'tes'}\n",
    "    \n",
    "    Returns:\n",
    "        old_end (str): Name of column to modify; {'Start', 'End'}\n",
    "        new_end (str): Name of column to pull new value from; {'Start_end', 'End_end'}\n",
    "        gene_func (str): What function to apply to new_end; {'min', 'max'}\n",
    "    \"\"\"\n",
    "    if mode == 'tss': \n",
    "        if strand == '+':\n",
    "            old_end = 'Start'\n",
    "            new_end = 'Start_end'\n",
    "            gene_func = 'min'\n",
    "        elif strand == '-':\n",
    "            old_end = 'End'\n",
    "            new_end = 'End_end'\n",
    "            gene_func = 'max'\n",
    "    elif mode == 'tes':\n",
    "        if strand == '+':\n",
    "            old_end = 'End'\n",
    "            new_end = 'End_end'\n",
    "            gene_func = 'max'\n",
    "        elif strand == '-':\n",
    "            old_end = 'Start'\n",
    "            new_end  = 'Start_end'\n",
    "            gene_func = 'min'\n",
    "            \n",
    "    return old_end, new_end, gene_func\n",
    "\n",
    "def update_transcript_ends(df, mode, strand):\n",
    "    \"\"\"\n",
    "    Update the ends of transcripts and the first / last exon\n",
    "    in a GTF. GTF must be sorted!\n",
    "    \n",
    "    Parameters:\n",
    "        df (pandas DataFrame): Sorted DF of GTF with 'Start_end', and 'End_end' \n",
    "            columns denoting the boundaries of each end region\n",
    "        mode (str): {'tss', 'tes'}\n",
    "        strand (str): {'+', '-'}\n",
    "    \n",
    "    Returns:\n",
    "        df (pandas DataFrame): DF of GTF with transcript and \n",
    "            exon ends modified\n",
    "    \"\"\"\n",
    "    old_col, new_col, gene_func = get_update_ends_settings(strand, mode)\n",
    "    \n",
    "    temp = df[['Feature', 'gene_id', 'transcript_id', 'Strand', 'Start', 'End', 'Start_end', 'End_end']].copy(deep=True)\n",
    "    temp = temp.loc[temp.Feature != 'gene']\n",
    "    if mode == 'tss':\n",
    "        inds = temp.groupby('transcript_id').head(2).index.tolist()\n",
    "    elif mode == 'tes':\n",
    "        inds = temp.groupby('transcript_id').head(1).index.tolist()\n",
    "        inds += temp.groupby('transcript_id').tail(1).index.tolist()\n",
    "\n",
    "    df.loc[inds, old_col] = df.loc[inds, new_col]\n",
    "    \n",
    "    # convert float dtypes\n",
    "    df.Start = df.Start.astype(int)\n",
    "    df.End = df.End.astype(int)\n",
    "    \n",
    "    return df\n",
    "\n",
    "def update_gene_ends(df, mode, strand):\n",
    "    \"\"\"\n",
    "    Update the ends of genes in a GTF. \n",
    "    \n",
    "    Parameters:\n",
    "        df (pandas DataFrame): GTF dataframe \n",
    "        mode (str): {'tss', 'tes'}\n",
    "        strand (str): {'+', '-'}\n",
    "        \n",
    "    Returns:\n",
    "        df (pandas DataFrame): DataFrame of GTF with gene ends updated\n",
    "    \"\"\"\n",
    "    # determine which ends we're updating and how we're doing so \n",
    "    old_col, new_col, gene_func = get_update_ends_settings(strand, mode)\n",
    "    \n",
    "    # get min or max of transcript ends depending on settings\n",
    "    temp = df[['Feature', 'gene_id', old_col]].copy(deep=True)\n",
    "    temp = temp.loc[temp.Feature == 'transcript']\n",
    "    temp = temp.groupby(['gene_id', 'Feature'], observed=True).agg(gene_func).reset_index()\n",
    "    temp.drop('Feature', axis=1, inplace=True)\n",
    "    \n",
    "    # add that coord to gene end\n",
    "    df = df.merge(temp, on='gene_id', suffixes=('', '_gene'))\n",
    "    inds = df.loc[df.Feature == 'gene'].index.tolist()\n",
    "    df.loc[inds, old_col] = df.loc[inds, '{}_gene'.format(old_col)]\n",
    "    df.drop('{}_gene'.format(old_col), axis=1, inplace=True)\n",
    "    \n",
    "    return df\n",
    "\n",
    "def update_gtf_ends(gtf, tss, tes):\n",
    "    \"\"\"\n",
    "    Update gene, transcript, and exon boundaries to be \n",
    "    furthest upstream or downstream entry for end used\n",
    "    \n",
    "    Parameters:\n",
    "        gtf (pandas DataFrame): DF of GTF\n",
    "        tss (pyranges PyRanges): PyRanges object of reference TSSs\n",
    "        tes (pyranges PyRanges): PyRanges object of reference TESs\n",
    "    \n",
    "    Returns: \n",
    "        gtf (pandas DataFrame): DF of GTF with updated ends\n",
    "            based on the TSSs and TESs used in the input beds\n",
    "    \"\"\"\n",
    "    gtf = gtf.copy(deep=True)\n",
    "\n",
    "    for mode, ends in zip(['tss', 'tes'], [tss, tes]):\n",
    "        ends = ends.df\n",
    "        ends = ends[['Start', 'End', '{}_id'.format(mode)]]\n",
    "        gtf = gtf.merge(ends, how='left',\n",
    "                        on='{}_id'.format(mode),\n",
    "                        suffixes=('', '_end'))\n",
    "\n",
    "        fwd, rev = get_stranded_gtf_dfs(gtf)\n",
    "        df = pd.DataFrame()\n",
    "        for strand, temp in zip(['+', '-'], [fwd, rev]):\n",
    "    \n",
    "            # fix exon, transcript, and gene boundaries\n",
    "            temp = update_transcript_ends(temp, mode, strand)\n",
    "            temp = update_gene_ends(temp, mode, strand)\n",
    "            df = pd.concat([df, temp])\n",
    "        \n",
    "        gtf = df.copy(deep=True)\n",
    "        gtf.drop(['Start_end', 'End_end'], axis=1, inplace=True)\n",
    "        \n",
    "    return gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "id": "7db16831-2564-4356-ba8d-799ef02cde42",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15\n"
     ]
    }
   ],
   "source": [
    "h5 = '/Users/fairliereese/mortazavi_lab/data/rnawg/lr_bulk/cerberus/human_cerberus.h5'\n",
    "gtf = '/Users/fairliereese/mortazavi_lab/data/rnawg/lr_bulk/talon/human_known_nic_nnc_talon.gtf'\n",
    "gtf = 'test_dupe_genes.gtf'\n",
    "agg = True\n",
    "update_ends = True\n",
    "\n",
    "# def replace_gtf_ids():\n",
    "\n",
    "df = pr.read_gtf(gtf).df\n",
    "print(len(df.index))\n",
    "entry_types = ['gene', 'transcript', 'exon']\n",
    "df = df.loc[df.Feature.isin(entry_types)]\n",
    "df = sort_gtf(df)\n",
    "\n",
    "if not update_ends:\n",
    "    _, _, _, m_df = read_h5(h5)\n",
    "else:\n",
    "    _, tss, tes, m_df = read_h5(h5)\n",
    "    # hack to remove duplicates for now\n",
    "    # tss = pr.PyRanges(tss.df[['Chromosome', 'Start', 'End', 'Strand', 'gene_id', 'tss']].drop_duplicates())\n",
    "    # tes = pr.PyRanges(tes.df[['Chromosome', 'Start', 'End', 'Strand', 'gene_id', 'tes']].drop_duplicates())\n",
    "    tss = tss.df\n",
    "    tes = tes.df\n",
    "    tss['tss_id'] = tss.gene_id+'_'+tss.tss.astype(str)\n",
    "    tes['tes_id'] = tes.gene_id+'_'+tes.tes.astype(str)\n",
    "    tss = pr.PyRanges(tss)\n",
    "    tes = pr.PyRanges(tes)\n",
    "\n",
    "\n",
    "m_df.drop(['transcript_triplet', \n",
    "           'gene_name', 'gene_id'], axis=1, inplace=True)\n",
    "\n",
    "df = df.merge(m_df, how='left', \n",
    "                left_on=['transcript_name', 'transcript_id'],\n",
    "                right_on=['original_transcript_name', 'original_transcript_id'],\n",
    "                suffixes=('', '_cerberus'))\n",
    "\n",
    "if update_ends:\n",
    "    df = update_gtf_ends(df, tss, tes)\n",
    "\n",
    "if agg:\n",
    "    if not update_ends:\n",
    "        raise ValueError('Must update ends to aggregate transcripts')\n",
    "    pass\n",
    "    # gtf = agg_gtf_transcripts()\n",
    "    \n",
    "df.drop(['transcript_id', 'transcript_name'], axis=1, inplace=True)\n",
    "df.rename({'transcript_id_cerberus': 'transcript_id',\n",
    "            'transcript_name_cerberus': 'transcript_name'},\n",
    "           axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "id": "e7145406-6e06-411b-b6a7-1692693ba40b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>...</th>\n",
       "      <th>original_transcript_id</th>\n",
       "      <th>ic</th>\n",
       "      <th>ic_id</th>\n",
       "      <th>tss_id</th>\n",
       "      <th>tss</th>\n",
       "      <th>tes_id</th>\n",
       "      <th>tes</th>\n",
       "      <th>original_transcript_name</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr6</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>1609864</td>\n",
       "      <td>1613947</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000054598.7</td>\n",
       "      <td>FOXC1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr6</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1609864</td>\n",
       "      <td>1613947</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000054598.7</td>\n",
       "      <td>FOXC1</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000380874.3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000054598_1</td>\n",
       "      <td>ENSG00000054598_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000054598_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>FOXC1-201</td>\n",
       "      <td>ENSG00000054598[1,1,1]</td>\n",
       "      <td>FOXC1[1,1,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr6</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>1609864</td>\n",
       "      <td>1613947</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000054598.7</td>\n",
       "      <td>FOXC1</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000380874.3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000054598_1</td>\n",
       "      <td>ENSG00000054598_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000054598_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>FOXC1-201</td>\n",
       "      <td>ENSG00000054598[1,1,1]</td>\n",
       "      <td>FOXC1[1,1,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr6</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1609864</td>\n",
       "      <td>1613947</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000054598.7</td>\n",
       "      <td>FOXC1</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000645831.1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000054598_1</td>\n",
       "      <td>ENSG00000054598_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000054598_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>FOXC1-202</td>\n",
       "      <td>ENSG00000054598[1,1,1]</td>\n",
       "      <td>FOXC1[1,1,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr6</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>1609864</td>\n",
       "      <td>1613947</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000054598.7</td>\n",
       "      <td>FOXC1</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000645831.1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000054598_1</td>\n",
       "      <td>ENSG00000054598_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000054598_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>FOXC1-202</td>\n",
       "      <td>ENSG00000054598[1,1,1]</td>\n",
       "      <td>FOXC1[1,1,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>1232186</td>\n",
       "      <td>1235091</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000176022.5</td>\n",
       "      <td>B3GALT6</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1232186</td>\n",
       "      <td>1235091</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000176022.5</td>\n",
       "      <td>B3GALT6</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000379198.4</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000176022_1</td>\n",
       "      <td>ENSG00000176022_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000176022_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>B3GALT6-201</td>\n",
       "      <td>ENSG00000176022[1,1,1]</td>\n",
       "      <td>B3GALT6[1,1,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>1232186</td>\n",
       "      <td>1235091</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000176022.5</td>\n",
       "      <td>B3GALT6</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000379198.4</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000176022_1</td>\n",
       "      <td>ENSG00000176022_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000176022_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>B3GALT6-201</td>\n",
       "      <td>ENSG00000176022[1,1,1]</td>\n",
       "      <td>B3GALT6[1,1,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1232186</td>\n",
       "      <td>1235091</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000176022.5</td>\n",
       "      <td>B3GALT6</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000647651.1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000176022_1</td>\n",
       "      <td>ENSG00000176022_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000176022_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>B3GALT6-202</td>\n",
       "      <td>ENSG00000176022[1,1,1]</td>\n",
       "      <td>B3GALT6[1,1,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>1232186</td>\n",
       "      <td>1235091</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000176022.5</td>\n",
       "      <td>B3GALT6</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000647651.1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000176022_1</td>\n",
       "      <td>ENSG00000176022_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000176022_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>B3GALT6-202</td>\n",
       "      <td>ENSG00000176022[1,1,1]</td>\n",
       "      <td>B3GALT6[1,1,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>83508239</td>\n",
       "      <td>83512177</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000196767.7</td>\n",
       "      <td>POU3F4</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>83508239</td>\n",
       "      <td>83512177</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000196767.7</td>\n",
       "      <td>POU3F4</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000373200.5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000196767_1</td>\n",
       "      <td>ENSG00000196767_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000196767_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>POU3F4-201</td>\n",
       "      <td>ENSG00000196767[1,1,1]</td>\n",
       "      <td>POU3F4[1,1,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>83508239</td>\n",
       "      <td>83512177</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000196767.7</td>\n",
       "      <td>POU3F4</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000373200.5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000196767_1</td>\n",
       "      <td>ENSG00000196767_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000196767_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>POU3F4-201</td>\n",
       "      <td>ENSG00000196767[1,1,1]</td>\n",
       "      <td>POU3F4[1,1,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>83508239</td>\n",
       "      <td>83512177</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000196767.7</td>\n",
       "      <td>POU3F4</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000644024.1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000196767_1</td>\n",
       "      <td>ENSG00000196767_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000196767_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>POU3F4-202</td>\n",
       "      <td>ENSG00000196767[1,1,1]</td>\n",
       "      <td>POU3F4[1,1,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>83508239</td>\n",
       "      <td>83512177</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000196767.7</td>\n",
       "      <td>POU3F4</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000644024.1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000196767_1</td>\n",
       "      <td>ENSG00000196767_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000196767_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>POU3F4-202</td>\n",
       "      <td>ENSG00000196767[1,1,1]</td>\n",
       "      <td>POU3F4[1,1,1]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15 rows × 38 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chromosome  Source     Feature     Start       End Score Strand Frame  \\\n",
       "0        chr6  HAVANA        gene   1609864   1613947     .      +     .   \n",
       "1        chr6  HAVANA  transcript   1609864   1613947     .      +     .   \n",
       "2        chr6  HAVANA        exon   1609864   1613947     .      +     .   \n",
       "3        chr6  HAVANA  transcript   1609864   1613947     .      +     .   \n",
       "4        chr6  HAVANA        exon   1609864   1613947     .      +     .   \n",
       "5        chr1  HAVANA        gene   1232186   1235091     .      +     .   \n",
       "6        chr1  HAVANA  transcript   1232186   1235091     .      +     .   \n",
       "7        chr1  HAVANA        exon   1232186   1235091     .      +     .   \n",
       "8        chr1  HAVANA  transcript   1232186   1235091     .      +     .   \n",
       "9        chr1  HAVANA        exon   1232186   1235091     .      +     .   \n",
       "10       chrX  HAVANA        gene  83508239  83512177     .      +     .   \n",
       "11       chrX  HAVANA  transcript  83508239  83512177     .      +     .   \n",
       "12       chrX  HAVANA        exon  83508239  83512177     .      +     .   \n",
       "13       chrX  HAVANA  transcript  83508239  83512177     .      +     .   \n",
       "14       chrX  HAVANA        exon  83508239  83512177     .      +     .   \n",
       "\n",
       "              gene_id gene_name  ... original_transcript_id   ic  \\\n",
       "0   ENSG00000054598.7     FOXC1  ...                    NaN  NaN   \n",
       "1   ENSG00000054598.7     FOXC1  ...      ENST00000380874.3  1.0   \n",
       "2   ENSG00000054598.7     FOXC1  ...      ENST00000380874.3  1.0   \n",
       "3   ENSG00000054598.7     FOXC1  ...      ENST00000645831.1  1.0   \n",
       "4   ENSG00000054598.7     FOXC1  ...      ENST00000645831.1  1.0   \n",
       "5   ENSG00000176022.5   B3GALT6  ...                    NaN  NaN   \n",
       "6   ENSG00000176022.5   B3GALT6  ...      ENST00000379198.4  1.0   \n",
       "7   ENSG00000176022.5   B3GALT6  ...      ENST00000379198.4  1.0   \n",
       "8   ENSG00000176022.5   B3GALT6  ...      ENST00000647651.1  1.0   \n",
       "9   ENSG00000176022.5   B3GALT6  ...      ENST00000647651.1  1.0   \n",
       "10  ENSG00000196767.7    POU3F4  ...                    NaN  NaN   \n",
       "11  ENSG00000196767.7    POU3F4  ...      ENST00000373200.5  1.0   \n",
       "12  ENSG00000196767.7    POU3F4  ...      ENST00000373200.5  1.0   \n",
       "13  ENSG00000196767.7    POU3F4  ...      ENST00000644024.1  1.0   \n",
       "14  ENSG00000196767.7    POU3F4  ...      ENST00000644024.1  1.0   \n",
       "\n",
       "                ic_id             tss_id  tss             tes_id  tes  \\\n",
       "0                 NaN                NaN  NaN                NaN  NaN   \n",
       "1   ENSG00000054598_1  ENSG00000054598_1  1.0  ENSG00000054598_1  1.0   \n",
       "2   ENSG00000054598_1  ENSG00000054598_1  1.0  ENSG00000054598_1  1.0   \n",
       "3   ENSG00000054598_1  ENSG00000054598_1  1.0  ENSG00000054598_1  1.0   \n",
       "4   ENSG00000054598_1  ENSG00000054598_1  1.0  ENSG00000054598_1  1.0   \n",
       "5                 NaN                NaN  NaN                NaN  NaN   \n",
       "6   ENSG00000176022_1  ENSG00000176022_1  1.0  ENSG00000176022_1  1.0   \n",
       "7   ENSG00000176022_1  ENSG00000176022_1  1.0  ENSG00000176022_1  1.0   \n",
       "8   ENSG00000176022_1  ENSG00000176022_1  1.0  ENSG00000176022_1  1.0   \n",
       "9   ENSG00000176022_1  ENSG00000176022_1  1.0  ENSG00000176022_1  1.0   \n",
       "10                NaN                NaN  NaN                NaN  NaN   \n",
       "11  ENSG00000196767_1  ENSG00000196767_1  1.0  ENSG00000196767_1  1.0   \n",
       "12  ENSG00000196767_1  ENSG00000196767_1  1.0  ENSG00000196767_1  1.0   \n",
       "13  ENSG00000196767_1  ENSG00000196767_1  1.0  ENSG00000196767_1  1.0   \n",
       "14  ENSG00000196767_1  ENSG00000196767_1  1.0  ENSG00000196767_1  1.0   \n",
       "\n",
       "   original_transcript_name           transcript_id transcript_name  \n",
       "0                       NaN                     NaN             NaN  \n",
       "1                 FOXC1-201  ENSG00000054598[1,1,1]    FOXC1[1,1,1]  \n",
       "2                 FOXC1-201  ENSG00000054598[1,1,1]    FOXC1[1,1,1]  \n",
       "3                 FOXC1-202  ENSG00000054598[1,1,1]    FOXC1[1,1,1]  \n",
       "4                 FOXC1-202  ENSG00000054598[1,1,1]    FOXC1[1,1,1]  \n",
       "5                       NaN                     NaN             NaN  \n",
       "6               B3GALT6-201  ENSG00000176022[1,1,1]  B3GALT6[1,1,1]  \n",
       "7               B3GALT6-201  ENSG00000176022[1,1,1]  B3GALT6[1,1,1]  \n",
       "8               B3GALT6-202  ENSG00000176022[1,1,1]  B3GALT6[1,1,1]  \n",
       "9               B3GALT6-202  ENSG00000176022[1,1,1]  B3GALT6[1,1,1]  \n",
       "10                      NaN                     NaN             NaN  \n",
       "11               POU3F4-201  ENSG00000196767[1,1,1]   POU3F4[1,1,1]  \n",
       "12               POU3F4-201  ENSG00000196767[1,1,1]   POU3F4[1,1,1]  \n",
       "13               POU3F4-202  ENSG00000196767[1,1,1]   POU3F4[1,1,1]  \n",
       "14               POU3F4-202  ENSG00000196767[1,1,1]   POU3F4[1,1,1]  \n",
       "\n",
       "[15 rows x 38 columns]"
      ]
     },
     "execution_count": 152,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "id": "fa5f32c9-937e-4fd1-a2a7-41cca19309df",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Chromosome', 'Source', 'Feature', 'Start', 'End', 'Score', 'Strand',\n",
       "       'Frame', 'gene_id', 'gene_name', 'gene_status', 'gene_type',\n",
       "       'talon_gene', 'havana_gene', 'level', 'tag', 'transcript_status',\n",
       "       'talon_transcript', 'source', 'transcript_type', 'havana_transcript',\n",
       "       'protein_id', 'exon_number', 'exon_id', 'talon_exon', 'exon_status',\n",
       "       'transcript_support_level', 'ccdsid', 'original_transcript_id', 'ic',\n",
       "       'ic_id', 'tss_id', 'tss', 'tes_id', 'tes', 'original_transcript_name',\n",
       "       'transcript_id', 'transcript_name'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 158,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "id": "86f6e384-e863-4a36-b280-8a3b53d82b76",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>...</th>\n",
       "      <th>original_transcript_id</th>\n",
       "      <th>ic</th>\n",
       "      <th>ic_id</th>\n",
       "      <th>tss_id</th>\n",
       "      <th>tss</th>\n",
       "      <th>tes_id</th>\n",
       "      <th>tes</th>\n",
       "      <th>original_transcript_name</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr6</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1609864</td>\n",
       "      <td>1613947</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000054598.7</td>\n",
       "      <td>FOXC1</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000380874.3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000054598_1</td>\n",
       "      <td>ENSG00000054598_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000054598_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>FOXC1-201</td>\n",
       "      <td>ENSG00000054598[1,1,1]</td>\n",
       "      <td>FOXC1[1,1,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr6</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>1609864</td>\n",
       "      <td>1613947</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000054598.7</td>\n",
       "      <td>FOXC1</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000380874.3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000054598_1</td>\n",
       "      <td>ENSG00000054598_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000054598_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>FOXC1-201</td>\n",
       "      <td>ENSG00000054598[1,1,1]</td>\n",
       "      <td>FOXC1[1,1,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr6</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1609864</td>\n",
       "      <td>1613947</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000054598.7</td>\n",
       "      <td>FOXC1</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000645831.1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000054598_1</td>\n",
       "      <td>ENSG00000054598_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000054598_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>FOXC1-202</td>\n",
       "      <td>ENSG00000054598[1,1,1]</td>\n",
       "      <td>FOXC1[1,1,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr6</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>1609864</td>\n",
       "      <td>1613947</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000054598.7</td>\n",
       "      <td>FOXC1</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000645831.1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000054598_1</td>\n",
       "      <td>ENSG00000054598_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000054598_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>FOXC1-202</td>\n",
       "      <td>ENSG00000054598[1,1,1]</td>\n",
       "      <td>FOXC1[1,1,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1232186</td>\n",
       "      <td>1235091</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000176022.5</td>\n",
       "      <td>B3GALT6</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000379198.4</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000176022_1</td>\n",
       "      <td>ENSG00000176022_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000176022_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>B3GALT6-201</td>\n",
       "      <td>ENSG00000176022[1,1,1]</td>\n",
       "      <td>B3GALT6[1,1,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>1232186</td>\n",
       "      <td>1235091</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000176022.5</td>\n",
       "      <td>B3GALT6</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000379198.4</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000176022_1</td>\n",
       "      <td>ENSG00000176022_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000176022_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>B3GALT6-201</td>\n",
       "      <td>ENSG00000176022[1,1,1]</td>\n",
       "      <td>B3GALT6[1,1,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1232186</td>\n",
       "      <td>1235091</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000176022.5</td>\n",
       "      <td>B3GALT6</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000647651.1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000176022_1</td>\n",
       "      <td>ENSG00000176022_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000176022_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>B3GALT6-202</td>\n",
       "      <td>ENSG00000176022[1,1,1]</td>\n",
       "      <td>B3GALT6[1,1,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>1232186</td>\n",
       "      <td>1235091</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000176022.5</td>\n",
       "      <td>B3GALT6</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000647651.1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000176022_1</td>\n",
       "      <td>ENSG00000176022_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000176022_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>B3GALT6-202</td>\n",
       "      <td>ENSG00000176022[1,1,1]</td>\n",
       "      <td>B3GALT6[1,1,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>83508239</td>\n",
       "      <td>83512177</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000196767.7</td>\n",
       "      <td>POU3F4</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000373200.5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000196767_1</td>\n",
       "      <td>ENSG00000196767_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000196767_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>POU3F4-201</td>\n",
       "      <td>ENSG00000196767[1,1,1]</td>\n",
       "      <td>POU3F4[1,1,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>83508239</td>\n",
       "      <td>83512177</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000196767.7</td>\n",
       "      <td>POU3F4</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000373200.5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000196767_1</td>\n",
       "      <td>ENSG00000196767_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000196767_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>POU3F4-201</td>\n",
       "      <td>ENSG00000196767[1,1,1]</td>\n",
       "      <td>POU3F4[1,1,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>83508239</td>\n",
       "      <td>83512177</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000196767.7</td>\n",
       "      <td>POU3F4</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000644024.1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000196767_1</td>\n",
       "      <td>ENSG00000196767_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000196767_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>POU3F4-202</td>\n",
       "      <td>ENSG00000196767[1,1,1]</td>\n",
       "      <td>POU3F4[1,1,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>83508239</td>\n",
       "      <td>83512177</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000196767.7</td>\n",
       "      <td>POU3F4</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000644024.1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000196767_1</td>\n",
       "      <td>ENSG00000196767_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000196767_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>POU3F4-202</td>\n",
       "      <td>ENSG00000196767[1,1,1]</td>\n",
       "      <td>POU3F4[1,1,1]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12 rows × 38 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chromosome  Source     Feature     Start       End Score Strand Frame  \\\n",
       "1        chr6  HAVANA  transcript   1609864   1613947     .      +     .   \n",
       "2        chr6  HAVANA        exon   1609864   1613947     .      +     .   \n",
       "3        chr6  HAVANA  transcript   1609864   1613947     .      +     .   \n",
       "4        chr6  HAVANA        exon   1609864   1613947     .      +     .   \n",
       "6        chr1  HAVANA  transcript   1232186   1235091     .      +     .   \n",
       "7        chr1  HAVANA        exon   1232186   1235091     .      +     .   \n",
       "8        chr1  HAVANA  transcript   1232186   1235091     .      +     .   \n",
       "9        chr1  HAVANA        exon   1232186   1235091     .      +     .   \n",
       "11       chrX  HAVANA  transcript  83508239  83512177     .      +     .   \n",
       "12       chrX  HAVANA        exon  83508239  83512177     .      +     .   \n",
       "13       chrX  HAVANA  transcript  83508239  83512177     .      +     .   \n",
       "14       chrX  HAVANA        exon  83508239  83512177     .      +     .   \n",
       "\n",
       "              gene_id gene_name  ... original_transcript_id   ic  \\\n",
       "1   ENSG00000054598.7     FOXC1  ...      ENST00000380874.3  1.0   \n",
       "2   ENSG00000054598.7     FOXC1  ...      ENST00000380874.3  1.0   \n",
       "3   ENSG00000054598.7     FOXC1  ...      ENST00000645831.1  1.0   \n",
       "4   ENSG00000054598.7     FOXC1  ...      ENST00000645831.1  1.0   \n",
       "6   ENSG00000176022.5   B3GALT6  ...      ENST00000379198.4  1.0   \n",
       "7   ENSG00000176022.5   B3GALT6  ...      ENST00000379198.4  1.0   \n",
       "8   ENSG00000176022.5   B3GALT6  ...      ENST00000647651.1  1.0   \n",
       "9   ENSG00000176022.5   B3GALT6  ...      ENST00000647651.1  1.0   \n",
       "11  ENSG00000196767.7    POU3F4  ...      ENST00000373200.5  1.0   \n",
       "12  ENSG00000196767.7    POU3F4  ...      ENST00000373200.5  1.0   \n",
       "13  ENSG00000196767.7    POU3F4  ...      ENST00000644024.1  1.0   \n",
       "14  ENSG00000196767.7    POU3F4  ...      ENST00000644024.1  1.0   \n",
       "\n",
       "                ic_id             tss_id  tss             tes_id  tes  \\\n",
       "1   ENSG00000054598_1  ENSG00000054598_1  1.0  ENSG00000054598_1  1.0   \n",
       "2   ENSG00000054598_1  ENSG00000054598_1  1.0  ENSG00000054598_1  1.0   \n",
       "3   ENSG00000054598_1  ENSG00000054598_1  1.0  ENSG00000054598_1  1.0   \n",
       "4   ENSG00000054598_1  ENSG00000054598_1  1.0  ENSG00000054598_1  1.0   \n",
       "6   ENSG00000176022_1  ENSG00000176022_1  1.0  ENSG00000176022_1  1.0   \n",
       "7   ENSG00000176022_1  ENSG00000176022_1  1.0  ENSG00000176022_1  1.0   \n",
       "8   ENSG00000176022_1  ENSG00000176022_1  1.0  ENSG00000176022_1  1.0   \n",
       "9   ENSG00000176022_1  ENSG00000176022_1  1.0  ENSG00000176022_1  1.0   \n",
       "11  ENSG00000196767_1  ENSG00000196767_1  1.0  ENSG00000196767_1  1.0   \n",
       "12  ENSG00000196767_1  ENSG00000196767_1  1.0  ENSG00000196767_1  1.0   \n",
       "13  ENSG00000196767_1  ENSG00000196767_1  1.0  ENSG00000196767_1  1.0   \n",
       "14  ENSG00000196767_1  ENSG00000196767_1  1.0  ENSG00000196767_1  1.0   \n",
       "\n",
       "   original_transcript_name           transcript_id transcript_name  \n",
       "1                 FOXC1-201  ENSG00000054598[1,1,1]    FOXC1[1,1,1]  \n",
       "2                 FOXC1-201  ENSG00000054598[1,1,1]    FOXC1[1,1,1]  \n",
       "3                 FOXC1-202  ENSG00000054598[1,1,1]    FOXC1[1,1,1]  \n",
       "4                 FOXC1-202  ENSG00000054598[1,1,1]    FOXC1[1,1,1]  \n",
       "6               B3GALT6-201  ENSG00000176022[1,1,1]  B3GALT6[1,1,1]  \n",
       "7               B3GALT6-201  ENSG00000176022[1,1,1]  B3GALT6[1,1,1]  \n",
       "8               B3GALT6-202  ENSG00000176022[1,1,1]  B3GALT6[1,1,1]  \n",
       "9               B3GALT6-202  ENSG00000176022[1,1,1]  B3GALT6[1,1,1]  \n",
       "11               POU3F4-201  ENSG00000196767[1,1,1]   POU3F4[1,1,1]  \n",
       "12               POU3F4-201  ENSG00000196767[1,1,1]   POU3F4[1,1,1]  \n",
       "13               POU3F4-202  ENSG00000196767[1,1,1]   POU3F4[1,1,1]  \n",
       "14               POU3F4-202  ENSG00000196767[1,1,1]   POU3F4[1,1,1]  \n",
       "\n",
       "[12 rows x 38 columns]"
      ]
     },
     "execution_count": 157,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dupe_cols = ['Chromosome', 'Source', \n",
    "             'Feature', \n",
    "             'Start', 'End', \n",
    "             'Score', 'Strand', 'Frame', 'gene_id', 'gene_name', \n",
    "             'gene_status', 'gene_type', 'talon_gene', \n",
    "             'ic', 'ic_id', 'tss_id', 'tss', 'tes_id', 'tes', 'transcript_id',\n",
    "             'transcript_name']\n",
    "df.loc[df[dupe_cols].duplicated(keep=False)]\n",
    "# df[dupe_cols]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1ecf767-7333-42d3-a0f6-71129ef07213",
   "metadata": {},
   "source": [
    "## testing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "id": "bd7b448d-a5ce-4e07-bc5e-fae62abcc96a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_end_df(c,s,st,e,n, source,mode):\n",
    "    df = pd.DataFrame()\n",
    "    cols = ['Chromosome', 'Strand', 'Start', 'End', 'Name']\n",
    "    var = [c,s,st,e,n]\n",
    "    for col, var in zip(cols, var):\n",
    "        if type(var) == list:\n",
    "            df[col] = var\n",
    "\n",
    "    # add source\n",
    "    df['source'] = source\n",
    "\n",
    "    df = format_end_df(df)\n",
    "\n",
    "    # get end # and gene id\n",
    "    if any(df.Name.isnull()):\n",
    "        df['gene_id'] = np.nan\n",
    "        df[mode] = np.nan\n",
    "    else:\n",
    "        df['gene_id'] = df.Name.str.split('_', expand=True)[0]\n",
    "        df[mode] = df.Name.str.split('_', expand=True)[1]\n",
    "\n",
    "    # get arbitrary unique ids\n",
    "    df['id'] = [i for i in range(len(df.index))]\n",
    "    \n",
    "    # get id\n",
    "    df['{}_id'.format(mode)] = df.gene_id+'_'+df[mode]\n",
    "\n",
    "    return df\n",
    "\n",
    "def format_end_df(df):\n",
    "    sort_cols = ['Chromosome', 'Start', 'End', 'Strand']\n",
    "    df = df.sort_values(by=sort_cols)\n",
    "    order = ['Chromosome', 'Start', 'End', 'Strand', 'Name', 'source']\n",
    "    order = [o for o in order if o in df.columns]\n",
    "    df = df[order]\n",
    "    df.reset_index(drop=True, inplace=True)\n",
    "    return df\n",
    "\n",
    "def make_exon_df(n,c,e,s,g,t):\n",
    "    df = pd.DataFrame()\n",
    "    df['Chromosome'] = c\n",
    "    df['Start'] = [i[0] for i in e]\n",
    "    df['End'] = [i[1] for i in e]\n",
    "    df['Strand'] = s\n",
    "    df['Feature'] = 'exon'\n",
    "    cols = ['gene_name', 'gene_id']\n",
    "    for c in cols:\n",
    "        df[c] = g\n",
    "    cols = ['transcript_id', 'transcript_name']\n",
    "    for c in cols:\n",
    "        df[c] = t\n",
    "\n",
    "    # reorder exons and starts/ stops if needed\n",
    "    df['new_Start'] = df[['Start', 'End']].min(axis=1)\n",
    "    df['new_End'] = df[['Start', 'End']].max(axis=1)\n",
    "    df.drop(['Start', 'End'], axis=1, inplace=True)\n",
    "    df.rename({'new_Start':'Start',\n",
    "               'new_End':'End'}, axis=1, inplace=True)\n",
    "    s = s[0]\n",
    "    if s == '+':\n",
    "        ascending = True\n",
    "    elif s == '-':\n",
    "        ascending = False\n",
    "    df.sort_values(by='Start', ascending=ascending, inplace=True)\n",
    "    return df\n",
    "\n",
    "def make_hier_entry(df, how='t'):\n",
    "    \"\"\"\n",
    "    kind {'g','t'}\n",
    "    \"\"\"\n",
    "    agg_dict = {'min_coord': 'min', 'max_coord': 'max'}\n",
    "    t_df = df.copy(deep=True)\n",
    "    t_df['min_coord'] = t_df[['Start', 'End']].min(axis=1)\n",
    "    t_df['max_coord'] = t_df[['Start', 'End']].max(axis=1)\n",
    "    if how == 't':\n",
    "        gb_cols = ['Chromosome', 'Strand', 'gene_name',\n",
    "                   'gene_id', 'transcript_id', 'transcript_name']\n",
    "        if 'tss_id' in t_df.columns:\n",
    "            gb_cols.append('tss_id')\n",
    "        if 'tes_id' in t_df.columns:\n",
    "            gb_cols.append('tes_id')\n",
    "    elif how == 'g':\n",
    "        gb_cols = ['Chromosome', 'Strand', 'gene_name',\n",
    "                   'gene_id']\n",
    "        \n",
    "    cols = gb_cols + ['min_coord', 'max_coord']\n",
    "    t_df = t_df[cols]\n",
    "    t_df = t_df.groupby(gb_cols).agg(agg_dict).reset_index()\n",
    "    t_df.rename({'min_coord': 'Start', 'max_coord': 'End'}, axis=1, inplace=True)\n",
    "    if how == 't':\n",
    "        t_df['Feature'] = 'transcript'\n",
    "    elif how == 'g':\n",
    "        t_df['Feature'] = 'gene'\n",
    "        \n",
    "    return t_df\n",
    "\n",
    "def make_test_gtf(ts):\n",
    "    df = pd.concat(ts)\n",
    "        # make transcript entries\n",
    "    t_df = make_hier_entry(df, how='t')\n",
    "    # make gene entries\n",
    "    g_df = make_hier_entry(df, how='g')\n",
    "\n",
    "    # concat everything and sort by gene id, transcript id, feature rank (gene =0, t =1, exon=2), then start coords\n",
    "    df = pd.concat([df, t_df, g_df])\n",
    "    df = sort_gtf(df)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "id": "fa9b6b78-9c93-4e35-948a-45398560e282",
   "metadata": {},
   "outputs": [],
   "source": [
    "# tests for update_gtf_ends\n",
    "def test_update_gtf_ends(print_dfs=True):\n",
    "\n",
    "    ts = []\n",
    "    # t1 - fwd strand transcript from gene w/ >2 transcripts\n",
    "    n = 3\n",
    "    c = ['1' for i in range(n)]\n",
    "    e = [[1,10], [14,20], [25,30]]\n",
    "    s = ['+' for i in range(n)]\n",
    "    g = 'g1'\n",
    "    t = 'g1_t1'\n",
    "    df = make_exon_df(n,c,e,s,g,t)\n",
    "    df['tss_id'] = 'g1_1'\n",
    "    df['tes_id'] = 'g1_1'\n",
    "    ts.append(df)\n",
    "    \n",
    "    # t1.5 - fwd strand transcript from gene that doesn't \n",
    "    # need boundaries updated\n",
    "    n = 3\n",
    "    c = ['1' for i in range(n)]\n",
    "    e = [[0, 10], [14,20], [25, 35]]\n",
    "    s = ['+' for i in range(n)]\n",
    "    g = 'g1'\n",
    "    t = 'g1_t1.5'\n",
    "    df = make_exon_df(n,c,e,s,g,t)\n",
    "    df['tss_id'] = 'g1_1'\n",
    "    df['tes_id'] = 'g1_1'\n",
    "    ts.append(df)\n",
    "\n",
    "    # t2 - rev. strand transcript from gene w/ >2 transcripts\n",
    "    n = 3\n",
    "    c = ['1' for i in range(n)]\n",
    "    e = [[90,60], [45,30], [10,8]]\n",
    "    s = ['-' for i in range(n)]\n",
    "    g = 'g2'\n",
    "    t = 'g2_t1'\n",
    "    df = make_exon_df(n,c,e,s,g,t)\n",
    "    df['tss_id'] = 'g2_1'\n",
    "    df['tes_id'] = 'g2_1'\n",
    "    ts.append(df)\n",
    "\n",
    "    # t3 - rev. strand transcript from gene w/ >2 transcripts\n",
    "    n = 3\n",
    "    c = ['1' for i in range(n)]\n",
    "    e = [[95,60], [45,30], [10,6]]\n",
    "    s = ['-' for i in range(n)]\n",
    "    g = 'g2'\n",
    "    t = 'g2_t2'\n",
    "    df = make_exon_df(n,c,e,s,g,t)\n",
    "    df['tss_id'] = 'g2_2'\n",
    "    df['tes_id'] = 'g2_2'\n",
    "    ts.append(df)\n",
    "    \n",
    "    # t4 - fwd strand gene w/ monoexonic transcript\n",
    "    n = 1\n",
    "    c = ['1' for i in range(n)]\n",
    "    e = [[20,30]]\n",
    "    s = ['+']\n",
    "    g = 'g3'\n",
    "    t = 'g3_t1'\n",
    "    df = make_exon_df(n,c,e,s,g,t)\n",
    "    df['tss_id'] = 'g3_1'\n",
    "    df['tes_id'] = 'g3_1'\n",
    "    ts.append(df)\n",
    "    \n",
    "    # t5 - rev strand gene w/ monoexonic transcript\n",
    "    n = 1\n",
    "    c = ['1' for i in range(n)]\n",
    "    e = [[50, 40]]\n",
    "    s = ['-']\n",
    "    g = 'g4'\n",
    "    t = 'g4_t1'\n",
    "    df = make_exon_df(n,c,e,s,g,t)\n",
    "    df['tss_id'] = 'g4_1'\n",
    "    df['tes_id'] = 'g4_1'\n",
    "    ts.append(df)\n",
    "\n",
    "    # tss reference\n",
    "    n = 5\n",
    "    mode = 'tss'\n",
    "    c = ['1' for i in range(n)]\n",
    "    s = ['+', '-', '-', '+', '-']\n",
    "    st = [0, 85, 91, 19, 50] \n",
    "    e = [3, 93, 98, 22, 55]\n",
    "    n = ['g1_1', 'g2_1', 'g2_2', 'g3_1', 'g4_1']\n",
    "    source = 'v1'\n",
    "    tss = make_end_df(c,s,st,e,n,source,mode)\n",
    "    tss = pr.PyRanges(tss)\n",
    "\n",
    "    # tes reference\n",
    "    n = 5\n",
    "    mode = 'tes'\n",
    "    c = ['1' for i in range(n)]\n",
    "    s = ['+', '-', '-', '+', '-']\n",
    "    st = [25, 7, 4, 30, 35]\n",
    "    e = [35, 9, 6, 31, 40]\n",
    "    n = ['g1_1', 'g2_1', 'g2_2', 'g3_1', 'g4_1']\n",
    "    source = 'v1'\n",
    "    tes = make_end_df(c,s,st,e,n,source,mode)\n",
    "    tes = pr.PyRanges(tes)\n",
    "\n",
    "    test_df = make_test_gtf(ts)\n",
    "    # test_df = pr.PyRanges(test_df)\n",
    "\n",
    "    test = update_gtf_ends(test_df, tss, tes)\n",
    "\n",
    "    # ctrl for update_gtf_ends\n",
    "    # tests for update_gtf_ends\n",
    "    ts = []\n",
    "    # t1 - fwd strand transcript\n",
    "    n = 3\n",
    "    c = ['1' for i in range(n)]\n",
    "    e = [[0,10], [14,20], [25,35]]\n",
    "    s = ['+' for i in range(n)]\n",
    "    g = 'g1'\n",
    "    t = 'g1_t1'\n",
    "    df = make_exon_df(n,c,e,s,g,t)\n",
    "    df['tss_id'] = 'g1_1'\n",
    "    df['tes_id'] = 'g1_1'\n",
    "    ts.append(df)\n",
    "    \n",
    "    # t1.5 - fwd strand transcript from gene that doesn't \n",
    "    # need boundaries updated\n",
    "    n = 3\n",
    "    c = ['1' for i in range(n)]\n",
    "    e = [[0, 10], [14,20], [25, 35]]\n",
    "    s = ['+' for i in range(n)]\n",
    "    g = 'g1'\n",
    "    t = 'g1_t1.5'\n",
    "    df = make_exon_df(n,c,e,s,g,t)\n",
    "    df['tss_id'] = 'g1_1'\n",
    "    df['tes_id'] = 'g1_1'\n",
    "    ts.append(df)\n",
    "\n",
    "    # t2 - rev. strand transcript from gene w/ >2 transcripts\n",
    "    n = 3\n",
    "    c = ['1' for i in range(n)]\n",
    "    e = [[93,60], [45,30], [10,7]]\n",
    "    s = ['-' for i in range(n)]\n",
    "    g = 'g2'\n",
    "    t = 'g2_t1'\n",
    "    df = make_exon_df(n,c,e,s,g,t)\n",
    "    df['tss_id'] = 'g2_1'\n",
    "    df['tes_id'] = 'g2_1'\n",
    "    ts.append(df)\n",
    "\n",
    "    # t3 - rev. strand transcript from gene w/ >2 transcripts\n",
    "    n = 3\n",
    "    c = ['1' for i in range(n)]\n",
    "    e = [[98,60], [45,30], [10,4]]\n",
    "    s = ['-' for i in range(n)]\n",
    "    g = 'g2'\n",
    "    t = 'g2_t2'\n",
    "    df = make_exon_df(n,c,e,s,g,t)\n",
    "    df['tss_id'] = 'g2_2'\n",
    "    df['tes_id'] = 'g2_2'\n",
    "    ts.append(df)\n",
    "    \n",
    "    # t4 - fwd strand gene w/ monoexonic transcript\n",
    "    n = 1\n",
    "    c = ['1' for i in range(n)]\n",
    "    e = [[19,31]]\n",
    "    s = ['+']\n",
    "    g = 'g3'\n",
    "    t = 'g3_t1'\n",
    "    df = make_exon_df(n,c,e,s,g,t)\n",
    "    df['tss_id'] = 'g3_1'\n",
    "    df['tes_id'] = 'g3_1'\n",
    "    ts.append(df)\n",
    "    \n",
    "    # t4 - rev strand gene w/ monoexonic transcript\n",
    "    n = 1\n",
    "    c = ['1' for i in range(n)]\n",
    "    e = [[55, 35]]\n",
    "    s = ['-']\n",
    "    g = 'g4'\n",
    "    t = 'g4_t1'\n",
    "    df = make_exon_df(n,c,e,s,g,t)\n",
    "    df['tss_id'] = 'g4_1'\n",
    "    df['tes_id'] = 'g4_1'\n",
    "    ts.append(df)\n",
    "\n",
    "    ctrl = make_test_gtf(ts)\n",
    "    ctrl.reset_index(inplace=True, drop=True)\n",
    "    test.reset_index(inplace=True, drop=True)\n",
    "\n",
    "    if print_dfs:\n",
    "        print('test')\n",
    "        print(test)\n",
    "        print(test.index)\n",
    "        print(test.dtypes)\n",
    "        print('ctrl')\n",
    "        print(ctrl)\n",
    "        print(ctrl.index)\n",
    "        print(ctrl.dtypes)\n",
    "\n",
    "    pd.testing.assert_frame_equal(ctrl, test, check_like=True)\n",
    "\n",
    "    assert len(ctrl.index) == len(test.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "id": "3bf42937-cc46-4762-9c3e-57500559e2c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "test_update_gtf_ends(False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 583,
   "id": "8e0ab4fa-925a-46c3-8e1f-13f7dc1d1ac1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# tests for map_gtf_ids\n",
    "# - transcript that does not have duplicate\n",
    "# - transcript that does have a duplicate\n",
    "\n",
    "ts = []\n",
    "# t1 - transcript that doesn't need to be merged\n",
    "n = 3\n",
    "c = ['1' for i in range(n)]\n",
    "e = [[1,10], [14,20], [25,30]]\n",
    "s = ['+' for i in range(n)]\n",
    "g = 'g1'\n",
    "t = 'g1_t1'\n",
    "ts.append(make_exon_df(n,c,e,s,g,t))\n",
    "\n",
    "# t2 - rev. strand transcript that needs to be merged\n",
    "n = 3\n",
    "c = ['1' for i in range(n)]\n",
    "e = [[90,60], [45,30], [10,8]]\n",
    "s = ['-' for i in range(n)]\n",
    "g = 'g2'\n",
    "t = 'g2_t1'\n",
    "ts.append(make_exon_df(n,c,e,s,g,t))\n",
    "\n",
    "# t3 - rev. strand transcript that needs to be merged\n",
    "n = 3\n",
    "c = ['1' for i in range(n)]\n",
    "e = [[95,60], [45,30], [10,6]]\n",
    "s = ['-' for i in range(n)]\n",
    "g = 'g2'\n",
    "t = 'g2_t2'\n",
    "ts.append(make_exon_df(n,c,e,s,g,t))\n",
    "\n",
    "test_df = make_test_gtf(ts)\n",
    "\n",
    "# map file\n",
    "otid = ['g1_t1', 'g2_t1', 'g2_t2',]\n",
    "otname = [o+'n' for o in otid]\n",
    "tid = ['g1[1,1,1]', 'g2[1,1,1]', 'g2[1,1,1]']\n",
    "tname = [o.split('[')[0]+'n['+o.split('[')[1] for o in tid]\n",
    "gid = ['g1', 'g2', 'g2']\n",
    "gname = ['g1n', 'g2n', 'g2n']\n",
    "m_df = pd.DataFrame()\n",
    "m_df['original_transcript_id'] = otid\n",
    "m_df['original_transcript_name'] = otname\n",
    "m_df['transcript_id'] = tid\n",
    "m_df['transcript_name'] = tname\n",
    "m_df['gene_name'] = gname\n",
    "m_df['transcript_triplet'] = m_df.transcript_id.str.slice(2)\n",
    "m_df['gene_id'] = gid\n",
    "\n",
    "# control\n",
    "otid = ['g1_t1', 'g2_t2']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "id": "d0044c70-1a2a-41bd-ad01-21da1137b860",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>original_transcript_id</th>\n",
       "      <th>original_transcript_name</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_triplet</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>g1_t1</td>\n",
       "      <td>g1_t1n</td>\n",
       "      <td>g1[1,1,1]</td>\n",
       "      <td>g1n[1,1,1]</td>\n",
       "      <td>g1n</td>\n",
       "      <td>[1,1,1]</td>\n",
       "      <td>g1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>g2_t1</td>\n",
       "      <td>g2_t1n</td>\n",
       "      <td>g2[1,1,1]</td>\n",
       "      <td>g2n[1,1,1]</td>\n",
       "      <td>g2n</td>\n",
       "      <td>[1,1,1]</td>\n",
       "      <td>g2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>g2_t2</td>\n",
       "      <td>g2_t2n</td>\n",
       "      <td>g2[1,1,1]</td>\n",
       "      <td>g2n[1,1,1]</td>\n",
       "      <td>g2n</td>\n",
       "      <td>[1,1,1]</td>\n",
       "      <td>g2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  original_transcript_id original_transcript_name transcript_id  \\\n",
       "0                  g1_t1                   g1_t1n     g1[1,1,1]   \n",
       "1                  g2_t1                   g2_t1n     g2[1,1,1]   \n",
       "2                  g2_t2                   g2_t2n     g2[1,1,1]   \n",
       "\n",
       "  transcript_name gene_name transcript_triplet gene_id  \n",
       "0      g1n[1,1,1]       g1n            [1,1,1]      g1  \n",
       "1      g2n[1,1,1]       g2n            [1,1,1]      g2  \n",
       "2      g2n[1,1,1]       g2n            [1,1,1]      g2  "
      ]
     },
     "execution_count": 209,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "m_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "id": "6d54891e-6260-4a1b-9782-3836b53b1005",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>feature_rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>gene</td>\n",
       "      <td>g1</td>\n",
       "      <td>g1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>transcript</td>\n",
       "      <td>g1</td>\n",
       "      <td>g1</td>\n",
       "      <td>g1_t1</td>\n",
       "      <td>g1_t1</td>\n",
       "      <td>1</td>\n",
       "      <td>30</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>g1</td>\n",
       "      <td>g1</td>\n",
       "      <td>g1_t1</td>\n",
       "      <td>g1_t1</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>g1</td>\n",
       "      <td>g1</td>\n",
       "      <td>g1_t1</td>\n",
       "      <td>g1_t1</td>\n",
       "      <td>14</td>\n",
       "      <td>20</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>g1</td>\n",
       "      <td>g1</td>\n",
       "      <td>g1_t1</td>\n",
       "      <td>g1_t1</td>\n",
       "      <td>25</td>\n",
       "      <td>30</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>gene</td>\n",
       "      <td>g2</td>\n",
       "      <td>g2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6</td>\n",
       "      <td>95</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>transcript</td>\n",
       "      <td>g2</td>\n",
       "      <td>g2</td>\n",
       "      <td>g2_t1</td>\n",
       "      <td>g2_t1</td>\n",
       "      <td>8</td>\n",
       "      <td>90</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>g2</td>\n",
       "      <td>g2</td>\n",
       "      <td>g2_t1</td>\n",
       "      <td>g2_t1</td>\n",
       "      <td>60</td>\n",
       "      <td>90</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>g2</td>\n",
       "      <td>g2</td>\n",
       "      <td>g2_t1</td>\n",
       "      <td>g2_t1</td>\n",
       "      <td>30</td>\n",
       "      <td>45</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>g2</td>\n",
       "      <td>g2</td>\n",
       "      <td>g2_t1</td>\n",
       "      <td>g2_t1</td>\n",
       "      <td>8</td>\n",
       "      <td>10</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>transcript</td>\n",
       "      <td>g2</td>\n",
       "      <td>g2</td>\n",
       "      <td>g2_t2</td>\n",
       "      <td>g2_t2</td>\n",
       "      <td>6</td>\n",
       "      <td>95</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>g2</td>\n",
       "      <td>g2</td>\n",
       "      <td>g2_t2</td>\n",
       "      <td>g2_t2</td>\n",
       "      <td>60</td>\n",
       "      <td>95</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>g2</td>\n",
       "      <td>g2</td>\n",
       "      <td>g2_t2</td>\n",
       "      <td>g2_t2</td>\n",
       "      <td>30</td>\n",
       "      <td>45</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>g2</td>\n",
       "      <td>g2</td>\n",
       "      <td>g2_t2</td>\n",
       "      <td>g2_t2</td>\n",
       "      <td>6</td>\n",
       "      <td>10</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome Strand     Feature gene_name gene_id transcript_id  \\\n",
       "0          1      +        gene        g1      g1           NaN   \n",
       "0          1      +  transcript        g1      g1         g1_t1   \n",
       "0          1      +        exon        g1      g1         g1_t1   \n",
       "1          1      +        exon        g1      g1         g1_t1   \n",
       "2          1      +        exon        g1      g1         g1_t1   \n",
       "1          1      -        gene        g2      g2           NaN   \n",
       "1          1      -  transcript        g2      g2         g2_t1   \n",
       "0          1      -        exon        g2      g2         g2_t1   \n",
       "1          1      -        exon        g2      g2         g2_t1   \n",
       "2          1      -        exon        g2      g2         g2_t1   \n",
       "2          1      -  transcript        g2      g2         g2_t2   \n",
       "0          1      -        exon        g2      g2         g2_t2   \n",
       "1          1      -        exon        g2      g2         g2_t2   \n",
       "2          1      -        exon        g2      g2         g2_t2   \n",
       "\n",
       "  transcript_name  Start  End  feature_rank  \n",
       "0             NaN      1   30             0  \n",
       "0           g1_t1      1   30             1  \n",
       "0           g1_t1      1   10             2  \n",
       "1           g1_t1     14   20             2  \n",
       "2           g1_t1     25   30             2  \n",
       "1             NaN      6   95             0  \n",
       "1           g2_t1      8   90             1  \n",
       "0           g2_t1     60   90             2  \n",
       "1           g2_t1     30   45             2  \n",
       "2           g2_t1      8   10             2  \n",
       "2           g2_t2      6   95             1  \n",
       "0           g2_t2     60   95             2  \n",
       "1           g2_t2     30   45             2  \n",
       "2           g2_t2      6   10             2  "
      ]
     },
     "execution_count": 210,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "id": "61e4c6a4-7e68-429c-9813-862c9e5eafb8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Chromosome         object\n",
       "Strand             object\n",
       "Feature            object\n",
       "gene_name          object\n",
       "gene_id            object\n",
       "transcript_id      object\n",
       "transcript_name    object\n",
       "Start               int64\n",
       "End                 int64\n",
       "feature_rank        int64\n",
       "dtype: object"
      ]
     },
     "execution_count": 163,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_df.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61f503ed-de47-4d30-94b6-9cfb9b500f14",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # tests for update_gtf_ends()\n",
    "# def make_gtf(c,e,s,g,t):\n",
    "\n",
    "\n",
    "# # t1 - transcript that doesn't need to be merged\n",
    "# n = 3\n",
    "# c = ['1' for i in range(n)]\n",
    "# e [[1,10], [14,20], [25,30]]\n",
    "# s = ['+' for i in range(n)]\n",
    "# g = 'g1'\n",
    "# t = 'g1_t1'\n",
    "\n",
    "# # t2 - rev. strand transcript that needs ends to be updated\n",
    "# n = 3\n",
    "# c = ['1' for i in range(n)]\n",
    "# e = [[90,60], [45,30], [10,8]]\n",
    "# s = ['-' for i in range(n)]\n",
    "# g = 'g2'\n",
    "# t = 'g1_t1'\n",
    "\n",
    "# df = pd.DataFrame()\n",
    "# df['Chromosome'] = c\n",
    "# df['Start'] = [i[0] for i in e]\n",
    "# df['End'] = [i[1] for i in e]\n",
    "# df['Strand'] = s\n",
    "# cols = ['gene_name', 'gene_id']\n",
    "# for c in cols:\n",
    "#     df[c] = g\n",
    "# cols = []\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
