{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7e4387d2-7683-4c42-93b4-5776e7107cac",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from cerberus.cerberus import *"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a830a56-85b7-48bd-ac02-102f46c48af2",
   "metadata": {},
   "source": [
    "## end aggregation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ead1e459-9ea7-40f0-9115-04a62e31866c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def test_agg_2_ends_1(print_dfs=False):\n",
    "    \"\"\"\n",
    "    Test agg_2_ends w/ and w/o end adding\n",
    "    \"\"\"\n",
    "    \n",
    "    def get_test(mode='tss'):\n",
    "        # example should have \n",
    "        # - entries that overlap\n",
    "        # - entries that don't overlap but are within a certain distance\n",
    "        # - entries that are unique to either\n",
    "        # - entries that overlap but aren't using the same gene_id / strand (these will be \n",
    "        #      equivalent situtations b/c the gene ids for things on different strands will always\n",
    "        #      differ\n",
    "\n",
    "        # later - beds that don't have strand\n",
    "        # later - beds that don't have gid\n",
    "\n",
    "        buffer = 20\n",
    "        \n",
    "        n = 4\n",
    "        c = ['1' for i in range(n)]\n",
    "        s = ['+' for i in range(n)]\n",
    "        st = [1, 200, 100, 300]\n",
    "        e = [15, 250, 110, 340]\n",
    "        n = ['gene1_1', 'gene1_2', 'gene1_3', 'gene1_4']\n",
    "        source = 'v1'\n",
    "        bed1 = make_end_df(c,s,st,e,n, source, mode)\n",
    "        bed1 = pr.PyRanges(bed1)\n",
    "\n",
    "        n = 4\n",
    "        c = ['1' for i in range(n)]\n",
    "        s = ['+' for i in range(n)]\n",
    "        st = [5, 120, 500, 200]\n",
    "        e = [10, 140, 550, 250]\n",
    "        n = ['gene1_1', 'gene1_2', 'gene1_3', 'gene2_1']\n",
    "        source = 'v2'\n",
    "        bed2 = make_end_df(c,s,st,e,n, source, mode)\n",
    "        bed2 = pr.PyRanges(bed2)\n",
    "\n",
    "        return bed1, bed2\n",
    "    \n",
    "    def get_ctrl(add=True):\n",
    "        \n",
    "        mode = 'tss'\n",
    "        \n",
    "        n = 6\n",
    "        c = ['1' for i in range(n)]\n",
    "        s = ['+' for i in range(n)]\n",
    "        st = [1,200,100,300, 200,500]\n",
    "        e = [15,250,110,340, 250,550]\n",
    "        n = ['gene1_1', 'gene1_2', 'gene1_3', 'gene1_4', 'gene2_1', 'gene1_5']\n",
    "        source = ['v1,v2','v1','v1,v2','v1', 'v2','v2']\n",
    "        df = make_end_df(c,s,st,e,n, source,mode)\n",
    "\n",
    "        # convert a few dtypes\n",
    "        df['Strand'] = df['Strand'].astype('category')\n",
    "        df['Chromosome'] = df['Chromosome'].astype('category')\n",
    "\n",
    "        # if we're not adding new ends\n",
    "        if not add:\n",
    "            df = df.loc[df.source != 'v2']\n",
    "            \n",
    "        # remove unnecessary columns\n",
    "        df.drop(['gene_id', 'id', mode], axis=1, inplace=True)\n",
    "        \n",
    "        # fix ids \n",
    "        df.reset_index(drop=True, inplace=True)\n",
    "\n",
    "        return df\n",
    "\n",
    "    tests = [True, False]\n",
    "    for add_ends in tests:\n",
    "        \n",
    "        buffer = 20\n",
    "        mode = 'tss'\n",
    "        sort_cols = ['Chromosome', 'Strand', 'gene_id', 'Start', 'End']\n",
    "        order = ['Chromosome', 'Start', 'End', 'Strand', 'Name',\n",
    "                 'gene_id', 'source', mode]\n",
    "        bed1, bed2 = get_test()\n",
    "        df = agg_2_ends(bed1, bed2,\n",
    "                        strand=True, \n",
    "                        gid=True,\n",
    "                        buffer=buffer,\n",
    "                        add_ends=add_ends,\n",
    "                        mode=mode)\n",
    "        \n",
    "        test = format_end_df(df)\n",
    "        ctrl = get_ctrl(add=add_ends)\n",
    "        \n",
    "        if print_dfs:\n",
    "            print('test')\n",
    "            print(test)\n",
    "            print(test.index)\n",
    "            print(test.dtypes)\n",
    "            print('ctrl')\n",
    "            print(ctrl)\n",
    "            print(ctrl.index)\n",
    "            print(ctrl.dtypes)\n",
    "\n",
    "        pd.testing.assert_frame_equal(ctrl, test, check_like=True)\n",
    "        \n",
    "        assert len(ctrl.index) == len(test.index)\n",
    "        \n",
    "def make_end_df(c,s,st,e,n, source,mode):\n",
    "    df = pd.DataFrame()\n",
    "    cols = ['Chromosome', 'Strand', 'Start', 'End', 'Name']\n",
    "    var = [c,s,st,e,n]\n",
    "    for col, var in zip(cols, var):\n",
    "        if type(var) == list:\n",
    "            df[col] = var\n",
    "            \n",
    "    # add source\n",
    "    df['source'] = source\n",
    "\n",
    "    df = format_end_df(df)\n",
    "\n",
    "    # get end # and gene id\n",
    "    if any(df.Name.isnull()):\n",
    "        df['gene_id'] = np.nan\n",
    "        df[mode] = np.nan\n",
    "    else:\n",
    "        df['gene_id'] = df.Name.str.split('_', expand=True)[0]\n",
    "        df[mode] = df.Name.str.split('_', expand=True)[1]\n",
    "    \n",
    "    # get arbitrary unique ids\n",
    "    df['id'] = [i for i in range(len(df.index))]\n",
    "\n",
    "    return df\n",
    "\n",
    "def format_end_df(df):\n",
    "    sort_cols = ['Chromosome', 'Start', 'End', 'Strand']\n",
    "    df = df.sort_values(by=sort_cols)\n",
    "    order = ['Chromosome', 'Start', 'End', 'Strand', 'Name', 'source']\n",
    "    order = [o for o in order if o in df.columns]\n",
    "    df = df[order]\n",
    "    df.reset_index(drop=True, inplace=True)\n",
    "    return df\n",
    "\n",
    "def test_agg_2_ends_2(print_dfs=False):\n",
    "    \"\"\"\n",
    "    Test agg_2_ends w/ and w/o end adding\n",
    "    \"\"\"\n",
    "    \n",
    "    def get_test(mode='tss'):\n",
    "        \n",
    "        # adding a bed file that doesn't have strand or gid info\n",
    "        # TODO add an entry in bed2 that is duplicated based on \n",
    "        # - strandedness\n",
    "        # - gene id\n",
    "        # in bed1\n",
    "\n",
    "        buffer = 20\n",
    "        \n",
    "        n = 6\n",
    "        c = ['1' for i in range(n)]\n",
    "        s = ['+' for i in range(n)]\n",
    "        s[-1] = '-'\n",
    "        st = [1, 200, 100, 300, 260, 260]\n",
    "        e = [15, 250, 110, 340, 290, 290]\n",
    "        n = ['gene1_1', 'gene1_2', 'gene1_3', 'gene1_4', 'gene2_1', 'gene3_1']\n",
    "        source = 'v1'\n",
    "        bed1 = make_end_df(c,s,st,e,n, source, mode)\n",
    "        bed1 = pr.PyRanges(bed1)\n",
    "\n",
    "        n = 4\n",
    "        c = ['1' for i in range(n)]\n",
    "        s = [np.nan for i in range(n)]\n",
    "        st = [5, 120, 500, 200]\n",
    "        e = [10, 140, 550, 250]\n",
    "        n = [np.nan for i in range(n)]\n",
    "        source = 'v2'\n",
    "        bed2 = make_end_df(c,s,st,e,n, source, mode)\n",
    "        bed2 = pr.PyRanges(bed2)\n",
    "\n",
    "        return bed1, bed2\n",
    "    \n",
    "    def get_ctrl(add=True):\n",
    "        \n",
    "        mode = 'tss'\n",
    "        \n",
    "        n = 6\n",
    "        c = ['1' for i in range(n)]\n",
    "        s = ['+' for i in range(n)]\n",
    "        s[-1] = '-'\n",
    "        st = [1,200,100,300, 260,260]\n",
    "        e = [15,250,110,340, 290,290]\n",
    "        n = ['gene1_1', 'gene1_2', 'gene1_3', 'gene1_4', 'gene2_1', 'gene3_1']\n",
    "        source = ['v1,v2','v1,v2','v1,v2','v1', 'v1,v2', 'v1,v2']\n",
    "        df = make_end_df(c,s,st,e,n, source,mode)\n",
    "\n",
    "        # convert a few dtypes\n",
    "        df['Strand'] = df['Strand'].astype('category')\n",
    "        df['Chromosome'] = df['Chromosome'].astype('category')\n",
    "\n",
    "        # if we're not adding new ends\n",
    "        if not add:\n",
    "            df = df.loc[df.source != 'v2']\n",
    "            \n",
    "        # remove unnecessary columns\n",
    "        df.drop(['gene_id', 'id', mode], axis=1, inplace=True)\n",
    "        \n",
    "        # fix ids \n",
    "        df.reset_index(drop=True, inplace=True)\n",
    "\n",
    "        return df\n",
    "\n",
    "    strand = False\n",
    "    gid = False\n",
    "    add_ends = False\n",
    "        \n",
    "    buffer = 20\n",
    "    mode = 'tss'\n",
    "    sort_cols = ['Chromosome', 'Strand', 'gene_id', 'Start', 'End']\n",
    "    order = ['Chromosome', 'Start', 'End', 'Strand', 'Name',\n",
    "             'gene_id', 'source', mode]\n",
    "    bed1, bed2 = get_test()\n",
    "    df = agg_2_ends(bed1, bed2,\n",
    "                    strand=strand, \n",
    "                    gid=gid,\n",
    "                    buffer=buffer,\n",
    "                    add_ends=add_ends,\n",
    "                    mode=mode)\n",
    "    test = format_end_df(df)\n",
    "    ctrl = get_ctrl(add=add_ends)\n",
    "\n",
    "    if print_dfs:\n",
    "        print('test')\n",
    "        print(test)\n",
    "        print(test.index)\n",
    "        print(test.dtypes)\n",
    "        print('ctrl')\n",
    "        print(ctrl)\n",
    "        print(ctrl.index)\n",
    "        print(ctrl.dtypes)\n",
    "\n",
    "    pd.testing.assert_frame_equal(ctrl, test, check_like=True)\n",
    "        \n",
    "    assert len(ctrl.index) == len(test.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "bc09f92b-4165-4a17-9cce-7b1c386523f2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "test\n",
      "  Chromosome  Start  End Strand     Name source\n",
      "0          1      1   15      +  gene1_1  v1,v2\n",
      "1          1    100  110      +  gene1_3  v1,v2\n",
      "2          1    200  250      +  gene1_2  v1,v2\n",
      "3          1    260  290      +  gene2_1  v1,v2\n",
      "4          1    260  290      -  gene3_1  v1,v2\n",
      "5          1    300  340      +  gene1_4     v1\n",
      "RangeIndex(start=0, stop=6, step=1)\n",
      "Chromosome    category\n",
      "Start            int64\n",
      "End              int64\n",
      "Strand        category\n",
      "Name            object\n",
      "source          object\n",
      "dtype: object\n",
      "ctrl\n",
      "  Chromosome  Start  End Strand     Name source\n",
      "0          1      1   15      +  gene1_1  v1,v2\n",
      "1          1    100  110      +  gene1_3  v1,v2\n",
      "2          1    200  250      +  gene1_2  v1,v2\n",
      "3          1    260  290      +  gene2_1  v1,v2\n",
      "4          1    260  290      -  gene3_1  v1,v2\n",
      "5          1    300  340      +  gene1_4     v1\n",
      "RangeIndex(start=0, stop=6, step=1)\n",
      "Chromosome    category\n",
      "Start            int64\n",
      "End              int64\n",
      "Strand        category\n",
      "Name            object\n",
      "source          object\n",
      "dtype: object\n"
     ]
    }
   ],
   "source": [
    "test_agg_2_ends_1(print_dfs=False)\n",
    "test_agg_2_ends_2(print_dfs=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "898d0edd-82be-4a4d-87ab-a231ddb400bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# def agg_2_ends(bed1, bed2,\n",
    "#                strand,\n",
    "#                gid,\n",
    "#                buffer,\n",
    "#                add_ends,\n",
    "#                mode):\n",
    "#     \"\"\"\n",
    "#     Parameters:\n",
    "#         bed1 (pyranges PyRanges): Bed PyRanges object for existing ends\n",
    "#         bed2 (pyranges PyRanges): Bed PyRanges object for new ends\n",
    "#         buffer (int): Maximum allowable distance between ends in bed1 and bed2 \n",
    "#             to call them the same end\n",
    "#         add_ends (bool): Whether to initialize new regions from bed2\n",
    "#     \"\"\"\n",
    "    \n",
    "#     source1 = bed1.df.source.unique().tolist()[0]\n",
    "#     source2 = bed2.df.source.unique().tolist()[0]\n",
    "    \n",
    "#     new_c = '{}_new'.format(mode)\n",
    "#     max_c = '{}_max'.format(mode)\n",
    "    \n",
    "#     # depending on whether the new bed has strand information, \n",
    "#     # construct the join call\n",
    "#     if strand:\n",
    "#         temp_joined = bed1.join(bed2, \n",
    "#             strandedness='same',\n",
    "#             suffix='_new',\n",
    "#             slack=buffer,\n",
    "#             how='left')\n",
    "#     elif not strand:\n",
    "#         temp_joined = bed1.join(bed2, \n",
    "#             strandedness=False,\n",
    "#             suffix='_new',\n",
    "#             slack=buffer,\n",
    "#             how='left')\n",
    "        \n",
    "\n",
    "#     temp_joined = temp_joined.df\n",
    "#     # print(temp_joined)\n",
    "#     df = pd.DataFrame()\n",
    "\n",
    "#     ### old ends ###\n",
    "\n",
    "#     # situation 1: ends match across the datasets in coord and gene id\n",
    "#     temp = temp_joined.loc[temp_joined.gene_id == temp_joined.gene_id_new].copy(deep=True)\n",
    "#     temp.source = temp.source+','+temp.source_new\n",
    "#     # print(temp)\n",
    "#     df = pd.concat([df, temp])\n",
    "#     # print(df)\n",
    "\n",
    "#     # situation 2: ends are only in the first dataset\n",
    "#     temp = temp_joined.loc[(temp_joined.gene_id_new == '-1')|(temp_joined.gene_id!=temp_joined.gene_id_new)].copy(deep=True)\n",
    "#     # print('here')\n",
    "#     # print(temp)\n",
    "#     df = pd.concat([df, temp])\n",
    "#     # print('df before add_aends')\n",
    "#     # print(df)\n",
    "#     # print()\n",
    "\n",
    "#     # restrict to relevant columns\n",
    "#     cols = ['Chromosome', 'Start', 'End', 'Strand',\n",
    "#             'Name', 'gene_id', 'source', mode, 'id_new']\n",
    "#     df = df[cols]\n",
    "#     df.rename({'id_new': 'id'}, axis=1, inplace=True)\n",
    "\n",
    "#     ### new ends, only add if we're allowing them to be independent\n",
    "#     ### end support\n",
    "#     if add_ends:\n",
    "        \n",
    "#         new_df = pd.DataFrame()\n",
    "        \n",
    "#         drop_cols = ['Start', 'End', 'Strand', 'gene_id', 'source', 'Name', mode]\n",
    "#         m = {'Start_new': 'Start', \n",
    "#              'End_new': 'End',\n",
    "#              'gene_id_new': 'gene_id',\n",
    "#              'Strand_new': 'Strand',\n",
    "#              'source_new': 'source',\n",
    "#              'Name_new': 'Name', \n",
    "#              new_c: mode}\n",
    "\n",
    "#         # situation 3: the ends overlapped, but the gene ids didn't match\n",
    "#         temp = temp_joined.loc[(temp_joined.gene_id!=temp_joined.gene_id_new)&(temp_joined.gene_id_new!='-1')].copy(deep=True)\n",
    "#         temp.drop(drop_cols, axis=1, inplace=True)\n",
    "#         temp.rename(m, axis=1, inplace=True)\n",
    "#         # print('hewwo')\n",
    "#         # print(temp.head())\n",
    "#         # df = pd.concat([df, temp])\n",
    "#         # print('situation 3 new ends: ')\n",
    "#         # print(temp)\n",
    "#         new_df = pd.concat([new_df, temp])\n",
    "#         # print(df)\n",
    "\n",
    "#         # situation 4: the ends are brand new and didn't overlap at all in the existing ends\n",
    "#         bed2 = bed2.df\n",
    "#         # print(bed2.id.dtypes)\n",
    "#         # print(temp.id.dtypes)\n",
    "#         # print(bed2.id.tolist())\n",
    "#         # print(df.id.tolist())\n",
    "#         # print('id check')\n",
    "#         # print('df')\n",
    "#         # print(df.head())\n",
    "    \n",
    "#         # print('bed2')\n",
    "#         # print(bed2.head())\n",
    "#         inds = list(set(bed2.id.tolist())-set(df.id.tolist()))\n",
    "#         # print(inds)\n",
    "#         temp = bed2.loc[inds]\n",
    "#         # print(temp)\n",
    "#         # df = pd.concat([df, temp])\n",
    "#         # print('situation 4 new ends: ')\n",
    "#         # print(temp)\n",
    "#         new_df = pd.concat([new_df, temp])\n",
    "        \n",
    "#         # print('bepis')\n",
    "#         # print(df.head())\n",
    "        \n",
    "#         # if we added new ends, number those according to which gene they're from \n",
    "#         # n = int(df[mode].max())+1\n",
    "#         # new_ends = df.loc[df.source == source2].index\n",
    "#         # nums = [i for i in range(n, \n",
    "        \n",
    "#         # print(df.head())\n",
    "#         # temp = df.loc[df.source.str.contains(source1)].copy(deep=True)\n",
    "#         # print(source1)\n",
    "#         # print(temp.head())\n",
    "        \n",
    "#         g_maxes = get_gene_feat_max(df, mode)\n",
    "#         new_df = renumber_new_feats(new_df, g_maxes, mode)\n",
    "        \n",
    "# #         # get max end number per gene from existing ends\n",
    "# #         df[mode] = df[mode].astype(int)\n",
    "# #         temp = df[['gene_id', mode]].copy(deep=True)\n",
    "# #         temp = temp.groupby('gene_id').max().reset_index()\n",
    "# #         temp.rename({mode: max_c}, axis=1, inplace=True)\n",
    "# # #         print('new_df')\n",
    "# # #         print(new_df.head())\n",
    "# # #         print('temp')\n",
    "# # #         print(temp.head())\n",
    "        \n",
    "# #         # merge with new ends that are being added\n",
    "# #         new_df = new_df.merge(temp, how='left', on='gene_id')\n",
    "# #         new_df[max_c].fillna(0, inplace=True)\n",
    "# #         # df = df.merge(temp, how='left', on='gene_id')\n",
    "# #         sort_cols = ['gene_id', mode]\n",
    "        \n",
    "# #         # renumber new ends \n",
    "# #         new_df[new_c] = new_df.sort_values(by=sort_cols,\n",
    "# #                                         ascending=[True, True])\\\n",
    "# #                                         .groupby(['gene_id'])\\\n",
    "# #                                         .cumcount()+1\n",
    "# #         new_df[new_c] = new_df[new_c].astype(int) + new_df[max_c].astype(int)\n",
    "        \n",
    "#         # some df formatting\n",
    "#         new_df.drop([mode, max_c], axis=1, inplace=True)\n",
    "#         new_df.rename({new_c: mode}, axis=1, inplace=True)\n",
    "#         # print(new_df.head())\n",
    "        \n",
    "#         # finally, concatenate new and old df\n",
    "#         df = pd.concat([df, new_df])\n",
    "        \n",
    "\n",
    "\n",
    "#         # print(df[['gene_id', mode]].dtypes)\n",
    "#         # print(temp[['gene_id', mode]].dtypes)\n",
    "# #         print(df[['Name', mode]])\n",
    "# #         print(df['Name'].map(m, na_action='ignore'))\n",
    "# #         df[mode] = df['Name'].map(m, na_action='ignore')\n",
    "# #         print(m)\n",
    "# #         print('hewwo?')                \n",
    "# #         print()\n",
    "# #         print()\n",
    "# #         print(temp)\n",
    "# #         print('break')\n",
    "# #         print(df)\n",
    "# #         df['Name'] = df.gene_id+'_'+df[mode]\n",
    "        \n",
    "    \n",
    "#     # add missing columns\n",
    "#     # ie mode, gene_id, Strand?\n",
    "\n",
    "#     # drop unnecessary columns and create new names\n",
    "#     # and do some extra formatting\n",
    "#     df.drop('id', axis=1, inplace=True)\n",
    "#     df['Name'] = df.gene_id+'_'+df[mode].astype(str)\n",
    "#     df[mode] = df[mode].astype(int)\n",
    "#     # keep_cols = ['Chromosome', 'Start', 'End', 'Strand', 'gene_id', 'source', mode]\n",
    "#     # df = df[keep_cols]\n",
    "    \n",
    "#     return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "37d66c99-931c-4480-a75b-eecf1262563a",
   "metadata": {},
   "outputs": [],
   "source": [
    "test_agg_2_ends_1(print_dfs=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "a47c700a-161c-4c02-a540-064a7005438d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def format_agg_2_ends_bed(bed, mode):\n",
    "    cols = ['Name', 'gene_id', mode, 'Strand']\n",
    "    \n",
    "    gid = True\n",
    "    strand = True\n",
    "    for col in cols:\n",
    "        if col not in bed.cols:\n",
    "            bed[col] = [np.nan for i in range(len(bed.index))]\n",
    "            \n",
    "            # record some bools\n",
    "            if col == 'gene_id':\n",
    "                gid = False\n",
    "            elif col == 'Strand':\n",
    "                strand = False\n",
    "                \n",
    "    return bed, gid, strand"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "id": "6d851cc6-2f86-4390-b569-ac884c1da05f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def aggregate_ends(beds, sources, add_ends, mode):\n",
    "    \n",
    "    df = pd.DataFrame()\n",
    "    i = 0\n",
    "    for bed_fname, source, add in zip(beds, sources, add_ends):\n",
    "        \n",
    "        bed = read_bed(bed_fname, mode)\n",
    "        bed = bed.df\n",
    "        bed['source'] = source\n",
    "        bed = pr.PyRanges(bed)\n",
    "        \n",
    "        # first bed; just accept all these ends\n",
    "        if len(df.index) == 0:\n",
    "            \n",
    "            if 'gene_id' not in bed.df.columns and 'Strand' not in bed.df.columns:\n",
    "                raise Exception('First bed must contain Strand and gene_id columns')\n",
    "                \n",
    "            df = bed.df\n",
    "        \n",
    "        # more than one bed; merge and reconcile ends\n",
    "        else:\n",
    "            \n",
    "            if 'gene_id' not in bed.df.columns or 'Strand' not in bed.df.columns:\n",
    "                if add:\n",
    "                    raise Exception('Cannot add new ends from {} because '+\\\n",
    "                                    'it does not contain gene_id information.')\n",
    "            \n",
    "            # add missing columns but keep track of what information we'll be \n",
    "            # able to merge on \n",
    "            bed, gid, strand = format_agg_2_ends_bed(bed, mode)\n",
    "            \n",
    "            # \n",
    "            df = agg_2_ends(df, bed,\n",
    "                            strand, gid,\n",
    "                            buffer, add, mode)\n",
    "        \n",
    "        # \n",
    "            i += 1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 312,
   "id": "79f19b07-b632-4fca-966c-6128ddffe065",
   "metadata": {},
   "outputs": [],
   "source": [
    "fname = 'tests/files/Canx_1_tss.bed'\n",
    "df = read_bed(fname, mode='tss')\n",
    "\n",
    "fname = 'tests/files/Canx_ccre_pels.bed'\n",
    "# df = read_bed(fname, mode='tss')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 313,
   "id": "5248f4a9-e81e-4b2e-b0d8-7f94ad4e546b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>tss</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr11</td>\n",
       "      <td>50298268</td>\n",
       "      <td>50298369</td>\n",
       "      <td>ENSMUSG00000020368_3</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000020368</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr11</td>\n",
       "      <td>50300939</td>\n",
       "      <td>50301040</td>\n",
       "      <td>ENSMUSG00000020368_2</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000020368</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr11</td>\n",
       "      <td>50325622</td>\n",
       "      <td>50325723</td>\n",
       "      <td>ENSMUSG00000020368_1</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000020368</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+----------------------+------------+-------+\n",
       "| Chromosome   |     Start |       End | Name                 | Score      | +3    |\n",
       "| (category)   |   (int32) |   (int32) | (object)             | (object)   | ...   |\n",
       "|--------------+-----------+-----------+----------------------+------------+-------|\n",
       "| chr11        |  50298268 |  50298369 | ENSMUSG00000020368_3 | .          | ...   |\n",
       "| chr11        |  50300939 |  50301040 | ENSMUSG00000020368_2 | .          | ...   |\n",
       "| chr11        |  50325622 |  50325723 | ENSMUSG00000020368_1 | .          | ...   |\n",
       "+--------------+-----------+-----------+----------------------+------------+-------+\n",
       "Stranded PyRanges object has 3 rows and 8 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "3 hidden columns: Strand, gene_id, tss"
      ]
     },
     "execution_count": 313,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c746f86-23b0-4efd-93f8-b6389dd2bc77",
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "unexpected EOF while parsing (<ipython-input-245-2533ef766177>, line 32)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-245-2533ef766177>\"\u001b[0;36m, line \u001b[0;32m32\u001b[0m\n\u001b[0;31m    \u001b[0m\n\u001b[0m    ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m unexpected EOF while parsing\n"
     ]
    }
   ],
   "source": [
    "# keeping entries from the old end df\n",
    "\n",
    "# if we have strand information, enforce that\n",
    "if 'Strand' in bed.columns:\n",
    "    temp = df.join(bed,\n",
    "                   strandedness='same',\n",
    "                   suffix='_new',\n",
    "                   slack=buffer,\n",
    "                   how='left')\n",
    "else:\n",
    "    temp = df.join(bed,\n",
    "                   strandedness=False,\n",
    "                   suffix='_new', \n",
    "                   slack=buffer,\n",
    "                   how='left')\n",
    "\n",
    "# convert back to df\n",
    "temp = temp.df\n",
    "print(temp.head())\n",
    "\n",
    "# if we have gene id information, also enforce that\n",
    "if 'gene_id' in bed.columns:\n",
    "    temp = temp.loc[(temp.gene_id == temp.gene_id_new)|(temp.gene_id_new == '-1')]\n",
    "\n",
    "# reconcile sources\n",
    "inds = temp.loc[temp.gene_id_new != '-1'].index\n",
    "new_sources = temp.loc[inds, 'source']+','+temp.loc[inds, 'source_new']\n",
    "temp.loc[inds, 'source'] = new_sources\n",
    "\n",
    "# if we're adding in new ends from here\n",
    "if add:\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d6565405-bf7b-4f4b-9cd4-ff59a4540e7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # first subset the human ccres for those that are promoters \n",
    "# # and are w/i genomic region of Canx\n",
    "# #  https://api.wenglab.org/screen_v13/fdownloads/V3/GRCh38-cCREs.bed\n",
    "# fname = '/Users/fairliereese/mortazavi_lab/ref/ccre_hg38/GRCh38-cCREs.bed'\n",
    "# df = pr.read_bed(fname)\n",
    "# df = df['chr11', 50293961:50325673]\n",
    "\n",
    "# # ok so the toy gtf I have is definitely mouse oops\n",
    "# # https://api.wenglab.org/screen_v13/fdownloads/V3/mm10-ccREs.bed\n",
    "# fname = '/Users/fairliereese/mortazavi_lab/ref/ccre_vM3/mm10-ccREs.bed'\n",
    "# df = pr.read_bed(fname)\n",
    "# df = df['chr11', 50293961:50325673].df\n",
    "# df = df.loc[df.Strand.str.contains('pELS')] \n",
    "# df.to_csv('tests/files/Canx_ccre_pels.bed', sep='\\t', index=False, header=None)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
