{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7e4387d2-7683-4c42-93b4-5776e7107cac",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import h5py \n",
    "import os\n",
    "\n",
    "from cerberus.cerberus import *\n",
    "from cerberus.main import *"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6af365ed-1ee0-4693-b68b-8d98db211794",
   "metadata": {},
   "source": [
    "## ddebugging agg_gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "1a750e53-3eb4-45dc-abc2-0aa6d6cc7562",
   "metadata": {},
   "outputs": [],
   "source": [
    "def agg_gtf(df):\n",
    "    \"\"\"\n",
    "    Deduplicate GTF transcripts that have the same triplet id\n",
    "\n",
    "    Parameters:\n",
    "        df (pandas DataFrame): DF of gtf from `update_ends`\n",
    "\n",
    "    Returns:\n",
    "        df (pandas DataFrame): DF of gtf with deduplicated\n",
    "            transcript / exon entries based on the triplet id\n",
    "    \"\"\"\n",
    "\n",
    "    def collapse_non_gb_col(x):\n",
    "        x = x.fillna('')\n",
    "        x = x.astype(str)\n",
    "        x = x.unique().tolist()\n",
    "        x = ','.join(x)\n",
    "        if x == '':\n",
    "            x = np.nan\n",
    "        return(x)\n",
    "\n",
    "    gb_cols = ['Chromosome',\n",
    "                 'Feature',\n",
    "                 'Start', 'End',\n",
    "                 'Score', 'Strand', 'Frame', 'gene_id', 'gene_name',\n",
    "                 'gene_status', 'gene_type', 'talon_gene',\n",
    "                 'ic', 'ic_id', 'tss_id', 'tss', 'tes_id', 'tes', 'transcript_id',\n",
    "                 'transcript_name']\n",
    "    gb_cols = list(set(df.columns)&set(gb_cols))\n",
    "\n",
    "    # get dictionary:function mapping for aggregation\n",
    "    agg_cols = list(set(df.columns)-set(gb_cols))\n",
    "    agg_dict = dict()\n",
    "    for c in agg_cols:\n",
    "        agg_dict[c] = collapse_non_gb_col\n",
    "\n",
    "    # get collapsed features to add to deduplicated df\n",
    "    t_df = df.loc[df.Feature == 'transcript'].copy(deep=True)\n",
    "    non_gb_cols = list(set(t_df.columns.tolist())-set(gb_cols))\n",
    "    t_df[non_gb_cols].fillna('', inplace=True)\n",
    "    t_df = t_df.groupby(gb_cols, observed=True).agg(agg_dict).reset_index()\n",
    "    t_df = t_df[['transcript_id']+non_gb_cols]\n",
    "    collapsed_feats = t_df.copy(deep=True)\n",
    "\n",
    "    # deduplicate df based only on transcript id\n",
    "    temp = df[['transcript_id', 'original_transcript_id']].drop_duplicates()\n",
    "    dupe_old_tids = temp.loc[temp.transcript_id.duplicated(keep='first'), 'original_transcript_id']\n",
    "    df = df.loc[~df.original_transcript_id.isin(dupe_old_tids)]\n",
    "\n",
    "    # replace the non gb columns with the ones that we already grouped\n",
    "    df.drop(non_gb_cols, axis=1, inplace=True)\n",
    "    df = df.merge(collapsed_feats, how='left', on='transcript_id')\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a8c0e72b-006f-464f-bf6d-bd9803e4c27c",
   "metadata": {},
   "outputs": [],
   "source": [
    "d = '/Users/fairliereese/Documents/programming/mortazavi_lab/data/rnawg/lr_bulk/cerberus/'\n",
    "h5 = '{}human_cerberus.h5'.format(d)\n",
    "gtf = '/Users/fairliereese/Documents/programming/mortazavi_lab/data/rnawg/lr_bulk/lapa/smol.gtf'\n",
    "update_ends = True\n",
    "agg = True\n",
    "o = '{}smol_replace.gtf'.format(d)\n",
    "\n",
    "# replace_gtf_ids(h5, gstf, update_ends, agg, o)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f4b58d58-9c4a-4a87-9ef1-5a15a3827983",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "9b85df4c-c263-45fc-9f16-fa57cbddf835",
   "metadata": {},
   "outputs": [],
   "source": [
    "print_dfs = False\n",
    "# def test_agg_gtf(print_dfs=True):\n",
    "\n",
    "# need:\n",
    "# transcripts that don't need to be aggregated\n",
    "# transcripts that do need to be aggregated\n",
    "\n",
    "ts = []\n",
    "\n",
    "# t1 - transcript that won't need to be aggregated\n",
    "n = 3\n",
    "c = ['1' for i in range(n)]\n",
    "e = [[1,10], [14,20], [25,30]]\n",
    "s = ['+' for i in range(n)]\n",
    "g = 'g1'\n",
    "nt = 'g1[1,1,1]'\n",
    "t = 'g1_t1'\n",
    "ag1='known'\n",
    "ag2='p1'\n",
    "df = make_exon_df(n,c,e,s,g,t,nt,ag1,ag2)\n",
    "df['tss_id'] = 'g1_1'\n",
    "df['tes_id'] = 'g1_1'\n",
    "ts.append(df)\n",
    "\n",
    "# t2 - similar transcript from g1 gene that won't need to be aggregated\n",
    "n = 3\n",
    "c = ['1' for i in range(n)]\n",
    "e = [[1,10], [14,20], [25,50]]\n",
    "s = ['+' for i in range(n)]\n",
    "g = 'g1'\n",
    "nt = 'g1[1,1,3]'\n",
    "t = 'g1_t2'\n",
    "ag1 = 'novel'\n",
    "ag2 = 'p1'\n",
    "df = make_exon_df(n,c,e,s,g,t,nt,ag1,ag2)\n",
    "df['tss_id'] = 'g1_1'\n",
    "df['tes_id'] = 'g1_3'\n",
    "ts.append(df)\n",
    "\n",
    "# t3 - transcript from a different gene that does need to be aggregated\n",
    "n = 3\n",
    "c = ['1' for i in range(n)]\n",
    "e = [[90,60], [45,30], [10,8]]\n",
    "s = ['-' for i in range(n)]\n",
    "g = 'g2'\n",
    "nt = 'g2[1,1,1]'\n",
    "t = 'g2_t1'\n",
    "ag1 = 'known'\n",
    "ag2 = 'p3'\n",
    "df = make_exon_df(n,c,e,s,g,t,nt,ag1,ag2)\n",
    "df['tss_id'] = 'g2_1'\n",
    "df['tes_id'] = 'g2_1'\n",
    "ts.append(df)\n",
    "\n",
    "# t4 - transcript that needs to be collapsed with t3\n",
    "n = 3\n",
    "c = ['1' for i in range(n)]\n",
    "e = [[90,60], [45,30], [10,8]]\n",
    "s = ['-' for i in range(n)]\n",
    "g = 'g2'\n",
    "nt = 'g2[1,1,1]'\n",
    "t = 'g2_t2'\n",
    "ag1 = 'novel'\n",
    "ag2 ='p4'\n",
    "df = make_exon_df(n,c,e,s,g,t,nt,ag1,ag2)\n",
    "df['tss_id'] = 'g2_1'\n",
    "df['tes_id'] = 'g2_1'\n",
    "ts.append(df)\n",
    "\n",
    "# make test gtf\n",
    "test_df = make_test_gtf(ts)\n",
    "test_df.rename({'transcript_id': 'original_transcript_id',\n",
    "                'transcript_name': 'original_transcript_name'},\n",
    "               axis=1, inplace=True)\n",
    "test_df.rename({'new_transcript_id': 'transcript_id'},\n",
    "               axis=1, inplace=True)\n",
    "test_df['transcript_name'] = test_df['transcript_id']\n",
    "\n",
    "# # make ctrl df\n",
    "# ts = []\n",
    "\n",
    "# # t1 - transcript that won't need to be aggregated\n",
    "# n = 3\n",
    "# c = ['1' for i in range(n)]\n",
    "# e = [[1,10], [14,20], [25,30]]\n",
    "# s = ['+' for i in range(n)]\n",
    "# g = 'g1'\n",
    "# t = 'g1[1,1,1]'\n",
    "# ag1='known'\n",
    "# ag2='p1'\n",
    "# df = make_exon_df(n,c,e,s,g,t,ag1=ag1,ag2=ag2)\n",
    "# df['original_transcript_id'] = 'g1_t1'\n",
    "# df['original_transcript_name'] = df['original_transcript_id']\n",
    "# df['tss_id'] = 'g1_1'\n",
    "# df['tes_id'] = 'g1_1'\n",
    "# ts.append(df)\n",
    "\n",
    "# # t2 - similar transcript from g1 gene that won't need to be aggregated\n",
    "# n = 3\n",
    "# c = ['1' for i in range(n)]\n",
    "# e = [[1,10], [14,20], [25,50]]\n",
    "# s = ['+' for i in range(n)]\n",
    "# g = 'g1'\n",
    "# t = 'g1[1,1,3]'\n",
    "# ag1 = 'novel'\n",
    "# ag2 = 'p1'\n",
    "# df = make_exon_df(n,c,e,s,g,t,ag1=ag1,ag2=ag2)\n",
    "# df['original_transcript_id'] = 'g1_t2'\n",
    "# df['original_transcript_name'] = df['original_transcript_id']\n",
    "# df['tss_id'] = 'g1_1'\n",
    "# df['tes_id'] = 'g1_3'\n",
    "# ts.append(df)\n",
    "\n",
    "# # t3 /t4 collapsed- transcript from a different gene that does need to be aggregated\n",
    "# n = 3\n",
    "# c = ['1' for i in range(n)]\n",
    "# e = [[90,60], [45,30], [10,8]]\n",
    "# s = ['-' for i in range(n)]\n",
    "# g = 'g2'\n",
    "# t = 'g2[1,1,1]'\n",
    "# ag1 = 'known,novel'\n",
    "# ag2 = 'p3,p4'\n",
    "# df = make_exon_df(n,c,e,s,g,t,ag1=ag1,ag2=ag2)\n",
    "# df['tss_id'] = 'g2_1'\n",
    "# df['tes_id'] = 'g2_1'\n",
    "# df['original_transcript_id'] = 'g2_t1,g2_t2'\n",
    "# df['original_transcript_name'] = df['original_transcript_id']\n",
    "# ts.append(df)\n",
    "\n",
    "\n",
    "# ctrl = pr.PyRanges(make_test_gtf(ts)).df\n",
    "# test = pr.PyRanges(agg_gtf(test_df)).df\n",
    "\n",
    "# ctrl.reset_index(inplace=True, drop=True)\n",
    "# test.reset_index(inplace=True, drop=True)\n",
    "\n",
    "# if print_dfs:\n",
    "#     print('test')\n",
    "#     print(test)\n",
    "#     print(test.index)\n",
    "#     print(test.dtypes)\n",
    "#     print('ctrl')\n",
    "#     print(ctrl)\n",
    "#     print(ctrl.index)\n",
    "#     print(ctrl.dtypes)\n",
    "\n",
    "# #     pd.testing.assert_frame_equal(ctrl, test, check_like=True)\n",
    "\n",
    "# #     assert len(ctrl.index) == len(test.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "299d626d-92a6-48e2-84a3-d720c11e3890",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>original_transcript_id</th>\n",
       "      <th>original_transcript_name</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>ag1</th>\n",
       "      <th>ag2</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>tss_id</th>\n",
       "      <th>tes_id</th>\n",
       "      <th>transcript_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>gene</td>\n",
       "      <td>g1</td>\n",
       "      <td>g1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>transcript</td>\n",
       "      <td>g1</td>\n",
       "      <td>g1</td>\n",
       "      <td>g1_t1</td>\n",
       "      <td>g1_t1</td>\n",
       "      <td>g1[1,1,1]</td>\n",
       "      <td>known</td>\n",
       "      <td>p1</td>\n",
       "      <td>1</td>\n",
       "      <td>30</td>\n",
       "      <td>g1_1</td>\n",
       "      <td>g1_1</td>\n",
       "      <td>g1[1,1,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>g1</td>\n",
       "      <td>g1</td>\n",
       "      <td>g1_t1</td>\n",
       "      <td>g1_t1</td>\n",
       "      <td>g1[1,1,1]</td>\n",
       "      <td>known</td>\n",
       "      <td>p1</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>g1_1</td>\n",
       "      <td>g1_1</td>\n",
       "      <td>g1[1,1,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>g1</td>\n",
       "      <td>g1</td>\n",
       "      <td>g1_t1</td>\n",
       "      <td>g1_t1</td>\n",
       "      <td>g1[1,1,1]</td>\n",
       "      <td>known</td>\n",
       "      <td>p1</td>\n",
       "      <td>14</td>\n",
       "      <td>20</td>\n",
       "      <td>g1_1</td>\n",
       "      <td>g1_1</td>\n",
       "      <td>g1[1,1,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>g1</td>\n",
       "      <td>g1</td>\n",
       "      <td>g1_t1</td>\n",
       "      <td>g1_t1</td>\n",
       "      <td>g1[1,1,1]</td>\n",
       "      <td>known</td>\n",
       "      <td>p1</td>\n",
       "      <td>25</td>\n",
       "      <td>30</td>\n",
       "      <td>g1_1</td>\n",
       "      <td>g1_1</td>\n",
       "      <td>g1[1,1,1]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome Strand     Feature gene_name gene_id original_transcript_id  \\\n",
       "0          1      +        gene        g1      g1                    NaN   \n",
       "1          1      +  transcript        g1      g1                  g1_t1   \n",
       "2          1      +        exon        g1      g1                  g1_t1   \n",
       "3          1      +        exon        g1      g1                  g1_t1   \n",
       "4          1      +        exon        g1      g1                  g1_t1   \n",
       "\n",
       "  original_transcript_name transcript_id    ag1  ag2  Start  End tss_id  \\\n",
       "0                      NaN           NaN    NaN  NaN      1   50    NaN   \n",
       "1                    g1_t1     g1[1,1,1]  known   p1      1   30   g1_1   \n",
       "2                    g1_t1     g1[1,1,1]  known   p1      1   10   g1_1   \n",
       "3                    g1_t1     g1[1,1,1]  known   p1     14   20   g1_1   \n",
       "4                    g1_t1     g1[1,1,1]  known   p1     25   30   g1_1   \n",
       "\n",
       "  tes_id transcript_name  \n",
       "0    NaN             NaN  \n",
       "1   g1_1       g1[1,1,1]  \n",
       "2   g1_1       g1[1,1,1]  \n",
       "3   g1_1       g1[1,1,1]  \n",
       "4   g1_1       g1[1,1,1]  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "025f9485-dc3f-4b75-bbac-d0ae17f6b611",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>original_transcript_id</th>\n",
       "      <th>original_transcript_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>g1[1,1,1]</td>\n",
       "      <td>g1[1,1,1]</td>\n",
       "      <td>g1_t1</td>\n",
       "      <td>g1_t1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>g1[1,1,1]</td>\n",
       "      <td>g1[1,1,1]</td>\n",
       "      <td>g1_t1</td>\n",
       "      <td>g1_t1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>g1[1,1,1]</td>\n",
       "      <td>g1[1,1,1]</td>\n",
       "      <td>g1_t1</td>\n",
       "      <td>g1_t1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>g1[1,1,1]</td>\n",
       "      <td>g1[1,1,1]</td>\n",
       "      <td>g1_t1</td>\n",
       "      <td>g1_t1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  transcript_id transcript_name original_transcript_id  \\\n",
       "0           NaN             NaN                    NaN   \n",
       "1     g1[1,1,1]       g1[1,1,1]                  g1_t1   \n",
       "2     g1[1,1,1]       g1[1,1,1]                  g1_t1   \n",
       "3     g1[1,1,1]       g1[1,1,1]                  g1_t1   \n",
       "4     g1[1,1,1]       g1[1,1,1]                  g1_t1   \n",
       "\n",
       "  original_transcript_name  \n",
       "0                      NaN  \n",
       "1                    g1_t1  \n",
       "2                    g1_t1  \n",
       "3                    g1_t1  \n",
       "4                    g1_t1  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_df[['transcript_id', 'transcript_name', 'original_transcript_id', 'original_transcript_name']].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "2c81baf1-e08d-4542-bcea-42765d673124",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_end_df(c,s,st,e,n, source,mode):\n",
    "    df = pd.DataFrame()\n",
    "    cols = ['Chromosome', 'Strand', 'Start', 'End', 'Name']\n",
    "    var = [c,s,st,e,n]\n",
    "    for col, var in zip(cols, var):\n",
    "        if type(var) == list:\n",
    "            df[col] = var\n",
    "\n",
    "    # add source\n",
    "    df['source'] = source\n",
    "\n",
    "    df = format_end_df(df)\n",
    "\n",
    "    # get end # and gene id\n",
    "    if any(df.Name.isnull()):\n",
    "        df['gene_id'] = np.nan\n",
    "        df[mode] = np.nan\n",
    "    else:\n",
    "        df['gene_id'] = df.Name.str.split('_', expand=True)[0]\n",
    "        df[mode] = df.Name.str.split('_', expand=True)[1]\n",
    "\n",
    "    # get arbitrary unique ids\n",
    "    df['id'] = [i for i in range(len(df.index))]\n",
    "\n",
    "    # get id\n",
    "    df['{}_id'.format(mode)] = df.gene_id+'_'+df[mode]\n",
    "\n",
    "    return df\n",
    "\n",
    "def format_end_df(df):\n",
    "    sort_cols = ['Chromosome', 'Start', 'End', 'Strand']\n",
    "    df = df.sort_values(by=sort_cols)\n",
    "    order = ['Chromosome', 'Start', 'End', 'Strand', 'Name', 'source']\n",
    "    order = [o for o in order if o in df.columns]\n",
    "    df = df[order]\n",
    "    df.reset_index(drop=True, inplace=True)\n",
    "    return df\n",
    "\n",
    "def make_exon_df(n,c,e,s,g,t,nt=None,ag1=None,ag2=None):\n",
    "    df = pd.DataFrame()\n",
    "    df['Chromosome'] = c\n",
    "    df['Start'] = [i[0] for i in e]\n",
    "    df['End'] = [i[1] for i in e]\n",
    "    df['Strand'] = s\n",
    "    df['Feature'] = 'exon'\n",
    "    cols = ['gene_name', 'gene_id']\n",
    "    for c in cols:\n",
    "        df[c] = g\n",
    "    cols = ['transcript_id', 'transcript_name']\n",
    "    for c in cols:\n",
    "        df[c] = t\n",
    "    if nt:\n",
    "        df['new_transcript_id'] = nt\n",
    "    if ag1:\n",
    "        df['ag1'] = ag1\n",
    "    if ag2:\n",
    "        df['ag2'] = ag2\n",
    "\n",
    "    # reorder exons and starts/ stops if needed\n",
    "    df['new_Start'] = df[['Start', 'End']].min(axis=1)\n",
    "    df['new_End'] = df[['Start', 'End']].max(axis=1)\n",
    "    df.drop(['Start', 'End'], axis=1, inplace=True)\n",
    "    df.rename({'new_Start':'Start',\n",
    "               'new_End':'End'}, axis=1, inplace=True)\n",
    "    s = s[0]\n",
    "    if s == '+':\n",
    "        ascending = True\n",
    "    elif s == '-':\n",
    "        ascending = False\n",
    "    df.sort_values(by='Start', ascending=ascending, inplace=True)\n",
    "    return df\n",
    "\n",
    "def make_hier_entry(df, how='t'):\n",
    "    \"\"\"\n",
    "    kind {'g','t'}\n",
    "    \"\"\"\n",
    "    agg_dict = {'min_coord': 'min', 'max_coord': 'max'}\n",
    "    t_df = df.copy(deep=True)\n",
    "    t_df['min_coord'] = t_df[['Start', 'End']].min(axis=1)\n",
    "    t_df['max_coord'] = t_df[['Start', 'End']].max(axis=1)\n",
    "    if how == 't':\n",
    "        gb_cols = ['Chromosome', 'Strand', 'gene_name',\n",
    "                   'gene_id', 'transcript_id', 'transcript_name',\n",
    "                   'tss_id', 'tes_id',\n",
    "                   'new_transcript_id', 'original_transcript_id',\n",
    "                   'original_transcript_name', 'ag1', 'ag2']\n",
    "        gb_cols = list(set(gb_cols)&(set(t_df.columns)))\n",
    "    elif how == 'g':\n",
    "        gb_cols = ['Chromosome', 'Strand', 'gene_name',\n",
    "                   'gene_id']\n",
    "\n",
    "    cols = gb_cols + ['min_coord', 'max_coord']\n",
    "    t_df = t_df[cols]\n",
    "    t_df = t_df.groupby(gb_cols).agg(agg_dict).reset_index()\n",
    "    t_df.rename({'min_coord': 'Start', 'max_coord': 'End'}, axis=1, inplace=True)\n",
    "    if how == 't':\n",
    "        t_df['Feature'] = 'transcript'\n",
    "    elif how == 'g':\n",
    "        t_df['Feature'] = 'gene'\n",
    "\n",
    "    return t_df\n",
    "\n",
    "def make_test_gtf(ts):\n",
    "    df = pd.concat(ts)\n",
    "        # make transcript entries\n",
    "    t_df = make_hier_entry(df, how='t')\n",
    "    # make gene entries\n",
    "    g_df = make_hier_entry(df, how='g')\n",
    "\n",
    "    # concat everything and sort by gene id, transcript id, feature rank (gene =0, t =1, exon=2), then start coords\n",
    "    df = pd.concat([df, t_df, g_df])\n",
    "    df = sort_gtf(df)\n",
    "    return df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fda3e212-2a3a-45ea-9eb9-877d2a5bee4d",
   "metadata": {},
   "source": [
    "## write h5 ref from input beds and ics tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "c97c6fff-29c7-4e14-8d92-37ff668f732e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>source</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>169804296</td>\n",
       "      <td>169804436</td>\n",
       "      <td>+</td>\n",
       "      <td>v40</td>\n",
       "      <td>ENSG00000000460_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>169807786</td>\n",
       "      <td>169807887</td>\n",
       "      <td>+</td>\n",
       "      <td>v40</td>\n",
       "      <td>ENSG00000000460_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>169821668</td>\n",
       "      <td>169821769</td>\n",
       "      <td>+</td>\n",
       "      <td>v40</td>\n",
       "      <td>ENSG00000000460_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>169852986</td>\n",
       "      <td>169853135</td>\n",
       "      <td>+</td>\n",
       "      <td>v40</td>\n",
       "      <td>ENSG00000000460_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>169854029</td>\n",
       "      <td>169854130</td>\n",
       "      <td>+</td>\n",
       "      <td>v40</td>\n",
       "      <td>ENSG00000000460_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome      Start        End Strand source               Name\n",
       "0       chr1  169804296  169804436      +    v40  ENSG00000000460_2\n",
       "1       chr1  169807786  169807887      +    v40  ENSG00000000460_4\n",
       "2       chr1  169821668  169821769      +    v40  ENSG00000000460_5\n",
       "3       chr1  169852986  169853135      +    v40  ENSG00000000460_3\n",
       "4       chr1  169854029  169854130      +    v40  ENSG00000000460_1"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ic = '/Users/fairliereese/Documents/programming/mortazavi_lab/data/rnawg/lr_bulk/cerberus/temp/talon_ic.tsv'\n",
    "tes = '/Users/fairliereese/Documents/programming/mortazavi_lab/data/rnawg/lr_bulk/cerberus/test_tes.bed'\n",
    "tss = '/Users/fairliereese/Documents/programming/mortazavi_lab/data/rnawg/lr_bulk/cerberus/test_tss.bed'\n",
    "tss_map = '/Users/fairliereese/Documents/programming/mortazavi_lab/data/rnawg/lr_bulk/cerberus/test_tes_source_map.bed'\n",
    "tes_map = '/Users/fairliereese/Documents/programming/mortazavi_lab/data/rnawg/lr_bulk/cerberus/test_tss_source_map.bed'\n",
    "\n",
    "# df = read_ic_ref(ic)\n",
    "# df.head()\n",
    "\n",
    "write_reference(tss, tes, ic, 'test.h5')\n",
    "\n",
    "# df = read_cerberus_ends(tss, mode='tss')\n",
    "# df\n",
    "\n",
    "df = read_cerberus_source_map(tss_map)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "594c1015-1b51-4152-8799-84c5fe5b2848",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
