{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "7e4387d2-7683-4c42-93b4-5776e7107cac",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from cerberus.cerberus import *"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a830a56-85b7-48bd-ac02-102f46c48af2",
   "metadata": {},
   "source": [
    "## end aggregation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "29a51c94-fdab-4c56-aced-14fc1d6b7505",
   "metadata": {},
   "outputs": [],
   "source": [
    "# def agg_2_ends(bed1, bed2,\n",
    "#                strand,\n",
    "#                gid,\n",
    "#                buffer,\n",
    "#                add_ends,\n",
    "#                mode):\n",
    "#     \"\"\"\n",
    "#     Parameters:\n",
    "#         bed1 (pyranges PyRanges): Bed PyRanges object for existing ends\n",
    "#         bed2 (pyranges PyRanges): Bed PyRanges object for new ends\n",
    "#         strand (bool): Whether bed2 has strand info\n",
    "#         gid (bool): Whether bed2 has gene id info\n",
    "#         buffer (int): Maximum allowable distance between ends in bed1 and bed2\n",
    "#             to call them the same end\n",
    "#         add_ends (bool): Whether to initialize new regions from bed2\n",
    "#         mode (str): {'tss', 'tes'}\n",
    "#     \"\"\"\n",
    "\n",
    "#     source1 = bed1.df.source.unique().tolist()[0]\n",
    "#     source2 = bed2.df.source.unique().tolist()[0]\n",
    "\n",
    "#     new_c = '{}_new'.format(mode)\n",
    "#     max_c = '{}_max'.format(mode)\n",
    "\n",
    "#     # convert into int64 \n",
    "#     bed1 = pr.PyRanges(bed1.df, int64=True)\n",
    "#     bed2 = pr.PyRanges(bed2.df, int64=True)\n",
    "    \n",
    "#     # depending on whether the new bed has strand information,\n",
    "#     # construct the join call\n",
    "#     if strand:\n",
    "#         temp_joined = bed1.join(bed2,\n",
    "#             strandedness='same',\n",
    "#             suffix='_new',\n",
    "#             slack=buffer,\n",
    "#             how='left')\n",
    "#     elif not strand:\n",
    "#         bed2 = bed2.df\n",
    "#         bed2.drop('Strand', axis=1, inplace=True)\n",
    "#         bed2 = pr.PyRanges(bed2)\n",
    "#         temp_joined = bed1.join(bed2,\n",
    "#             strandedness=False,\n",
    "#             suffix='_new',\n",
    "#             slack=buffer,\n",
    "#             how='left')\n",
    "\n",
    "#     # format null starts as actual nans b/c of join\n",
    "#     temp_joined = temp_joined.df\n",
    "#     temp_joined.loc[temp_joined.Start_new == -1, 'Start_new'] = np.nan\n",
    "\n",
    "#     # df to hold final end annotations\n",
    "#     df = pd.DataFrame()\n",
    "\n",
    "#     ### old ends ###\n",
    "\n",
    "#     # situation 1: ends match across the datasets in coord and gene id\n",
    "#     if gid:\n",
    "#         temp = temp_joined.loc[temp_joined.gene_id == temp_joined.gene_id_new].copy(deep=True)\n",
    "#     else:\n",
    "#         temp = temp_joined.loc[~temp_joined.Start_new.isnull()].copy(deep=True)\n",
    "#     temp.source = temp.source+','+temp.source_new\n",
    "#     df = pd.concat([df, temp])\n",
    "\n",
    "#     # situation 2: ends are only in the first dataset\n",
    "#     if gid:\n",
    "#         temp = temp_joined.loc[(temp_joined.Start_new.isnull())|(temp_joined.gene_id!=temp_joined.gene_id_new)].copy(deep=True)\n",
    "#     else:\n",
    "#         temp = temp_joined.loc[temp_joined.Start_new.isnull()].copy(deep=True)\n",
    "#     df = pd.concat([df, temp])\n",
    "\n",
    "#     # restrict to relevant columns\n",
    "#     cols = ['Chromosome', 'Start', 'End', 'Strand',\n",
    "#             'Name', 'gene_id', 'source', mode, 'id_new']\n",
    "#     df = df[cols]\n",
    "#     df.rename({'id_new': 'id'}, axis=1, inplace=True)\n",
    "\n",
    "#     ### new ends, only add if we're allowing them to be independent\n",
    "#     ### end support\n",
    "#     if add_ends and strand and gid:\n",
    "\n",
    "#         new_df = pd.DataFrame()\n",
    "\n",
    "#         drop_cols = ['Start', 'End', 'Strand', 'gene_id', 'source', 'Name', mode]\n",
    "#         m = {'Start_new': 'Start',\n",
    "#              'End_new': 'End',\n",
    "#              'gene_id_new': 'gene_id',\n",
    "#              'Strand_new': 'Strand',\n",
    "#              'source_new': 'source',\n",
    "#              'Name_new': 'Name',\n",
    "#              new_c: mode}\n",
    "\n",
    "#         # situation 3: the ends overlapped, but the gene ids didn't match\n",
    "#         temp = temp_joined.loc[(temp_joined.gene_id!=temp_joined.gene_id_new)&(temp_joined.gene_id_new!='-1')].copy(deep=True)\n",
    "#         temp.drop(drop_cols, axis=1, inplace=True)\n",
    "#         temp.rename(m, axis=1, inplace=True)\n",
    "#         new_df = pd.concat([new_df, temp])\n",
    "\n",
    "#         # situation 4: the ends are brand new and didn't overlap at all in the existing ends\n",
    "#         bed2 = bed2.df\n",
    "#         inds = list(set(bed2.id.tolist())-set(df.id.tolist()))\n",
    "#         temp = bed2.loc[inds]\n",
    "#         new_df = pd.concat([new_df, temp])\n",
    "\n",
    "#         g_maxes = get_gene_feat_max(df, mode)\n",
    "#         new_df = renumber_new_feats(new_df, g_maxes, mode)\n",
    "\n",
    "#         # some df formatting\n",
    "#         new_df.drop([mode, max_c], axis=1, inplace=True)\n",
    "#         new_df.rename({new_c: mode}, axis=1, inplace=True)\n",
    "\n",
    "#         # finally, concatenate new and old df\n",
    "#         df = pd.concat([df, new_df])\n",
    "\n",
    "#     # drop unnecessary columns and create new names\n",
    "#     # and do some extra formatting\n",
    "#     df.drop('id', axis=1, inplace=True)\n",
    "#     df['Name'] = df.gene_id+'_'+df[mode].astype(str)\n",
    "#     df[mode] = df[mode].astype(int)\n",
    "#     df['Start'] = df.Start.astype(int)\n",
    "#     keep_cols = ['Chromosome', 'Start', 'End', 'Strand', 'Name', 'gene_id', mode, 'source']\n",
    "#     df = df[keep_cols]\n",
    "#     df['id'] = [i for i in range(len(df.index))]\n",
    "\n",
    "#     return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "6d851cc6-2f86-4390-b569-ac884c1da05f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def aggregate_ends(beds, sources, add_ends, buffer, mode):\n",
    "    \"\"\"\n",
    "    Aggregate ends from more than one bed source.\n",
    "    \n",
    "    Parameters:\n",
    "        beds (list of str): List of bed file names\n",
    "        sources (list of str): List of source names for each bed\n",
    "        add_ends (list of bool): List of booleans indicating whether\n",
    "            to add novel ends for each bed file\n",
    "        buffer (int): Allowable distance to an existing end for new\n",
    "            ends to be called the same end\n",
    "        mode (str): {'tss', 'tes'}\n",
    "        \n",
    "    Returns:\n",
    "        df (pandas DataFrame): DataFrame of regions from \n",
    "            aggregated bed files\n",
    "    \"\"\"\n",
    "    \n",
    "    df = pd.DataFrame()\n",
    "    i = 0\n",
    "    for bed_fname, source, add in zip(beds, sources, add_ends):\n",
    "        \n",
    "        # read in bed file and do some formatting\n",
    "        bed = read_bed(bed_fname, mode)\n",
    "        bed = bed.df\n",
    "        bed['source'] = source\n",
    "        bed['id'] = [i for i in range(len(bed.index))]\n",
    "        bed = pr.PyRanges(bed)\n",
    "        \n",
    "        # first bed; just accept all these ends\n",
    "        if len(df.index) == 0:\n",
    "            \n",
    "            if not add:\n",
    "                raise Exception('Must add ends from first bed file')\n",
    "            \n",
    "            if 'gene_id' not in bed.df.columns and 'Strand' not in bed.df.columns:\n",
    "                raise Exception('First bed must contain Strand and gene_id columns')\n",
    "                \n",
    "            df = bed.df\n",
    "        \n",
    "        # more than one bed; merge and reconcile ends\n",
    "        else:\n",
    "\n",
    "            if 'gene_id' not in bed.df.columns or 'Strand' not in bed.df.columns:\n",
    "                if add:\n",
    "                    raise Exception('Cannot add new ends from {} because '+\\\n",
    "                                    'it does not contain gene_id information.')\n",
    "            \n",
    "            # add missing columns but keep track of what information we'll be \n",
    "            # able to merge on \n",
    "            bed, gid, strand = format_agg_2_ends_bed(bed, mode)\n",
    "            df = pr.PyRanges(df)\n",
    "            df = agg_2_ends(df, bed,\n",
    "                            strand, gid,\n",
    "                            buffer, add, mode)\n",
    "            i += 1\n",
    "     \n",
    "    df.drop('id', axis=1, inplace=True)\n",
    "    return df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "a47c700a-161c-4c02-a540-064a7005438d",
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "EOF while scanning triple-quoted string literal (<ipython-input-41-bb5c361109af>, line 18)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-41-bb5c361109af>\"\u001b[0;36m, line \u001b[0;32m18\u001b[0m\n\u001b[0;31m    return bed, gid, strand\u001b[0m\n\u001b[0m                           \n^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m EOF while scanning triple-quoted string literal\n"
     ]
    }
   ],
   "source": [
    "def format_agg_2_ends_bed(bed, mode):\n",
    "    \"\"\"\n",
    "    Format bed file for agg_2_ends\n",
    "    \n",
    "    Parameters:\n",
    "        bed (pandas DataFrame): DataFrame of bed file\n",
    "        mode (str): {'tss', 'tes'}\n",
    "        \n",
    "    Returns:\n",
    "        bed (pandas DataFrame): DataFrame with added gene id\n",
    "            and strand info\n",
    "        gid (bool): Whether or not the bed file contained a gene id\n",
    "        strand (bool): Whether or not the bed file contained a strand\n",
    "    \"\"\"\n",
    "    cols = ['Name', 'gene_id', mode, 'Strand']\n",
    "    \n",
    "    gid = True\n",
    "    strand = True\n",
    "    for col in cols:\n",
    "        if col not in bed.columns:\n",
    "            bed[col] = [np.nan for i in range(len(bed.index))]\n",
    "            \n",
    "            # record some bools\n",
    "            if col == 'gene_id':\n",
    "                gid = False\n",
    "            elif col == 'Strand':\n",
    "                strand = False    \n",
    "                \n",
    "    return bed, gid, strand"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "79f19b07-b632-4fca-966c-6128ddffe065",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chromosome    category\n",
      "Start            int64\n",
      "End              int64\n",
      "Name            object\n",
      "Score           object\n",
      "Strand        category\n",
      "gene_id         object\n",
      "tss             object\n",
      "source          object\n",
      "id               int64\n",
      "dtype: object\n",
      "Chromosome    category\n",
      "Start            int64\n",
      "End              int64\n",
      "Name            object\n",
      "Score           object\n",
      "Strand        category\n",
      "gene_id         object\n",
      "tss             object\n",
      "source          object\n",
      "id               int64\n",
      "dtype: object\n",
      "\n",
      "Chromosome    category\n",
      "Start            int64\n",
      "End              int64\n",
      "Strand        category\n",
      "Name            object\n",
      "gene_id         object\n",
      "tss              int64\n",
      "source          object\n",
      "id               int64\n",
      "dtype: object\n",
      "Chromosome    category\n",
      "Start            int64\n",
      "End              int64\n",
      "Name            object\n",
      "Score           object\n",
      "Strand        category\n",
      "gene_id         object\n",
      "tss             object\n",
      "source          object\n",
      "id               int64\n",
      "dtype: object\n",
      "\n"
     ]
    }
   ],
   "source": [
    "d = '/Users/fairliereese/Documents/programming/mortazavi_lab/data/rnawg/lr_bulk/cerberus/'\n",
    "v29_bed = '{}v29_tss.bed'.format(d)\n",
    "v39_bed = '{}v39_tss.bed'.format(d)\n",
    "talon_bed = '{}talon_tss.bed'.format(d)\n",
    "\n",
    "# beds = [v39_bed, v29_bed, talon_bed] #problem w/ 2nd pair\n",
    "# beds = [v29_bed, talon_bed, v39_bed] #problem on 1st pair\n",
    "beds = [v29_bed, v39_bed, talon_bed] # problem on 1st pair\n",
    "sources = ['v39', 'v29', 'talon']\n",
    "add_ends = [True, True, True]\n",
    "\n",
    "mode = 'tss'\n",
    "buffer = 20\n",
    "\n",
    "df = aggregate_ends(beds, sources, add_ends, buffer, mode)\n",
    "\n",
    "# fname = 'tests/files/Canx_1_tss.bed'\n",
    "# df = read_bed(fname, mode='tss')\n",
    "\n",
    "# fname = 'tests/files/Canx_ccre_pels.bed'\n",
    "# # df = read_bed(fname, mode='tss')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "5248f4a9-e81e-4b2e-b0d8-7f94ad4e546b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>tss</th>\n",
       "      <th>source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>169794679</td>\n",
       "      <td>169794780</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000000460_3</td>\n",
       "      <td>ENSG00000000460</td>\n",
       "      <td>3</td>\n",
       "      <td>v39,v29,talon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>169794992</td>\n",
       "      <td>169795129</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000000460_1</td>\n",
       "      <td>ENSG00000000460</td>\n",
       "      <td>1</td>\n",
       "      <td>v39,v29,talon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>169795358</td>\n",
       "      <td>169795459</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000000460_2</td>\n",
       "      <td>ENSG00000000460</td>\n",
       "      <td>2</td>\n",
       "      <td>v39,v29,talon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>196651994</td>\n",
       "      <td>196652106</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000000971_2</td>\n",
       "      <td>ENSG00000000971</td>\n",
       "      <td>2</td>\n",
       "      <td>v39,v29,talon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>196676937</td>\n",
       "      <td>196677038</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000000971_3</td>\n",
       "      <td>ENSG00000000971</td>\n",
       "      <td>3</td>\n",
       "      <td>v39,v29,talon</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome      Start        End Strand               Name          gene_id  \\\n",
       "0       chr1  169794679  169794780      +  ENSG00000000460_3  ENSG00000000460   \n",
       "1       chr1  169794992  169795129      +  ENSG00000000460_1  ENSG00000000460   \n",
       "2       chr1  169795358  169795459      +  ENSG00000000460_2  ENSG00000000460   \n",
       "3       chr1  196651994  196652106      +  ENSG00000000971_2  ENSG00000000971   \n",
       "4       chr1  196676937  196677038      +  ENSG00000000971_3  ENSG00000000971   \n",
       "\n",
       "   tss         source  \n",
       "0    3  v39,v29,talon  \n",
       "1    1  v39,v29,talon  \n",
       "2    2  v39,v29,talon  \n",
       "3    2  v39,v29,talon  \n",
       "4    3  v39,v29,talon  "
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c746f86-23b0-4efd-93f8-b6389dd2bc77",
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "unexpected EOF while parsing (<ipython-input-245-2533ef766177>, line 32)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-245-2533ef766177>\"\u001b[0;36m, line \u001b[0;32m32\u001b[0m\n\u001b[0;31m    \u001b[0m\n\u001b[0m    ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m unexpected EOF while parsing\n"
     ]
    }
   ],
   "source": [
    "# keeping entries from the old end df\n",
    "\n",
    "# if we have strand information, enforce that\n",
    "if 'Strand' in bed.columns:\n",
    "    temp = df.join(bed,\n",
    "                   strandedness='same',\n",
    "                   suffix='_new',\n",
    "                   slack=buffer,\n",
    "                   how='left')\n",
    "else:\n",
    "    temp = df.join(bed,\n",
    "                   strandedness=False,\n",
    "                   suffix='_new', \n",
    "                   slack=buffer,\n",
    "                   how='left')\n",
    "\n",
    "# convert back to df\n",
    "temp = temp.df\n",
    "print(temp.head())\n",
    "\n",
    "# if we have gene id information, also enforce that\n",
    "if 'gene_id' in bed.columns:\n",
    "    temp = temp.loc[(temp.gene_id == temp.gene_id_new)|(temp.gene_id_new == '-1')]\n",
    "\n",
    "# reconcile sources\n",
    "inds = temp.loc[temp.gene_id_new != '-1'].index\n",
    "new_sources = temp.loc[inds, 'source']+','+temp.loc[inds, 'source_new']\n",
    "temp.loc[inds, 'source'] = new_sources\n",
    "\n",
    "# if we're adding in new ends from here\n",
    "if add:\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d6565405-bf7b-4f4b-9cd4-ff59a4540e7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # first subset the human ccres for those that are promoters \n",
    "# # and are w/i genomic region of Canx\n",
    "# #  https://api.wenglab.org/screen_v13/fdownloads/V3/GRCh38-cCREs.bed\n",
    "# fname = '/Users/fairliereese/mortazavi_lab/ref/ccre_hg38/GRCh38-cCREs.bed'\n",
    "# df = pr.read_bed(fname)\n",
    "# df = df['chr11', 50293961:50325673]\n",
    "\n",
    "# # ok so the toy gtf I have is definitely mouse oops\n",
    "# # https://api.wenglab.org/screen_v13/fdownloads/V3/mm10-ccREs.bed\n",
    "# fname = '/Users/fairliereese/mortazavi_lab/ref/ccre_vM3/mm10-ccREs.bed'\n",
    "# df = pr.read_bed(fname)\n",
    "# df = df['chr11', 50293961:50325673].df\n",
    "# df = df.loc[df.Strand.str.contains('pELS')] \n",
    "# df.to_csv('tests/files/Canx_ccre_pels.bed', sep='\\t', index=False, header=None)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
