{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7e4387d2-7683-4c42-93b4-5776e7107cac",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "import pandas as pd\n",
    "from cerberus.cerberus import *"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a830a56-85b7-48bd-ac02-102f46c48af2",
   "metadata": {},
   "source": [
    "## end aggregation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "4d2d9027-1026-4744-9ed0-11db6cbf6bfd",
   "metadata": {},
   "outputs": [],
   "source": [
    "beds = ['tests/files/Canx_1_tss.bed', 'tests/files/Canx_2_tss.bed', 'tests/files/Canx_ccre_pels.bed']\n",
    "sources = ['v1', 'v2', 'pels_ccre']\n",
    "add_ends = [True, True, True]\n",
    "buffer = 300\n",
    "mode = 'tss'\n",
    "\n",
    "df = pd.DataFrame()\n",
    "i = 0\n",
    "for bed_fname, source, add in zip(beds, sources, add_ends):\n",
    "\n",
    "    bed = read_bed(bed_fname, mode)\n",
    "    bed = bed.df \n",
    "    bed['source'] = source\n",
    "    bed = pr.PyRanges(bed)\n",
    "    \n",
    "    # first bed; just use ends\n",
    "    if len(df.index) == 0:\n",
    "        df = bed.df\n",
    "        \n",
    "    # construct the join call based on what information is present in the new bed\n",
    "    if 'Strand' in bed.columns:\n",
    "        pass\n",
    "    if 'gene_id' in bed.columns:\n",
    "        pass\n",
    "\n",
    "    # if we're adding new ends, make sure to do a right merge and renumber\n",
    "\n",
    "            \n",
    "        \n",
    "    i +=1\n",
    "    if i ==2: break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e29934c-49d9-4560-b25e-0a6ed102f64f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "fe865132-f4b3-4d00-bbbc-c01492247102",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pr.PyRanges(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "ddb0c6d1-717b-4f00-948a-613872647c1a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>tss</th>\n",
       "      <th>source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr11</td>\n",
       "      <td>50298268</td>\n",
       "      <td>50298369</td>\n",
       "      <td>ENSMUSG00000020368_3</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000020368</td>\n",
       "      <td>3</td>\n",
       "      <td>v1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr11</td>\n",
       "      <td>50300939</td>\n",
       "      <td>50301040</td>\n",
       "      <td>ENSMUSG00000020368_2</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000020368</td>\n",
       "      <td>2</td>\n",
       "      <td>v1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr11</td>\n",
       "      <td>50325622</td>\n",
       "      <td>50325723</td>\n",
       "      <td>ENSMUSG00000020368_1</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000020368</td>\n",
       "      <td>1</td>\n",
       "      <td>v1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+----------------------+------------+-------+\n",
       "| Chromosome   |     Start |       End | Name                 | Score      | +4    |\n",
       "| (category)   |   (int32) |   (int32) | (object)             | (object)   | ...   |\n",
       "|--------------+-----------+-----------+----------------------+------------+-------|\n",
       "| chr11        |  50298268 |  50298369 | ENSMUSG00000020368_3 | .          | ...   |\n",
       "| chr11        |  50300939 |  50301040 | ENSMUSG00000020368_2 | .          | ...   |\n",
       "| chr11        |  50325622 |  50325723 | ENSMUSG00000020368_1 | .          | ...   |\n",
       "+--------------+-----------+-----------+----------------------+------------+-------+\n",
       "Stranded PyRanges object has 3 rows and 9 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "4 hidden columns: Strand, gene_id, tss, source"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "3dcb1ca5-a673-4b62-82cb-4e186e9f854e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>tss</th>\n",
       "      <th>source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr11</td>\n",
       "      <td>50299327</td>\n",
       "      <td>50299428</td>\n",
       "      <td>ENSMUSG00000020368_2</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000020368</td>\n",
       "      <td>2</td>\n",
       "      <td>v2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr11</td>\n",
       "      <td>50300939</td>\n",
       "      <td>50301040</td>\n",
       "      <td>ENSMUSG00000020368_1</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000020368</td>\n",
       "      <td>1</td>\n",
       "      <td>v2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+----------------------+------------+-------+\n",
       "| Chromosome   |     Start |       End | Name                 | Score      | +4    |\n",
       "| (category)   |   (int32) |   (int32) | (object)             | (object)   | ...   |\n",
       "|--------------+-----------+-----------+----------------------+------------+-------|\n",
       "| chr11        |  50299327 |  50299428 | ENSMUSG00000020368_2 | .          | ...   |\n",
       "| chr11        |  50300939 |  50301040 | ENSMUSG00000020368_1 | .          | ...   |\n",
       "+--------------+-----------+-----------+----------------------+------------+-------+\n",
       "Stranded PyRanges object has 2 rows and 9 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "4 hidden columns: Strand, gene_id, tss, source"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "cb07e035-9dd7-4959-ae94-2e3547392318",
   "metadata": {},
   "outputs": [],
   "source": [
    "# def merge_ends(old_ends, new_ends, strand, gene_id, how):\n",
    "#     \"\"\"\n",
    "#     Parameters:\n",
    "#         strand (bool)\n",
    "#         gene_id (bool)\n",
    "#         how (str): {'left', 'right'}\n",
    "#     \"\"\"\n",
    "    \n",
    "#     # if we have strand information, enforce that\n",
    "#     if strand:\n",
    "#         temp = df.join(bed,\n",
    "#                        strandedness='same',\n",
    "#                        suffix='_new',\n",
    "#                        slack=buffer,\n",
    "#                        how=how)\n",
    "#     else:\n",
    "#         temp = df.join(bed,\n",
    "#                        strandedness=False,\n",
    "#                        suffix='_new', \n",
    "#                        slack=buffer,\n",
    "#                        how=how)\n",
    "    \n",
    "#     if gene_id:\n",
    "#         temp = temp.loc[(temp.gene_id == temp.gene_id_new)|(temp.gene_id_new == '-1')]\n",
    "    \n",
    "#     return temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 271,
   "id": "6de0a47c-aa07-4f35-aa96-0e57dce34a4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_test_agg_2_bed():\n",
    "    # example should have \n",
    "    # - entries that overlap\n",
    "    # - entries that don't overlap but are within a certain distance\n",
    "    # - entries that are unique to either\n",
    "    # - entries that overlap but aren't using the same gene_id\n",
    "\n",
    "    # beds that don't have strand\n",
    "    # beds that don't have gid\n",
    "\n",
    "    buffer = 20\n",
    "\n",
    "    chroms = [1,1,1,1]\n",
    "    starts = [1, 200, 100, 300]\n",
    "    ends = [15, 250, 110,340]\n",
    "    strands = ['+','+','+', '+']\n",
    "    gids = ['gene_1', 'gene_1', 'gene_1', 'gene_1']\n",
    "    source = 'v1'\n",
    "    bed1 = pr.PyRanges(chromosomes=chroms, starts=starts, ends=ends).as_df()\n",
    "    bed1['Strand'] = strands\n",
    "    bed1['gene_id'] = gids\n",
    "    bed1['source'] = source\n",
    "    bed1 = pr.PyRanges(df=bed1)\n",
    "\n",
    "    chroms = [1,1,1,1]\n",
    "    starts = [5, 120, 500, 200]\n",
    "    ends = [10, 140, 550, 250]\n",
    "    strands=['+','+','+','+']\n",
    "    gids = ['gene_1', 'gene_1', 'gene_1', 'gene_2']\n",
    "    temp_ids = [i for i in range(len(chroms))]\n",
    "    source = 'v2'\n",
    "    bed2 = pyranges.PyRanges(chromosomes=chroms, starts=starts, ends=ends).as_df()\n",
    "    bed2['gene_id'] = gids\n",
    "    bed2['Strand'] = strands\n",
    "    bed2['source'] = source\n",
    "    bed2['id'] = temp_ids\n",
    "    bed2 = pyranges.PyRanges(df=bed2)\n",
    "    \n",
    "    return bed1, bed2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 278,
   "id": "d7ffb069-2534-4b99-a1b4-80ab41f2fd91",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_ctrl_ends(add=True):\n",
    "    df = pd.DataFrame()\n",
    "    rows = [1,200,100,300, 200,500]\n",
    "    df['Start'] = rows\n",
    "    rows = ['1' for i in range(len(df.index))]\n",
    "    df['Chromosome'] = rows\n",
    "    rows = [15,250,110,340, 250,550]\n",
    "    df['End'] = rows\n",
    "    rows = ['gene_1','gene_1','gene_1','gene_1', 'gene_2','gene_1']\n",
    "    df['gene_id'] = rows\n",
    "    rows = ['+' for i in range(len(df.index))]\n",
    "    df['Strand'] = rows\n",
    "    rows = ['v1,v2','v1','v1,v2','v1', 'v2','v2']\n",
    "    df['source'] = rows\n",
    "    \n",
    "    # convert a few dtypes\n",
    "    df['Strand'] = df['Strand'].astype('category')\n",
    "    df['Chromosome'] = df['Chromosome'].astype('category')\n",
    "    \n",
    "    # if we're not adding new ends\n",
    "    if not add:\n",
    "        df = df.loc[df.source != 'v2']\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 288,
   "id": "898d0edd-82be-4a4d-87ab-a231ddb400bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "def agg_2_ends(bed1, bed2, buffer, add_ends):\n",
    "    \"\"\"\n",
    "    Parameters:\n",
    "        bed1 (pyranges PyRanges): Bed PyRanges object for existing ends\n",
    "        bed2 (pyranges PyRanges): Bed PyRanges object for new ends\n",
    "        buffer (int): Maximum allowable distance between ends in bed1 and bed2 \n",
    "            to call them the same end\n",
    "        add_ends (bool): Whether to initialize new regions from bed2\n",
    "    \"\"\"\n",
    "    \n",
    "    temp_joined = bed1.join(bed2, \n",
    "        strandedness='same',\n",
    "        suffix='_new',\n",
    "        slack=buffer,\n",
    "        how='left')\n",
    "\n",
    "\n",
    "    temp_joined = temp_joined.df\n",
    "    # print(temp_joined)\n",
    "    df = pd.DataFrame()\n",
    "\n",
    "    ### old ends ###\n",
    "\n",
    "    # situation 1: ends match across the datasets in coord and gene id\n",
    "    temp = temp_joined.loc[temp_joined.gene_id == temp_joined.gene_id_new].copy(deep=True)\n",
    "    temp.source = temp.source+','+temp.source_new\n",
    "    # print(temp)\n",
    "    df = pd.concat([df, temp])\n",
    "    # print(df)\n",
    "\n",
    "    # situation 2: ends are only in the first dataset\n",
    "    temp = temp_joined.loc[(temp_joined.gene_id_new == '-1')|(temp_joined.gene_id!=temp_joined.gene_id_new)].copy(deep=True)\n",
    "    # print('here')\n",
    "    # print(temp)\n",
    "    df = pd.concat([df, temp])\n",
    "    # print(df)\n",
    "    # print()\n",
    "\n",
    "    ### new ends, only add if we're allowing them to be independent\n",
    "    ### end support\n",
    "    if add_ends:\n",
    "        drop_cols = ['Start', 'End', 'Strand', 'gene_id', 'source']\n",
    "        m = {'Start_new': 'Start', \n",
    "             'End_new': 'End',\n",
    "             'gene_id_new': 'gene_id',\n",
    "             'Strand_new': 'Strand',\n",
    "             'source_new': 'source'}\n",
    "\n",
    "        # situation 3: the ends overlapped, but the gene ids didn't match\n",
    "        temp = temp_joined.loc[(temp_joined.gene_id!=temp_joined.gene_id_new)&(temp_joined.gene_id_new!='-1')].copy(deep=True)\n",
    "        temp.drop(drop_cols, axis=1, inplace=True)\n",
    "        temp.rename(m, axis=1, inplace=True)\n",
    "        # print('hewwo')\n",
    "        # print(temp.head())\n",
    "        df = pd.concat([df, temp])\n",
    "        # print(df)\n",
    "\n",
    "        # situation 4: the ends are brand new and didn't overlap at all in the existing ends\n",
    "        bed2 = bed2.df\n",
    "        # print(bed2.id.dtypes)\n",
    "        # print(temp.id.dtypes)\n",
    "        # print(bed2.id.tolist())\n",
    "        # print(df.id.tolist())\n",
    "        inds = list(set(bed2.id.tolist())-set(df.id.tolist()))\n",
    "        # print(inds)\n",
    "        temp = bed2.loc[inds]\n",
    "        # print(temp)\n",
    "        df = pd.concat([df, temp])\n",
    "\n",
    "    # drop unnecessary columns\n",
    "    drop_cols = ['id', 'Start_new', 'End_new', 'gene_id_new', 'Strand_new', 'source_new']\n",
    "    df.drop(drop_cols, axis=1, inplace=True)\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 299,
   "id": "ead1e459-9ea7-40f0-9115-04a62e31866c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def test_agg_2_ends():\n",
    "    \"\"\"\n",
    "    Test agg_2_ends w/ end adding\n",
    "    \"\"\"\n",
    "    tests = [True, False]\n",
    "    for add_ends in tests:\n",
    "        \n",
    "        buffer = 20\n",
    "        sort_cols = ['Chromosome', 'Strand', 'gene_id', 'Start', 'End']\n",
    "        bed1, bed2 = get_test_agg_2_bed()\n",
    "\n",
    "        df = agg_2_ends(bed1, bed2,\n",
    "                        buffer=buffer,\n",
    "                        add_ends=add_ends)\n",
    "        df = df.sort_values(by=sort_cols)\n",
    "        df.reset_index(drop=True, inplace=True)\n",
    "\n",
    "        ctrl = get_ctrl_ends(add=add_ends)\n",
    "        ctrl = ctrl.sort_values(by=sort_cols)\n",
    "        ctrl.reset_index(drop=True, inplace=True)\n",
    "\n",
    "        pd.testing.assert_frame_equal(ctrl, df, check_like=True)\n",
    "        assert len(ctrl.index) == len(df.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 300,
   "id": "37d66c99-931c-4480-a75b-eecf1262563a",
   "metadata": {},
   "outputs": [],
   "source": [
    "test_agg_2_ends()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3671b215-d1df-4952-a3cc-c4b22b1a6f47",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Start</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>End</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>Strand</th>\n",
       "      <th>source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>gene_1</td>\n",
       "      <td>+</td>\n",
       "      <td>v1,v2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>100</td>\n",
       "      <td>1</td>\n",
       "      <td>110</td>\n",
       "      <td>gene_1</td>\n",
       "      <td>+</td>\n",
       "      <td>v1,v2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>200</td>\n",
       "      <td>1</td>\n",
       "      <td>250</td>\n",
       "      <td>gene_1</td>\n",
       "      <td>+</td>\n",
       "      <td>v1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>300</td>\n",
       "      <td>1</td>\n",
       "      <td>340</td>\n",
       "      <td>gene_1</td>\n",
       "      <td>+</td>\n",
       "      <td>v1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>500</td>\n",
       "      <td>1</td>\n",
       "      <td>550</td>\n",
       "      <td>gene_1</td>\n",
       "      <td>+</td>\n",
       "      <td>v2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>200</td>\n",
       "      <td>1</td>\n",
       "      <td>250</td>\n",
       "      <td>gene_2</td>\n",
       "      <td>+</td>\n",
       "      <td>v2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Start Chromosome  End gene_id Strand source\n",
       "0      1          1   15  gene_1      +  v1,v2\n",
       "1    100          1  110  gene_1      +  v1,v2\n",
       "2    200          1  250  gene_1      +     v1\n",
       "3    300          1  340  gene_1      +     v1\n",
       "4    500          1  550  gene_1      +     v2\n",
       "5    200          1  250  gene_2      +     v2"
      ]
     },
     "execution_count": 291,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ctrl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08668c33-f085-46af-a84b-93a33e730c89",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>+</td>\n",
       "      <td>gene_1</td>\n",
       "      <td>v1,v2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>110</td>\n",
       "      <td>+</td>\n",
       "      <td>gene_1</td>\n",
       "      <td>v1,v2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>250</td>\n",
       "      <td>+</td>\n",
       "      <td>gene_1</td>\n",
       "      <td>v1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>300</td>\n",
       "      <td>340</td>\n",
       "      <td>+</td>\n",
       "      <td>gene_1</td>\n",
       "      <td>v1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>500</td>\n",
       "      <td>550</td>\n",
       "      <td>+</td>\n",
       "      <td>gene_1</td>\n",
       "      <td>v2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>250</td>\n",
       "      <td>+</td>\n",
       "      <td>gene_2</td>\n",
       "      <td>v2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome  Start  End Strand gene_id source\n",
       "0          1      1   15      +  gene_1  v1,v2\n",
       "1          1    100  110      +  gene_1  v1,v2\n",
       "2          1    200  250      +  gene_1     v1\n",
       "3          1    300  340      +  gene_1     v1\n",
       "4          1    500  550      +  gene_1     v2\n",
       "5          1    200  250      +  gene_2     v2"
      ]
     },
     "execution_count": 292,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2b283d5-4533-4b83-895b-e3804a28ed58",
   "metadata": {},
   "outputs": [],
   "source": [
    "buffer = 20\n",
    "add_ends = False\n",
    "bed1, bed2 = get_test_agg_2_bed()\n",
    "\n",
    "df = agg_2_ends(bed1, bed2,\n",
    "                buffer=buffer,\n",
    "                add_ends=add_ends)\n",
    "df = df.sort_values(by=sort_cols)\n",
    "df.reset_index(drop=True, inplace=True)\n",
    "\n",
    "ctrl = get_ctrl_ends(add=add_ends)\n",
    "ctrl = ctrl.sort_values(by=sort_cols)\n",
    "ctrl.reset_index(drop=True, inplace=True)\n",
    "\n",
    "pd.testing.assert_frame_equal(ctrl, df, check_like=True)\n",
    "assert len(ctrl.index) == len(df.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "109db6e8-8bb7-4f0c-8d68-586a323c4c07",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Start</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>End</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>Strand</th>\n",
       "      <th>source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>gene_1</td>\n",
       "      <td>+</td>\n",
       "      <td>v1,v2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>100</td>\n",
       "      <td>1</td>\n",
       "      <td>110</td>\n",
       "      <td>gene_1</td>\n",
       "      <td>+</td>\n",
       "      <td>v1,v2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>200</td>\n",
       "      <td>1</td>\n",
       "      <td>250</td>\n",
       "      <td>gene_1</td>\n",
       "      <td>+</td>\n",
       "      <td>v1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>300</td>\n",
       "      <td>1</td>\n",
       "      <td>340</td>\n",
       "      <td>gene_1</td>\n",
       "      <td>+</td>\n",
       "      <td>v1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Start Chromosome  End gene_id Strand source\n",
       "0      1          1   15  gene_1      +  v1,v2\n",
       "1    100          1  110  gene_1      +  v1,v2\n",
       "2    200          1  250  gene_1      +     v1\n",
       "3    300          1  340  gene_1      +     v1"
      ]
     },
     "execution_count": 294,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ctrl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c8fa71a-733e-4527-bbe8-e1402549ad96",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>+</td>\n",
       "      <td>gene_1</td>\n",
       "      <td>v1,v2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>110</td>\n",
       "      <td>+</td>\n",
       "      <td>gene_1</td>\n",
       "      <td>v1,v2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>250</td>\n",
       "      <td>+</td>\n",
       "      <td>gene_1</td>\n",
       "      <td>v1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>300</td>\n",
       "      <td>340</td>\n",
       "      <td>+</td>\n",
       "      <td>gene_1</td>\n",
       "      <td>v1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome  Start  End Strand gene_id source\n",
       "0          1      1   15      +  gene_1  v1,v2\n",
       "1          1    100  110      +  gene_1  v1,v2\n",
       "2          1    200  250      +  gene_1     v1\n",
       "3          1    300  340      +  gene_1     v1"
      ]
     },
     "execution_count": 295,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b726dc92-56c6-405d-9092-1c17f4543ffb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  Chromosome  Start  End Strand gene_id source  Start_new  End_new  \\\n",
      "0          1      1   15      +  gene_1  v1,v2          5       10   \n",
      "2          1    100  110      +  gene_1  v1,v2        120      140   \n",
      "\n",
      "  gene_id_new Strand_new source_new  id  \n",
      "0      gene_1          +         v2   0  \n",
      "2      gene_1          +         v2   1  \n",
      "  Chromosome  Start  End Strand gene_id source  Start_new  End_new  \\\n",
      "0          1      1   15      +  gene_1  v1,v2          5       10   \n",
      "2          1    100  110      +  gene_1  v1,v2        120      140   \n",
      "1          1    200  250      +  gene_1     v1        200      250   \n",
      "3          1    300  340      +  gene_1     v1         -1       -1   \n",
      "\n",
      "  gene_id_new Strand_new source_new  id  \n",
      "0      gene_1          +         v2   0  \n",
      "2      gene_1          +         v2   1  \n",
      "1      gene_2          +         v2   3  \n",
      "3          -1          +         -1  -1  \n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/fairliereese/miniconda3/lib/python3.7/site-packages/pandas/core/generic.py:5516: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self[name] = value\n"
     ]
    }
   ],
   "source": [
    "temp_joined = bed1.join(bed2, \n",
    "        strandedness='same',\n",
    "        suffix='_new',\n",
    "        slack=buffer,\n",
    "        how='left')\n",
    "\n",
    "\n",
    "temp_joined = temp_joined.df\n",
    "# print(temp_joined)\n",
    "df = pd.DataFrame()\n",
    "\n",
    "### old ends ###\n",
    "\n",
    "# situation 1: ends match across the datasets in coord and gene id\n",
    "temp = temp_joined.loc[temp_joined.gene_id == temp_joined.gene_id_new]\n",
    "temp.source = temp.source+','+temp.source_new\n",
    "# print(temp)\n",
    "df = pd.concat([df, temp])\n",
    "print(df)\n",
    "\n",
    "\n",
    "# situation 2: ends are only in the first dataset\n",
    "temp = temp_joined.loc[(temp_joined.gene_id_new == '-1')|(temp_joined.gene_id!=temp_joined.gene_id_new)]\n",
    "# print('here')\n",
    "# print(temp)\n",
    "df = pd.concat([df, temp])\n",
    "print(df)\n",
    "print()\n",
    "\n",
    "### new ends, only add if we're allowing them to be independent\n",
    "### end support\n",
    "if add:\n",
    "    drop_cols = ['Start', 'End', 'Strand', 'gene_id', 'source']\n",
    "    m = {'Start_new': 'Start', \n",
    "         'End_new': 'End',\n",
    "         'gene_id_new': 'gene_id',\n",
    "         'Strand_new': 'Strand',\n",
    "         'source_new': 'source'}\n",
    "    \n",
    "    # situation 3: the ends overlapped, but the gene ids didn't match\n",
    "    temp = temp_joined.loc[(temp_joined.gene_id!=temp_joined.gene_id_new)&(temp_joined.gene_id_new!='-1')].copy(deep=True)\n",
    "    temp.drop(drop_cols, axis=1, inplace=True)\n",
    "    temp.rename(m, axis=1, inplace=True)\n",
    "    # print('hewwo')\n",
    "    # print(temp.head())\n",
    "    df = pd.concat([df, temp])\n",
    "    # print(df)\n",
    "    \n",
    "    # situation 4: the ends are brand new and didn't overlap at all in the existing ends\n",
    "    bed2 = bed2.df\n",
    "    # print(bed2.id.dtypes)\n",
    "    # print(temp.id.dtypes)\n",
    "    # print(bed2.id.tolist())\n",
    "    # print(df.id.tolist())\n",
    "    inds = list(set(bed2.id.tolist())-set(df.id.tolist()))\n",
    "    # print(inds)\n",
    "    temp = bed2.loc[inds]\n",
    "    # print(temp)\n",
    "    df = pd.concat([df, temp])\n",
    "\n",
    "# drop unnecessary columns\n",
    "drop_cols = ['id', 'Start_new', 'End_new', 'gene_id_new', 'Strand_new', 'source_new']\n",
    "df.drop(drop_cols, axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 254,
   "id": "e97938f8-bfe6-450a-be38-a7e216a258ef",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chromosome    category\n",
      "Start            int64\n",
      "End              int64\n",
      "Strand        category\n",
      "gene_id         object\n",
      "source          object\n",
      "dtype: object\n"
     ]
    }
   ],
   "source": [
    "print(df.dtypes)\n",
    "sort_cols = ['Chromosome', 'Strand', 'gene_id', 'Start', 'End']\n",
    "df = df.sort_values(by=sort_cols)\n",
    "df.reset_index(drop=True, inplace=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 255,
   "id": "16f1c0e3-7112-4b31-a06e-3a20e5160864",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Start            int64\n",
      "Chromosome    category\n",
      "End              int64\n",
      "gene_id         object\n",
      "Strand        category\n",
      "source          object\n",
      "dtype: object\n"
     ]
    }
   ],
   "source": [
    "ctrl = bepis()\n",
    "print(ctrl.dtypes)\n",
    "ctrl = ctrl.sort_values(by=sort_cols)\n",
    "ctrl.reset_index(drop=True, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 256,
   "id": "d36b0b93-6c61-4239-8599-b8c2c931e1ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.testing.assert_frame_equal(ctrl, df, check_like=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 245,
   "id": "9c746f86-23b0-4efd-93f8-b6389dd2bc77",
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "unexpected EOF while parsing (<ipython-input-245-2533ef766177>, line 32)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-245-2533ef766177>\"\u001b[0;36m, line \u001b[0;32m32\u001b[0m\n\u001b[0;31m    \u001b[0m\n\u001b[0m    ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m unexpected EOF while parsing\n"
     ]
    }
   ],
   "source": [
    "# keeping entries from the old end df\n",
    "\n",
    "# if we have strand information, enforce that\n",
    "if 'Strand' in bed.columns:\n",
    "    temp = df.join(bed,\n",
    "                   strandedness='same',\n",
    "                   suffix='_new',\n",
    "                   slack=buffer,\n",
    "                   how='left')\n",
    "else:\n",
    "    temp = df.join(bed,\n",
    "                   strandedness=False,\n",
    "                   suffix='_new', \n",
    "                   slack=buffer,\n",
    "                   how='left')\n",
    "\n",
    "# convert back to df\n",
    "temp = temp.df\n",
    "print(temp.head())\n",
    "\n",
    "# if we have gene id information, also enforce that\n",
    "if 'gene_id' in bed.columns:\n",
    "    temp = temp.loc[(temp.gene_id == temp.gene_id_new)|(temp.gene_id_new == '-1')]\n",
    "\n",
    "# reconcile sources\n",
    "inds = temp.loc[temp.gene_id_new != '-1'].index\n",
    "new_sources = temp.loc[inds, 'source']+','+temp.loc[inds, 'source_new']\n",
    "temp.loc[inds, 'source'] = new_sources\n",
    "\n",
    "# if we're adding in new ends from here\n",
    "if add:\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "d6136ef0-5eee-4f11-b25d-6a8623bb2d44",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>tss</th>\n",
       "      <th>source</th>\n",
       "      <th>Start_new</th>\n",
       "      <th>End_new</th>\n",
       "      <th>Name_new</th>\n",
       "      <th>Score_new</th>\n",
       "      <th>Strand_new</th>\n",
       "      <th>gene_id_new</th>\n",
       "      <th>tss_new</th>\n",
       "      <th>source_new</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr11</td>\n",
       "      <td>50300939</td>\n",
       "      <td>50301040</td>\n",
       "      <td>ENSMUSG00000020368_2</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000020368</td>\n",
       "      <td>2</td>\n",
       "      <td>v1,v2</td>\n",
       "      <td>50300939</td>\n",
       "      <td>50301040</td>\n",
       "      <td>ENSMUSG00000020368_1</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000020368</td>\n",
       "      <td>1</td>\n",
       "      <td>v2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr11</td>\n",
       "      <td>50298268</td>\n",
       "      <td>50298369</td>\n",
       "      <td>ENSMUSG00000020368_3</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000020368</td>\n",
       "      <td>3</td>\n",
       "      <td>v1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr11</td>\n",
       "      <td>50325622</td>\n",
       "      <td>50325723</td>\n",
       "      <td>ENSMUSG00000020368_1</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000020368</td>\n",
       "      <td>1</td>\n",
       "      <td>v1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome     Start       End                  Name Score Strand  \\\n",
       "0      chr11  50300939  50301040  ENSMUSG00000020368_2     .      -   \n",
       "1      chr11  50298268  50298369  ENSMUSG00000020368_3     .      -   \n",
       "2      chr11  50325622  50325723  ENSMUSG00000020368_1     .      -   \n",
       "\n",
       "              gene_id tss source  Start_new   End_new              Name_new  \\\n",
       "0  ENSMUSG00000020368   2  v1,v2   50300939  50301040  ENSMUSG00000020368_1   \n",
       "1  ENSMUSG00000020368   3     v1         -1        -1                    -1   \n",
       "2  ENSMUSG00000020368   1     v1         -1        -1                    -1   \n",
       "\n",
       "  Score_new Strand_new         gene_id_new tss_new source_new  \n",
       "0         .          -  ENSMUSG00000020368       1         v2  \n",
       "1        -1          -                  -1      -1         -1  \n",
       "2        -1          -                  -1      -1         -1  "
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40eb4e21-5480-4a02-ba41-6c252f4be13e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# adding entries from the new df\n",
    "if 'Strand' in bed.columns:\n",
    "    temp2 = df.join(bed,\n",
    "                     strandedness='same',\n",
    "                     suffix\n",
    "                     "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d6565405-bf7b-4f4b-9cd4-ff59a4540e7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # first subset the human ccres for those that are promoters \n",
    "# # and are w/i genomic region of Canx\n",
    "# #  https://api.wenglab.org/screen_v13/fdownloads/V3/GRCh38-cCREs.bed\n",
    "# fname = '/Users/fairliereese/mortazavi_lab/ref/ccre_hg38/GRCh38-cCREs.bed'\n",
    "# df = pr.read_bed(fname)\n",
    "# df = df['chr11', 50293961:50325673]\n",
    "\n",
    "# # ok so the toy gtf I have is definitely mouse oops\n",
    "# # https://api.wenglab.org/screen_v13/fdownloads/V3/mm10-ccREs.bed\n",
    "# fname = '/Users/fairliereese/mortazavi_lab/ref/ccre_vM3/mm10-ccREs.bed'\n",
    "# df = pr.read_bed(fname)\n",
    "# df = df['chr11', 50293961:50325673].df\n",
    "# df = df.loc[df.Strand.str.contains('pELS')] \n",
    "# df.to_csv('tests/files/Canx_ccre_pels.bed', sep='\\t', index=False, header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "1956d547-74d8-40ee-a58c-1e502da87c90",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  Chromosome  Start  End gene_id Name  Start_b  End_b gene_id_b Name_b\n",
      "0          1     11   11  gene_1    d        1     15    gene_1      a\n",
      "1          1     11   11  gene_1    d       10     12    gene_2      b\n",
      "2          1     11   11  gene_2    e        1     15    gene_1      a\n",
      "3          1     11   11  gene_2    e       10     12    gene_2      b\n"
     ]
    }
   ],
   "source": [
    "import pyranges\n",
    "\n",
    "starts = [1, 10, 20]\n",
    "ends = [15, 12, 25]\n",
    "chroms = [1, 1, 1]\n",
    "gids = ['gene_1', 'gene_2', 'gene_2']\n",
    "names = ['a', 'b', 'c']\n",
    "gr = pyranges.PyRanges(chromosomes=chroms, starts=starts, ends=ends).as_df()\n",
    "gr['gene_id'] = gids\n",
    "gr['Name'] = names\n",
    "gr = pyranges.PyRanges(df=gr)\n",
    "\n",
    "starts = [11, 11]\n",
    "ends = [11, 11]\n",
    "chroms = [1, 1]\n",
    "names = ['d', 'e']\n",
    "gids = ['gene_1', 'gene_2']\n",
    "gr2 = pyranges.PyRanges(chromosomes=chroms, starts=starts, ends=ends).as_df()\n",
    "gr2['gene_id'] = gids\n",
    "gr2['Name'] = names\n",
    "gr2 = pyranges.PyRanges(df=gr2)\n",
    "\n",
    "temp = gr2.join(gr, how='left', strandedness=None).df\n",
    "print(temp)\n",
    "temp = temp.loc[temp.gene_id == temp.gene_id_b]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "e32b1564-aa74-486b-a92a-b6940ca01c7a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>Name</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>gene_id_b</th>\n",
       "      <th>Name_b</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>gene_1</td>\n",
       "      <td>d</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>gene_1</td>\n",
       "      <td>a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>gene_2</td>\n",
       "      <td>e</td>\n",
       "      <td>10</td>\n",
       "      <td>12</td>\n",
       "      <td>gene_2</td>\n",
       "      <td>b</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome  Start  End gene_id Name  Start_b  End_b gene_id_b Name_b\n",
       "0          1     11   11  gene_1    d        1     15    gene_1      a\n",
       "3          1     11   11  gene_2    e       10     12    gene_2      b"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d404633-6926-4d7c-b610-a8b03d94b446",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
