{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7e4387d2-7683-4c42-93b4-5776e7107cac",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import h5py \n",
    "import os\n",
    "\n",
    "from cerberus.cerberus import *\n",
    "from cerberus.main import *\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3c24a57-cb3a-4fe0-84ca-59b2926c949a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def replace_gtf_ids(gtf, h5, agg):\n",
    "    \"\"\"\n",
    "    Replace transcript ids and names in a gtf with the triplets\n",
    "    calculated from assign_triplets\n",
    "\n",
    "    Parameters:\n",
    "        gtf (str): Path to gtf file\n",
    "        h5 (str): Path to h5 annotation (output from assign)\n",
    "        agg (bool): Whether or not to collapse transcripts with\n",
    "            duplicate triplets\n",
    "\n",
    "    Returns:\n",
    "        df (pyranges PyRanges): PyRanges gtf table with updated ids\n",
    "    \"\"\"\n",
    "    \n",
    "    # keep only the gene, exon, and transcript entries\n",
    "    df = pr.read_gtf(gtf).df\n",
    "    entry_types = ['gene', 'transcript', 'exon']\n",
    "    df = df.loc[df.Feature.isin(entry_types)]\n",
    "   \n",
    "    if not update_ends:\n",
    "        _, _, _, m_df = read_h5(h5)\n",
    "    else:\n",
    "        _, tss, tes, m_df = read_h5(h5)\n",
    "\n",
    "    # groupby transcripts that are the same\n",
    "    gb_cols = ['gene_name', 'gene_id', 'transcript_triplet',\n",
    "               'transcript_id', 'transcript_name']\n",
    "    temp = m_df[['transcript_id',\n",
    "                 'original_transcript_id',\n",
    "                 'original_transcript_name']].copy(deep=True)\n",
    "    m_df = m_df.groupby(gb_cols).agg({'original_transcript_id': ','.join,\n",
    "                                      'original_transcript_name': ','.join}).reset_index()\n",
    "    m_df = m_df.merge(temp, on='transcript_id', suffixes=('','_merge'))\n",
    "    m_df.drop(['gene_name', 'gene_id', 'transcript_triplet'],\n",
    "              axis=1, inplace=True)\n",
    "\n",
    "    # add new transcript ids\n",
    "    df = df.merge(m_df, left_on=['transcript_id', 'transcript_name'],\n",
    "                  right_on=['original_transcript_id_merge',\n",
    "                            'original_transcript_name_merge'],\n",
    "                 suffixes=('_x', ''))\n",
    "\n",
    "    # drop old tids\n",
    "    df.drop(['transcript_id_x', 'transcript_name_x',\n",
    "             'original_transcript_name_merge'],\n",
    "            axis=1, inplace=True)\n",
    "\n",
    "    # remove duplicated transcripts; just keeping the first one\n",
    "    if agg:\n",
    "        temp = df[['transcript_id', 'original_transcript_id_merge']].drop_duplicates()\n",
    "        dupe_old_tids = temp.loc[temp.transcript_id.duplicated(keep='first'), 'original_transcript_id_merge']\n",
    "        df = df.loc[~df.original_transcript_id_merge.isin(dupe_old_tids)]\n",
    "\n",
    "    # drop last column\n",
    "    df.drop('original_transcript_id_merge', axis=1, inplace=True)\n",
    "\n",
    "    df = pr.PyRanges(df)\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "751133e1-1dfe-4fe2-b195-f1f5ec8fcac9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def replace_gtf_ids(gtf, h5, agg, update_ends, o): \n",
    "    \"\"\"\n",
    "    Replace transcript ids and names in a gtf with the triplets\n",
    "    calculated from convert_transcriptome\n",
    "    \n",
    "    Parameters:\n",
    "        gtf (str): Path to GTF file to update\n",
    "        h5 (str): Path to cerberus reference with tid map\n",
    "        agg (bool): Aggregate / collapse transcripts with the same \n",
    "            triplets\n",
    "        update_ends (bool): Change the ends of transcripts based on \n",
    "            ends seen in the h5 annotation\n",
    "        o (str): Output file name\n",
    "    \"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0faa232-0c52-4e30-9591-96406c346324",
   "metadata": {},
   "source": [
    "## modify / collapse transcripts in TALON abundance file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "4c6b0b01-d85e-4d2c-af62-08d3122061c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "h5 = '/Users/fairliereese/mortazavi_lab/data/rnawg/lr_bulk/cerberus/human_cerberus.h5'\n",
    "gtf = '/Users/fairliereese/mortazavi_lab/data/rnawg/lr_bulk/talon/human_known_nic_nnc_talon.gtf'\n",
    "agg = True\n",
    "update_ends = True\n",
    "\n",
    "# replace_gtf_ids(gtf, h5, agg, update_ends)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "7db16831-2564-4356-ba8d-799ef02cde42",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pr.read_gtf(gtf).df\n",
    "entry_types = ['gene', 'transcript', 'exon']\n",
    "df = df.loc[df.Feature.isin(entry_types)]\n",
    "\n",
    "if not update_ends:\n",
    "    _, _, _, m_df = read_h5(h5)\n",
    "else:\n",
    "    _, tss, tes, m_df = read_h5(h5)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "1e4df3fb-bae0-4a23-8112-ec6a3d7bccca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>original_transcript_id</th>\n",
       "      <th>ic</th>\n",
       "      <th>ic_id</th>\n",
       "      <th>tss_id</th>\n",
       "      <th>tss</th>\n",
       "      <th>tes_id</th>\n",
       "      <th>tes</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>original_transcript_name</th>\n",
       "      <th>transcript_triplet</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENCODEHT000206942</td>\n",
       "      <td>1</td>\n",
       "      <td>ENCODEHG000058846_1</td>\n",
       "      <td>ENCODEHG000058846_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENCODEHG000058846_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>ENCODEHT000206942</td>\n",
       "      <td>[1,1,1]</td>\n",
       "      <td>ENCODEHG000058846[1,1,1]</td>\n",
       "      <td>ENCODEHG000058846[1,1,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENCODEHT000206867</td>\n",
       "      <td>4</td>\n",
       "      <td>ENCODEHG000058837_4</td>\n",
       "      <td>ENCODEHG000058837_2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENCODEHG000058837_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>ENCODEHT000206867</td>\n",
       "      <td>[2,4,1]</td>\n",
       "      <td>ENCODEHG000058837[2,4,1]</td>\n",
       "      <td>ENCODEHG000058837[2,4,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENCODEHT000206868</td>\n",
       "      <td>2</td>\n",
       "      <td>ENCODEHG000058837_2</td>\n",
       "      <td>ENCODEHG000058837_2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENCODEHG000058837_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>ENCODEHT000206868</td>\n",
       "      <td>[2,2,1]</td>\n",
       "      <td>ENCODEHG000058837[2,2,1]</td>\n",
       "      <td>ENCODEHG000058837[2,2,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENCODEHT000206870</td>\n",
       "      <td>3</td>\n",
       "      <td>ENCODEHG000058837_3</td>\n",
       "      <td>ENCODEHG000058837_2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENCODEHG000058837_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>ENCODEHT000206870</td>\n",
       "      <td>[2,3,1]</td>\n",
       "      <td>ENCODEHG000058837[2,3,1]</td>\n",
       "      <td>ENCODEHG000058837[2,3,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENCODEHT000206886</td>\n",
       "      <td>1</td>\n",
       "      <td>ENCODEHG000058837_1</td>\n",
       "      <td>ENCODEHG000058837_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENCODEHG000058837_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>ENCODEHT000206886</td>\n",
       "      <td>[1,1,1]</td>\n",
       "      <td>ENCODEHG000058837[1,1,1]</td>\n",
       "      <td>ENCODEHG000058837[1,1,1]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  original_transcript_id  ic                ic_id               tss_id  tss  \\\n",
       "0      ENCODEHT000206942   1  ENCODEHG000058846_1  ENCODEHG000058846_1    1   \n",
       "1      ENCODEHT000206867   4  ENCODEHG000058837_4  ENCODEHG000058837_2    2   \n",
       "2      ENCODEHT000206868   2  ENCODEHG000058837_2  ENCODEHG000058837_2    2   \n",
       "3      ENCODEHT000206870   3  ENCODEHG000058837_3  ENCODEHG000058837_2    2   \n",
       "4      ENCODEHT000206886   1  ENCODEHG000058837_1  ENCODEHG000058837_1    1   \n",
       "\n",
       "                tes_id  tes            gene_id          gene_name  \\\n",
       "0  ENCODEHG000058846_1    1  ENCODEHG000058846  ENCODEHG000058846   \n",
       "1  ENCODEHG000058837_1    1  ENCODEHG000058837  ENCODEHG000058837   \n",
       "2  ENCODEHG000058837_1    1  ENCODEHG000058837  ENCODEHG000058837   \n",
       "3  ENCODEHG000058837_1    1  ENCODEHG000058837  ENCODEHG000058837   \n",
       "4  ENCODEHG000058837_1    1  ENCODEHG000058837  ENCODEHG000058837   \n",
       "\n",
       "  original_transcript_name transcript_triplet             transcript_id  \\\n",
       "0        ENCODEHT000206942            [1,1,1]  ENCODEHG000058846[1,1,1]   \n",
       "1        ENCODEHT000206867            [2,4,1]  ENCODEHG000058837[2,4,1]   \n",
       "2        ENCODEHT000206868            [2,2,1]  ENCODEHG000058837[2,2,1]   \n",
       "3        ENCODEHT000206870            [2,3,1]  ENCODEHG000058837[2,3,1]   \n",
       "4        ENCODEHT000206886            [1,1,1]  ENCODEHG000058837[1,1,1]   \n",
       "\n",
       "            transcript_name  \n",
       "0  ENCODEHG000058846[1,1,1]  \n",
       "1  ENCODEHG000058837[2,4,1]  \n",
       "2  ENCODEHG000058837[2,2,1]  \n",
       "3  ENCODEHG000058837[2,3,1]  \n",
       "4  ENCODEHG000058837[1,1,1]  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "m_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "b0d6c29f-5bf5-4992-8f58-8504568b6b98",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>...</th>\n",
       "      <th>tag</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>NNC_transcript</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>intergenic_novel</th>\n",
       "      <th>antisense_gene</th>\n",
       "      <th>gene_antisense_to_IDs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>TALON</td>\n",
       "      <td>gene</td>\n",
       "      <td>10647</td>\n",
       "      <td>11606</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>TALON</td>\n",
       "      <td>transcript</td>\n",
       "      <td>10647</td>\n",
       "      <td>11606</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>10647</td>\n",
       "      <td>10791</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>10882</td>\n",
       "      <td>11057</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>11434</td>\n",
       "      <td>11606</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 36 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome Source     Feature  Start    End Score Strand Frame  \\\n",
       "0      SIRV1  TALON        gene  10647  11606     .      +     .   \n",
       "1      SIRV1  TALON  transcript  10647  11606     .      +     .   \n",
       "2      SIRV1  TALON        exon  10647  10791     .      +     .   \n",
       "3      SIRV1  TALON        exon  10882  11057     .      +     .   \n",
       "4      SIRV1  TALON        exon  11434  11606     .      +     .   \n",
       "\n",
       "             gene_id          gene_name  ...  tag transcript_type  \\\n",
       "0  ENCODEHG000058846  ENCODEHG000058846  ...  NaN             NaN   \n",
       "1  ENCODEHG000058846  ENCODEHG000058846  ...  NaN             NaN   \n",
       "2  ENCODEHG000058846  ENCODEHG000058846  ...  NaN             NaN   \n",
       "3  ENCODEHG000058846  ENCODEHG000058846  ...  NaN             NaN   \n",
       "4  ENCODEHG000058846  ENCODEHG000058846  ...  NaN             NaN   \n",
       "\n",
       "  transcript_support_level havana_transcript NNC_transcript protein_id ccdsid  \\\n",
       "0                      NaN               NaN            NaN        NaN    NaN   \n",
       "1                      NaN               NaN           TRUE        NaN    NaN   \n",
       "2                      NaN               NaN            NaN        NaN    NaN   \n",
       "3                      NaN               NaN            NaN        NaN    NaN   \n",
       "4                      NaN               NaN            NaN        NaN    NaN   \n",
       "\n",
       "  intergenic_novel antisense_gene gene_antisense_to_IDs  \n",
       "0             TRUE            NaN                   NaN  \n",
       "1              NaN            NaN                   NaN  \n",
       "2              NaN            NaN                   NaN  \n",
       "3              NaN            NaN                   NaN  \n",
       "4              NaN            NaN                   NaN  \n",
       "\n",
       "[5 rows x 36 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "d748e55d-2caf-4853-8f9b-9686889f2bb3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def map_gtf_transcripts(df, m_df):\n",
    "    \"\"\"\n",
    "    Add cerberus transcript id to each entry in gtf\n",
    "    \"\"\"\n",
    "    # groupby transcripts that are the same\n",
    "    gb_cols = ['gene_name', 'gene_id', 'transcript_triplet',\n",
    "               'transcript_id', 'transcript_name']\n",
    "    temp = m_df[['transcript_id',\n",
    "                 'original_transcript_id',\n",
    "                 'original_transcript_name']].copy(deep=True)\n",
    "    m_df = m_df.groupby(gb_cols).agg({'original_transcript_id': ','.join,\n",
    "                                      'original_transcript_name': ','.join}).reset_index()\n",
    "    m_df = m_df.merge(temp, on='transcript_id', suffixes=('','_merge'))\n",
    "    m_df.drop(['gene_name', 'gene_id', 'transcript_triplet'],\n",
    "              axis=1, inplace=True)\n",
    "\n",
    "    # add new transcript ids\n",
    "    df = df.merge(m_df, left_on=['transcript_id', 'transcript_name'],\n",
    "                  right_on=['original_transcript_id_merge',\n",
    "                            'original_transcript_name_merge'],\n",
    "                 suffixes=('_x', ''))\n",
    "\n",
    "    # drop old tids\n",
    "    df.drop(['transcript_id_x', 'transcript_name_x',\n",
    "             'original_transcript_name_merge'],\n",
    "            axis=1, inplace=True)\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "c80991fc-edd8-4e25-bbb5-ec1f730936e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "test = map_gtf_transcripts(df, m_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "1ef7aa35-4670-4d04-8932-587527c2c1a7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>...</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>intergenic_novel</th>\n",
       "      <th>antisense_gene</th>\n",
       "      <th>gene_antisense_to_IDs</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>original_transcript_id</th>\n",
       "      <th>original_transcript_name</th>\n",
       "      <th>original_transcript_id_merge</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>663</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1232225</td>\n",
       "      <td>1235041</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000176022.5</td>\n",
       "      <td>B3GALT6</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSP00000496787.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000176022[1,1,1]</td>\n",
       "      <td>B3GALT6[1,1,1]</td>\n",
       "      <td>ENST00000379198.4,ENST00000647651.1</td>\n",
       "      <td>B3GALT6-201,B3GALT6-202</td>\n",
       "      <td>ENST00000647651.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>665</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1232258</td>\n",
       "      <td>1235041</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000176022.5</td>\n",
       "      <td>B3GALT6</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSP00000368496.2</td>\n",
       "      <td>CCDS13.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000176022[1,1,1]</td>\n",
       "      <td>B3GALT6[1,1,1]</td>\n",
       "      <td>ENST00000379198.4,ENST00000647651.1</td>\n",
       "      <td>B3GALT6-201,B3GALT6-202</td>\n",
       "      <td>ENST00000379198.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>369779</th>\n",
       "      <td>chr6</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1609971</td>\n",
       "      <td>1613897</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000054598.7</td>\n",
       "      <td>FOXC1</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSP00000370256.2</td>\n",
       "      <td>CCDS4473.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000054598[1,1,1]</td>\n",
       "      <td>FOXC1[1,1,1]</td>\n",
       "      <td>ENST00000380874.3,ENST00000645831.1</td>\n",
       "      <td>FOXC1-201,FOXC1-202</td>\n",
       "      <td>ENST00000380874.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>369781</th>\n",
       "      <td>chr6</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1610066</td>\n",
       "      <td>1613897</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000054598.7</td>\n",
       "      <td>FOXC1</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSP00000493906.1</td>\n",
       "      <td>CCDS4473.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000054598[1,1,1]</td>\n",
       "      <td>FOXC1[1,1,1]</td>\n",
       "      <td>ENST00000380874.3,ENST00000645831.1</td>\n",
       "      <td>FOXC1-201,FOXC1-202</td>\n",
       "      <td>ENST00000645831.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1155612</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>83508260</td>\n",
       "      <td>83512127</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000196767.7</td>\n",
       "      <td>POU3F4</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSP00000362296.2</td>\n",
       "      <td>CCDS14450.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000196767[1,1,1]</td>\n",
       "      <td>POU3F4[1,1,1]</td>\n",
       "      <td>ENST00000373200.5,ENST00000644024.1</td>\n",
       "      <td>POU3F4-201,POU3F4-202</td>\n",
       "      <td>ENST00000373200.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1155614</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>83508292</td>\n",
       "      <td>83512127</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000196767.7</td>\n",
       "      <td>POU3F4</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSP00000495996.1</td>\n",
       "      <td>CCDS14450.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000196767[1,1,1]</td>\n",
       "      <td>POU3F4[1,1,1]</td>\n",
       "      <td>ENST00000373200.5,ENST00000644024.1</td>\n",
       "      <td>POU3F4-201,POU3F4-202</td>\n",
       "      <td>ENST00000644024.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6 rows × 39 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chromosome  Source     Feature     Start       End Score Strand Frame  \\\n",
       "663           chr1  HAVANA  transcript   1232225   1235041     .      +     .   \n",
       "665           chr1  HAVANA  transcript   1232258   1235041     .      +     .   \n",
       "369779        chr6  HAVANA  transcript   1609971   1613897     .      +     .   \n",
       "369781        chr6  HAVANA  transcript   1610066   1613897     .      +     .   \n",
       "1155612       chrX  HAVANA  transcript  83508260  83512127     .      +     .   \n",
       "1155614       chrX  HAVANA  transcript  83508292  83512127     .      +     .   \n",
       "\n",
       "                   gene_id gene_name  ...         protein_id       ccdsid  \\\n",
       "663      ENSG00000176022.5   B3GALT6  ...  ENSP00000496787.1          NaN   \n",
       "665      ENSG00000176022.5   B3GALT6  ...  ENSP00000368496.2     CCDS13.1   \n",
       "369779   ENSG00000054598.7     FOXC1  ...  ENSP00000370256.2   CCDS4473.1   \n",
       "369781   ENSG00000054598.7     FOXC1  ...  ENSP00000493906.1   CCDS4473.1   \n",
       "1155612  ENSG00000196767.7    POU3F4  ...  ENSP00000362296.2  CCDS14450.1   \n",
       "1155614  ENSG00000196767.7    POU3F4  ...  ENSP00000495996.1  CCDS14450.1   \n",
       "\n",
       "        intergenic_novel antisense_gene gene_antisense_to_IDs  \\\n",
       "663                  NaN            NaN                   NaN   \n",
       "665                  NaN            NaN                   NaN   \n",
       "369779               NaN            NaN                   NaN   \n",
       "369781               NaN            NaN                   NaN   \n",
       "1155612              NaN            NaN                   NaN   \n",
       "1155614              NaN            NaN                   NaN   \n",
       "\n",
       "                  transcript_id transcript_name  \\\n",
       "663      ENSG00000176022[1,1,1]  B3GALT6[1,1,1]   \n",
       "665      ENSG00000176022[1,1,1]  B3GALT6[1,1,1]   \n",
       "369779   ENSG00000054598[1,1,1]    FOXC1[1,1,1]   \n",
       "369781   ENSG00000054598[1,1,1]    FOXC1[1,1,1]   \n",
       "1155612  ENSG00000196767[1,1,1]   POU3F4[1,1,1]   \n",
       "1155614  ENSG00000196767[1,1,1]   POU3F4[1,1,1]   \n",
       "\n",
       "                      original_transcript_id original_transcript_name  \\\n",
       "663      ENST00000379198.4,ENST00000647651.1  B3GALT6-201,B3GALT6-202   \n",
       "665      ENST00000379198.4,ENST00000647651.1  B3GALT6-201,B3GALT6-202   \n",
       "369779   ENST00000380874.3,ENST00000645831.1      FOXC1-201,FOXC1-202   \n",
       "369781   ENST00000380874.3,ENST00000645831.1      FOXC1-201,FOXC1-202   \n",
       "1155612  ENST00000373200.5,ENST00000644024.1    POU3F4-201,POU3F4-202   \n",
       "1155614  ENST00000373200.5,ENST00000644024.1    POU3F4-201,POU3F4-202   \n",
       "\n",
       "        original_transcript_id_merge  \n",
       "663                ENST00000647651.1  \n",
       "665                ENST00000379198.4  \n",
       "369779             ENST00000380874.3  \n",
       "369781             ENST00000645831.1  \n",
       "1155612            ENST00000373200.5  \n",
       "1155614            ENST00000644024.1  \n",
       "\n",
       "[6 rows x 39 columns]"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test.loc[(test.original_transcript_id.str.contains(','))&(test.Feature == 'transcript')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "9e623498-d5cb-42e0-86d9-96b0b6c81974",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "663        ENST00000379198.4,ENST00000647651.1\n",
       "665        ENST00000379198.4,ENST00000647651.1\n",
       "369779     ENST00000380874.3,ENST00000645831.1\n",
       "369781     ENST00000380874.3,ENST00000645831.1\n",
       "1155612    ENST00000373200.5,ENST00000644024.1\n",
       "1155614    ENST00000373200.5,ENST00000644024.1\n",
       "Name: original_transcript_id, dtype: object"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test.loc[(test.original_transcript_id.str.contains(','))&(test.Feature == 'transcript'), 'original_transcript_id']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "0b15d849-f299-4880-9d99-ef56b31937c7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    ENCODEHG000058846[1,1,1]\n",
       "1    ENCODEHG000058846[1,1,1]\n",
       "2    ENCODEHG000058846[1,1,1]\n",
       "3    ENCODEHG000058846[1,1,1]\n",
       "4    ENCODEHG000058837[2,4,1]\n",
       "Name: transcript_id, dtype: object"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test.transcript_id.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "8e0ab4fa-925a-46c3-8e1f-13f7dc1d1ac1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# tests for map_gtf_ids\n",
    "# - transcript that does not have duplicate\n",
    "# - transcript that does have a duplicate\n",
    "\n",
    "def make_exon_df(n,c,e,s,g,t):\n",
    "    df = pd.DataFrame()\n",
    "    df['Chromosome'] = c\n",
    "    df['Start'] = [i[0] for i in e]\n",
    "    df['End'] = [i[1] for i in e]\n",
    "    df['Strand'] = s\n",
    "    df['Feature'] = 'exon'\n",
    "    cols = ['gene_name', 'gene_id']\n",
    "    for c in cols:\n",
    "        df[c] = g\n",
    "    cols = ['transcript_id', 'transcript_name']\n",
    "    for c in cols:\n",
    "        df[c] = t\n",
    "\n",
    "    # reorder exons and starts/ stops if needed\n",
    "    df['new_Start'] = df[['Start', 'End']].min(axis=1)\n",
    "    df['new_End'] = df[['Start', 'End']].max(axis=1)\n",
    "    df.drop(['Start', 'End'], axis=1, inplace=True)\n",
    "    df.rename({'new_Start':'Start',\n",
    "               'new_End':'End'}, axis=1, inplace=True)\n",
    "    s = s[0]\n",
    "    if s == '+':\n",
    "        ascending = True\n",
    "    elif s == '-':\n",
    "        ascending = False\n",
    "    df.sort_values(by='Start', ascending=ascending, inplace=True)\n",
    "    return df\n",
    "\n",
    "def make_hier_entry(df, how='t'):\n",
    "    \"\"\"\n",
    "    kind {'g','t'}\n",
    "    \"\"\"\n",
    "    agg_dict = {'min_coord': 'min', 'max_coord': 'max'}\n",
    "    t_df = df.copy(deep=True)\n",
    "    t_df['min_coord'] = t_df[['Start', 'End']].min(axis=1)\n",
    "    t_df['max_coord'] = t_df[['Start', 'End']].max(axis=1)\n",
    "    if how == 't':\n",
    "        gb_cols = ['Chromosome', 'Strand', 'gene_name',\n",
    "                   'gene_id', 'transcript_id', 'transcript_name']\n",
    "    elif how == 'g':\n",
    "        gb_cols = ['Chromosome', 'Strand', 'gene_name',\n",
    "                   'gene_id']\n",
    "        \n",
    "    cols = gb_cols + ['min_coord', 'max_coord']\n",
    "    t_df = t_df[cols]\n",
    "    t_df = t_df.groupby(gb_cols).agg(agg_dict).reset_index()\n",
    "    t_df.rename({'min_coord': 'Start', 'max_coord': 'End'}, axis=1, inplace=True)\n",
    "    if how == 't':\n",
    "        t_df['Feature'] = 'transcript'\n",
    "    elif how == 'g':\n",
    "        t_df['Feature'] = 'gene'\n",
    "        \n",
    "    return t_df\n",
    "\n",
    "def make_test_gtf(ts):\n",
    "    df = pd.concat(ts)\n",
    "        # make transcript entries\n",
    "    t_df = make_hier_entry(df, how='t')\n",
    "    # make gene entries\n",
    "    g_df = make_hier_entry(df, how='g')\n",
    "\n",
    "    # concat everything and sort by gene id, transcript id, feature rank (gene =0, t =1, exon=2), then start coords\n",
    "    df = pd.concat([df, t_df, g_df])\n",
    "    df['feature_rank'] = df.Feature.map({'gene':0, 'transcript':1, 'exon':2})\n",
    "\n",
    "    rev = df.loc[df.Strand == '-'].copy(deep=True)\n",
    "    fwd = df.loc[df.Strand == '+'].copy(deep=True)\n",
    "\n",
    "    df = pd.DataFrame()\n",
    "    for temp in [fwd, rev]:\n",
    "        if len(temp.index) > 0:\n",
    "            strand = temp.Strand.values.tolist()[0]\n",
    "            if strand == '+':\n",
    "                ascending = True\n",
    "            elif strand == '-':\n",
    "                ascending = False\n",
    "            # temp.sort_values(by=['feature_rank', 'gene_id', 'transcript_id', 'Start'],\n",
    "            #                ascending=[True, True, True, ascending], inplace=True)\n",
    "            temp.sort_values(by=['transcript_id', 'gene_id', 'feature_rank', 'Start'], \n",
    "                             ascending=[True, True, True, ascending],\n",
    "                             na_position='first', inplace=True)\n",
    "            \n",
    "            df = pd.concat([df, temp])\n",
    "    # df.drop('feature_rank', axis=1, inplace=True)\n",
    "    return df\n",
    "\n",
    "ts = []\n",
    "# t1 - transcript that doesn't need to be merged\n",
    "n = 3\n",
    "c = ['1' for i in range(n)]\n",
    "e = [[1,10], [14,20], [25,30]]\n",
    "s = ['+' for i in range(n)]\n",
    "g = 'g1'\n",
    "t = 'g1_t1'\n",
    "ts.append(make_exon_df(n,c,e,s,g,t))\n",
    "\n",
    "# t2 - rev. strand transcript that needs to be merged\n",
    "n = 3\n",
    "c = ['1' for i in range(n)]\n",
    "e = [[90,60], [45,30], [10,8]]\n",
    "s = ['-' for i in range(n)]\n",
    "g = 'g2'\n",
    "t = 'g2_t1'\n",
    "ts.append(make_exon_df(n,c,e,s,g,t))\n",
    "\n",
    "# t3 - rev. strand transcript that needs to be merged\n",
    "n = 3\n",
    "c = ['1' for i in range(n)]\n",
    "e = [[95,60], [45,30], [10,6]]\n",
    "s = ['-' for i in range(n)]\n",
    "g = 'g2'\n",
    "t = 'g2_t2'\n",
    "ts.append(make_exon_df(n,c,e,s,g,t))\n",
    "\n",
    "df = make_test_gtf(ts)\n",
    "\n",
    "# # make transcript entries\n",
    "# t_df = make_hier_entry(df, how='t')\n",
    "# # make gene entries\n",
    "# g_df = make_hier_entry(df, how='g')\n",
    "\n",
    "# # concat everything and sort by gene id, transcript id, feature rank (gene =0, t =1, exon=2), then start coords\n",
    "# df = pd.concat([df, t_df, g_df])\n",
    "# df['feature_rank'] = df.Feature.map({'gene':0, 'transcript':1, 'exon':2})\n",
    "\n",
    "# rev = df.loc[df.Strand == '-'].copy(deep=True)\n",
    "# fwd = df.loc[df.Strand == '+'].copy(deep=True)\n",
    "\n",
    "# df = pd.DataFrame()\n",
    "# for temp in [fwd, rev]:\n",
    "#     if len(temp.index) > 0:\n",
    "#         strand = temp.Strand.values.tolist()[0]\n",
    "#         if strand == '+':\n",
    "#             ascending = True\n",
    "#         elif strand == '-':\n",
    "#             ascending = False\n",
    "#         temp.sort_values(by=['feature_rank', 'gene_id', 'transcript_id', 'Start'],\n",
    "#                        ascending=[True, True, True, ascending], inplace=True)\n",
    "#         df = pd.concat([df, temp])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "id": "d0044c70-1a2a-41bd-ad01-21da1137b860",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>feature_rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>gene</td>\n",
       "      <td>g1</td>\n",
       "      <td>g1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>transcript</td>\n",
       "      <td>g1</td>\n",
       "      <td>g1</td>\n",
       "      <td>g1_t1</td>\n",
       "      <td>g1_t1</td>\n",
       "      <td>1</td>\n",
       "      <td>30</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>g1</td>\n",
       "      <td>g1</td>\n",
       "      <td>g1_t1</td>\n",
       "      <td>g1_t1</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>g1</td>\n",
       "      <td>g1</td>\n",
       "      <td>g1_t1</td>\n",
       "      <td>g1_t1</td>\n",
       "      <td>14</td>\n",
       "      <td>20</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>g1</td>\n",
       "      <td>g1</td>\n",
       "      <td>g1_t1</td>\n",
       "      <td>g1_t1</td>\n",
       "      <td>25</td>\n",
       "      <td>30</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>gene</td>\n",
       "      <td>g2</td>\n",
       "      <td>g2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6</td>\n",
       "      <td>95</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>transcript</td>\n",
       "      <td>g2</td>\n",
       "      <td>g2</td>\n",
       "      <td>g2_t1</td>\n",
       "      <td>g2_t1</td>\n",
       "      <td>8</td>\n",
       "      <td>90</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>g2</td>\n",
       "      <td>g2</td>\n",
       "      <td>g2_t1</td>\n",
       "      <td>g2_t1</td>\n",
       "      <td>60</td>\n",
       "      <td>90</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>g2</td>\n",
       "      <td>g2</td>\n",
       "      <td>g2_t1</td>\n",
       "      <td>g2_t1</td>\n",
       "      <td>30</td>\n",
       "      <td>45</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>g2</td>\n",
       "      <td>g2</td>\n",
       "      <td>g2_t1</td>\n",
       "      <td>g2_t1</td>\n",
       "      <td>8</td>\n",
       "      <td>10</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>transcript</td>\n",
       "      <td>g2</td>\n",
       "      <td>g2</td>\n",
       "      <td>g2_t2</td>\n",
       "      <td>g2_t2</td>\n",
       "      <td>6</td>\n",
       "      <td>95</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>g2</td>\n",
       "      <td>g2</td>\n",
       "      <td>g2_t2</td>\n",
       "      <td>g2_t2</td>\n",
       "      <td>60</td>\n",
       "      <td>95</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>g2</td>\n",
       "      <td>g2</td>\n",
       "      <td>g2_t2</td>\n",
       "      <td>g2_t2</td>\n",
       "      <td>30</td>\n",
       "      <td>45</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>g2</td>\n",
       "      <td>g2</td>\n",
       "      <td>g2_t2</td>\n",
       "      <td>g2_t2</td>\n",
       "      <td>6</td>\n",
       "      <td>10</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome Strand     Feature gene_name gene_id transcript_id  \\\n",
       "0          1      +        gene        g1      g1           NaN   \n",
       "0          1      +  transcript        g1      g1         g1_t1   \n",
       "0          1      +        exon        g1      g1         g1_t1   \n",
       "1          1      +        exon        g1      g1         g1_t1   \n",
       "2          1      +        exon        g1      g1         g1_t1   \n",
       "1          1      -        gene        g2      g2           NaN   \n",
       "1          1      -  transcript        g2      g2         g2_t1   \n",
       "0          1      -        exon        g2      g2         g2_t1   \n",
       "1          1      -        exon        g2      g2         g2_t1   \n",
       "2          1      -        exon        g2      g2         g2_t1   \n",
       "2          1      -  transcript        g2      g2         g2_t2   \n",
       "0          1      -        exon        g2      g2         g2_t2   \n",
       "1          1      -        exon        g2      g2         g2_t2   \n",
       "2          1      -        exon        g2      g2         g2_t2   \n",
       "\n",
       "  transcript_name  Start  End  feature_rank  \n",
       "0             NaN      1   30             0  \n",
       "0           g1_t1      1   30             1  \n",
       "0           g1_t1      1   10             2  \n",
       "1           g1_t1     14   20             2  \n",
       "2           g1_t1     25   30             2  \n",
       "1             NaN      6   95             0  \n",
       "1           g2_t1      8   90             1  \n",
       "0           g2_t1     60   90             2  \n",
       "1           g2_t1     30   45             2  \n",
       "2           g2_t1      8   10             2  \n",
       "2           g2_t2      6   95             1  \n",
       "0           g2_t2     60   95             2  \n",
       "1           g2_t2     30   45             2  \n",
       "2           g2_t2      6   10             2  "
      ]
     },
     "execution_count": 137,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d54891e-6260-4a1b-9782-3836b53b1005",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61f503ed-de47-4d30-94b6-9cfb9b500f14",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # tests for update_gtf_ends()\n",
    "# def make_gtf(c,e,s,g,t):\n",
    "\n",
    "\n",
    "# # t1 - transcript that doesn't need to be merged\n",
    "# n = 3\n",
    "# c = ['1' for i in range(n)]\n",
    "# e [[1,10], [14,20], [25,30]]\n",
    "# s = ['+' for i in range(n)]\n",
    "# g = 'g1'\n",
    "# t = 'g1_t1'\n",
    "\n",
    "# # t2 - rev. strand transcript that needs ends to be updated\n",
    "# n = 3\n",
    "# c = ['1' for i in range(n)]\n",
    "# e = [[90,60], [45,30], [10,8]]\n",
    "# s = ['-' for i in range(n)]\n",
    "# g = 'g2'\n",
    "# t = 'g1_t1'\n",
    "\n",
    "# df = pd.DataFrame()\n",
    "# df['Chromosome'] = c\n",
    "# df['Start'] = [i[0] for i in e]\n",
    "# df['End'] = [i[1] for i in e]\n",
    "# df['Strand'] = s\n",
    "# cols = ['gene_name', 'gene_id']\n",
    "# for c in cols:\n",
    "#     df[c] = g\n",
    "# cols = []\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
