{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f132506d-8571-4973-ab21-2eae8930bd12",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "cdc92cb0-55e9-4f78-b4ac-3f92256eaaab",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame()\n",
    "\n",
    "chroms = [1,1,1]\n",
    "starts = [1,5,10]\n",
    "ends = [3, 8, 15]\n",
    "strands = ['+' for i in range(len(ends))]\n",
    "\n",
    "df['Chromosome'] = chroms\n",
    "df['Start'] = starts\n",
    "df['End'] = ends\n",
    "df['Strand'] = strands\n",
    "\n",
    "df = pd.read_csv('my_file.csv')\n",
    "\n",
    "df['Feature'] = 'exon'\n",
    "df['transcript_id'] = 'transcript_1'\n",
    "df['gene_id'] = 'Clu'\n",
    "\n",
    "df = pr.PyRanges(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "7f991826-d816-4c93-8ddd-a3f15f279b7a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>transcript_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>8</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>transcript_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>15</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>transcript_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+-----------------+\n",
       "|   Chromosome |     Start |       End | Strand       | Feature    | transcript_id   |\n",
       "|   (category) |   (int32) |   (int32) | (category)   | (object)   | (object)        |\n",
       "|--------------+-----------+-----------+--------------+------------+-----------------|\n",
       "|            1 |         1 |         3 | +            | exon       | transcript_1    |\n",
       "|            1 |         5 |         8 | +            | exon       | transcript_1    |\n",
       "|            1 |        10 |        15 | +            | exon       | transcript_1    |\n",
       "+--------------+-----------+-----------+--------------+------------+-----------------+\n",
       "Stranded PyRanges object has 3 rows and 6 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "87063b93-04fb-4c4a-97b5-4bcf5b327d3b",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_gtf('my_test_gtf.gtf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "4594a2db-c698-4deb-a5c8-a7122f65a076",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('/Users/fairliereese/Downloads/Clu-h2kbKI-1A.txt', sep='\\t', header=None, names=)\n",
    "# df = df.loc[df.Feature == 'exon']\n",
    "\n",
    "# df['transcript_id'] = 'transcript_1'\n",
    "# df['gene_id'] = 'Clu'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "6bd272ad-6da7-47eb-ad3d-654daa6221df",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Clu-h2kBKI</td>\n",
       "      <td>.</td>\n",
       "      <td>igr</td>\n",
       "      <td>1</td>\n",
       "      <td>1468</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=intergenic_region</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Clu-h2kbKI</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>1469</td>\n",
       "      <td>1674</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon1A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Clu-h2kbKI</td>\n",
       "      <td>.</td>\n",
       "      <td>intron</td>\n",
       "      <td>1675</td>\n",
       "      <td>4717    .</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=intron1A</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Clu-h2kbKI</td>\n",
       "      <td>.</td>\n",
       "      <td>exon    4718</td>\n",
       "      <td>4838</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Clu-h2kbKI</td>\n",
       "      <td>.</td>\n",
       "      <td>intron</td>\n",
       "      <td>4839</td>\n",
       "      <td>6310</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=intron2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            0  1             2     3          4  5  6            7  \\\n",
       "0  Clu-h2kBKI  .           igr     1       1468  .  +            .   \n",
       "1  Clu-h2kbKI  .          exon  1469       1674  .  +            .   \n",
       "2  Clu-h2kbKI  .        intron  1675  4717    .  +  .  ID=intron1A   \n",
       "3  Clu-h2kbKI  .  exon    4718  4838          .  +  .     ID=exon2   \n",
       "4  Clu-h2kbKI  .        intron  4839       6310  .  +            .   \n",
       "\n",
       "                      8  \n",
       "0  ID=intergenic_region  \n",
       "1             ID=exon1A  \n",
       "2                   NaN  \n",
       "3                   NaN  \n",
       "4            ID=intron2  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0d3994b-c215-4dee-938d-97d0200c718f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
